---
title: Authors
tbl-colwidths: [5,15,80]
layout-valign: middle
---

The creation of this text book was developed through a series of ...


| | | |
|-|:-:|-|
| 2022 | ![](assets/images/headshots/FELLOWS_YATES_James.jpg){width="100px" fig-align="center"} | ðŸ‡¬ðŸ‡§ **James Fellows Yates** is an archaeology-trained biomolecular archaeologist and convert to palaeogenomics, and is recently pivoting to bioinformatics. He specialises in ancient metagenomics analysis, generating tools and high-throughput approaches and high-quality pipelines for validating and analysing ancient (oral) microbiomes and palaeogenomic data. |
| 2022 | ![](assets/images/headshots/WARINNER_Christina.png){width="100px" fig-align="center"} | ðŸ‡ºðŸ‡¸ **Christina Warinner** is Group Leader of Microbiome Sciences at the Max Planck Institute for Evolutionary Anthropology in Leipzig, Germany, and Associate Professor of Anthropology at Harvard University. She serves on the Leadership Team of the Max Planck-Harvard Research Center for the Archaeoscience of the Ancient Mediterranean (MHAAM), and is a Professor in the Faculty of Biological Sciences at Friedrich Schiller University in Jena, Germany. Her research focuses on the use of metagenomics and paleoproteomics to better understand past human diet, health, and the evolution of the human microbiome. |
| 2022 | ![](assets/images/headshots/ANDRADES_VALTUENA_Aida.jpg){width="100px" layout-valign="middle"} | ðŸ‡ªðŸ‡¸ **Aida Andrades ValtueÃ±a** is a geneticist interested in pathogen evolution, with particular interest in prehistoric pathogens. She has been exploring new methods to analyse ancient pathogen data to understand their past function and ecology to inform models of pathogen emergence. |
| 2022 | ![](assets/images/headshots/HERBIG_Alexander.jpeg){width="100px" fig-align="center"} | ðŸ‡©ðŸ‡ª **Alexander Herbig** is a bioinformatician and group leader for Computational Pathogenomics at the Max Planck Institute for Evolutionary Anthropology. His main interest is in studying the evolution of human pathogens and in methods development for pathogen detection and bacterial genomics. |
| 2022 | ![](assets/images/headshots/HUEBNER_Alex.jpg){width="100px" fig-align="center"} | ðŸ‡©ðŸ‡ª **Alex HÃ¼bner** is a computational biologist, who originally studied biotechnology, before switching to evolutionary biology during his PhD. For his postdoc in the Warinner lab, he focuses on investigating whether new methods in the field of modern metagenomics can be directly applied to ancient DNA data. Here, he is particularly interested in the _de novo_ assembly of ancient metagenomic sequencing data and the subsequent analysis of its results. |
| 2022 | ![](assets/images/headshots/HISS_Alina.jpg){width="100px" fig-align="center"} | ðŸ‡©ðŸ‡ª **Alina Hiss** is a PhD student in the Computational Pathogenomics group at the Max Planck Institute for Evolutionary Anthropology. She is interested in the evolution of human pathogens and working on material from the Carpathian basin to gain insights about the presence and spread of pathogens in the region during the Early Medieval period. |
| 2022 | ![](assets/images/headshots/KOCHER_Arthur.jpg){width="100px" fig-align="center"} | ðŸ‡«ðŸ‡· **Arthur Kocher** initially trained as a veterinarian. He then pursued a PhD in the field of disease ecology, during which he studied the impact of biodiversity changes on the transmission of zoonotic diseases using molecular tools such as DNA metabarcoding. During his Post-Docs, he extended his research focus to evolutionary aspects of pathogens, which he currently investigates using ancient genomic data and Bayesian phylogenetics. |
| 2022 | ![](assets/images/headshots/SCHMID_Clemens.JPG){width="100px" fig-align="center"} | ðŸ‡©ðŸ‡ª **Clemens Schmid** is a computational archaeologist pursuing a PhD in the group of Stephan Schiffels at the department of Archaeogenetics at the Max Planck Institute for Evolutionary Anthropology. He is trained both in archaeology and computer science and currently develops computational methods for the spatiotemporal co-analysis of archaeological and ancient genomic data. He worked in research projects on the European Neolithic, Copper and Bronze age and maintains research software in R, C++ and Haskell. |
| 2022 | ![](assets/images/headshots/VELSKO_Irina.jpeg){width="100px" fig-align="center"} | ðŸ‡ºðŸ‡¸ **Irina Velsko** is a postdoc in the Microbiome group of the department of Archaeogenetics at the Max Planck Institute for Evolutionary Anthropology. She did her PhD work on oral microbiology and immunology of the living, and now works on oral microbiomes of the living and the dead. Her work focuses on the evolution and ecology of dental plaque biofilms, both modern and ancient, and the complex interplay between microbiomes and their hosts. |
| 2022 | ![](assets/images/headshots/BORRY_Maxime.png){width="100px" fig-align="center"} | ðŸ‡«ðŸ‡· **Maxime Borry** is a doctoral researcher in bioinformatics at the Max Planck Institute for Evolutionary Anthropology in Germany. After an undergraduate in life sciences and a master in Ecology, followed by a master in bioinformatics, he is now working on the completion of his PhD, focused on developing new tools and data analysis of ancient metagenomic samples. |
| 2022 | ![](assets/images/headshots/MICHEL_Megan.jpg){width="100px" fig-align="center"} | ðŸ‡ºðŸ‡¸ **Megan Michel** is a PhD student jointly affiliated with the Archaeogenetics Department at the Max Planck Institute for Evolutionary Anthropology and the Human Evolutionary Biology Department at Harvard University. Her research focuses on using computational genomic analyses to understand how pathogens have co-evolved with their hosts over the course of human history. |
| 2022 | ![](assets/images/headshots/OSKOLKOV_Nikolay.jpg){width="100px" fig-align="center"} | ðŸ‡·ðŸ‡º **Nikolay Oskolkov** is a bioinformatician at Lund University and the bioinformatics platform of SciLifeLab, Sweden. He defended his PhD in theoretical physics in 2007, and switched to life sciences in 2012. His research interests include mathematical statistics and machine learning applied to genetics and genomics, single cell and ancient metagenomics data analysis. |
| 2022 | ![](assets/images/headshots/DUCHENE_Sebastian.jpeg){width="100px" fig-align="center"} | ðŸ‡¦ðŸ‡º **Sebastian Duchene** is an Australian Research Council Fellow at the Doherty Institute for Infection and Immunity at the University of Melbourne, Australia. Prior to joining the University of Melbourne he obtained his PhD and conducted postdoctoral work at the University of Sydney. His research is in molecular evolution and epidemiology of infectious pathogens, notably viruses and bacteria, and developing Bayesian phylodynamic methods. |
| 2022 | ![](assets/images/headshots/LAMNIDIS_Thiseas.jpg){width="100px" fig-align="center"} | ðŸ‡¬ðŸ‡· **Thiseas Lamnidis** is a human population geneticist interested in European population history after the Bronze Age. To gain the required resolution to differentiate between Iron Age European populations, he is developing analytical methods based on the sharing of rare variation between individuals. He has also contributed to pipelines that streamline the processing and analysis of genetic data in a reproducible manner, while also facilitating dissemination of information among interdisciplinary colleagues. |