<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Alexander Hübner">

<title>Introduction to Ancient Metagenomics - 12&nbsp; Introduction to de novo Genome Assembly</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./authentication-decontamination.html" rel="next">
<link href="./functional-profiling.html" rel="prev">
<link href="./favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<meta property="og:title" content="Introduction to Ancient Metagenomics - 12&nbsp; Introduction to de novo Genome Assembly">
<meta property="og:description" content="">
<meta property="og:image" content="https://spaam-community.github.org/intro-to-ancient-metagenomics-book/assets/images/chapters/denovo-assembly/metagenome_assembly_scheme.png">
<meta property="og:site-name" content="Introduction to Ancient Metagenomics">
<meta property="og:image:height" content="987">
<meta property="og:image:width" content="1564">
</head>

<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-ancient-metagenomics.html">Ancient Metagenomics</a></li><li class="breadcrumb-item"><a href="./denovo-assembly.html"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to _de novo_ Genome Assembly</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main tools-wide">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-evolutionary-biology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Evolutionary Biology</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication-decontamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Authentication and Decontamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Introduction to Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancientmetagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lecture" id="toc-lecture" class="nav-link active" data-scroll-target="#lecture"><span class="header-section-number">12.0.1</span> Lecture</a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction"><span class="header-section-number">12.1</span> Introduction</a></li>
  <li><a href="#practical-course" id="toc-practical-course" class="nav-link" data-scroll-target="#practical-course"><span class="header-section-number">12.2</span> Practical course</a>
  <ul class="collapse">
  <li><a href="#sample-overview" id="toc-sample-overview" class="nav-link" data-scroll-target="#sample-overview"><span class="header-section-number">12.2.1</span> Sample overview</a></li>
  <li><a href="#preparing-the-sequencing-data-for-de-novo-assembly" id="toc-preparing-the-sequencing-data-for-de-novo-assembly" class="nav-link" data-scroll-target="#preparing-the-sequencing-data-for-de-novo-assembly"><span class="header-section-number">12.2.2</span> Preparing the sequencing data for <em>de novo</em> assembly</a></li>
  <li><a href="#de-novo-assembly" id="toc-de-novo-assembly" class="nav-link" data-scroll-target="#de-novo-assembly"><span class="header-section-number">12.2.3</span> <em>De novo</em> assembly</a></li>
  <li><a href="#aligning-the-short-read-data-against-the-contigs" id="toc-aligning-the-short-read-data-against-the-contigs" class="nav-link" data-scroll-target="#aligning-the-short-read-data-against-the-contigs"><span class="header-section-number">12.2.4</span> Aligning the short-read data against the contigs</a></li>
  <li><a href="#reconstructing-metagenome-assembled-genomes" id="toc-reconstructing-metagenome-assembled-genomes" class="nav-link" data-scroll-target="#reconstructing-metagenome-assembled-genomes"><span class="header-section-number">12.2.5</span> Reconstructing metagenome-assembled genomes</a></li>
  <li><a href="#taxonomic-assignment-of-contigs" id="toc-taxonomic-assignment-of-contigs" class="nav-link" data-scroll-target="#taxonomic-assignment-of-contigs"><span class="header-section-number">12.2.6</span> Taxonomic assignment of contigs</a></li>
  <li><a href="#taxonomic-assignment-of-mags" id="toc-taxonomic-assignment-of-mags" class="nav-link" data-scroll-target="#taxonomic-assignment-of-mags"><span class="header-section-number">12.2.7</span> Taxonomic assignment of MAGs</a></li>
  <li><a href="#evaluating-the-amount-of-ancient-dna-damage" id="toc-evaluating-the-amount-of-ancient-dna-damage" class="nav-link" data-scroll-target="#evaluating-the-amount-of-ancient-dna-damage"><span class="header-section-number">12.2.8</span> Evaluating the amount of ancient DNA damage</a></li>
  <li><a href="#annotating-genomes-for-function" id="toc-annotating-genomes-for-function" class="nav-link" data-scroll-target="#annotating-genomes-for-function"><span class="header-section-number">12.2.9</span> Annotating genomes for function</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">12.2.10</span> Summary</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Alexander Hübner </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to create the <a href="./before-you-start.html#creating-a-conda-environment">conda environment</a> from the <code>yml</code> file in the following <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book/raw/main/assets/envs/denovo-assembly.yml">link</a> (right click and save as to download), and once created, activate the environment with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate denovo-assembly</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="lecture" class="level3" data-number="12.0.1">
<h3 data-number="12.0.1" class="anchored" data-anchor-id="lecture"><span class="header-section-number">12.0.1</span> Lecture</h3>
<p>Lecture slides and video from the <a href="https://www.spaam-community.org/wss-summer-school/#/2022/README">2022 edition of the summer school</a>.</p>
<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vQiyXe-6yn1hfaoibwI8DBYHubQxEd_WEvXte6JqyyjnuY8DXfP64360tRvhmRxOPAoM4Gt6hqKG5ts/embed?start=false&amp;loop=true&amp;delayms=10000" frameborder="0" width="100%" height="400px" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true">
</iframe>
<p>PDF version of these slides can be downloaded from <a href="https://github.com/SPAAM-community/wss-summer-school/raw/main/docs/assets/slides/2022/4c-intro-to-denovoassembly/SPAAM%20Summer%20School%202022%20-%204C%20-%20Genome%20Assembly.pdf">here</a>.</p>
<iframe width="100%" height="400px" src="https://www.youtube.com/embed/1-o4449Hu4o" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="">
</iframe>
</section>
<section id="introduction" class="level2" data-number="12.1">
<h2 data-number="12.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">12.1</span> Introduction</h2>
<p>First of all, what is a <strong>metagenomic</strong> sample? Metagenomic sample is a sample that consists of DNA from more than one source. The number and the type of sources might vary widely between different samples. Typical sources for ancient remains are e.g.&nbsp;the host organism and the microbial species. The important part is that we generally do not know the origin of a DNA molecule prior to analysing the sequencing data generated from the DNA library of this sample. In the example presented in (<a href="#fig-denovoassembly-overview">Figure&nbsp;<span>12.1</span></a>), our metagenomic sample has DNA from three different sources, here coloured in blue, red, and yellow.</p>
<div id="fig-denovoassembly-overview" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/metagenome_assembly_scheme.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.1: Overview of the ways how to analyse metagenomic sequencing data.</figcaption>
</figure>
</div>
<p>How can we determine the sources of the DNA that we have in our metagenomic sample? There are three main options whose <em>pros</em> and <em>cons</em> are summarised in (<a href="#tbl-denovoassembly-proscons">Table&nbsp;<span>12.1</span></a>).</p>
<div class="cell tbl-cap-location-top">
<div id="tbl-denovoassembly-proscons" class="anchored">
<table class="table table-sm table-striped small">
<caption>Table&nbsp;12.1: Pros and cons of the major methods for determining the sources of metagenomic DNA.</caption>
<colgroup>
<col style="width: 30%">
<col style="width: 34%">
<col style="width: 35%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">method</th>
<th style="text-align: center;">pros</th>
<th style="text-align: center;">cons</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">reference-based alignment</td>
<td style="text-align: center;">highly sensitive, applicable to aDNA samples</td>
<td style="text-align: center;">requires all sources to be represented in database</td>
</tr>
<tr class="even">
<td style="text-align: center;">single-genome assembly</td>
<td style="text-align: center;">high-quality genomes from cultivated bacteria</td>
<td style="text-align: center;">not available for ancient DNA samples</td>
</tr>
<tr class="odd">
<td style="text-align: center;">metagenome assembly</td>
<td style="text-align: center;">able to recover unknown diversity present in sample</td>
<td style="text-align: center;">highly dependent on preservation of ancient DNA</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Until recently, the only option for ancient DNA samples was to take the short-read sequencing data and align them against some known reference genomes. However, this approach is heavily relying on whether all sources of our samples are represented in the available reference genomes. If a source is missing in the reference database - in our toy example, this is the case for the yellow source (<a href="#fig-denovoassembly-overview">Figure&nbsp;<span>12.1</span></a>) -, then we won’t be able to detect it using this reference database.</p>
<p>While there is a potential workaround for modern metagenomic samples, <em>single-genome assembly</em> relies on being able to cultivate a microbial species to obtain an isolate. This is unfeasible for ancient metagenomic samples because there are no more viable microbial cells available that could be cultivated.</p>
<p>Around 2015, a technical revolution started when the first programs, e.g.&nbsp;MEGAHIT <span class="citation" data-cites="LiMegahit2015">(<a href="introduction-to-ngs-sequencing.html#ref-LiMegahit2015" role="doc-biblioref">Li et al. 2015</a>)</span> and metaSPAdes <span class="citation" data-cites="Nurk2017">(<a href="introduction-to-ngs-sequencing.html#ref-Nurk2017" role="doc-biblioref">Nurk et al. 2017</a>)</span>, were published that could successfully perform <em>de novo</em> assembly from metagenomic data. Since then, tens of thousands metagenomic samples have been assembled and it was revealed that even well studied environments, such as the human gut microbiome, have a lot of additional microbial diversity that has not been observed previously via culturing and isolation <span class="citation" data-cites="Almeida2021">(<a href="introduction-to-ngs-sequencing.html#ref-Almeida2021" role="doc-biblioref">Almeida et al. 2021</a>)</span>.</p>
<p>The technical advancement of being able to perform <em>de novo</em> assembly on metagenomic samples led to an explosion of studies that analysed samples that were considered almost impossible to study beforehand. For researchers that are exposed to ancient DNA, the imminent question arises: can we apply the same methods to ancient DNA data? In this practical course, we will walk through all required steps that are necessary to successfully perform <em>de novo</em> assembly from ancient DNA metagenomic sequencing data and show you what you can do once you have obtained the data.</p>
</section>
<section id="practical-course" class="level2" data-number="12.2">
<h2 data-number="12.2" class="anchored" data-anchor-id="practical-course"><span class="header-section-number">12.2</span> Practical course</h2>
<section id="sample-overview" class="level3" data-number="12.2.1">
<h3 data-number="12.2.1" class="anchored" data-anchor-id="sample-overview"><span class="header-section-number">12.2.1</span> Sample overview</h3>
<p>For this practical course, I selected a palaeofaeces sample from the study by <span class="citation" data-cites="Maixner2021">Maixner et al. (<a href="introduction-to-ngs-sequencing.html#ref-Maixner2021" role="doc-biblioref">2021</a>)</span>, who generated deep metagenomic sequencing data for four palaeofaeces samples that were excavated from an Austrian salt mine in Hallstatt and were associated with the Celtic Iron Age. We will focus on the youngest sample, <strong>2612</strong>, which was dated to be just a few hundred years old (<a href="#fig-denovoassembly-maixner">Figure&nbsp;<span>12.2</span></a>).</p>
<div id="fig-denovoassembly-maixner" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/Maixner2021_GraphicalAbstract.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.2: The graphical abstract of Maixner et al.&nbsp;(2021).</figcaption>
</figure>
</div>
<p>However, because the sample was very deeply sequenced for more than 250 million paired-end reads and we do not want to wait for days for the analysis to finish, we will not use all data but just a sub-sample of them.</p>
<p>You can access the sub-sampled data by changing to the folder</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /vol/volume/denovo-assembly/2612_R1.fastq.gz 2612_R1.fastq.gz</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /vol/volume/denovo-assembly/2612_R2.fastq.gz 2612_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>on the cluster. If you would like to repeat the practical on your own computational infrastructure, you will find information on how to access the files in the following boxes with the title “Self-guided: data preparation”:</p>
<div class="callout callout-style-default callout-note callout-titled" title="Self-guided: data preparation">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Self-guided: data preparation
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>For everyone, who runs this practical on their own infrastructure, you can download the sequencing data from here:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://share.eva.mpg.de/index.php/s/CtLq2R9iqEcAFyg/download/2612_R1.fastq.gz</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://share.eva.mpg.de/index.php/s/mc5JrpDWdL4rC24/download/2612_R2.fastq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How many sequences are in each FastQ file?</strong></p>
<p>Hint: You can run either <code>seqtk size &lt;FastQ file&gt;</code> or <code>bioawk -c fastx 'END{print NR}' &lt;FastQ file&gt;</code> to find this out.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are about 3.25 million paired-end sequences in these files.</p>
</div>
</div>
</div>
</section>
<section id="preparing-the-sequencing-data-for-de-novo-assembly" class="level3" data-number="12.2.2">
<h3 data-number="12.2.2" class="anchored" data-anchor-id="preparing-the-sequencing-data-for-de-novo-assembly"><span class="header-section-number">12.2.2</span> Preparing the sequencing data for <em>de novo</em> assembly</h3>
<p>Before running the actual assembly, we need to pre-process our sequencing data. Typical pre-processing steps include the trimming of adapter sequences and barcodes from the sequencing data and the removal of host or contaminant sequences, such as the bacteriophage PhiX, which is commonly sequenced as a quality control.</p>
<p>Many assembly pipelines, such as <a href="https://nf-co.re/mag">nf-core/mag</a>, have these steps automatically included, however, these steps can be performed prior to it, too. For this practical course, I have performed these steps for us and we could directly continue with the <em>de novo</em> assembly.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The characteristic of ancient DNA samples that pre-determines the success of the <em>de novo</em> assembly is the <strong>distribution of the DNA molecule length</strong>. Determine this distribution prior to running the <em>de novo</em> assembly to be able to predict the results of the <em>de novo</em> assembly.</p>
</div>
</div>
<p>However, since the average length of the insert size of sequencing data (<a href="#fig-denovoassembly-illumina">Figure&nbsp;<span>12.3</span></a>) is highly correlated with the success of the assembly, we want to first evaluate it. For this we can make use of the program fastp <span class="citation" data-cites="Chen2018">(<a href="introduction-to-ngs-sequencing.html#ref-Chen2018" role="doc-biblioref">Chen et al. 2018</a>)</span>.</p>
<div id="fig-denovoassembly-illumina" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/insert_size_scheme.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.3: Scheme of a read pair of Illumina sequencing data.</figcaption>
</figure>
</div>
<p>Fastp will try to overlap the two mates of a read pair and, if this is possible, return the length of the merged sequence, which is identical to insert size or the DNA molecule length. If the two mates cannot be overlapped, we will not be able to know the exact DNA molecule length but only know that it is longer than 290 bp (each read has a length of 150 bp and FastP requires a 11 bp overlap between the reads).</p>
<p>The final histogram of the insert sizes that is returned by FastP can tell us how well preserved the DNA of an ancient sample is (<a href="#fig-denovoassembly-lendist">Figure&nbsp;<span>12.4</span></a>). The more the distribution is skewed to the right, i.e.&nbsp;the longer the DNA molecules are, the more likely we are to obtain long contiguous DNA sequences from the <em>de novo</em> assembly. A distribution that is skewed to the left means that the DNA molecules are more highly degraded and this lowers our chances for obtaining long continuous sequences.</p>
<div id="fig-denovoassembly-lendist" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/Orlando2013_FigS4.5.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.4: Example of a DNA molecule length distribution of a well-preserved ancient DNA sample. This histogram belongs to a 700,000-year-old horse that was preserved in permafrost, as reported in <span class="citation" data-cites="Orlando2013">Orlando et al. (<a href="introduction-to-ngs-sequencing.html#ref-Orlando2013" role="doc-biblioref">2013</a>)</span>, Fig. S4.</figcaption>
</figure>
</div>
<p>To infer the distribution of the DNA molecules, we can run the command</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="at">--in1</span> 2612_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">--in2</span> 2612_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">--stdout</span> <span class="at">--merge</span> <span class="at">-A</span> <span class="at">-G</span> <span class="at">-Q</span> <span class="at">-L</span> <span class="at">--json</span> /dev/null <span class="at">--html</span> overlaps.html <span class="dt">\</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;</span> /dev/null</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Infer the distribution of the DNA molecule length of the sequencing data. Is this sample well-preserved?</strong></p>
<p>Hint: You can easily inspect the distribution by opening the HTML report <code>overlaps.html</code>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Here is the histogram of the insert sizes determined by fastp (<a href="#fig-denovoassembly-insertsize">Figure&nbsp;<span>12.5</span></a>). By default, fastp will only keep reads that are longer than 30 bp and requires an overlap between the read mates of 30 bp. The maximum read length is 150 bp, therefore, the histogram only spreads from 31 to 271 bp in total.</p>
<div id="fig-denovoassembly-insertsize" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/2612_insertsize_hist.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.5: Histogram of the insert sizes determined by fastp</figcaption>
</figure>
</div>
<p>The sequencing data for the sample <strong>2612</strong> were generated across eight different sequencing runs, which differed in their nominal length. Some sequencing runs were 2x 100 bp, while others were 2x 150 bp. This is the reason why we observe two peaks just short of 100 and 150 bp. The difference to the nominal length is caused by the quality trimming of the data.</p>
<p>Overall, we have almost no short DNA molecules (&lt; 50 bp) but most DNA molecules are longer than 80 bp. Additionally, there were &gt; 200,000 read pairs that could not be overlapped. Therefore, we can conclude that the sample <strong>2612</strong> is moderately degraded ancient DNA sample and has many long DNA molecules.</p>
</div>
</div>
</div>
</section>
<section id="de-novo-assembly" class="level3" data-number="12.2.3">
<h3 data-number="12.2.3" class="anchored" data-anchor-id="de-novo-assembly"><span class="header-section-number">12.2.3</span> <em>De novo</em> assembly</h3>
<p>Now, we will actual perform the <em>de novo</em> assembly on the sequencing data. For this, we will use the program MEGAHIT <span class="citation" data-cites="LiMegahit2015">(<a href="introduction-to-ngs-sequencing.html#ref-LiMegahit2015" role="doc-biblioref">Li et al. 2015</a>)</span>, a <em>de Bruijn</em>-graph assembler.</p>
<div id="fig-denovoassembly-kmers" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/kmer_counting.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.6: Overview of inferring <em>k</em>-mers from a DNA sequence. Credit: <a href="https://medium.com/swlh/bioinformatics-1-k-mer-counting-8c1283a07e29">https://medium.com/swlh/bioinformatics-1-k-mer-counting-8c1283a07e29</a></figcaption>
</figure>
</div>
<p><em>De Bruijn</em> graph assemblers are together with overlap assemblers the predominant group of assemblers. They use <em>k</em>-mers (see <a href="#fig-denovoassembly-kmers">Figure&nbsp;<span>12.6</span></a> for an example of <em>4</em>-mers) extracted from the short-read sequencing data to build the graph. For this, they connect each <em>k</em>-mer to adjacent <em>k</em>-mer using a directional edge. By walking along the edges and visiting each <em>k</em>-mer or node once, longer continuous sequences are reconstructed. This is a very rough explanation and I would advise you to watch this excerpt of a <a href="https://www.youtube.com/watch?v=OY9Q_rUCGDw">lecture</a> by Rob Edwards from San Diego State University and a <a href="https://youtu.be/TNYZZKrjCSk?t=112">Coursera lecture</a> by Ben Langmead from Johns Hopkins University, if you would like to learn more about it.</p>
<p>All <em>de Bruijn</em> graph assemblers work in a similar way so the question is why do we use MEGAHIT and not other programs, such as metaSPAdes?</p>
<div class="callout callout-style-default callout-note callout-titled" title="Pros and cons of MEGAHIT">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pros and cons of MEGAHIT
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Pros</strong>:</p>
<ul>
<li>low-memory footprint: can be run on computational infrastructure that does not have large memory resources</li>
<li>can assembly both single-end and paired-end sequencing data</li>
<li>the assembly algorithm can cope with the presence of high amounts of ancient DNA damage</li>
</ul>
<p><strong>Cons</strong>:</p>
<ul>
<li>lower assembly quality on modern comparative data, particularly a higher rate of mis-assemblies (CAMI II challenge)</li>
</ul>
</div>
</div>
<p>In the <a href="https://christinawarinner.com/about-us/">Warinner group</a>, we realised after some tests that MEGAHIT has a clear advantage when ancient DNA damage is present at higher quantities. While it produced a higher number of mis-assemblies compared to metaSPAdes when being evaluated on simulated modern metagenomic data <span class="citation" data-cites="Meyer2022">(Critical Assessment of Metagenome Interpretation II challenge, <a href="introduction-to-ngs-sequencing.html#ref-Meyer2022" role="doc-biblioref">Meyer et al. 2022</a>)</span>, it produces more long contigs when ancient DNA damage is present.</p>
<p>To <em>de novo</em> assemble the short-read sequencing data of the sample <strong>2612</strong> using MEGAHIT, we can run the command</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">megahit</span> <span class="at">-1</span> 2612_R1.fastq.gz <span class="dt">\</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">-2</span> 2612_R2.fastq.gz <span class="dt">\</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">-t</span> 14 <span class="at">--min-contig-len</span> 500 <span class="dt">\</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">--out-dir</span> megahit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This will use the paired-end sequencing data as input and return all contigs that are at least 500 bp long.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>While MEGAHIT is able to use merged sequencing data, it is advised to use the unmerged paired-end data as input. In tests using simulated data I have observed that MEGAHIT performed slightly better when using the unmerged data and it likely has something to do with its internal algorithm to infer insert sizes from the paired-end data.</p>
</div>
</div>
<p>While we are waiting for MEGAHIT to finish, here is a question:</p>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Which <em>k</em>-mer lengths did MEGAHIT select for the <em>de novo</em> assembly?</strong></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Based on the maximum read length, MEGAHIT decided to use the <em>k</em>-mer lengths of 21, 29, 39, 59, 79, 99, 119, and 141.</p>
</div>
</div>
</div>
<p>Now, as MEGAHIT has finished, we want to evaluate the quality of the assembly results. MEGAHIT has written the contiguous sequences (contigs) into a single FastA file stored in the folder <code>megahit</code>. We will process this FastA file with a small script, calN50, which will count the number of contigs and give us an overview of their length distribution.</p>
<p>To download the script and run it, we can execute the following commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://raw.githubusercontent.com/lh3/calN50/master/calN50.js</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">k8</span> ./calN50.js megahit/final.contigs.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How many contigs were assembled? What is the sum of the lengths of all contigs? What is the maximum, the median, and the minimum contig length?</strong></p>
<p>Hint: The maximum contig length is indicated by the label “N0”, the median by the label “N50”, and the minimum by the label “N100”?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>MEGAHIT assembled 3,606 contigs and their lengths sum up to 11.4 Mb. The maximum contig length was 448 kb, the median length was 15.6 kb, and the minimum length was 500 bp.</p>
</div>
</div>
</div>
<p>There is a final caveat when assembling ancient metagenomic data with MEGAHIT: while it is able to assemble sequencing data with a high percentage of C-to-T substitutions, it tends to introduce these changes into the contig sequences, too.</p>
<div id="fig-denovoassembly-debruijnbubble" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/Leggett2013_Fig1a.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.7: <em>De Bruijn</em> graph with a bubble caused by a second allele. Adapted from <span class="citation" data-cites="Leggett2013">(<a href="introduction-to-ngs-sequencing.html#ref-Leggett2013" role="doc-biblioref">Leggett et al. 2013, fig. 1a</a>)</span></figcaption>
</figure>
</div>
<p>These C-to-T substitutions are similar to biological single-nucleotide polymorphisms in the sequencing data. Both lead the introduction of bubbles in the <em>de Bruijn</em> graph when two alleles are present in the k-mer sequences (<a href="#fig-denovoassembly-debruijnbubble">Figure&nbsp;<span>12.7</span></a>) and the assembler decides during its pruning steps which allele to keep in the contig sequence.</p>
<p>While it does not really matter which allele is kept for biological polymorphisms, it does matter for technical artefacts that are introduced by the presence of ancient DNA damage. In our group we realised that the gene sequences that were annotated on the contigs of MEGAHIT tended to have a higher number of nonsense mutations compared to the known variants of the genes. After manual inspection, we observed that many of these mutations appeared because MEGAHIT chose the damage-derived T allele over the C allele or the damage-derived A allele over a G allele <span class="citation" data-cites="Klapper2023">(see <a href="introduction-to-ngs-sequencing.html#ref-Klapper2023" role="doc-biblioref">Klapper et al. 2023, fig. S1</a>)</span>.</p>
<p>To overcome this issue, my colleagues Maxime Borry, James Fellows Yates and I developed a strategy to replace these damage-derived alleles in the contig sequences. This approach consists of aligning the short-read data against the contigs, performing genotyping along them, and finally replacing all alleles for which we have strong support for an allele that is different from the one selected by MEGAHIT.</p>
<p>We standardised this approach and added it to the Nextflow pipeline nf-core/mag <span class="citation" data-cites="Krakau2022">(<a href="introduction-to-ngs-sequencing.html#ref-Krakau2022" role="doc-biblioref">Krakau et al. 2022</a>)</span>. It can simply be activated by providing the parameter <code>--ancient_dna</code>.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>While MEGAHIT is able to assemble ancient metagenomic sequencing data with high amounts of ancient DNA damage, it tends to introduce damage-derived T and A alleles into the contig sequences instead of the true C and G alleles. This can lead to a higher number of nonsense mutations in coding sequences. We strongly advise you to correct such mutations, e.g.&nbsp;by using the ancient DNA workflow of the Nextflow pipeline <a href="https://nf-co.re/mag">nf-core/mag</a>.</p>
</div>
</div>
</section>
<section id="aligning-the-short-read-data-against-the-contigs" class="level3" data-number="12.2.4">
<h3 data-number="12.2.4" class="anchored" data-anchor-id="aligning-the-short-read-data-against-the-contigs"><span class="header-section-number">12.2.4</span> Aligning the short-read data against the contigs</h3>
<p>After the assembly, the next detrimental step that is required for many subsequent analyses is the alignment of the short-read sequencing data back to assembled contigs.</p>
<p>Analyses that require these alignment information are for example:</p>
<ul>
<li>the correction of the contig sequences to remove damage-derived alleles</li>
<li>the non-reference binning of contigs into MAGs for inferring the coverage along the contigs</li>
<li>the quantification of the presence of ancient DNA damage</li>
</ul>
<p>Aligning the short-read data to the contigs requires multiple steps:</p>
<ol type="1">
<li>Build an index from the contigs for the alignment program BowTie2</li>
<li>Align the short-read data against the contigs using this index with BowTie2</li>
<li>Calculate the mismatches and deletions of the short-read data against the contig sequences</li>
<li>Sort the aligned reads by the contig name and the coordinate they were aligned to</li>
<li>Index the resulting alignment file for faster access</li>
</ol>
<p>To execute the alignment step we can run the following commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> alignment</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2-build</span> <span class="at">-f</span> megahit/final.contigs.fa alignment/2612</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2</span> <span class="at">-p</span> 14 –-very-sensitive <span class="at">-N</span> 1 <span class="at">-x</span> alignment/2612 <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-1</span> 2612_R1.fastq.gz <span class="at">-2</span> 2612_R2.fastq.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> view <span class="at">-Sb</span> <span class="at">-</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> calmd <span class="at">-u</span> /dev/stdin megahit/final.contigs.fa <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> sort <span class="at">-o</span> alignment/2612.sorted.calmd.bam <span class="at">-</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index alignment/2612.sorted.calmd.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>However, these steps are rather time-consuming, even when we just have so little sequencing data as we do for our course example. The alignment is rather slow because we allow a single mismatch in the seeds that are used by the aligner BowTie2 to quickly determine the position of a read along the contig sequences (parameter <code>-N 1</code>). This is necessary because otherwise we might not be able to align reads with ancient DNA damage present on them. Secondly, the larger the resulting alignment file is the longer it takes to sort it by coordinate.</p>
<p>To save us some time and continue with the more interesting analyses, I prepared the resulting files for us. For this, I also corrected damage-derived alleles in the contig sequences. You can access these files on the cluster by running the following commands:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> alignment</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /vol/volume/denovo-assembly/2612.sorted.calmd.bam alignment/</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /vol/volume/denovo-assembly/2612.sorted.calmd.bam.bai alignment/</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> /vol/volume/denovo-assembly/2612.fa alignment/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Self-guided: data preparation">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Self-guided: data preparation
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>For everyone, who runs this practical on their own infrastructure, you can download the files:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> alignment</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> alignment/2612.sorted.calmd.bam <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    https://share.eva.mpg.de/index.php/s/bDKgFLj9GpRFdPg/download/2612.sorted.calmd.bam</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> alignment/2612.sorted.calmd.bam.bai <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    https://share.eva.mpg.de/index.php/s/HWqg6fJj6ZEEBAL/download/2612.sorted.calmd.bam.bai</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> alignment/2612.fa <span class="dt">\</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    https://share.eva.mpg.de/index.php/s/z6ZAai42RPribX5/download/final.contigs.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="reconstructing-metagenome-assembled-genomes" class="level3" data-number="12.2.5">
<h3 data-number="12.2.5" class="anchored" data-anchor-id="reconstructing-metagenome-assembled-genomes"><span class="header-section-number">12.2.5</span> Reconstructing metagenome-assembled genomes</h3>
<p>There are typically two major approaches on how to study biological diversity of samples using the results obtained from the <em>de novo</em> assembly. The first one is to reconstruct metagenome-assembled genomes (MAGs) and to study the species diversity.</p>
<p>“Metagenome-assembled genome” is a convoluted term that means that we reconstructed a genome from metagenomic data via <em>de novo</em> assembly. While these reconstructed genomes, particularly the high-quality ones, are most likely a good representation of the genomes of the organisms present in the sample, the scientific community refrains from calling them a species or a strain. The reason is that for calling a genome a species or a strain additional analyses would be necessary, of which many would include the cultivation of the organism. For many samples, this is not feasible and therefore the community stuck to the term MAG instead.</p>
<p>The most commonly applied method to obtain MAGs is the so-called “non-reference binning”. Non-reference binning means that we do not try to identify contigs by aligning them against known reference genomes, but only use the characteristics of the contigs themselves to cluster them (<a href="#fig-denovoassembly-binning">Figure&nbsp;<span>12.8</span></a>).</p>
<div id="fig-denovoassembly-binning" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/mag_binning_scheme.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.8: Scheme of binning <em>de novo</em> assembled contigs into metagenome-assembled genomes. During the binning contigs are grouped into clusters based on their characteristics, such as tetra-nucleotide frequency and the coverage along the contigs. Clusters of contigs that fulfil a minimum of quality criteria are then considered as metagenome-assembled genomes. However, depending on the sample, a number of contigs will remain unbinned.</figcaption>
</figure>
</div>
<p>The two most commonly used characteristics are:</p>
<ul>
<li>the tetra-nucleotide frequency: the frequency of all 4-mers (e.g.&nbsp;AAAA, AAAC, AAAG etc.) in the contig sequence</li>
<li>the coverage along the contig</li>
</ul>
<p>The idea here is that two contigs that are derived from the same bacterial genome will likely have a similar nucleotide composition and coverage.</p>
<p>This approach works very well when the contigs are longer and they strongly differ in the nucleotide composition and the coverage from each other. However, if this is not the case, e.g.&nbsp;there is more than one strain of the same species in the sample, these approaches will likely not be able to assign some contigs to clusters: these contigs remain unbinned. In case there is a high number of unbinned contigs, one can also employ the more sensitive reference-based binning strategies but we will not cover this in this practical course.</p>
<p>There are a number of different binning tools out there that can be used for this task. Since this number is constantly growing, there have been attempts to standardise the test data sets that these tools are run on so that their performances can be easily compared. The most well-known attempt is the Critical Assessment of Metagenome Interpretation <span class="citation" data-cites="Meyer2022">(Critical Assessment of Metagenome Interpretation II challenge, <a href="introduction-to-ngs-sequencing.html#ref-Meyer2022" role="doc-biblioref">Meyer et al. 2022</a>)</span>, which released the latest comparison of commonly used tools for assembly, binning, and bin refinement in 2022. I recommend that you first check the performance of a new tool against the CAMI datasets before testing it to see whether it is worth using it.</p>
<p>Three commonly used binners are:</p>
<ul>
<li>metaBAT2 <span class="citation" data-cites="Kang2019">(<a href="introduction-to-ngs-sequencing.html#ref-Kang2019" role="doc-biblioref">Kang et al. 2019</a>, more than 1,300 citations)</span></li>
<li>MaxBin2 <span class="citation" data-cites="WuMaxbin2016">(<a href="introduction-to-ngs-sequencing.html#ref-WuMaxbin2016" role="doc-biblioref">Wu, Simmons, and Singer 2016</a>, more than 1,200 citations)</span></li>
<li>CONCOCT <span class="citation" data-cites="Alneberg2014">(<a href="introduction-to-ngs-sequencing.html#ref-Alneberg2014" role="doc-biblioref">Alneberg et al. 2014</a>, more than 1,000 citation)</span></li>
</ul>
<p>Each of these three binners employs a slightly different strategy. While metaBAT2 simply uses the two previously mentioned metrics, the tetra-nucleotide frequency and the coverage along the contigs, MaxBin2 additionally uses single-copy marker genes to infer the taxonomic origin of contigs. In contrast, CONCOCT also just uses the two aforementioned metrics but first performs a Principal Component Analysis (PCA) on these metrics and uses the PCA results for the clustering.</p>
<p>The easiest way to run all these three programs is the program metaWRAP <span class="citation" data-cites="Uritskiy2018">(<a href="introduction-to-ngs-sequencing.html#ref-Uritskiy2018" role="doc-biblioref">Uritskiy, DiRuggiero, and Taylor 2018</a>)</span>. metaWRAP is in fact a pipeline that allows you to assemble your contigs, bin them, and subsequently refine the resulting MAGs. However, the pipeline is not very well written and does not contain any strategies to deal with ancient metagenomic sequencing data. Therefore, I prefer to use different pipelines, such as nf-core/mag for the assembly, and only use metaWRAP for binning and bin refinement.</p>
<p>To skip the first steps of metaWRAP and start straight with the binning, we need to create the folder structure and files that metaWRAP expects:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> metawrap/INITIAL_BINNING/2612/work_files</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ln</span> <span class="at">-s</span> <span class="va">$PWD</span>/alignment/2612.sorted.calmd.bam <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    metawrap/INITIAL_BINNING/2612/work_files/2612.bam</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> metawrap/faux_reads</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"@"</span> <span class="op">&gt;</span> metawrap/faux_reads/2612_1.fastq</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"@"</span> <span class="op">&gt;</span> metawrap/faux_reads/2612_2.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can start to run the binning. In this practical course, we will focus on metaBAT2 and MaxBin2. To bin the contigs with these binners, we execute:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate metawrap-env</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">metawrap</span> binning <span class="at">-o</span> metawrap/INITIAL_BINNING/2612 <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> 14 <span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> alignment/2612.fa <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--metabat2</span> <span class="at">--maxbin2</span> <span class="at">--universal</span> <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    metawrap/faux_reads/2612_1.fastq metawrap/faux_reads/2612_2.fastq</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>MetaWRAP is not a well-written Python software and has not been update for more than three years. It still relies on the deprecated Python v2.7. This is in conflict with many other tools and therefore it requires its own conda environment, <code>metawrap-env</code>. Do not forget to deactivate this environment afterwards again!</p>
</div>
</div>
<p>MetaWRAP will run metaBAT2 and MaxBin2 for us and store their respective output into sub-folders in the folder <code>metawrap/INITIAL_BINNING/2612</code>.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How many bins did metaBAT2 and MaxBin2 reconstruct, respectively? Is there a difference in the genome sizes of these reconstructed bins?</strong></p>
<p>Hint: You can use the previously introduced script <code>k8 ./calN50.js</code> to analyse the genome size of the individual bins.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>metaBAT2 reconstructed seven bins, while MaxBin2 reconstructed only five bins.</p>
<p>When comparing the genome sizes of these bins, we can see that despite having reconstructed fewer bins, MaxBin2’s bins have on average larger genome size and all of them are at least 1.5 Mb. In contrast, five out of seven metaBAT2 bins are shorter than 1.5 Mb.</p>
</div>
</div>
</div>
<p>While we could have run these two binning softwares manually ourselves, there is another reason why we should use metaWRAP: it has a powerful bin refinement algorithm.</p>
<p>As we just observed, binning tools come to different results when performing non-reference binning on the same contigs. So how do we know which binning tool delivered the better or even correct results?</p>
<p>A standard approach is to identify single-copy marker genes that are specific for certain taxonomic lineages, e.g.&nbsp;to all members of the family <em>Prevotellaceae</em> or to all members of the kingdom archaea. If we find lineage-specific marker genes from more than one lineage in our bin, something likely went wrong. While in certain cases horizontal gene transfer could explain such a finding, it is much more common that a binning tool clustered two contigs from two different taxons.</p>
<p>During its bin refinement, metaWRAP first combines the results of all binning tools in all combinations. So it would merge the results of metaBAT2 and MaxBin2, metaBAT2 and CONCOCT, MaxBin2 and CONCOCT, and all three together. Afterwards, it evaluates the presence of lineage-specific marker genes on the contigs of the bins for every combination and the individual binning tools themselves. In the case that it would find marker genes of more than one lineage in a bin, it would split the bin into two. After having evaluated everything, metaWRAP selects the refined bins that have the highest quality score across the whole set of bins.</p>
<div id="fig-denovoassembly-metawrap" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/MetaWrap_Fig4.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.9: Performance of metaWRAP’s bin refinement algorithm compared to other tools. Adapted from <span class="citation" data-cites="Uritskiy2018">Uritskiy, DiRuggiero, and Taylor (<a href="introduction-to-ngs-sequencing.html#ref-Uritskiy2018" role="doc-biblioref">2018</a>)</span>, Fig. 4</figcaption>
</figure>
</div>
<p>Using this approach, the authors of metaWRAP could show that they can outperform the individual binning tools and other bin refinement algorithms both regarding the <strong>completeness</strong> and <strong>contamination</strong> that was estimated for the MAGs (<a href="#fig-denovoassembly-metawrap">Figure&nbsp;<span>12.9</span></a>).</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>Running metaWRAP’s bin refinement module requires about 72 GB of memory because it has to load a larger reference database containing the lineage-specific marker genes of checkM.</p>
<p>If your computer infrastructure cannot provide so much memory, I have prepared the results of metaWRAP’s bin refinement algorithm, <code>metawrap_50_10_bins.stats</code>, which can be found in the folder <code>/vol/volume/denovo-assembly</code>.</p>
</div>
</div>
<p>To apply metaWRAP’s bin refinement to the bins that we obtained from metaBAT2 and MaxBin2, we first need to install the software checkM <span class="citation" data-cites="Parks2015">(<a href="introduction-to-ngs-sequencing.html#ref-Parks2015" role="doc-biblioref">Parks et al. 2015</a>)</span> that will provide the lineage-specific marker gene catalogue:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> checkM</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> checkM/checkm_data_2015_01_16.tar.gz <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvf checkM/checkm_data_2015_01_16.tar.gz <span class="at">-C</span> checkM</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> checkM <span class="kw">|</span> <span class="ex">checkm</span> data setRoot checkM</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Afterwards, we can execute metaWRAP’s bin refinement module:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> metawrap/BIN_REFINEMENT/2612</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">metawrap</span> bin_refinement <span class="at">-o</span> metawrap/BIN_REFINEMENT/2612 <span class="dt">\</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-t</span> 14 <span class="dt">\</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-c</span> 50 <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">-x</span> 10 <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">-A</span> metawrap/INITIAL_BINNING/2612/maxbin2_bins <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-B</span> metawrap/INITIAL_BINNING/2612/metabat2_bins</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The latter step will produce a summary file, <code>metawrap_50_10_bins.stats</code>, that lists all retained bins and some key characteristics, such as the genome size, the completeness estimate, and the contamination estimate. The latter two can be used to assign a quality score according to the Minimum Information for MAG (MIMAG; see info box).</p>
<div class="callout callout-style-default callout-note callout-titled" title="The Minimum Information for MAG (MIMAG)">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
The Minimum Information for MAG (MIMAG)
</div>
</div>
<div class="callout-body-container callout-body">
<p>The two most common metrics to evaluate the quality of MAGs are:</p>
<ul>
<li>the <strong>completeness</strong>: how many of the expected lineage-specific single-copy marker genes were present in the MAG?</li>
<li>the <strong>contamination</strong>: how many of the expected lineage-specific single-copy marker genes were present more than once in the MAG?</li>
</ul>
<p>These metric is usually calculated using the marker-gene catalogue of checkM <span class="citation" data-cites="Parks2015">(<a href="introduction-to-ngs-sequencing.html#ref-Parks2015" role="doc-biblioref">Parks et al. 2015</a>)</span>, also if there are other estimates from other tools such as BUSCO <span class="citation" data-cites="Manni2021">(<a href="introduction-to-ngs-sequencing.html#ref-Manni2021" role="doc-biblioref">Manni et al. 2021</a>)</span>, GUNC <span class="citation" data-cites="Orakov2021">(<a href="introduction-to-ngs-sequencing.html#ref-Orakov2021" role="doc-biblioref">Orakov et al. 2021</a>)</span> or checkM2 <span class="citation" data-cites="Chklovski2022">(<a href="introduction-to-ngs-sequencing.html#ref-Chklovski2022" role="doc-biblioref">Chklovski et al. 2022</a>)</span>.</p>
<p>Depending on the estimates on completeness and contamination plus the presence of RNA genes, MAGs are assigned to the quality category following the Minimum Information for MAG criteria <span class="citation" data-cites="Bowers2017">(<a href="introduction-to-ngs-sequencing.html#ref-Bowers2017" role="doc-biblioref">Bowers et al. 2017</a>)</span> You can find the overview <a href="https://www.nature.com/articles/nbt.3893/tables/1">here</a>.</p>
</div>
</div>
<p>As these two steps will run rather long and need a large amount of memory and disk space, I have provided the results of metaWRAP’s bin refinement. You can find the file here: <code>/vol/volume/denovo-assembly/metawrap_50_10_bins.stats</code>. Be aware that these results are based on the bin refinement of the results of three binning tools and include CONCOCT.</p>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>How many bins were retained after the refinement with metaWRAP? How many high-quality and medium-quality MAGs did the refinement yield following the MIMAG criteria?</strong></p>
<p>Hint: You can more easily visualise tables on the terminal using the Python program <code>visidata</code>. After installing it with <code>pip install visidata</code>, you can open a table using <code>vd -f tsv /vol/volume/denovo-assembly/metawrap_50_10_bins.stats</code>. Next to separating the columns nicely, it allows you to perform a lot of operations like sorting conveniently. Check the cheat sheet <a href="https://jsvine.github.io/visidata-cheat-sheet/en/">here</a>.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In total, metaWRAP retained five bins, similarly to MaxBin2. Of these five bins, the bins <code>bin.3</code> and <code>bin.4</code> had completeness and contamination estimates that would qualify them for being high-quality MAGs. However, we would need to check the presence of rRNA and tRNA genes. The other three bins are medium-quality bins because their completeness estimate was &lt; 90%.</p>
</div>
</div>
</div>
</section>
<section id="taxonomic-assignment-of-contigs" class="level3" data-number="12.2.6">
<h3 data-number="12.2.6" class="anchored" data-anchor-id="taxonomic-assignment-of-contigs"><span class="header-section-number">12.2.6</span> Taxonomic assignment of contigs</h3>
<p>What should we do when we simply want to know to which taxon a certain contig most likely belongs to?</p>
<p>Reconstructing metagenome-assembled genomes requires multiple steps and might not even provide the answer in the case that the contig of interest is not binned into a MAG. Instead, it is sufficient to perform a sequence search against a reference database.</p>
<p>There are plenty of tools available for this task, such as:</p>
<ul>
<li>BLAST/DIAMOND</li>
<li>Kraken2</li>
<li>Centrifuge</li>
<li>MMSeqs2</li>
</ul>
<p>For each tool, we can either use pre-computed reference databases or compute our own one. The two taxonomic classification systems that are most commonly used are:</p>
<ul>
<li>NCBI Taxonomy</li>
<li>GTDB</li>
</ul>
<p>As for any task that involves the alignment of sequences against a reference database, the chosen reference database should fit the sequences you are searching for. If your reference database does not capture the diversity of your samples, you will not be able to assign a subset of the contigs. There is also a trade-off between a large reference database that contains all sequences and its memory requirement. <span class="citation" data-cites="Wright2023">Wright, Comeau, and Langille (<a href="introduction-to-ngs-sequencing.html#ref-Wright2023" role="doc-biblioref">2023</a>)</span> elaborated on this quite extensively when comparing Kraken2 against MetaPhlAn.</p>
<p>While all of these tools can do the job, I typically prefer to use the program MMSeqs2 <span class="citation" data-cites="Steinegger2017">(<a href="introduction-to-ngs-sequencing.html#ref-Steinegger2017" role="doc-biblioref">Steinegger and Söding 2017</a>)</span> because it comes along with a very fast algorithm based on aminoacid sequence alignment and implements a lowest common ancestor (LCA) algorithm (<a href="#fig-denovoassembly-mmseqs2">Figure&nbsp;<span>12.10</span></a>). Recently, they implemented a <em>taxonomy</em> workflow <span class="citation" data-cites="Mirdita2021">(<a href="introduction-to-ngs-sequencing.html#ref-Mirdita2021" role="doc-biblioref">Mirdita et al. 2021</a>)</span> that allows to efficiently assign contigs to taxons. Luckily, it comes with multiple pre-computed reference databases, such as the GTDB v207 reference database <span class="citation" data-cites="Parks2020">(<a href="introduction-to-ngs-sequencing.html#ref-Parks2020" role="doc-biblioref">Parks et al. 2020</a>)</span>, and therefore it is even more accessible for users.</p>
<div id="fig-denovoassembly-mmseqs2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/MMSeqs2_classify_Fig1.jpeg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.10: Scheme of the <em>taxonomy</em> workflow implemented into MMSeqs2. Adapted from <span class="citation" data-cites="Mirdita2021">Mirdita et al. (<a href="introduction-to-ngs-sequencing.html#ref-Mirdita2021" role="doc-biblioref">2021</a>)</span>, Fig. 1.</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>The latest version of the pre-computed GTDB reference database (r207) requires about 105 GB of harddisk space and 700 GB of memory for running.</p>
<p>As our computer infrastructure does not provide so much memory for every user, I pre-computed the results. You can find the results <code>2612.mmseqs2_gtdb.tsv</code> in the folder <code>/vol/volume/denovo-assembly</code>.</p>
<p>An alternative for users with a less powerful infrastructure is the program <a href="https://github.com/fbreitwieser/krakenuniq">KrakenUniq</a>.</p>
</div>
</div>
<p>Before running MMSeqs2’s <em>taxonomy</em> workflow against the GTDB reference database, we need to install it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> refdbs/mmseqs2/gtdb</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> databases GTDB <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    refdbs/mmseqs2/gtdb /tmp <span class="at">--threads</span> 14</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Subsequently, we can align all the contigs of the sample 2612 against the GTDB r207 with MMSeqs2:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> mmseqs2</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> createdb alignment/2612.fa mmseqs2/2612.contigs</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> taxonomy  mmseqs2/2612.contigs <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    refdbs/mmseqs2/gtdb/mmseqs2_gtdb <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>    mmseqs2/2612.mmseqs2_gtdb <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>    /tmp <span class="dt">\</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">-a</span> <span class="at">--tax-lineage</span> 1 <span class="dt">\</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--lca-ranks</span> kingdom,phylum,class,order,family,genus,species <span class="dt">\</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--orf-filter</span> 1 <span class="dt">\</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">--remove-tmp-files</span> 1 <span class="dt">\</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">--threads</span> 14</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> createtsv mmseqs2/2612.contigs <span class="dt">\</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    mmseqs2/2612.mmseqs2_gtdb <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>What is the proportion of contigs that could be assigned to the different taxonomic ranks, such as species or genus? What are the dominant taxa?</strong></p>
<p>Hint: You can access this information easily by opening the file using visidata: <code>vd 2612.mmseqs2_gtdb.tsv</code></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>From the 3,606 contigs, MMSeqs2’s <em>taxonomy</em> workflow assigned 3,523 contigs to any taxonomy. For the rest, there was not enough information and they were discarded.</p>
<p>From the 3,523 assigned contigs, 2,013 were assigned to the rank “species”, while 1,137 could only be assigned to the rank “genus”.</p>
<p>The most contigs were assigned the archael species <em>Halococcus morrhuae</em> (n=386), followed by the bacterial species <em>Olsenella E sp003150175</em> (n=298) and <em>Collinsella sp900768795</em> (n=186).</p>
</div>
</div>
</div>
</section>
<section id="taxonomic-assignment-of-mags" class="level3" data-number="12.2.7">
<h3 data-number="12.2.7" class="anchored" data-anchor-id="taxonomic-assignment-of-mags"><span class="header-section-number">12.2.7</span> Taxonomic assignment of MAGs</h3>
<p>MMSeqs2’s <em>taxonomy</em> workflow is very useful to classify all contigs taxonomically. However, how would we determine which species we reconstructed by binning our contigs?</p>
<p>The simplest approach would be that we could summarise MMSeqs2’s taxonomic assignments of all contigs of a MAG and then determine which lineage is the most frequent one. Although this would work in general, there is another approach that is more sophisticated: GTDB toolkit <span class="citation" data-cites="Chaumeil2020">(GTDBTK, <a href="introduction-to-ngs-sequencing.html#ref-Chaumeil2020" role="doc-biblioref">Chaumeil et al. 2020</a>)</span>.</p>
<p>GTDBTK performs three steps to assign a MAG to a taxonomic lineage:</p>
<ol type="1">
<li><strong>Lineage identification</strong> based on single-copy marker genes using Hidden Markov Models (HMMs)</li>
<li><strong>Multi-sequence alignment</strong> of the identified marker genes</li>
<li>Placement of the MAG genome into a <strong>fixed reference tree</strong> at class level</li>
</ol>
<p>The last step is particularly clever. Based on the known diversity of a lineage present in the GTDB, it will construct a reference tree with all known taxa of this lineage. Afterwards, the tree structure is fixed and an algorithm attempts to create a new branch in the tree for placing the unknown MAG based on both the tree structure and the multi-sequence alignment.</p>
<p>In most cases, both the simple approach of taking the majority of the MMSeqs2’s <em>taxonomy</em> results and the GTDBTK approach lead to very similar results. However, GTDBTK performs better when determining whether a new MAG potentially belongs to a new genus or even a new family.</p>
<p>To infer which taxa our five reconstructed MAGs represent, we can run the GTDBTK.</p>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p>The latest version of the database used by GTDBTK (r207) requires about 70 GB of harddisk space and 80 GB of memory for running.</p>
<p>As our computer infrastructure does not provide so much memory for every user, I pre-computed the results. You can find the results <code>2612.gtdbtk_archaea.tsv</code> and <code>2612.gtdbtk_bacteria.tsv</code> in the folder <code>/vol/volume/denovo-assembly</code>.</p>
</div>
</div>
<p>First, we need to install the GTDB database:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> refdbs/gtdbtk</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> refdbs/gtdbtk/gtdbtk_v2_data.tar.gz <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    https://data.gtdb.ecogenomic.org/releases/latest/auxillary_files/gtdbtk_v2_data.tar.gz</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvf refdbs/gtdbtk/gtdbtk_v2_data.tar.gz <span class="at">-C</span> refdbs/gtdbtk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Afterwards, we can run GTDBTK’s <em>classify</em> workflow:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> gtdbtk</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="va">GTDBTK_DATA_PATH</span><span class="op">=</span><span class="st">"</span><span class="va">$PWD</span><span class="st">/refdbs/gtdbtk/gtdbtk_r207_v2"</span> <span class="dt">\</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="ex">gtdbtk</span> classify_wf <span class="at">--cpu</span> 14 <span class="at">--extension</span> fa <span class="dt">\</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--genome_dir</span> metawrap/BIN_REFINEMENT/2612/metawrap_50_10_bins <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--out_dir</span> gtdbtk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Do the classifications obtained by GTDBTK match the classifications that were assigned to the contigs using MMSeqs2? Would you expect these taxa given the archaeological context of the samples?</strong></p>
<p>Hint: You can access the classification results of GTDBTK easily by opening the file using visidata: <code>vd 2612.gtdbtk_archaea.tsv</code> and <code>vd 2612.gtdbtk_bacteria.tsv</code></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The five MAGs reconstructed from the sample 2612 were assigned to the taxa:</p>
<ul>
<li><code>bin.1</code>: <em>Agathobacter rectalis</em></li>
<li><code>bin.2</code>: <em>Halococcus morrhuae</em></li>
<li><code>bin.3</code>: <em>Methanobrevibacter smithii</em></li>
<li><code>bin.4</code>: <em>Ruminococcus bromii</em></li>
<li><code>bin.5</code>: <em>Bifidobacterium longum</em></li>
</ul>
<p>All of these species were among the most frequent lineages that were identified by MMSeqs2’s <em>taxonomy</em> workflow highlighting the large overlap between the methods.</p>
<p>We would expect all five species to be present in our sample. All MAGs but <code>bin.2</code> were assigned to human gut microbiome commensal species that are typically found in healthy humans. The MAG <code>bin.2</code> was assigned to a halophilic archaeal species, which is typically found in salt mines.</p>
</div>
</div>
</div>
</section>
<section id="evaluating-the-amount-of-ancient-dna-damage" class="level3" data-number="12.2.8">
<h3 data-number="12.2.8" class="anchored" data-anchor-id="evaluating-the-amount-of-ancient-dna-damage"><span class="header-section-number">12.2.8</span> Evaluating the amount of ancient DNA damage</h3>
<p>One of the common questions that remain at this point of our analysis is whether the contigs that we assembled show evidence for the presence of ancient DNA damage. If yes, we could argue that these microbes are indeed ancient, particularly when their DNA fragment length distribution is rather short, too.</p>
<p>MAGs typically consist of either tens or hundreds of contigs. For this case, many of the commonly used tools for quantifying the amount of ancient DNA damage, such as damageprofiler <span class="citation" data-cites="Neukamm2021">(<a href="introduction-to-ngs-sequencing.html#ref-Neukamm2021" role="doc-biblioref">Neukamm, Peltzer, and Nieselt 2021</a>)</span> or mapDamage2 <span class="citation" data-cites="Jonsson2013">(<a href="introduction-to-ngs-sequencing.html#ref-Jonsson2013" role="doc-biblioref">Jónsson et al. 2013</a>)</span>, are not very well suited because they would require us to manually check each of their generated “smiley plots”, which visualise the amount of C-to-T substitutions at the end of reads.</p>
<div id="fig-denovoassembly-pydamage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="assets/images/chapters/denovo-assembly/Pydamage_NZ_JHCB02000011.1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;12.11: Overview of the model comparison performed by pyDamage. The green line represents the null model, i.e.&nbsp;the absence of ancient DNA damage, while the orange line represents the alternative model, i.e.&nbsp;the presence of ancient DNA damage.</figcaption>
</figure>
</div>
<p>Instead, we will use the program pyDamage <span class="citation" data-cites="Borry2021">(<a href="introduction-to-ngs-sequencing.html#ref-Borry2021" role="doc-biblioref">Borry et al. 2021</a>)</span> that was written with the particular use-case of metagenome-assembled contigs in mind. Although pyDamage can visualise the amount of C-to-T substitutions at the 5’ end of reads, it goes a step further and fits two models upon the substitution frequency (<a href="#fig-denovoassembly-pydamage">Figure&nbsp;<span>12.11</span></a>). The null hypothesis is that the observed distribution of C-to-T substitutions at the 5’ end of reads reflects a flat line, i.e.&nbsp;a case when no ancient DNA damage is present. The alternative model assumes that the distribution resembles an exponential decay, i.e.&nbsp;a case when ancient DNA damage is present. By comparing the fit of these two models to the observed data for each contig, pyDamage can quickly identify contigs that are likely of ancient origin without requiring the user to inspect the plots visually.</p>
<p>We can run pyDamage directly on the alignment data in BAM format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pydamage</span> analyze <span class="at">-w</span> 30 <span class="at">-p</span> 14 alignment/2612.sorted.calmd.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Evaluate the pyDamage results with respect to the amount of C-to-T substitutions observed on the contigs! How many contigs does pyDamage consider to show evidence for ancient DNA damage? How much power (prediction accuracy) does it have for this decision? Which MAGs are strongly “ancient” or “modern”?</strong></p>
<p>Hint: You can access pyDamage’s results easily by opening the file using visidata: <code>vd pydamage_results/pydamage_results.tsv</code></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-30-contents" aria-controls="callout-30" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-30" class="callout-30-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>From the 3,606 contigs, pyDamage inferred a q-value, i.e.&nbsp;a p-value corrected for multiple testing, of &lt; 0.5 to 26 contigs. This is partially due to a high fraction of the contigs with no sufficient information to discriminate between the models (predicted accuracy of &lt; 0.5). However, the majority of the contigs with a prediction accuracy of &gt; 0.5 still had q-values of 0.05 and higher. This suggests that overall the sample did not show large evidence of ancient DNA damage.</p>
<p>This reflects also on the MAGs. Although four of the five MAGs were human gut microbiome taxa, they did not show strong evidence of ancient DNA damage. This suggests that the sample is too young and is well preserved.</p>
</div>
</div>
</div>
</section>
<section id="annotating-genomes-for-function" class="level3" data-number="12.2.9">
<h3 data-number="12.2.9" class="anchored" data-anchor-id="annotating-genomes-for-function"><span class="header-section-number">12.2.9</span> Annotating genomes for function</h3>
<p>The second approach on how to study biological diversity of samples using the assembly results is to compare the reconstructed genes and their functions with each other.</p>
<p>While it is very interesting to reconstruct the metagenome-assembled genomes from contigs and speculate about the evolution of a certain taxon, this does not always help when studying the ecology of a microbiome. Studying the ecology of a microbiome means to understand which taxa are present in an environment, what type of community do they form, what kind of natural products do they produce etc.?</p>
<p>With the lack of any data from culture isolates, it is rather difficult to discriminate from which species a reconstructed gene sequence is coming from, particularly when the contigs are short. Many microbial species exchange genes among each other via horizontal gene transfer which leads to multiple copies of a gene to be present in our metagenome and increases the level of difficulty further.</p>
<p>Because of this, many researchers tend to annotate all genes of a MAGs and compare the presence and absence of these genes across other genomes that were taxonomically assigned to the same taxon. This analysis, called pan-genome analysis, allows us to estimate the diversity of a microbial species with respect to their repertoire of protein-coding genes.</p>
<p>One of the most commonly used annotation tools for MAGs is Prokka <span class="citation" data-cites="Seemann2014">(<a href="introduction-to-ngs-sequencing.html#ref-Seemann2014" role="doc-biblioref">Seemann 2014</a>)</span>, although it has recently been challenged by Bakta <span class="citation" data-cites="Schwengers2021">(<a href="introduction-to-ngs-sequencing.html#ref-Schwengers2021" role="doc-biblioref">Schwengers et al. 2021</a>)</span>. The latter provides the same functionality as Prokka but incorporates more up-to-date reference databases for the annotation. Therefore, the scientific community is slowly shifting to Bakta.</p>
<p>Next to returning information on the protein-coding genes, Prokka also returns the annotation of RNA genes (tRNAs and rRNAs), which will help us to evaluate the quality of MAGs regarding the MIMAG criteria.</p>
<p>For this practical course, we will use Prokka and we will focus on annotating the MAG <code>bin.3.fa</code> that we reconstructed from the sample 2612.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">prokka</span> <span class="at">--outdir</span> prokka <span class="dt">\</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--prefix</span> 2612_003 <span class="dt">\</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--compliant</span> <span class="at">--metagenome</span> <span class="at">--cpus</span> 14 <span class="dt">\</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>    metawrap_50_10_bins/bin.3.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Question" data-appearence="simple">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Prokka has annotated our MAG. What type of files does Prokka return? How many genes/tRNAs/rRNAs were detected?</strong></p>
<p>Hint: Check the file <code>prokka/2612_003.txt</code> for the number of annotated elements.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-32-contents" aria-controls="callout-32" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-32" class="callout-32-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Prokka returns the following files:</p>
<ul>
<li><code>.faa</code>: the amino acid sequences of all identified coding sequences</li>
<li><code>.ffn</code>: the nucleotide sequences of all identified coding sequences</li>
<li><code>.fna</code>: all contigs in nucleotide sequence format renamed following Prokka’s naming scheme</li>
<li><code>.gbk</code>: all annotations and sequences in GenBank format</li>
<li><code>.gff</code>: all annotations and sequences in GFF format</li>
<li><code>.tsv</code>: the tabular overview of all annotations</li>
<li><code>.txt</code>: the short summary of the annotation results</li>
</ul>
<p>Prokka found 1,797 coding sequences, 32 tRNAs, but no rRNAs. Finding no rRNAs is a common issues when trying to assemble MAGs without long-read sequencing data and is not just characteristic for ancient DNA samples. However, this means that we cannot technically call this MAG a high-quality MAG due to the lack of the rRNA genes.</p>
</div>
</div>
</div>
</section>
<section id="summary" class="level3" data-number="12.2.10">
<h3 data-number="12.2.10" class="anchored" data-anchor-id="summary"><span class="header-section-number">12.2.10</span> Summary</h3>
<p>In this practical course you have gone through all the important steps that are necessary for <em>de novo</em> assembling ancient metagenomic sequencing data to obtain contiguous DNA sequences with little error. Furthermore, you have learned how to cluster these sequences into bins without using any references and how to refine them based on lineage-specific marker genes. For these refined bins, you have evaluated their quality regarding common standards set by the scientific community and assigned the MAGs to its most likely taxon. Finally, we learned how to infer the presence of ancient DNA damage and annotate them for RNA genes and protein-coding sequences.</p>
<div class="callout callout-style-default callout-caution callout-titled" title="Cautionary note - sequencing depth">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Cautionary note - sequencing depth
</div>
</div>
<div class="callout-body-container callout-body">
<p>Be aware of the sequencing depth when you assemble your sample. This sample used in this practical course was not obtained by randomly subsampling but I subsampled the sample so that we are able to reconstruct MAGs.</p>
<p>The original sample had almost 200 million reads, however, I subsampled it to less than 5 million reads. You usually need a lot of sequencing data for <em>de novo</em> assembly and definitely more data than for reference-alignment based profiling. However, it also heavily depends on the complexity of the sample. So the best advice is: <strong>just give it a try</strong>!</p>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Almeida2021" class="csl-entry" role="listitem">
Almeida, Alexandre, Stephen Nayfach, Miguel Boland, Francesco Strozzi, Martin Beracochea, Zhou Jason Shi, Katherine S. Pollard, et al. 2021. <span>“A Unified Catalog of 204,938 Reference Genomes from the Human Gut Microbiome.”</span> <em>Nature Biotechnology</em> 39 (1): 105–14. <a href="https://doi.org/10.1038/s41587-020-0603-3">https://doi.org/10.1038/s41587-020-0603-3</a>.
</div>
<div id="ref-Alneberg2014" class="csl-entry" role="listitem">
Alneberg, Johannes, Brynjar Smári Bjarnason, Ino de Bruijn, Melanie Schirmer, Joshua Quick, Umer Z. Ijaz, Leo Lahti, Nicholas J. Loman, Anders F. Andersson, and Christopher Quince. 2014. <span>“Binning Metagenomic Contigs by Coverage and Composition.”</span> <em>Nature Methods</em> 11 (11): 1144–46. <a href="https://doi.org/10.1038/nmeth.3103">https://doi.org/10.1038/nmeth.3103</a>.
</div>
<div id="ref-Borry2021" class="csl-entry" role="listitem">
Borry, Maxime, Alexander Hübner, Adam B. Rohrlach, and Christina Warinner. 2021. <span>“<span>PyDamage</span>: Automated Ancient Damage Identification and Estimation for Contigs in Ancient <span>DNA</span> de Novo Assembly.”</span> <em>PeerJ</em> 9: e11845. <a href="https://doi.org/10.7717/peerj.11845">https://doi.org/10.7717/peerj.11845</a>.
</div>
<div id="ref-Bowers2017" class="csl-entry" role="listitem">
Bowers, Robert M., Nikos C. Kyrpides, Ramunas Stepanauskas, Miranda Harmon-Smith, Devin Doud, T. B. K. Reddy, Frederik Schulz, et al. 2017. <span>“Minimum Information about a Single Amplified Genome (<span>MISAG</span>) and a Metagenome-Assembled Genome (<span>MIMAG</span>) of Bacteria and Archaea.”</span> <em>Nature Biotechnology</em> 35 (8): 725–31. <a href="https://doi.org/10.1038/nbt.3893">https://doi.org/10.1038/nbt.3893</a>.
</div>
<div id="ref-Chaumeil2020" class="csl-entry" role="listitem">
Chaumeil, Pierre-Alain, Aaron J Mussig, Philip Hugenholtz, and Donovan H Parks. 2020. <span>“<span>GTDB-Tk</span>: A Toolkit to Classify Genomes with the <span>Genome Taxonomy Database</span>.”</span> <em>Bioinformatics</em> 36 (6): 1925–27. <a href="https://doi.org/10.1093/bioinformatics/btz848">https://doi.org/10.1093/bioinformatics/btz848</a>.
</div>
<div id="ref-Chen2018" class="csl-entry" role="listitem">
Chen, Shifu, Yanqing Zhou, Yaru Chen, and Jia Gu. 2018. <span>“Fastp: An Ultra-Fast All-in-One <span>FASTQ</span> Preprocessor.”</span> <em>Bioinformatics</em> 34 (17): i884–90. <a href="https://doi.org/10.1093/bioinformatics/bty560">https://doi.org/10.1093/bioinformatics/bty560</a>.
</div>
<div id="ref-Chklovski2022" class="csl-entry" role="listitem">
Chklovski, Alex, Donovan H. Parks, Ben J. Woodcroft, and Gene W. Tyson. 2022. <span>“<span>CheckM2</span>: A Rapid, Scalable and Accurate Tool for Assessing Microbial Genome Quality Using Machine Learning.”</span> <span>bioRxiv</span>. <a href="https://doi.org/10.1101/2022.07.11.499243">https://doi.org/10.1101/2022.07.11.499243</a>.
</div>
<div id="ref-Jonsson2013" class="csl-entry" role="listitem">
Jónsson, Hákon, Aurélien Ginolhac, Mikkel Schubert, Philip L. F. Johnson, and Ludovic Orlando. 2013. <span>“<span class="nocase">mapDamage2</span>.0: Fast Approximate <span>Bayesian</span> Estimates of Ancient <span>DNA</span> Damage Parameters.”</span> <em>Bioinformatics</em> 29 (13): 1682–84. <a href="https://doi.org/10.1093/bioinformatics/btt193">https://doi.org/10.1093/bioinformatics/btt193</a>.
</div>
<div id="ref-Kang2019" class="csl-entry" role="listitem">
Kang, Dongwan D., Feng Li, Edward Kirton, Ashleigh Thomas, Rob Egan, Hong An, and Zhong Wang. 2019. <span>“<span>MetaBAT</span> 2: An Adaptive Binning Algorithm for Robust and Efficient Genome Reconstruction from Metagenome Assemblies.”</span> <em>PeerJ</em> 7 (July): e7359. <a href="https://doi.org/10.7717/peerj.7359">https://doi.org/10.7717/peerj.7359</a>.
</div>
<div id="ref-Klapper2023" class="csl-entry" role="listitem">
Klapper, Martin, Alexander Hübner, Anan Ibrahim, Ina Wasmuth, Maxime Borry, Veit G. Haensch, Shuaibing Zhang, et al. 2023. <span>“Natural Products from Reconstructed Bacterial Genomes of the <span>Middle</span> and <span>Upper Paleolithic</span>.”</span> <em>Science</em> 380 (6645): 619–24. <a href="https://doi.org/10.1126/science.adf5300">https://doi.org/10.1126/science.adf5300</a>.
</div>
<div id="ref-Krakau2022" class="csl-entry" role="listitem">
Krakau, Sabrina, Daniel Straub, Hadrien Gourlé, Gisela Gabernet, and Sven Nahnsen. 2022. <span>“Nf-Core/Mag: A Best-Practice Pipeline for Metagenome Hybrid Assembly and Binning.”</span> <em>NAR Genomics and Bioinformatics</em> 4 (1): lqac007. <a href="https://doi.org/10.1093/nargab/lqac007">https://doi.org/10.1093/nargab/lqac007</a>.
</div>
<div id="ref-Leggett2013" class="csl-entry" role="listitem">
Leggett, Richard M., Ricardo H. Ramirez-Gonzalez, Walter Verweij, Cintia G. Kawashima, Zamin Iqbal, Jonathan D. G. Jones, Mario Caccamo, and Daniel MacLean. 2013. <span>“Identifying and <span>Classifying Trait Linked Polymorphisms</span> in <span>Non-Reference Species</span> by <span>Walking Coloured</span> de <span>Bruijn Graphs</span>.”</span> <em>PLOS ONE</em> 8 (3): e60058. <a href="https://doi.org/10.1371/journal.pone.0060058">https://doi.org/10.1371/journal.pone.0060058</a>.
</div>
<div id="ref-LiMegahit2015" class="csl-entry" role="listitem">
Li, Dinghua, Chi-Man Liu, Ruibang Luo, Kunihiko Sadakane, and Tak-Wah Lam. 2015. <span>“<span>MEGAHIT</span>: An Ultra-Fast Single-Node Solution for Large and Complex Metagenomics Assembly via Succinct de <span>Bruijn</span> Graph.”</span> <em>Bioinformatics</em> 31 (10): 1674–76. <a href="https://doi.org/10.1093/bioinformatics/btv033">https://doi.org/10.1093/bioinformatics/btv033</a>.
</div>
<div id="ref-Maixner2021" class="csl-entry" role="listitem">
Maixner, Frank, Mohamed S. Sarhan, Kun D. Huang, Adrian Tett, Alexander Schoenafinger, Stefania Zingale, Aitor Blanco-Míguez, et al. 2021. <span>“Hallstatt Miners Consumed Blue Cheese and Beer During the <span>Iron Age</span> and Retained a Non-<span>Westernized</span> Gut Microbiome Until the <span>Baroque</span> Period.”</span> <em>Current Biology</em> 31 (23): 5149–5162.e6. <a href="https://doi.org/10.1016/j.cub.2021.09.031">https://doi.org/10.1016/j.cub.2021.09.031</a>.
</div>
<div id="ref-Manni2021" class="csl-entry" role="listitem">
Manni, Mosè, Matthew R Berkeley, Mathieu Seppey, Felipe A Simão, and Evgeny M Zdobnov. 2021. <span>“<span>BUSCO Update</span>: <span>Novel</span> and <span>Streamlined Workflows</span> Along with <span>Broader</span> and <span>Deeper Phylogenetic Coverage</span> for <span>Scoring</span> of <span>Eukaryotic</span>, <span>Prokaryotic</span>, and <span>Viral Genomes</span>.”</span> <em>Molecular Biology and Evolution</em> 38 (10): 4647–54. <a href="https://doi.org/10.1093/molbev/msab199">https://doi.org/10.1093/molbev/msab199</a>.
</div>
<div id="ref-Meyer2022" class="csl-entry" role="listitem">
Meyer, Fernando, Adrian Fritz, Zhi-Luo Deng, David Koslicki, Till Robin Lesker, Alexey Gurevich, Gary Robertson, et al. 2022. <span>“Critical <span>Assessment</span> of <span>Metagenome Interpretation</span>: The Second Round of Challenges.”</span> <em>Nature Methods</em> 19 (4): 429–40. <a href="https://doi.org/10.1038/s41592-022-01431-4">https://doi.org/10.1038/s41592-022-01431-4</a>.
</div>
<div id="ref-Mirdita2021" class="csl-entry" role="listitem">
Mirdita, M, M Steinegger, F Breitwieser, J Söding, and E Levy Karin. 2021. <span>“Fast and Sensitive Taxonomic Assignment to Metagenomic Contigs.”</span> <em>Bioinformatics</em> 37 (18): 3029–31. <a href="https://doi.org/10.1093/bioinformatics/btab184">https://doi.org/10.1093/bioinformatics/btab184</a>.
</div>
<div id="ref-Neukamm2021" class="csl-entry" role="listitem">
Neukamm, Judith, Alexander Peltzer, and Kay Nieselt. 2021. <span>“<span>DamageProfiler</span>: Fast Damage Pattern Calculation for Ancient <span>DNA</span>.”</span> <em>Bioinformatics</em> 37 (20): 3652–53. <a href="https://doi.org/10.1093/bioinformatics/btab190">https://doi.org/10.1093/bioinformatics/btab190</a>.
</div>
<div id="ref-Nurk2017" class="csl-entry" role="listitem">
Nurk, Sergey, Dmitry Meleshko, Anton Korobeynikov, and Pavel A. Pevzner. 2017. <span>“<span class="nocase">metaSPAdes</span>: A New Versatile Metagenomic Assembler.”</span> <em>Genome Research</em> 27 (5): 824–34. <a href="https://doi.org/10.1101/gr.213959.116">https://doi.org/10.1101/gr.213959.116</a>.
</div>
<div id="ref-Orakov2021" class="csl-entry" role="listitem">
Orakov, Askarbek, Anthony Fullam, Luis Pedro Coelho, Supriya Khedkar, Damian Szklarczyk, Daniel R. Mende, Thomas S. B. Schmidt, and Peer Bork. 2021. <span>“<span>GUNC</span>: Detection of Chimerism and Contamination in Prokaryotic Genomes.”</span> <em>Genome Biology</em> 22 (1): 178. <a href="https://doi.org/10.1186/s13059-021-02393-0">https://doi.org/10.1186/s13059-021-02393-0</a>.
</div>
<div id="ref-Orlando2013" class="csl-entry" role="listitem">
Orlando, Ludovic, Aurélien Ginolhac, Guojie Zhang, Duane Froese, Anders Albrechtsen, Mathias Stiller, Mikkel Schubert, et al. 2013. <span>“Recalibrating <span>Equus</span> Evolution Using the Genome Sequence of an Early <span>Middle Pleistocene</span> Horse.”</span> <em>Nature</em> 499 (7456): 74–78. <a href="https://doi.org/10.1038/nature12323">https://doi.org/10.1038/nature12323</a>.
</div>
<div id="ref-Parks2020" class="csl-entry" role="listitem">
Parks, Donovan H., Maria Chuvochina, Pierre-Alain Chaumeil, Christian Rinke, Aaron J. Mussig, and Philip Hugenholtz. 2020. <span>“A Complete Domain-to-Species Taxonomy for <span>Bacteria</span> and <span>Archaea</span>.”</span> <em>Nature Biotechnology</em> 38 (9): 1079–86. <a href="https://doi.org/10.1038/s41587-020-0501-8">https://doi.org/10.1038/s41587-020-0501-8</a>.
</div>
<div id="ref-Parks2015" class="csl-entry" role="listitem">
Parks, Donovan H., Michael Imelfort, Connor T. Skennerton, Philip Hugenholtz, and Gene W. Tyson. 2015. <span>“<span>CheckM</span>: Assessing the Quality of Microbial Genomes Recovered from Isolates, Single Cells, and Metagenomes.”</span> <em>Genome Research</em> 25 (7): 1043–55. <a href="https://doi.org/10.1101/gr.186072.114">https://doi.org/10.1101/gr.186072.114</a>.
</div>
<div id="ref-Schwengers2021" class="csl-entry" role="listitem">
Schwengers, Oliver, Lukas Jelonek, Marius Alfred Dieckmann, Sebastian Beyvers, Jochen Blom, and AlexanderYR 2021 Goesmann. 2021. <span>“Bakta: Rapid and Standardized Annotation of Bacterial Genomes via Alignment-Free Sequence Identification.”</span> <em>Microbial Genomics</em> 7 (11): 000685. <a href="https://doi.org/10.1099/mgen.0.000685">https://doi.org/10.1099/mgen.0.000685</a>.
</div>
<div id="ref-Seemann2014" class="csl-entry" role="listitem">
Seemann, Torsten. 2014. <span>“Prokka: Rapid Prokaryotic Genome Annotation.”</span> <em>Bioinformatics</em> 30 (14): 2068–69. <a href="https://doi.org/10.1093/bioinformatics/btu153">https://doi.org/10.1093/bioinformatics/btu153</a>.
</div>
<div id="ref-Steinegger2017" class="csl-entry" role="listitem">
Steinegger, Martin, and Johannes Söding. 2017. <span>“<span>MMseqs2</span> Enables Sensitive Protein Sequence Searching for the Analysis of Massive Data Sets.”</span> <em>Nature Biotechnology</em> 35 (11): 1026–28. <a href="https://doi.org/10.1038/nbt.3988">https://doi.org/10.1038/nbt.3988</a>.
</div>
<div id="ref-Uritskiy2018" class="csl-entry" role="listitem">
Uritskiy, Gherman V., Jocelyne DiRuggiero, and James Taylor. 2018. <span>“<span>MetaWRAP</span>a Flexible Pipeline for Genome-Resolved Metagenomic Data Analysis.”</span> <em>Microbiome</em> 6 (1): 158. <a href="https://doi.org/10.1186/s40168-018-0541-1">https://doi.org/10.1186/s40168-018-0541-1</a>.
</div>
<div id="ref-Wright2023" class="csl-entry" role="listitem">
Wright, Robyn J., Andrè M. Comeau, and Morgan G. I. Langille. 2023. <span>“From Defaults to Databases: Parameter and Database Choice Dramatically Impact the Performance of Metagenomic Taxonomic Classification Tools.”</span> <em>Microbial Genomics</em> 9 (3): 000949. <a href="https://doi.org/10.1099/mgen.0.000949">https://doi.org/10.1099/mgen.0.000949</a>.
</div>
<div id="ref-WuMaxbin2016" class="csl-entry" role="listitem">
Wu, Yu-Wei, Blake A. Simmons, and Steven W. Singer. 2016. <span>“<span>MaxBin</span> 2.0: An Automated Binning Algorithm to Recover Genomes from Multiple Metagenomic Datasets.”</span> <em>Bioinformatics</em> 32 (4): 605–7. <a href="https://doi.org/10.1093/bioinformatics/btv638">https://doi.org/10.1093/bioinformatics/btv638</a>.
</div>
</div>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./functional-profiling.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./authentication-decontamination.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Authentication and Decontamination</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">© 2023 SPAAM Community &amp; Authors with ❤️. Available under <a href="http://creativecommons.org/licenses/by-sa/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>