---
title: Introduction to NGS Sequencing
author: James A. Fellows Yates
number-depth: 2
bibliography: assets/references/introduction-to-ngs-sequencing.bib
---

::: {.callout-important}
ðŸš§ This page is still under construction ðŸš§
:::

::: {.callout-important}
Parts of this text are a raw transcription of the 2022 lecture from the Werner Siemens-Stiftung funded SPAAM Summer School: Introduction to Ancient Metagenomics (2022). Extensive editing is still underway. Readability will be low.
:::

Next generation sequencing (NGS) revolutionised the whole of biology by providing rapid and cheap access to huge amounts of DNA sequence data. 
One unexpected benefit of this technology was that the technique used by Illumina sequencers was that it was also _perfect_ for sequencing ancient DNA.

In this chapter, we will do a brief overview of how DNA is structured, how DNA sequencing works, how most NGS sequenced DNA sequences are digitally represented, and then cover some of the important considerations for ancient metagenomic sequencing datasets.

## Basic structure of DNA

To understand how sequencing works, we need to understand the basic structure of DNA.
Deoxyribonucleic acid (DNA) is a biological molecule famous for it's 'double helix' structure ([@fig-intro-ngs-fig-3dhelix]), and is present in the vast majority of cells in your body . 
It encodes all the genetic information required for the development, growth, and function of living organisms.

The each strand of the double helix structure consist of four main components called nucleotides, which bind together in different orders.
These nucleotides consist of two groups:

- Pyrimidines: cytosine (C) and thymine (T)
- Purines: guanine (G) and adenine (A)

The nucleotides on one strand are connected via a 'sugar phosphate backbone', however the two strands are held together by the interaction of the hydrogen bond pairing of four different nucleobases on each strand ([@fig-intro-ngs-fig-2dhelix]).


:::{layout-ncol=2}
![Animation of 3D DNA double helix. Source: Erin Rod, [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/deed), via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:The_Structure_of_DNA_GIF.gif)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-3dhelix.gif){#fig-intro-ngs-fig-3dhelix height=300px}

![2D molecular diagram of DNA helix, with sugar-phosphate backbone and amine groups labelled indicated. Source: Pradana Aumars, [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/), via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Structure_ADN.png)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-2dhelix.png){#fig-intro-ngs-fig-2dhelix height=300px}
:::

This pairing consists of one pyrimidine and one purine, which are complementary to each other.

- C with G (to remember: thing CGI in animated movies)
- A and T (to remember: think AT-AT Walker from Star Wars, [@fig-intro-ngs-fig-atatwalker])

Therefore, whenever you find a C on one strand, you will normally find a G on the other strand (and vice versa), and the same for A and T.
If you find an A on one strand, you will find a T on the other strand, or if a T on one strand, you will have another.

What this means is that because of this complementary pairing, depending on which strand you are reading, you can always reconstruct the order bases on the _other_ strand.

![Lego construction of AT AT walker from Star Wars. Source: Tim Moreillon, [CC BY-SA 2.0 Generic](https://creativecommons.org/licenses/by-sa/2.0/deed), via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:LEGO_Star_Wars_AT-AT_Walker_(6740909421).jpg)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-atatwalker.jpg){#fig-intro-ngs-fig-atatwalker height=300px}

## DNA replication

This characteristic of the DNA molecule is a critical component on how DNA gets replicated (or copied) within a cell.

At it's core, DNA replication consists of:

1. Unwinding the DNA helix and separate the two strands by breaking the hydrogen bonds between the complementary bases using a helicase enzyme
2. Use a polymerase enzyme to add complementary new 'free' nucleotides floating in the cell to the exposed base on the strand (repeating for each base until a 'stop' signal is received)
3. Construct bonds on the sugar-phosphate backbone of the new strand using a ligase enzyme

A graphical representation can be seen in using a helicase enzyme in the image below ([@fig-intro-ngs-fig-dnareplication]).

![Schematic diagram of DNA replication, with a helicase unwinding DNA strand with DNA polymerases on the leading and lagging strand incorporating free nucleotides. Source: Christine Imiller, [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/deed), via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:DNA_Polymerase_DNA_Replication.png)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-dnareplication.jpg){#fig-intro-ngs-fig-dnareplication height=300px}

This concept is important because it is the basis of how DNA sequencing works, and how we can reconstruct the sequence of a DNA molecule.

## Extracting DNA

To understand why NGS sequencing revolutionised the field of palaeogenomics, we need to quickly need to compare a difference between how we get modern and ancient DNA.

### Modern DNA

To get DNA from 'modern' samples (i.e., living organisms), you first get your tissue.
You then have to break down (lyse) the cell membrane and/or walls, to release the molecular contents of the cell.
Extraction protocols then use a variety of enzymes to degrade the other biomolecules in the cell (e.g., proteins, lipids, RNA) so that they do not 'interfere' with the extraction of the DNA itself.
Finally, the DNA is separated and pulled out from the rest of the now-broken cell contents (purification).

This extracted DNA are typically in very long and intact strands that you can sort of imagine as long, soft spaghetti pasta like structures ([@fig-intro-ngs-fig-moderndnaextraction]).

![Simplified schematic diagram of the process of DNA extraction from living cells. Source: CNX OpenStax, [CC BY 4.0 International](https://creativecommons.org/licenses/by/4.0/deed), via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Figure_17_01_02.jpg)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-moderndnaextraction.jpg){#fig-intro-ngs-fig-moderndnaextraction width=90%}

### Ancient DNA

In contrast, while the concept of ancient DNA extraction is the similar, the end product of the extraction is different.

Taking a skeleton as an example, once you have your bone sample, cell lysis is normally not necessary as the cells are already broken down.
However, we have to instead 'liberate' the DNA from the mineral matrix of the bone - so palaeogenomicsts first demineralise the bone to release the DNA and then then degrade the remaining more robust molecules in the sample (typically just protein) before purifying the DNA.

The resulting DNA molecules are quite different from the modern DNA.
Rather than well cooked, soft spaghetti structures - ancient DNA molecules are more like extremely overcooked spaghetti. 
These molecules are highly degraded, broken down to very small fragments, and also often have 'damage' at the ends in the form of 'modified nucleotides' that do not represent the original sequence  (see the [Introduction to Ancient DNA chapter](introduction-to-ancient-dna.qmd) for more information).  
Finally, the small amount of tiny and damaged DNA molecules typically sits in a 'soup' of 'contaminating' high-quality modern DNA from the surrounding burial and storage environment ([@fig-intro-ngs-fig-ancientdnainbone]).

As we will find out in the next section, this short fragments of DNA is not a disadvantage for NGS sequencing, but rather a benefit.

![Â© [Lucy Reading-Ikkanda](https://www.lucyreading.co.uk/project/ancient-dna/) for The Scientist Magazine, June 2015 (reused here with permission). Very small fragmented 'endogenous' DNA is preserved in bone, but alongside more recent and modern contaminating microbial and recent human DNA](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-ancientdnainbone.png){#fig-intro-ngs-fig-ancientdnainbone height=600px}


## DNA sequencing

Now that we have our DNA we want to be able to 'read' the order of the nucleotide sequence that makes up the molecule.
Sequencing is the processing of converting the sequence of the four chemical nucleotides to the `ACTG` you can see on your computer screen.
The result of this process is the essentially what we do pretty much all of our analyses on in genetics and genomics.

The core concept of most NGS sequencing methods ^[Newer techniques such as Nanopore sequencers use different methods.] is to replicate the strand, but when adding a nucleotide, you instead supply a modified nucleotide with a fluorophore attached to it.
This fluorophore is a small molecule that emits a specific colour when excited by a laser.
As you have four different nucleotides, each will emit a different colour.
Therefore if you record the colour emitted light each time the modified nucleotide is added to a strand, you can reconstruct the sequence of the DNA molecule ([@fig-intro-ngs-fig-colourednucleotides]).

![Molecular diagram of the four DNA nucleotides with coloured fluorophore amine groups. Source: [@Ju2006-cl]](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-colourednucleotides.jpeg){#fig-intro-ngs-fig-colourednucleotides height=600px}

### Sanger sequencing

Historically, the first mass-production sequenced method was Sanger sequencing ([@fig-intro-ngs-fig-sangersequencing]) [@Sanger1977-zw].
This was the method used for sequenced the first human genome under the Human Genome Project [@Heather2015-zo]

The concept of Sanger sequencing involves making lots of copies of a single DNA molecule using replication. 
But when adding the nucleotides, you include within the mixture of (normal) free nucleotides, you include a few modified nucleotides that include fluorophore, but also that also have a 'blocking' component that prevents any additional nucleotides from being added to the strand of that _particular copy_ by the polymerase.
Critically, the point at which a blocking nucleotide is incorporated is random.
As there is only a small proportion of blocking nucleotides in amongst pool of free nucleotides,the incorporation of the blocking nucleotides occurs at a slow rate and at different points during the replication process.
This results in many replicated molecules but all of different lengths ([@fig-intro-ngs-fig-sangersequencing]).

![Diagram of sanger sequencing, with a primer, template and free blocking nucleotides being added to ends of molecules at random lengths, and then being sent through a size-selecting capillary with a laser dector to identify which of the four blocking nucleotides are passing through the capillary and thus ready ina  chromatopgraph. Source: Estevezj, [CC BY-SA 3.0 Unported](https://creativecommons.org/licenses/by-sa/3.0/) via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Sanger-sequencing.svg)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-sangersequencing.png){#fig-intro-ngs-fig-sangersequencing height=300px}

Once you have the replicated molecules, Sanger sequencing involves running the mixture through a capillary gel.
As all the molecules are of different lengths, they will move through the gel at different speeds, and therefore separate out.
A laser is then fired at the gel at regular intervals, and the light emitted from the fluorophores is detected.
Shorter molecules will move faster through the gel, and therefore will be detected first, whereas longer molecules will be detected later.
Therefore, by detecting the light emitted as the incrementally longer molecules pass through the gel, you can reconstruct the sequence of the nucleotides along the module.

The result is a chromatogram, which is a graph that records the strength of each colour at each point in the gel band ([@fig-intro-ngs-fig-chromatogram]).

![Example of gel-based chromatogram with four columns with bands indicating which colour is the blocking nucleotide of that particular sized DNA molecule, and on the size an intensity graph of each colour given off by each flourophore to indicate which colour it is. Source: Morse Phoque (Abizar), [CC BY-SA 3.0 Unported](https://creativecommons.org/licenses/by-sa/3.0/) via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Radioactive_Fluorescent_Seq.jpg)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-chromatogram.jpg){#fig-intro-ngs-fig-chromatogram height=300px}

While this method was used for the first sequencing human genome, it was very slow and very expensive - it does not scale well due to having to make many copies of a single molecule and sequence each nucleotide sequence one at a time.

Furthermore, the method requires a large amount of starting DNA.
However due to the degraded nature of ancient DNA, it is very rare to have sufficient concentration of DNA to be able to use Sanger sequencing.

### Next generation sequencing

However, in 2005, the first 'Next Generation Sequencing' (NGS) machines were released ^[With the more recent emergence of Nanopore sequencing, really NGS should be equated to 'second generation sequencing', a term you will see in some contexts.] in the form of Roche's 454 Pyrosequencing [@Heather2015-zo].

These new sequencers were able to quickly and cheaply sequence millions of DNA molecules at once, revolutionising genetics and enabling 'accessible' genomics to a much wider set of scientists.

While a few companies emerged with their own NGS machines, such as PacBio and IonTorrent, sequencers from Illumina have been by far the been the most successful.
These machines have been so successful, they have almost exclusively sequenced the entire corpus of ancient DNA sequence data.
Therefore the rest of this chapter will focus just on Illumina sequencing.

## Illumina sequencing 

Illumina sequencing relies on a process termed 'Sequencing by synthesis' (SBS) using reversible terminators [@Bentley2008-oj].
This follows a similar concept as Sanger Sequencing in that you add modified nucleotides with fluorophores, however instead of using multiple molecules with different blocking separated via gel, it uses nucleotides that can be reversibly blocked and thus continue incorporating nucleotides in a single molecule.

This works by immobilising the DNA so it stays on one location on a special type of plate.
The machine then performs out a typical DNA replication, adding the flourophore modified nucleotides one at a time, and then taking a photo of the emitted light.
However the key difference is the blocking part of the oligo is removed, so the next modified oligo is added.

To visualise how this looks like, you can imagine a plate upon which you see millions of coloured tiny dots, each cycling in realtime through different colours as each new nucleotide is added. 


<!-- TODO MAKE A PCITURE TO REPLACE GIF -->

To follow this however, the DNA needs to be prepared in a specific way.

### Flow cells and adapters

To ensure the DNA molecules do not move while taking photos, in Illumina sequencing, the DNA molecules are bound to a special glass slide called a flowcell.
This also ensures the DNA does not get washed away during the sequencing process as different solutions wash through the machine between each step.

<!-- TODO SEE IF CAN REPRODUCE OR MAKE OWN VERSION OF FLOW CELL IMAGE -->

To immobilise the DNA, the flow cell is coated with a 'lawn' of synthetic oligonucleotides (i.e., custom sequences made in a lab).
By attaching (ligating) the _complementary_ sequence of the synthetic oligonucleotides to your DNA molecules from your sample, when your DNA molecules flow over the lawn, the complementary bonds between the nucleotide pairs will form.
This then will hold the DNA in place.

The process of adding the complementary synthetic oligonucleotides (often referred to as oligos), called 'Adapters', is called 'Library preparation'.
These adapters sequences do not only include the complementary sequence to flowcell oligos, but also an additional sequence that acts as a priming site for polymerases to start the replication process
Furthermore, researchers have exploited the highly-multiplex nature of Illumina sequence to allow them to sequence multiple samples at once.
They do this by adding additional sequences to the adapter sequence construct that consist of short sample-specific 'indices' or 'barcodes'.
These indices allow DNA molecules from different samples to be sequenced at the same time and then separated back into the groups of molecules from the same samples during data analysis.
.

<!-- TODO INSERT ILLUMINA DNA CONSTRUCT FROM TINA -->


In the image above <!-- TODO TO DESCRIBE ONCE WE HAVE THE IMAGE And also, when you're sequencing multiple samples at once on the same sequencing run, you can add things called indices, or known as barcodes, which are basically sample specific. So it allows you to mix multiple samples at once into one sequencing run, sequencing all at the same time, later on separate them out. So this is ultimately a slightly simplified version of a Illuminate DNA construct. So the X, X, X, X, X in the middle, this is your target. So this is actually your DNA sequence from your sample. And then at both ends, you essentially have a target primer. So this is where your polymerase will bind onto to start, then actually going, reading into your target, or your insert is another phrase that you can call it. Then prior to the target primer, you also have an index. So this is actually your sample specific barcode. This is actually added typically onto the adapter, and so this happens in the, no, sorry. And the library construction, sorry. And then also you have the adapter and index primer right at the beginning. So this is both what binds onto your flow cell, but also acts as a primer for actually reading the index because you also need a sequence index to know which DNA molecule, to which sample DNA molecule it's coming from. --> 

### Clustering

Now that each DNA molecule is bound to it's own section of the flowcell, we have the problem that the light emitted from a single tiny fluorophore will not be strong enough to be captured by the camera in the sequencer.

Illumina machines get around this through a process called 'clustering'.
The clustering process will make many copies of the exact same DNA molecule right next to the original template molecule.
When the cluster of DNA copies emit the light from the same nucleotide along their sequence at the same time, collectively the light emitted will now be sufficiently strong to be captured by the camera.

To make copies sufficient copies of the 'template' molecule, clustering is carried out using 'bridge amplification' [@Bentley2008-oj].

The process of bridge amplification is as follows:

1. A single-stranded DNA molecule is immobilised on the flow cell via the oligo adapter on _one_ end of the module
2. The DNA is bent over
3. The oligo adapter on the other end of the molecule binds to the other complementary flow-cell oligo adapter in the close vicinity to the template (forming a bridge)
4. A priming sequence, free nucleotides, and a polymerase is added to the flow cell
5. The reverse complement of the DNA molecule is made via replication, and the polymerase mixture is washed away
6. The two strands are then separated to form two single-stranded DNA molecules (the new one the reverse complement of the original template molecule)
7. Repeat 1-5 for the two new strands until sufficient copies are made

This process is also depicted in [@fig-intro-ngs-fig-bridgeamplification.png].

![Diagram of bridge amplification. Source: DMLapato, [CC BY-SA 4.0 International](https://creativecommons.org/licenses/by-sa/4.0/) via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Cluster_Generation.png)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-bridgeamplification.png){#fig-intro-ngs-fig-bridgeamplification.png height=600px}

Of course, at the stage that you've had many cycles of replication, you have in a single cluster both forward and reverse complement copies of the same DNA molecule.
As this would emit different colours at each position along the sequence, one of the strand directions are removed but 'trimming' the flow cell lawn at one of the complementary oligo adapter sequences

### Sequencing-by-synthesis

With our cluster of many copies of the same DNA molecule, we can now start the 'Sequencing by synthesis' (SBS) process itself.

As with replication and Sanger sequencing, the process SBS involves adding free nucleotides to the template strand (with only the complementary base being able to be incorporated to the exposed strand), some of which are modified to emit light when excited with a laser, and taking a picture.

The main difference with Sanger sequencing is you can actually reuse the same DNA molecule to add more nucleotides along the same strand - rather than 'discarding' it with permanently blocking nucleotides. 
This makes the SBS technique more resource sufficient, both in reagents (you need fewer copies of the molecule and recycle the same original copies of template molecules), but also in space - you don't have to separate out the molecules in a gel - they are all fixed on the flowcell in a tightly packed layout.

![Diagram of sequencing by synthesis. Source: Abizar Lakdawalla, [CC BY-SA 3.0 Unported](https://creativecommons.org/licenses/by-sa/3.0/) via [Wikimedia Commons](https://en.wikipedia.org/wiki/File:Sequencing_by_synthesis_Reversible_terminators.png)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-sequencingbysynthesis.png){#fig-intro-ngs-fig-sequencingbysynthesis height=300px}


The process as in @fig-intro-ngs-fig-sequencingbysynthesis can be broken down to:

1. On a single stranded molecule, bind a primer on the molecule's adapter priming site
2. Add fluorescently labelled nucleotides (with a reversible terminator) to the flow cell
3. Only complementary nucleotides will bind to the template strand at the first exposed base
4. Wash away unbound nucleotides
5. Fire a laser to excite the modified nucleotides to emit the corresponding colour, and take a picture]
6. Clip off the fluorophore part of the nucleotide
7. Repeat 2-6 until the entire strand is sequenced

On Illumina sequencers, the number of repetitions (known as cycles) and images typically happens either 50, 75, or 125 times, depending on the machine and the type of sequencing chemistry kit.

The reason why Illumina sequencers (and similar NGS technologies) have been so successful is that this process is happening across millions of clusters at the same time.

![Diagram of sequencing by synthesis. Source: EBI Training, [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/) via [EBI Training](https://www.ebi.ac.uk/training/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/next-generation-sequencing/second-generation-sequencing/illumina-sequencing/)](assets/images/chapters/intro-to-ngs/fig-intro-ngs-fig-sbsimagecapture.png){#fig-intro-ngs-fig-sbsimagecapture height=300px}

You can see a small fraction of this flow cell in @fig-intro-ngs-fig-sbsimagecapture.
Each coloured dot corresponds to a cluster of DNA molecules.
At each cycle (each photo), a new nucleotide is added to the strand, and a laser is fired to excite the fluorophores.
You can see two different clusters emit different lights, as they are different DNA molecules thus have different nucleotides at that particular 'cycle' of the 'replication' process.
By converting the emitted light to the known corresponding A, C, G, T, at each photo you can reconstruct the sequence of the DNA molecule.

By now you probably are starting to feel comfortable with the concept for one-colour equals one nucleotide.
To throw a curve ball, Illumina sequencers actually have _three_ different 'colour chemistries' they use on different machines.

### Colour chemistry

<-- TODO got to here -->

But there's also on particularly the next seek and no seek machines, a slight different system, where they called, they tried to make it a bit cheaper by only using two colors or two dyes. And the approach they take there is that, again, a T is green, red is, C is a red. However, in this case, an A is actually two colors mixed at once. So if you emit that both, so if the machine picks up two colors, two wavelengths are being emitted, that is an A. However, if there was no color being emitted at all in that cluster, the machine reaches as a G, or a no detected dye. So this is very important for some caveats or things you have to consider when processing your data a bit later on. 

### Paired-end sequencing

So something you have to consider though is we're dealing with biology. It is not like, I don't know, chemistry or even sort of physics, where things are perfect and wonderful. Over time, areas start happening, we're not perfect. And essentially the imaging reagents start getting tired, the polymerases start adding mistakes, or don't bind on properly and more areas will occur. So sometimes your nucleotides will not bind, meaning you'll skip a base or you'll get multiple nucleotides being added once and you go forward one. And essentially within your cluster, the DNA molecules being sort of replicated and emitting light will get desynced. So the color will get a bit blurry and less clear. 

And so first the machine does calculate some sort of base quality. So it captures the probability that it thinks it captured the right color, so the right nucleotide of each photo. But to the point where if it has no idea, it will call a dead base call, which will be reported as an N. And this became more and more of a problem, particularly in the early days of sequencing. And so people thought, how can we improve or correct such sort of errors? And the idea that came up was paired end sequencing. 

So what this means is you do one sequencing in one direction, then you flip the entire read over and then read it from the other end. So whereas you have to consider that when you're in the forward direction, over time you're getting more and more errors. So the further you get into the molecule, the less confident you are. And so the more errors are gonna be. You can then turn it over and start the whole thing from scratch with fresh reagents. By going from the reverse end forward, you basically can correct the mistakes that were occurring at the end of the forward sequencing, but get the high quality calls from the other end from the beginning. So you can sort of, that's a bad explanation. But you can sort of see here, you read it once in this direction. So you read the DNA insert and the prime, the index. You turn it over and you do the same thing, but from the beginning. So you basically sequence the same DNA molecule twice. An added bonus of this is also you get more cycles. So if your DNA molecule is a single molecule, it's actually longer than the cycles. It's 50% fair, five base pairs. By going from the other end, you can capture any DNA nucleotides which you're missing from the forward sequencing. This is not so necessarily relevant to DNA. We will typically very short DNA molecules, but in some cases that will apply. 

### Demultiplexing

Then it comes on to biological to computational sequence. How do we take these sort of very raw nucleotide sequences and put it into a format that a computer can read? This typically happens in a step of demultiplexing. It's actually very rare that you yourself as students will have to deal with this or do with this. Typically this is done by sequencing centers or by your lab team. But what this essentially consists of is normally or rather often nowadays, there is two steps in which you're stuck together, which colloquially are no demultiplexing. 

The first step is called base calling. This is where you basically take your photo files of every single nucleotide and convert this into an ACTNG. So taking image files, putting into a text file. But also in most cases, we have multiple sequence, multiple samples at once. And all the samples have these barcodes or indices. And we need to actually separate these out. So you know that all of these DNA molecules come from this sample or these DNA sample, sorry, DNA molecules from this sample. And this happens in this demultiplexing step. So essentially a computer program will read in each DNA molecule whether it finds such a barcode across from one sample, let's say an example here, the red one that corresponds to sample one and the reverse here. And then the machine will basically sort or order the DNA molecules accordingly. So all of the DNA molecules which have these combinations and indices will put into sample one, sample two. So the blue and yellow will go here and so on. So this is something you, again, rarely will have to do yourself, but it's something just to keep in mind. 

### FASTQ File

And the output of this demultiplexing step is something called a fastq file. So this is a text-based format for storing biological sequences itimes of cases nucleotide sequences, but also with the base quality scores. So these are the things where the computer tries to estimate the probability that it thinks that the nucleotide call was correct. And both are encoded basically with ASCII text characters. Doesn't really matter what they are, if you're not familiar with that. So this is a very, very small example. These files can be gigabytes in size, so huge, huge text files typically compressed, but still a gigabyte start to being compressed. 

But what they, all they are made of, of basically a repeating set of four different lines. You have the first line up here, and this is called the ID line. This stores multiple information from the sequencer. So the sequencing, or the sequencing machine ID, then a run number, a flow cell ID, so each flow cell will have an ID from the manufacturer. And then you have essentially a bunch of coordinates going on here, which basically tells you from which cluster on their flow cell has the DNA molecule come from. Then you can have a bunch of extra information. This is often quite random depending on the sequencing and what they put in here, but often people put things like barcodes, maybe sort of how many errors were allowed during gene multiplexing, because you could do a certain enough filtering during that step. On the second line, you have the DNA sequence itself. So in this case, it starts with an N, this is dead-based call. This is actually quite often common, because this is when the camera is still calibrating itself. And so the first base is often a bit rubbish, but then the rest of the molecule, as you can see, sorry, the sequence here is A, C, T, and G, so it's usually a new molecule. You'll then have a plus, which is a separator. And the fourth line of this repeating set of four is then the base quality scores. So these are random, sort of a random set of characters, which I'll explain in a second. But basically this tells you the confidence of that, how good we think that that nucleotide call was. 

And then you can basically see the same thing here on the next line, and it repeats as follows. So you can see, for example, this number is a bit different, because it's from a slightly different coordinate cluster. So these quality scores, they are not uniform, I have to say. So it depends on both the age of the machine and what the manufacturer selected. But typically they will look something like this, where there's a fixed order in the ASCII characters. And each one will correspond to a different probability of Fred's score. I won't explain exactly what that is, it's a bit mazzy. But essentially what it means is that the higher the character along this score, the more confident you are of the base call, because the probability that it's incorrect is sort of low. 

## Sequencing recap

So to recap, DNA molecules are essentially made of nucleotides, A, C, T, and Gs. We have two strands, which is complementary based pairings, C, G, CGI, AT, Atatoka, Star Wars, it's great. Modern DNA is very long, but AT DNA is very short. And this is very good for NGS sequencing, where we do this massive multiplexing. So where rather than trying to sequence few very long DNA molecules, we sequence lots and lots and lots of very short ones at very high accuracy, which we can then reconstruct the long sequences later on, if you've got a good DNA. So the main steps are adding adapters to create something called a library, which allows your DNA molecules to bind to the glass slide, the flow cell, or something called a lawn. You then basically make a new strand, each cycle, you basically add a fluorescent nucleotide, which you can fire a laser, AT, which emits a colour, you can take a photo, and by basically recording, the order of the colours being emitted in a single point in the flow cell, you can reconstruct the DNA sequence. This de-syncing of clusters results in lower based quality scores over time, so you can also improve this by paired end sequencing, where you basically sequence from one end, and then do turn it around, then you sequence from the other end with fresh reagents. 

## Sequencing and considerations for ancient metagenomics

So for the last section of this lecture, I want to give you a few ideas of things you should consider when you're dealing with DNA for ancient genomics. Some of those are applicable to modern genomics too, but it's things I find that, through my career, people forget about in some cases. 

### Low DNA preservation

So firstly is low DNA preservation. So when you're dealing with an ancient DNA, your samples are very old, and only have very little DNA in the sample, and during library preparation, you may have to do lots and lots and lots of amplification, make lots and lots and lots of copies of your DNA to make a sufficient amount to actually put sequencing. And this is important, and it will be discussed later during the week, but this is important because during library construction, you can actually inflate your counts in terms of DNA molecules that come from a particular taxon, micro-brothaxon, for example, and basically skew your estimate of which species are in your sample or not, or were in your sample because they're now dead. Also, by overamplifying your DNA molecules, you reduce the number of sequencers you actually get out there. By having these duplicate molecules, you're not actually providing any more extra information about your DNA library, and your sequencing flow cell only has a fixed number of sequencing slots, which basically, if you've overamplified your library, will basically fill up your slots and you will not sequence as many unique reads, and so the amount of information you're getting out of your DNA molecule can be problematic. So this is actually quite, I've jumped ahead quite a bit in terms of detail here, but this is, I wanted to have the slides here for you to go back to and recap later. 

### Index hopping

Another thing during sequencing, you have to consider is something called index hopping. So this is a problem, particularly with Illumina Sequ, or people are aware of with Illumina sequencing, and it's a challenge when you're doing multiplex sequencing. When you're sequencing lots and lots of samples at once, and you have to have these indexes or barcodes, which allow you to identify this DNA molecule comes from this sample. This seems to happen more often on a type of flow cell, which you find on Illumina HiSeq X and NovaSeq machines, and is ultimately caused by free floating index primers. And why that is a problem, is that if you do not sufficiently clean up your primers, during the clustering process, you can accidentally start adding on or switching barcodes between DNA molecules. This does happen at a relatively low rate, but it does happen. And so what it means is that essentially you switch the barcodes and you may accidentally assign a DNA molecule from one sample into another sample when you're doing demultiplexing. And so let's say you're dealing with a microbiome sample, and you have lots of, let's say, unless you have an oral microbiome sample and a gut microbiome sample, you may start seeing, for example, oral species popping up in your, or sort of oral species which are only found in the mouth, ending up in your gut samples, which may be a bit weird. This can also be a particular problem if you're doing microbial genomics, so doing, let's say, pathogen reconstruction, if you're working on that. And you mix capture results with your shotgun samples because then you may start picking up the high amount of capture results in your shotgun samples where you're doing screening. So you may start getting false positives there. There's quite a few papers on this, also in the context of ancient DNA, so van der Waag has got quite a good paper to understand that and also had to correct such, or estimate the level of this in your studies. 

### Sequencing errors

They go back to the sequencing, you have to consider your sequencing errors. So if you don't sufficiently quality control and check for these errors happening on DNA sequencing, you may actually start incorporating errors into your analysis downstream. So for example, what may happen is if you have a low base quality score, the machine may have picked up the wrong base or the wrong nucleotide. And this means that your DNA molecule or sequence, when you compare to reference genome databases or reference genomes, may start going to the wrong place or match the wrong reference genome because you have the wrong nucleotides, the wrong sequence. Which can be a problem. Also, it can reduce your chance of getting sufficient overlap during the assembly, which is where you basically stick together all of the overlapping DNA molecules together to try and reconstruct the entire molecule. This is something that Alex will present on Thursday, like Subna. And also if you're doing variant calling for phylogenomics and you have very low coverage, this may also start increasing, increasing the add errors and you will do the wrong SNP call, which means that basically your, the relationship between your genomes in your tree, for example, will be skewed. And so it's always very important to check for such errors and check that your sequencing run was high quality.

### Dirty Genomes

You also have to consider, so this is again sort of jumping ahead, but there's a reason why I'm putting in this presentation, is dirty genomes. So unfortunately, there are many reference genomes which are very dirty. So dirty, I mean, for example, having a lot of adapters still in there. And this means you have the problem where if you yourself have not sufficiently cleaned up your DNA library to remove the adapters and remove also pre-processing, you will start seeing weird results. For example, I very, very highly expect that if any of you screen against the NTPI-NT nucleotide database, you will start seeing carp everywhere. And the reason why is because people somehow got onto the NCBI a whole carp genome without removing any of their adapters. So there's adapters sequences everywhere in the genome. And so whenever you have an adapter in your library, this will basically align to the carp genome and then you'll get carp, which particularly if you're trying to look at diet, for example, or ancient diet in, like say microbiome studies, where you look at some calculus, you may start seeing carp even if you're, I don't know, from, got samples from, I don't know, Chile or somewhere where carp is not expected to find. You also often will find this with zebrafish. So often many, many, many, I think it's Darius Varini or something like that, which makes no sense because all of these fish comes from one lake in Africa, but you see it everywhere in your metronome examples. And again, it's because of dirty genomes where they think of adapters or vectors in the genome. 

### Low Sequence Diversity

Another thing is low sequence diversity. So what I mean by this is mononucleotide reads, like GGGGG, or dinoucleotide repeats. This is not so much of an error necessarily when you're doing metagenomics, but when you come into genomics, this can be a problem. So the problem with such DNA molecules, like this one here, GGGGG, is they're very unspecific. They give you no information. That can come from any species everywhere because they are very common across all genomes. So the problem here is that firstly, it slows down your processing because basically you are aligning against, or comparing of DNA sequence against many, many, many different genomes to ultimately say, I don't know which one it comes from, which is unnecessary. And also in some cases, it can inflate counts at higher nodes when you're doing an LCA. This will be described later on. But this can be very common. So if you remember this two-color chemistry which I mentioned earlier, where you don't have one color per base, but rather if there is no color emitted, the machine uses a G, particularly with an ek-seq and no-seq data, you will have a lot of these Gs, particularly as we're dealing with ancient DNA, which is very short. When you have very short ancient DNA and you don't reach all of your cycles, so let's say you're doinbase pair cycles, but your DNA molecules are onlbase pairs, you will basically get to the end of your DNA molecules and not add anything else. So you start getting these very long tails of Gs at the end of your molecules. And if you don't remove these, this will make it very difficult to correctly align your DNA molecule to a reference genome or sequence. So be aware to look for these and remove these. Also because it will speed up your processing. So to recap the considerations, a lot of this will make more sense later on in the other sessions, but consider your duplication rates. You check for lots of copies of same DNA molecule in your library. It's a good idea to check for index hopping, so making sure that the index combinations that you have in your library are correct and you've sorted your samples correctly. Always check for sequencing error and remove low quality bases if possible. Check for adapters, so to make sure you don't start finding CARP everywhere. And also it's a good idea to look for low-seq and diversity reads. For example, particularly if you put next-seq or no-seq data, because it just slows down your results and you get lower quality text-long assignment. 

### Q and A
 
#### How to design barcodes

Okay. How to design the barcodes. That's a question from UD. That is quite tricky because there's a lot of considerations you have to make when making sure there's a balance and they're not too similar to each other. Often manufacturers have tools which allow you to basically generate this. I believe they also to sentence and have standard sets which they can also send you that you request. So you yourself do not have to necessarily design these. It depends on your lab. So often I'd say speak to your sequencing center if you have them. Because they all have advice. All check manufacturers, I think most like Agilent and Illumina will also basically have such things for you then. Yeah, sometimes they make it a little bit defined but you can't find them. And if you read the Meyer and Chercher article for buildinnew libraries, it would be provided in that set. Yeah. Could everyone hear Tina then? No? Okay. So she said that often the manufacturer make it a bit hard to find such functionality on their websites and stuff, but you can often do that. But also if you read the sort of classic paper by Meyer and Chercher 2011, 12, they actually have the set of barcodes that you can use yourself. So I think that link, that paper is on the website somewhere but we can also share with you. 

#### How many indicies can you use

How many indices can you use? So this is a good question. This depends on your strategy and is slowly changing over time. So you can actually choose one index if you want, which is at the beginning of your, this is from Laura, which is the beginning of your molecule. What has been recommended and what the Meyer and Chercher paper introduces double indexing where you have two at the end. And these are attached to your adapters. What people are commonly doing now is actually adding additional barcodes called inline barcodes. So these are very short sequences, about seven base pairs I think, which you actually attach immediately to the DNA molecule before library preparation. So after extraction before library preparation. And this actually helps you with, sorry, with correcting for index hopping. So if you read the van der Waal paper, which I mentioned earlier, we can send the link again later. They also describe how they use these internal barcodes to separate out. So for example, you can get from a manufacturer about 100, let's say barcodes, but you will normally per library can have somewhere between one to four separate identifiers. 

#### Why Gs called more often in two colour chemistry sequencing

So I have a question regarding the gene calling for Gs, right? You mentioned that Gs are a characteristic of ancient DNA or sequencing errors. Why is it specifically Gs that are called more often rather than the other bases? The reason, okay, so that is because your DNA molecule is very short. So let's say your DNA sequence is onlbase pairs long, but your sequencing cycles, so the number of cycles of imaging your machine is gonna do is let's sabase pairs. Once you've got through thbase pairs, there is nothing to sequence anymore. Your lights are not going to emit anything. So when you're on nobody-connected data, if no, sorry, machines, if no light is emitted, it reads it as a G. And you have to remember that the machine is not going to stop imaging once that one DNA molecule is finished. The DNA, the sequencing machine will keep taking photos until it's reached to the number of cycles you've set, whether in this case, 75. So once you've exhausted your DNA molecule, there's nothing to sequence, nothing to image anymore. So basically the machine will just keep picking up G for every remaining cycle of the run. Does that make sense? Yes. Yes, it does. But still I don't get why G and not like AT. Why is it specifically the space? Because on nobody can make sick data, whatever reason they've decided Gs means nothing. There's no color. Once it runs out of the... 

Yeah, so basically the problem with ancient DNA is we often have very short reads. So you might do say two bsequencing, which is very common, but you might have a read that's onlbases long. And so once it kind of runs out of DNA, it will just, it won't sequence anymore. So there'll be no more fluorophores. And because the NovaSeq and the next you can interpret that as a G, you'll just get these polyG tails, but actually it just means no more data. And another thing to... Could you make it up for yourself? And I think this may also happen in modern data as well. If you've fragmented your modern data too short, you'll also get that. It's just that in the complex of ancient DNA, the reason why I said that is because we are naturally already very, very short because of the degradation. And I'll also say this is, if you're used to doing modern DNA, this is where it's really different because let's say you're sequencing a regular library. What you would normally do for modern DNA is you have genomic DNA, which is huge. So for your microbial DNA, each genome is something likmillion bases long, and that's way too big for an aluminum machine. So what you would do is you would shear it either enzymatically or by sonication, usually to an average size of aboubases. And then you do your alumina sequencing usually two by 150. So you kind of measure one side, then you measure the other side, and you get a total obases sequenced out of thbase pair read. You never run out of DNA. You never actually get to the end of the molecule when you're sequencing. And so for most people that do modern DNA sequencing, they've never dealt with this problem before because they never see it because they're always sequencing a DNA molecule longer than what their sequencing chemistry can actually do. For ancient DNA, it's very different. We actually don't shear. We take advantage of the fact that because our DNA is short, we don't have to shear. And because we don't shear, it actually allows us to exclude some of the modern contamination because any modern DNA that's in there will be so long that it won't build a proper library and it won't be sequenced. And so it kind of helps us clear out some of the modern DNA that might be present. And so we will only sequence the short DNA sequences, which are more likely to actually be ancient. But the problem there is we're dealing with the real size of the ancient DNA, which might bbases, 30 basesbasesbases. We don't have necessarily the kind of consistency you would have if you were intentionally shearing modern DNA. So we do have some sequences that are very short.

#### What is full genome seuqencing

Okay, so Liasat asked, when we're talking about whole genome sequencing, so WGS and full genome sequencing, FGS, is it the same? I've never heard of FGS. So yes, I would say it probably is. 

#### Tools for generating indicies 
 
And then Jaime asks, is, I hope I'm saying that right. Is this kind of program publicly available? Is this to? Sorry? It was the index checker to make sure your pool is not. It was the index checker, yes. So again, lots of tools online, I think, basically, to make sure there's no overlap. Normally the manufacturer will offer such thing.

## Readings

### Reviews

[@Schuster2008-qx]

[@Shendure2008-fh]

[@Slatko2018-hg]

[@Van_Dijk2014-ep]

### Sequencing Library Construction

[@Kircher2012-fg]

[@Meyer2010-qc]

### Errors and Considerations

[@Ma2019-lg]

[@Sinha2017-zo]

[@Van_der_Valk2019-to]

## Questions to think about

- Why is Illumina sequencing technologies useful for aDNA?
- What problems can the 2-colour chemistry technology of NextSeq and NovaSeqs cause in downstream analysis?
- Why is 'Index-Hopping' a problem?
- What is good software to evaluate the quality of your sequencing runs?

## References