<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Vilma Pérez">

<title>5&nbsp; Introduction to Environmental DNA – Introduction to Ancient Metagenomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./section-useful-skills.html" rel="next">
<link href="./introduction-to-microbial-genomics.html" rel="prev">
<link href="./favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-9dcef9bc6179ae6ca98dd89ff182b1a7.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<meta property="og:title" content="5&nbsp; Introduction to Environmental DNA – Introduction to Ancient Metagenomics">
<meta property="og:description" content="">
<meta property="og:image" content="https://spaam-community.github.org/intro-to-ancient-metagenomics-book/assets/images/chapters/introduction-to-edna/figure1.jpg">
<meta property="og:site_name" content="Introduction to Ancient Metagenomics">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-theory.html">Theory</a></li><li class="breadcrumb-item"><a href="./introduction-to-edna.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Environmental DNA</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./foreword.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Foreword</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-edna.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Environmental DNA</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title"><em>De novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Authentication</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./contamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Contamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancient-metagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./deprecated-chapters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Deprecated Chapters</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-writing-guidelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Chapter Writing Guidelines</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-template.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Chapter Page Template</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-checklist.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Chapter Checklist</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#what-is-environmental-dna" id="toc-what-is-environmental-dna" class="nav-link active" data-scroll-target="#what-is-environmental-dna"><span class="header-section-number">5.0.1</span> What is environmental DNA?</a></li>
  <li><a href="#what-can-we-find-in-sedimentary-dna-records" id="toc-what-can-we-find-in-sedimentary-dna-records" class="nav-link" data-scroll-target="#what-can-we-find-in-sedimentary-dna-records"><span class="header-section-number">5.1</span> What can we find in sedimentary DNA records?</a></li>
  <li><a href="#taphonomic-processes-of-sedimentary-dna" id="toc-taphonomic-processes-of-sedimentary-dna" class="nav-link" data-scroll-target="#taphonomic-processes-of-sedimentary-dna"><span class="header-section-number">5.2</span> Taphonomic processes of sedimentary DNA</a>
  <ul class="collapse">
  <li><a href="#origin-of-edna-in-sediments" id="toc-origin-of-edna-in-sediments" class="nav-link" data-scroll-target="#origin-of-edna-in-sediments"><span class="header-section-number">5.2.1</span> Origin of eDNA in sediments</a></li>
  <li><a href="#fate-of-edna-in-the-environment" id="toc-fate-of-edna-in-the-environment" class="nav-link" data-scroll-target="#fate-of-edna-in-the-environment"><span class="header-section-number">5.2.2</span> Fate of eDNA in the environment</a></li>
  <li><a href="#mineralogical-influence-on-preservation" id="toc-mineralogical-influence-on-preservation" class="nav-link" data-scroll-target="#mineralogical-influence-on-preservation"><span class="header-section-number">5.2.3</span> Mineralogical influence on preservation</a></li>
  </ul></li>
  <li><a href="#dna-analysis-laboratory-processing" id="toc-dna-analysis-laboratory-processing" class="nav-link" data-scroll-target="#dna-analysis-laboratory-processing"><span class="header-section-number">5.3</span> DNA analysis: laboratory processing</a></li>
  <li><a href="#precautions-and-considerations-in-ancient-edna-research" id="toc-precautions-and-considerations-in-ancient-edna-research" class="nav-link" data-scroll-target="#precautions-and-considerations-in-ancient-edna-research"><span class="header-section-number">5.4</span> Precautions and considerations in ancient eDNA research</a>
  <ul class="collapse">
  <li><a href="#contamination-control-in-field-and-laboratory-sampling" id="toc-contamination-control-in-field-and-laboratory-sampling" class="nav-link" data-scroll-target="#contamination-control-in-field-and-laboratory-sampling"><span class="header-section-number">5.4.1</span> Contamination control in field and laboratory sampling</a></li>
  <li><a href="#tracking-contaminants-in-field-sampling" id="toc-tracking-contaminants-in-field-sampling" class="nav-link" data-scroll-target="#tracking-contaminants-in-field-sampling"><span class="header-section-number">5.4.2</span> Tracking contaminants in field sampling</a></li>
  <li><a href="#taxonomic-assignment-and-community-reconstruction" id="toc-taxonomic-assignment-and-community-reconstruction" class="nav-link" data-scroll-target="#taxonomic-assignment-and-community-reconstruction"><span class="header-section-number">5.4.3</span> Taxonomic assignment and community reconstruction</a></li>
  <li><a href="#spatial-heterogeneity-of-sedadna" id="toc-spatial-heterogeneity-of-sedadna" class="nav-link" data-scroll-target="#spatial-heterogeneity-of-sedadna"><span class="header-section-number">5.4.4</span> Spatial heterogeneity of sedaDNA</a></li>
  <li><a href="#vertical-migration-of-dna" id="toc-vertical-migration-of-dna" class="nav-link" data-scroll-target="#vertical-migration-of-dna"><span class="header-section-number">5.4.5</span> Vertical migration of DNA</a></li>
  <li><a href="#potential-applications" id="toc-potential-applications" class="nav-link" data-scroll-target="#potential-applications"><span class="header-section-number">5.4.6</span> Potential applications</a></li>
  <li><a href="#palaeoenvironmental-reconstruction" id="toc-palaeoenvironmental-reconstruction" class="nav-link" data-scroll-target="#palaeoenvironmental-reconstruction"><span class="header-section-number">5.4.7</span> Palaeoenvironmental reconstruction</a></li>
  </ul></li>
  <li><a href="#insights-provided-by-microbial-shifts-over-time" id="toc-insights-provided-by-microbial-shifts-over-time" class="nav-link" data-scroll-target="#insights-provided-by-microbial-shifts-over-time"><span class="header-section-number">5.5</span> Insights provided by microbial shifts over time</a>
  <ul class="collapse">
  <li><a href="#phytoplankton-changes-in-response-to-climate-in-polar-ecosystems" id="toc-phytoplankton-changes-in-response-to-climate-in-polar-ecosystems" class="nav-link" data-scroll-target="#phytoplankton-changes-in-response-to-climate-in-polar-ecosystems"><span class="header-section-number">5.5.1</span> Phytoplankton changes in response to climate in Polar ecosystems</a></li>
  <li><a href="#environmental-microbiome-associated-with-vegetational-changes" id="toc-environmental-microbiome-associated-with-vegetational-changes" class="nav-link" data-scroll-target="#environmental-microbiome-associated-with-vegetational-changes"><span class="header-section-number">5.5.2</span> Environmental microbiome associated with vegetational changes</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">5.6</span> Summary</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">5.7</span> References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-theory.html">Theory</a></li><li class="breadcrumb-item"><a href="./introduction-to-edna.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Environmental DNA</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Environmental DNA</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Vilma Pérez </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<section id="what-is-environmental-dna" class="level3" data-number="5.0.1">
<h3 data-number="5.0.1" class="anchored" data-anchor-id="what-is-environmental-dna"><span class="header-section-number">5.0.1</span> What is environmental DNA?</h3>
<p>DNA can be recovered from virtually anywhere. In fact, DNA can be found in the environment from cellular material shed by organisms (e.g., skin flakes, urine, faeces, hair, mucus, or their natural demise) that have accumulated in the surrounding water, soil, air, sediment (<a href="#fig-introductiontoedna-figure1" class="quarto-xref">Figure&nbsp;<span>5.1</span></a>). We call this type of DNA, <strong>environmental</strong> (eDNA).</p>
<p>Environmental DNA (eDNA) refers to the genetic material that can be extracted and studied from environmental samples without the need to isolate target organism first <span class="citation" data-cites="Taberlet2018-xa">(<a href="#ref-Taberlet2018-xa" role="doc-biblioref">Taberlet et al. 2018</a>)</span>.</p>
<div id="fig-introductiontoedna-figure1" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure1.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.1: Examples of where environmental DNA can be recovered from the environment.
</figcaption>
</figure>
</div>
<p>A defining feature of eDNA is its complexity. In environmental samples, we encounter a diverse mix of genetic material from multiple species and individuals, presenting unique challenges in the analysis compared to the analysis of genetic material of a single organism (<a href="#fig-introductiontoedna-figure2" class="quarto-xref">Figure&nbsp;<span>5.2</span></a>). Additionally, in some environmental substrates, DNA from certain organisms can be preserved for long periods even after the organisms have died. This preserved DNA is referred to as ancient eDNA <span class="citation" data-cites="Pedersen2015">(<a href="#ref-Pedersen2015" role="doc-biblioref">Pedersen et al. 2015</a>)</span>.</p>
<p>Introducing the element of time to the already complex eDNA scenario further intensifies its complexity. Time can be unkind to DNA, and the deeper we venture into the past, the more challenging it becomes to recover intact genetic material (<a href="#fig-introductiontoedna-figure3" class="quarto-xref">Figure&nbsp;<span>5.3</span></a>). Ancient environmental DNA embodies the intertwining challenges of temporal degradation and environmental complexity. Consequently, the earliest studies in this field have emerged recently in the early 2000s, only two decades ago <span class="citation" data-cites="Coolen1998-qi Willerslev2003-cv">(<a href="#ref-Coolen1998-qi" role="doc-biblioref">Coolen and Overmann 1998</a>; <a href="#ref-Willerslev2003-cv" role="doc-biblioref">Willerslev et al. 2003</a>)</span>.</p>
<div id="fig-introductiontoedna-figure2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure2.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.2: Ancient eDNA embodies the intertwining challenges of temporal degradation and environmental complexity.
</figcaption>
</figure>
</div>
<p>Over the past decade, there has been a growing interest in researching ancient eDNA, particularly the DNA present in sedimentary archives <span class="citation" data-cites="Capo2023-og">(<a href="#ref-Capo2023-og" role="doc-biblioref">E. Capo, Barouillet, and Smol 2023</a>)</span>.Sedimentary DNA can be well-preserved in freshwater and marine sediments, palaeo-soils, caves and permafrost. This preservation is facilitated by oxygen-depleted and cold sediments in aquatic environments, as well as the sheltered conditions provided by caves, which can offer relatively stable temperature and humidity <span class="citation" data-cites="Capo2023-og">(<a href="#ref-Capo2023-og" role="doc-biblioref">E. Capo, Barouillet, and Smol 2023</a>)</span>. These preservation conditions allows for significant potential in reconstructing past environments and tracking ecological changes over time <span class="citation" data-cites="Kjaer2022-bf">(<a href="#ref-Kjaer2022-bf" role="doc-biblioref">Kjær et al. 2022</a>)</span>.</p>
</section>
<section id="what-can-we-find-in-sedimentary-dna-records" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="what-can-we-find-in-sedimentary-dna-records"><span class="header-section-number">5.1</span> What can we find in sedimentary DNA records?</h2>
<p>Sediments consist of inorganic and organic matter, including eDNA, that accumulate over time, creating temporally resolved archives <span class="citation" data-cites="Picard2024-gi">(<a href="#ref-Picard2024-gi" role="doc-biblioref">Picard et al. 2024</a>)</span>. The vertical stratigraphy of these sediments allows for the reconstruction of temporal transects through the sediment core. In aquatic environments, the sedimentary DNA record comprises a mixture of genetic materials from organisms that live within the lake (autochthonous material) and those from the surrounding area and its catchment (allochthonous material) <span class="citation" data-cites="Capo2022-qv">(<a href="#ref-Capo2022-qv" role="doc-biblioref">Eric Capo et al. 2022</a>)</span>. In terrestrial enviroments like caves, the sources of DNA that form the sedimentary record are different. Caves can act as natural traps where organisms repeatedly fall in and perish, or as places of refuge where organisms seek shelter <span class="citation" data-cites="Murchie2023">(<a href="#ref-Murchie2023" role="doc-biblioref">Murchie et al. 2023</a>)</span>. Consequently, the sedimentary records in caves can include genetic material from these organisms. Additionally, microorganisms can live within cave sediments, integrating their genetic material into the sedimentary DNA record, similar to aquatic environments.</p>
<p>The pool of DNA preserved within sediments can be categorised into two primary fractions <span class="citation" data-cites="Capo2022-qv">(<a href="#ref-Capo2022-qv" role="doc-biblioref">Eric Capo et al. 2022</a>)</span> (<a href="#fig-introductiontoedna-figure3" class="quarto-xref">Figure&nbsp;<span>5.3</span></a>):</p>
<ol type="1">
<li><strong>Ancient molecules:</strong> This category comprises DNA found within dead cells or as extracellular DNA, either in its free from or attached to particles</li>
<li><strong>DNA within viable or living cells:</strong> This includes DNA in cysts, spores, pollen, and eggs These cells may be actively growing within the sediments or in a dormant state, capable of reactivation under favourable environmental conditions</li>
</ol>
<p>Therefore, <span class="citation" data-cites="Capo2021">(<a href="#ref-Capo2021" role="doc-biblioref">Eric Capo et al. 2021</a>)</span> defined the two distinct fractions of DNA found in sediment based on their degradation states as:</p>
<ol type="1">
<li><strong>Sedimentary DNA (sedDNA):</strong> includes DNA that is relatively young and better preserved</li>
<li><strong>Sedimentary ancient DNA (sedaDNA):</strong> comprises older DNA that tends to be more poorly preserved</li>
</ol>
<div id="fig-introductiontoedna-figure3" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure3.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.3: Composition of the DNA pool preserved in aquatic sediments. Modified from <span class="citation" data-cites="Capo2022-qv">(<a href="#ref-Capo2022-qv" role="doc-biblioref">Eric Capo et al. 2022</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="taphonomic-processes-of-sedimentary-dna" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="taphonomic-processes-of-sedimentary-dna"><span class="header-section-number">5.2</span> Taphonomic processes of sedimentary DNA</h2>
<p>The reliability of DNA records is influenced by taphonomic processes <span class="citation" data-cites="Sand2024 Giguet-Covex2023-qr">(<a href="#ref-Sand2024" role="doc-biblioref">Sand et al. 2024</a>; <a href="#ref-Giguet-Covex2023-qr" role="doc-biblioref">Giguet-Covex et al. 2023</a>)</span>. Taphonomic processes refer to the various mechanisms that regulate the production, transfer, incorporation, and preservation of sedimentary DNA. In this section, we will focus on aquatic environments (<a href="#fig-introductiontoedna-figure4" class="quarto-xref">Figure&nbsp;<span>5.4</span></a>), but similar processes can also affect sediment records in caves.</p>
<section id="origin-of-edna-in-sediments" class="level3" data-number="5.2.1">
<h3 data-number="5.2.1" class="anchored" data-anchor-id="origin-of-edna-in-sediments"><span class="header-section-number">5.2.1</span> Origin of eDNA in sediments</h3>
<p>Several factors come into play in determining how DNA is incorporated and buried within the sediments <span class="citation" data-cites="Giguet-Covex2023-qr">(<a href="#ref-Giguet-Covex2023-qr" role="doc-biblioref">Giguet-Covex et al. 2023</a>)</span>. These include the abundance and spatial distributio of organisms, their ability to form cysts or spores, and their edibility <span class="citation" data-cites="Capo2021">(<a href="#ref-Capo2021" role="doc-biblioref">Eric Capo et al. 2021</a>)</span>. For terrestrial species living in the surrounding area of aquatic environments, understanding material and DNA transport from the original habitat is essential to ensure reliable interpretation of the genomic records. For example, more terrestrial plant DNA is expected to be transported to lakes with high soil erosion and well developed hydrological connections <span class="citation" data-cites="Garces2023">(<a href="#ref-Garces2023" role="doc-biblioref">Garcés-Pastor et al. 2023</a>)</span>.</p>
</section>
<section id="fate-of-edna-in-the-environment" class="level3" data-number="5.2.2">
<h3 data-number="5.2.2" class="anchored" data-anchor-id="fate-of-edna-in-the-environment"><span class="header-section-number">5.2.2</span> Fate of eDNA in the environment</h3>
<p>Upon release into the environment, DNA molecules can undergo various processes <span class="citation" data-cites="Torti2015">(<a href="#ref-Torti2015" role="doc-biblioref">Torti, Lever, and Jørgensen 2015</a>)</span>.</p>
<ul>
<li><strong>Biotic degradation</strong>: Extracellular DNA becomes susceptible to extracellular and cell-associated DNases, which are likely the most immediate cause of DNA degradation <span class="citation" data-cites="Torti2015">(<a href="#ref-Torti2015" role="doc-biblioref">Torti, Lever, and Jørgensen 2015</a>)</span>. These enzymes are widespread in most environments and catalyse the hydrolysis of the phosphodiester bond between the phosphate and the deoxyribose moieties of DNA.</li>
<li><strong>Natural transformation</strong>: Living cells can take up extracellular DNA molecules from the environment, even if the DNA is extensively damaged, and integrate them into their own genomes <span class="citation" data-cites="Overballe-Petersen2013-kw">(<a href="#ref-Overballe-Petersen2013-kw" role="doc-biblioref">Overballe-Petersen et al. 2013</a>)</span>.</li>
<li><strong>Abiotic decay</strong>: As you already saw in the ancient DNA lecture, DNA repair mechanisms cease to be active, and DNA will start accumulating lesions inflicted by various abiotic (physical and chemical) factors <span class="citation" data-cites="Orlando2021">(<a href="#ref-Orlando2021" role="doc-biblioref">Orlando et al. 2021</a>)</span>.</li>
<li><strong>Long-term preservation in the environment</strong>: Despite the occurrence of the previous processes, extracellular DNA can still persist in the environment. Optimal DNA preservation conditions in aquatic environments include systems with cooler temperatures, lakes with minimal bioturbation, anoxic conditions at the sediment-water interface, and freshwater and neutral to slightly alkaline lakes (pH 7-9) <span class="citation" data-cites="Jia2022-vv">(<a href="#ref-Jia2022-vv" role="doc-biblioref">Jia et al. 2022</a>)</span>. DNA can be protected from nucleases by binding to humic acids due to a negative surface charge, and therefore prolong DNA survival <span class="citation" data-cites="Stotzky2000">(<a href="#ref-Stotzky2000" role="doc-biblioref">Stotzky 2000</a>)</span>. DNA is also protected from degradation upon adsorption onto mineral grains in the sediment matrix, which reduces DNA accessibility to nucleases. This adsorption process is crucial for preserving DNA molecules in sediments for millennia.</li>
</ul>
<p>Let’s delve deeper into this preservation process.</p>
<div id="fig-introductiontoedna-figure4" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure4.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.4: Taphonomic processes affecting sedimentary DNA signals from terrestrial, water and sedimentary environments <span class="citation" data-cites="Giguet-Covex2023-qr">(<a href="#ref-Giguet-Covex2023-qr" role="doc-biblioref">Giguet-Covex et al. 2023</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="mineralogical-influence-on-preservation" class="level3" data-number="5.2.3">
<h3 data-number="5.2.3" class="anchored" data-anchor-id="mineralogical-influence-on-preservation"><span class="header-section-number">5.2.3</span> Mineralogical influence on preservation</h3>
<p>Different minerals have varying adsorption capacities, yet research has demonstrated that specific clays appear to be the most effective at preserving eDNA compared to other common soil and sediment silicates <span class="citation" data-cites="Lorenz1981 Kjaer2022-bf">(<a href="#ref-Lorenz1981" role="doc-biblioref">Lorenz, Aardema, and Krumbein 1981</a>; <a href="#ref-Kjaer2022-bf" role="doc-biblioref">Kjær et al. 2022</a>)</span>. For example, clay minerals such as montmorillonite (belonging to smectite clays) can absorb more than their own weight in DNA, because of their relatively large negatively charged surface area <span class="citation" data-cites="Pedersen2015">(<a href="#ref-Pedersen2015" role="doc-biblioref">Pedersen et al. 2015</a>)</span>. Compared with clays, sand has been found less effective in binding DNA, but adsorption increases with divalent cation concentrations (Ca2+ and Mg2+) <span class="citation" data-cites="Lorenz1987">(<a href="#ref-Lorenz1987" role="doc-biblioref">Lorenz and Wackernagel 1987</a>)</span>. We can use this information to target layers in sediment records that have high clay content as hotspots for ancient eDNA <span class="citation" data-cites="Kanbar2020-fk">(<a href="#ref-Kanbar2020-fk" role="doc-biblioref">Kanbar et al. 2020</a>)</span>.</p>
<p>How do they work? Let’s take smectite clay as an example (<a href="#fig-introductiontoedna-figure5" class="quarto-xref">Figure&nbsp;<span>5.5</span></a>). Smectite clays belongs to the 2:1 clay type, characterised by two tetrahedral layers surrounding an octahedral layer. In the case of smectites, the middle octahedral layer contains a variable cation content. When spaces become available, a charge is generated. This charge allows both organic and inorganic compounds to intercalate into this layer, effectively ‘nano-confining’ and shielding the DNA from the various factors that might otherwise cause its degradation over time, such as nucleases. These nucleases can themselves adsorb to sand and clay, potentially inhibiting their ability to hydrolyse extracellular DNA <span class="citation" data-cites="Beall2009">(<a href="#ref-Beall2009" role="doc-biblioref">Beall et al. 2009</a>)</span>.</p>
<div id="fig-introductiontoedna-figure5" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure5.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.5: DNA binding to smectite clay can protect it from nucleases and degradation.
</figcaption>
</figure>
</div>
</section>
</section>
<section id="dna-analysis-laboratory-processing" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="dna-analysis-laboratory-processing"><span class="header-section-number">5.3</span> DNA analysis: laboratory processing</h2>
<p>The reliability of DNA records is influenced not only by taphonomic processes but also methodological aspects. As with any aDNA study, all laboratory processes must be conducted in specialised aDNA facilities, adhering to strict aDNA protocols when handling samples, as covered in previous lectures.</p>
<p>Once the sediment sample has been collected, the first laboratory processing step involves DNA extraction <span class="citation" data-cites="Heintzman2023 Picard2024-gi">(<a href="#ref-Heintzman2023" role="doc-biblioref">Heintzman et al. 2023</a>; <a href="#ref-Picard2024-gi" role="doc-biblioref">Picard et al. 2024</a>)</span> (<a href="#fig-introductiontoedna-figure6" class="quarto-xref">Figure&nbsp;<span>5.6</span></a>). This process involves isolating and purifying DNA from various biological samples and separating it from other cellular or environmental components.</p>
<p>DNA extraction methods can vary depending on the DNA source and the specific downstream applications. In the field of eDNA and therefore, sedimentary DNA, the choice of DNA extraction method can significantly impact the resulting DNA composition. Some methods may fail to recover a significant number of DNA molecules from a sediment sample, affecting the accuracy of the reconstruction of the genetic record <span class="citation" data-cites="Capo2021">(<a href="#ref-Capo2021" role="doc-biblioref">Eric Capo et al. 2021</a>)</span>.</p>
<p>A wide range of methods have been used in sedimentary DNA research, ranging from commercial kits (e.g., Powersoil® and PowerMax® from Qiagen) to custom protocols optimised for recovering short DNA fragments, a characteristic feature of ancient DNA <span class="citation" data-cites="Epp2019-cz">(<a href="#ref-Epp2019-cz" role="doc-biblioref">Epp, Zimmermann, and Stoof-Leichsenring 2019</a>)</span>. Studies comparing different extraction methods have emphasised the importance of selecting methods optimised for different sediment types and taxonomic targets <span class="citation" data-cites="Slon2017 Armbrecht2020 Murchie2021 Perez2022">(e.g., <a href="#ref-Slon2017" role="doc-biblioref">Slon et al. 2017</a>; <a href="#ref-Armbrecht2020" role="doc-biblioref">Armbrecht et al. 2020</a>; <a href="#ref-Murchie2021" role="doc-biblioref">Murchie et al. 2021</a>; <a href="#ref-Perez2022" role="doc-biblioref">Pérez et al. 2022</a>)</span>.</p>
<p>After extracting eDNA, various analytical methods have been used to analyse the DNA <span class="citation" data-cites="Heintzman2023 Picard2024-gi">(<a href="#ref-Heintzman2023" role="doc-biblioref">Heintzman et al. 2023</a>; <a href="#ref-Picard2024-gi" role="doc-biblioref">Picard et al. 2024</a>)</span>. These methods encompass the following approaches:</p>
<ol type="1">
<li>Techniques designed to detect or quantify specific target organisms using endpoint PCR assays, quantitative real-time qPCR, and specialised applications like droplet digital ddPCR.</li>
<li>DNA metabarcoding methods that involve PCR amplification of marker loci, often referred to as “barcodes,” coupled with high-throughput sequencing.</li>
<li>Metagenomic strategies centred on untargeted shotgun sequencing of the entire pool of DNA retrieved from sediment.</li>
<li>Hybridization-based target enrichment methods for the recovery of DNA fragments of interest from the sediment metagenome. Hybridization capture involves the use of short RNA probes, often referred to as “baits,” which are designed to complement specific DNA sequences of interest, such as taxonomic marker genes. These baits can then be employed to selectively extract the target DNA fragments from the DNA extract for subsequent sequencing.</li>
</ol>
<p>Recovered DNA sequences are then taxonomically and/or functionally annotated using a suite of bioinformatic tools to answer palaeoecological questions.</p>
<div id="fig-introductiontoedna-figure6" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure6.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure6-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.6: Sedimentary DNA workflow suggested by <span class="citation" data-cites="Picard2024-gi">(<a href="#ref-Picard2024-gi" role="doc-biblioref">Picard et al. 2024</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="precautions-and-considerations-in-ancient-edna-research" class="level2" data-number="5.4">
<h2 data-number="5.4" class="anchored" data-anchor-id="precautions-and-considerations-in-ancient-edna-research"><span class="header-section-number">5.4</span> Precautions and considerations in ancient eDNA research</h2>
<p>While the application of ancient eDNA may appear promising, it is, in fact, quite challenging. There are numerous precautions and considerations that must be taken into account when conducting this type of research to accurately reconstruct past biological communities.</p>
<section id="contamination-control-in-field-and-laboratory-sampling" class="level3" data-number="5.4.1">
<h3 data-number="5.4.1" class="anchored" data-anchor-id="contamination-control-in-field-and-laboratory-sampling"><span class="header-section-number">5.4.1</span> Contamination control in field and laboratory sampling</h3>
<p>Contamination control measures begin at the sample collection stage <span class="citation" data-cites="Rawlence2014-ex">(<a href="#ref-Rawlence2014-ex" role="doc-biblioref">Rawlence et al. 2014</a>)</span>. Sampling must follow rigorous protocols, which include the use of protective gear such as gloves, facemasks, or full-body suits, and the utilization of sterilised equipment (<a href="#fig-introductiontoedna-figure7" class="quarto-xref">Figure&nbsp;<span>5.7</span></a>).</p>
<div id="fig-introductiontoedna-figure7" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure7.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure7-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.7: Sample collection following aDNA sterility protocols. Left: Field sampling, <span class="citation" data-cites="Rawlence2014-ex">(<a href="#ref-Rawlence2014-ex" role="doc-biblioref">Rawlence et al. 2014</a>)</span> Journal of Quaternary Science. Right: Laboratory sub-sampling: <span class="citation" data-cites="Crump2021">(<a href="#ref-Crump2021" role="doc-biblioref">Crump 2021</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="tracking-contaminants-in-field-sampling" class="level3" data-number="5.4.2">
<h3 data-number="5.4.2" class="anchored" data-anchor-id="tracking-contaminants-in-field-sampling"><span class="header-section-number">5.4.2</span> Tracking contaminants in field sampling</h3>
<p>In coring procedures, a visible chemical (e.g., perfluoromethyldecalin, PFMD) or DNA-based tracer can be applied to the sediment core to assess the extent to which potential external contamination infiltrates the extracted core (<a href="#fig-introductiontoedna-figure8" class="quarto-xref">Figure&nbsp;<span>5.8</span></a>). The tracer can be applied to the surface of the core before sampling, or it can be applied during core retrieval in the field. After applying the tracer, the core surface is removed, and the sample is collected <span class="citation" data-cites="Epp2019-cz">(<a href="#ref-Epp2019-cz" role="doc-biblioref">Epp, Zimmermann, and Stoof-Leichsenring 2019</a>)</span>. The presence of the tracer DNA on both the removed surface and the collected samples is then verified using PCR or qPCR <span class="citation" data-cites="Bang-Andreasen2017-zr">(<a href="#ref-Bang-Andreasen2017-zr" role="doc-biblioref">Bang-Andreasen et al. 2017</a>)</span>.</p>
<div id="fig-introductiontoedna-figure8" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure8.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.8: Example of a procedure for tracer quantification. (A) Outer sides of the permafrost core were covered with tracers. (B) The outer layers (dashed lines) were removed one by one using sterile tools and extreme caution to avoid cross contamination between layers. (C) Resulting inner core, smaller in diameter and free of tracers and potential microbial contamination. (D) The inner core was transversely split in two using sterile chisel. (E) From the newly exposed surface, soil was collected from the center of the inner core for quantification of tracer elements. (F) Quantification of tracers in all the collected layers and from the center of the permafrost core (quantitative PCR: qPCR). <span class="citation" data-cites="Bang-Andreasen2017-zr">(<a href="#ref-Bang-Andreasen2017-zr" role="doc-biblioref">Bang-Andreasen et al. 2017</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="taxonomic-assignment-and-community-reconstruction" class="level3" data-number="5.4.3">
<h3 data-number="5.4.3" class="anchored" data-anchor-id="taxonomic-assignment-and-community-reconstruction"><span class="header-section-number">5.4.3</span> Taxonomic assignment and community reconstruction</h3>
<p>One crucial factor when creating lists of species from DNA sequences found in sediment is having good reference databases. If these databases are not comprehensive and meticulously maintained, you may miss the identification of certain species in your samples. Some databases may even contain errors, like contaminants or wrongly identified sequences, which could lead to incorrect results <span class="citation" data-cites="Cribdon2020">(<a href="#ref-Cribdon2020" role="doc-biblioref">Cribdon et al. 2020</a>)</span>.</p>
<p>Another essential consideration is the analysis process, known as the bioinformatic pipeline. This includes the specific software tools and steps used to turn your DNA sequences into a list of species and their quantities. Setting the right parameters and thresholds in this pipeline is crucial for accurate results. So, to generate reliable species lists from sediment DNA data, you need both a solid reference database and a well-designed analysis pipeline <span class="citation" data-cites="Cribdon2020">(<a href="#ref-Cribdon2020" role="doc-biblioref">Cribdon et al. 2020</a>)</span>.</p>
<p>A good example, though from a modern metagenomic study, is the widely reported metagenomic analysis of the New York subway, which initially suggested that pathogens like <em>Yersinia pestis</em> and <em>Bacillus anthracis</em> were part of the “normal subway microbiome” <span class="citation" data-cites="Afshinnekoo2015-rq">(<a href="#ref-Afshinnekoo2015-rq" role="doc-biblioref">Afshinnekoo et al. 2015</a>)</span>. However, subsequent reanalysis using more appropriate methods did not detect these pathogens <span class="citation" data-cites="Ackelsberg2015-hv">(<a href="#ref-Ackelsberg2015-hv" role="doc-biblioref">Ackelsberg et al. 2015</a>)</span>. Commonly used software pipelines can produce results that lack prima facie validity (e.g., reporting widespread distribution of notorious endemic species) but appropriate use of inclusion and exclusion sets can avoid this issue.</p>
</section>
<section id="spatial-heterogeneity-of-sedadna" class="level3" data-number="5.4.4">
<h3 data-number="5.4.4" class="anchored" data-anchor-id="spatial-heterogeneity-of-sedadna"><span class="header-section-number">5.4.4</span> Spatial heterogeneity of sedaDNA</h3>
<p>Spatial distribution patterns of eDNA can be strongly heterogenous within sediments even at a microscale. For instance, eDNA from animals may originate not only from microscopic bone and teeth fragments but also from extracellular DNA present in bodily fluids or as a result of tissue decomposition. This distribution can be uneven across different sections of a sedimentary profile.</p>
<p>A study conducted by Massillani and colleagues in 2021 (<a href="#fig-introductiontoedna-figure9" class="quarto-xref">Figure&nbsp;<span>5.9</span></a>) illustrated the varying taxonomic compositions of mammalian mitochondrial DNA on a microscale within resin-impregnated archaeological sediment blocks collected from prehistoric sites in Europe, Asia, Africa, and North America (<span class="citation" data-cites="Massilani2022-zc">Massilani et al. (<a href="#ref-Massilani2022-zc" role="doc-biblioref">2022</a>)</span>). Their findings highlighted that DNA may be concentrated in specific sources such as bone and coprolites within the sediments.</p>
<p>Researchers should consider this factor when designing studies based on sedaDNA, particularly during the sample collection phase. Neglecting this heterogeneity could potentially lead to false negative results, especially for rare species. Modifying the number of sampling locations and increasing natural replicates per location in the study design could enhance the reliability of species detection.</p>
<div id="fig-introductiontoedna-figure9" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure9.jpg" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure9-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.9: Microstratigraphic distribution of mitochondrial DNA of ancient mammalian taxa in impregnated sediment. <span class="citation" data-cites="Massilani2022-zc">(<a href="#ref-Massilani2022-zc" role="doc-biblioref">Massilani et al. 2022</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="vertical-migration-of-dna" class="level3" data-number="5.4.5">
<h3 data-number="5.4.5" class="anchored" data-anchor-id="vertical-migration-of-dna"><span class="header-section-number">5.4.5</span> Vertical migration of DNA</h3>
<p>Another challenge associated with the sedaDNA approach, especially in non-frozen environments such as cave sediment records, is the potential risk of DNA moving vertically between different layers (DNA leaching) and contaminate lower (older) layers.</p>
<p>To investigate this issue, Haile and colleagues <span class="citation" data-cites="Haile2007-ch">(<a href="#ref-Haile2007-ch" role="doc-biblioref">2007</a>)</span> extracted aDNA from sediments up to 3300 years old at two cave sites on New Zealand’s North Island (<a href="#fig-introductiontoedna-figure10" class="quarto-xref">Figure&nbsp;<span>5.10</span></a>). These sites offered a valuable opportunity to study vertical DNA migration, as the sediment layers spanned both pre- and post-European periods. This allowed them to test for the presence of non-native species, such as sheep, in sediments deposited before European settlement, providing an indicator of DNA movement within the strata. Their findings revealed that sheep DNA was found alongside moa DNA, an indigenous species, in pre-European strata, suggesting that genetic material had migrated downward through the sediment (<a href="#fig-introductiontoedna-figure10" class="quarto-xref">Figure&nbsp;<span>5.10</span></a>). However, the amount of sheep DNA reduced as sediment age increased. This research highlighted the potential of sedimentary DNA as a valuable resource for understanding past environments but also suggested that considerable caution should be taken when interpreting DNA profiles from sediment records.</p>
<p>Downward infiltration of water can transport DNA through the porous, granular sediments of caves and potentially soil profiles <span class="citation" data-cites="Haile2007-ch Jenkins2012-ac">(<a href="#ref-Haile2007-ch" role="doc-biblioref">Haile et al. 2007</a>; <a href="#ref-Jenkins2012-ac" role="doc-biblioref">Jenkins et al. 2012</a>)</span>. However, other studies indicate that this phenomenon is less likely in lake sediments or permanently frozen sediments. Once lake sediments are compacted, vertical pore water movement is minimal, leading to the immobilization of multivalent metals and organic compounds in the sediment matrix <span class="citation" data-cites="Anderson-Carpenter2011-wa">(<a href="#ref-Anderson-Carpenter2011-wa" role="doc-biblioref">Anderson-Carpenter et al. 2011</a>)</span>.</p>
<div id="fig-introductiontoedna-figure10" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-introductiontoedna-figure10-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/introduction-to-edna/figure10.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-introductiontoedna-figure10-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5.10: Stratigraphic sections in two cave sites in the North Island of New Zealand. The amount of moa and sheep DNA is shown in the figure across the sediment records. <span class="citation" data-cites="Haile2007-ch">(<a href="#ref-Haile2007-ch" role="doc-biblioref">Haile et al. 2007</a>)</span>
</figcaption>
</figure>
</div>
</section>
<section id="potential-applications" class="level3" data-number="5.4.6">
<h3 data-number="5.4.6" class="anchored" data-anchor-id="potential-applications"><span class="header-section-number">5.4.6</span> Potential applications</h3>
<p>The use of ancient environmental DNA has enabled the reconstruction of palaeoenvironments, facilitating the detection of hominins and the study of biotic turnovers in ecosystems. This has also provided insights into ecosystem restoration targets.</p>
</section>
<section id="palaeoenvironmental-reconstruction" class="level3" data-number="5.4.7">
<h3 data-number="5.4.7" class="anchored" data-anchor-id="palaeoenvironmental-reconstruction"><span class="header-section-number">5.4.7</span> Palaeoenvironmental reconstruction</h3>
<p>Classical palaeoecology has explored the responses of biota to environmental change, but it has not fully captured the diversity of organisms, their co-occurrences, and interactions. SedaDNA offers an opportunity to achieve this, as <em>theoretically</em> allows the reconstruction of an ecosystem diversity across all domains of life.</p>
<p>A good example of this type of this application is the study published by <span class="citation" data-cites="Kjaer2022-bf">(<a href="#ref-Kjaer2022-bf" role="doc-biblioref">Kjær et al. 2022</a>)</span>. The Late Pliocene and Early Pleistocene epochs (3.6 to 0.8 million years ago) exhibited climates resembling future warming predictions. However, knowledge of the biological communities in the Arctic during this period remains limited due to the scarcity of fossils. In this study, the authors recovered ancient DNA from permafrost sediment samples in North Greenland, unveiling a diverse ecosystem that included an open boreal forest, various vegetation types, and a wide range of animals such as mastodons, reindeer, rodents, and geese. The presence of certain marine species (e.g., horseshow crab and green algae) suggests a warmer climate compared to the present. This use of ancient eDNA and traditional palaeoecological proxies (e.g., pollen) opened up new avenues for tracing the ecology and evolution of biological communities dating back <em>two million years</em> (a new record in the aDNA field!). The long-term preservation of this ancient eDNA is most likely attributed to its binding to mineral surfaces.</p>
</section>
</section>
<section id="insights-provided-by-microbial-shifts-over-time" class="level2" data-number="5.5">
<h2 data-number="5.5" class="anchored" data-anchor-id="insights-provided-by-microbial-shifts-over-time"><span class="header-section-number">5.5</span> Insights provided by microbial shifts over time</h2>
<p>Understanding how microbial diversity and function change over time and space is crucial for comprehending how ecosystems respond to global changes. Microorganisms, particularly prokaryotes, play essential roles in ecosystems by facilitating biogeochemical cycles and they are also the fastest responders to environmental changes. However, prokaryotes have been overlooked in long-term ancient eDNA studies due to the difficulty in determining whether their DNA signals originate from sediment-living organisms or from dead microorganisms deposited over time <span class="citation" data-cites="Vuillemin2023">(<a href="#ref-Vuillemin2023" role="doc-biblioref">Vuillemin et al. 2023</a>)</span>. Recent advances in molecular genetic methods have enabled the identification of past microbial communities, including prokaryotes, and the tracing of their evolution and adaptation to environmental changes <span class="citation" data-cites="Capo2021">(<a href="#ref-Capo2021" role="doc-biblioref">Eric Capo et al. 2021</a>)</span>. This information is crucial for predicting future ecosystem shifts. The following studies showcase the type of information that can be obtained from microbial ancient eDNA and their interpretations.</p>
<section id="phytoplankton-changes-in-response-to-climate-in-polar-ecosystems" class="level3" data-number="5.5.1">
<h3 data-number="5.5.1" class="anchored" data-anchor-id="phytoplankton-changes-in-response-to-climate-in-polar-ecosystems"><span class="header-section-number">5.5.1</span> Phytoplankton changes in response to climate in Polar ecosystems</h3>
<p>Polar ecosystems are highly vulnerable to the ongoing climate change. The melting ice sheets and changes in oceanography in marine ecosystems are impacting all levels of the food web, especially primary producers. Changes in sea temperature and atmospheric CO2 levels are linked to shifts in the composition and size structure of phytoplankton communities, making them valuable palaeoindicators of climate change and ecosystem responses. However, some diatoms with heavily silicified structures preserve better than others, potentially skewing fossil analyses. This highlights the importance of sedaDNA techniques in complementing microfossil and other palaeo-proxy analyses for detailed investigations of periods undergoing environmental change, such as glacial-to-interglacial transitions.</p>
<p>Different studies have utilised sedimentary DNA from the Arctic and Antarctic environments to describe changes in phytoplankton communities, aiming to better understand the past and present responses of marine ecosystems to climate change. For example, a study by Buchwald and colleagues focused on the subarctic western Bering Sea, analysing metagenomic shotgun and diatom <em>rbcL</em> amplicon sequencing data over glacial–interglacial cycles, including the Eemian interglacial period <span class="citation" data-cites="Buchwald2024-wd">(<a href="#ref-Buchwald2024-wd" role="doc-biblioref">Buchwald et al. 2024</a>)</span>. They identified distinct plankton communities during these periods: the Holocene was dominated by picosized cyanobacteria and bacteria-feeding protists, the Eemian featured eukaryotic picosized chlorophytes and Triparmaceae, and the glacial period had microsized phototrophic protists, including sea ice-associated diatoms and diatom-feeding crustaceous zooplankton. This long-term record revealed a decrease in phytoplankton cell size with rising temperatures, similar to current changes in the warming Bering Sea. These shifts suggest that future warming may increase productivity but reduce carbon export, weakening the Bering Sea’s role as a carbon sink.</p>
<p>On the other hand, Armbrecht and colleagues examined a sedimentary DNA record spanning ~1 million years from the Scotia Sea in Antarctica to understand changes in phytoplankton communities over time <span class="citation" data-cites="Armbrecht2022-dc">(<a href="#ref-Armbrecht2022-dc" role="doc-biblioref">Armbrecht et al. 2022</a>)</span> . They found records of diatom and chlorophyte sedaDNA dating back ~540 thousand years, reconstructing the oldest marine eukaryote sedaDNA record to date. When analysing the results, the authors concluded that warm phases were associated with high diatom abundance and a significant shift from sea-ice to open-ocean species around 14.5 thousand years ago, following Meltwater Pulse 1A.</p>
<p>Together, these studies illustrate the profound impact of climate change on marine ecosystems in both the Arctic and Antarctic regions and demonstrate the potential of sedaDNA tools to study long-term marine ecosystem shifts and palaeo-productivity across glacial-interglacial cycles.</p>
</section>
<section id="environmental-microbiome-associated-with-vegetational-changes" class="level3" data-number="5.5.2">
<h3 data-number="5.5.2" class="anchored" data-anchor-id="environmental-microbiome-associated-with-vegetational-changes"><span class="header-section-number">5.5.2</span> Environmental microbiome associated with vegetational changes</h3>
<p>In this study, researchers used shotgun metagenomics to analyse ancient eDNA from lake sediments in northern Siberia, aiming to understand ecosystem changes over the last 6700 years. They combined taxonomic and functional gene analysis to gain insights into how the genetic records of biological communities preserved in the sediment evolved over time. Additionally, they described changes in eco-physiological adaptations and ecosystem functions of microbial and vegetational communities in response to past climate variations.</p>
<p>The researchers found that 6700 years ago, the area was an open boreal forest, which gradually transitioned into tundra. The changes in the plant community reconstructed using sedimentary DNA matched perfectly with those described in previous studies using other palaeoecological proxies such as pollen, DNA metabarcoding, and hybridization capture <span class="citation" data-cites="Schulte2021">(<a href="#ref-Schulte2021" role="doc-biblioref">Schulte et al. 2021</a>)</span>, further validating their observations. These shifts in palaeovegetation significantly impacted the environmental microbiome, with clear implications for ecosystem functioning. For example, there was an enrichment of certain bacteria and archaea that thrive in tundra conditions, particularly bacterial and archaeal ammonia oxidisers like Nitrospira, Nitrosopumilus, and Ca. Nitrosocosmicus.</p>
<p>This study demonstrated that microbial taxa, functions, and eco-physiological traits are valuable indicators of environmental conditions, highlighting their use as proxies in palaeoecosystem reconstruction.</p>
</section>
</section>
<section id="summary" class="level2" data-number="5.6">
<h2 data-number="5.6" class="anchored" data-anchor-id="summary"><span class="header-section-number">5.6</span> Summary</h2>
<ol type="1">
<li><p>Environmental DNA (eDNA) refers to genetic material shed by organisms into their surroundings, including water, soil, and sediments. It enables species identification without direct organism sampling, offering valuable insights into both current and historical ecosystems.</p></li>
<li><p>Environmental samples are inherently complex, containing a mix of genetic material from multiple species. This complexity is further amplified in ancient eDNA, where modern and degraded DNA are combined, making sample recovery and analysis more challenging.</p></li>
<li><p>The preservation of eDNA depends largely on environmental conditions. Cold, anoxic environments and clay-rich sediments help protect DNA for millennia. Taphonomic processes play a key role in the reliability of the DNA records.</p></li>
<li><p>Analysing ancient eDNA requires stringent laboratory protocols, with contamination control being critical during both field sampling and lab work to ensure the accuracy and integrity of the research.</p></li>
<li><p>Ancient eDNA provides a powerful tool for reconstructing past environments and tracking long-term ecological changes, such as climate-driven shifts in microbial, plant, and animal communities. This information is crucial for understanding ecological responses and guiding conservation efforts.</p></li>
</ol>
</section>
<section id="references" class="level2" data-number="5.7">
<h2 data-number="5.7" class="anchored" data-anchor-id="references"><span class="header-section-number">5.7</span> References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Ackelsberg2015-hv" class="csl-entry" role="listitem">
Ackelsberg, Joel, Jennifer Rakeman, Scott Hughes, Jeannine Petersen, Paul Mead, Martin Schriefer, Luke Kingry, Alex Hoffmaster, and Jay E Gee. 2015. <span>“Lack of Evidence for Plague or Anthrax on the New York City Subway.”</span> <em>Cell Systems</em> 1 (1): 4–5. <a href="https://doi.org/10.1016/j.cels.2015.07.008">https://doi.org/10.1016/j.cels.2015.07.008</a>.
</div>
<div id="ref-Afshinnekoo2015-rq" class="csl-entry" role="listitem">
Afshinnekoo, Ebrahim, Cem Meydan, Shanin Chowdhury, Dyala Jaroudi, Collin Boyer, Nick Bernstein, Julia M Maritz, et al. 2015. <span>“Geospatial Resolution of Human and Bacterial Diversity with City-Scale Metagenomics.”</span> <em>Cell Systems</em> 1 (1): 72–87. <a href="https://doi.org/10.1016/j.cels.2015.01.001">https://doi.org/10.1016/j.cels.2015.01.001</a>.
</div>
<div id="ref-Anderson-Carpenter2011-wa" class="csl-entry" role="listitem">
Anderson-Carpenter, Lynn L, Jason S McLachlan, Stephen T Jackson, Melanie Kuch, Candice Y Lumibao, and Hendrik N Poinar. 2011. <span>“Ancient <span>DNA</span> from Lake Sediments: Bridging the Gap Between Paleoecology and Genetics.”</span> <em>BMC Evolutionary Biology</em> 11 (1): 30. <a href="https://doi.org/10.1186/1471-2148-11-30">https://doi.org/10.1186/1471-2148-11-30</a>.
</div>
<div id="ref-Armbrecht2020" class="csl-entry" role="listitem">
Armbrecht, Linda, Salvador Herrando-Pérez, Raphael Eisenhofer, Gustaaf M Hallegraeff, Christopher J S Bolch, and Alan Cooper. 2020. <span>“An Optimized Method for the Extraction of Ancient Eukaryote <span>DNA</span> from Marine Sediments.”</span> <em>Molecular Ecology Resources</em> 20 (4): 906–19. <a href="https://doi.org/10.1111/1755-0998.13162">https://doi.org/10.1111/1755-0998.13162</a>.
</div>
<div id="ref-Armbrecht2022-dc" class="csl-entry" role="listitem">
Armbrecht, Linda, Michael E Weber, Maureen E Raymo, Victoria L Peck, Trevor Williams, Jonathan Warnock, Yuji Kato, et al. 2022. <span>“Ancient Marine Sediment <span>DNA</span> Reveals Diatom Transition in Antarctica.”</span> <em>Nature Communications</em> 13 (1): 5787. <a href="https://doi.org/10.1038/s41467-022-33494-4">https://doi.org/10.1038/s41467-022-33494-4</a>.
</div>
<div id="ref-Bang-Andreasen2017-zr" class="csl-entry" role="listitem">
Bang-Andreasen, Toke, Morten Schostag, Anders Priemé, Bo Elberling, and Carsten S Jacobsen. 2017. <span>“Potential Microbial Contamination During Sampling of Permafrost Soil Assessed by Tracers.”</span> <em>Scientific Reports</em> 7 (1): 1–11. <a href="https://doi.org/10.1038/srep43338">https://doi.org/10.1038/srep43338</a>.
</div>
<div id="ref-Beall2009" class="csl-entry" role="listitem">
Beall, Gary W, Drew S Sowersby, Rachel D Roberts, Michael H Robson, and L Kevin Lewis. 2009. <span>“Analysis of Oligonucleotide <span>DNA</span> Binding and Sedimentation Properties of Montmorillonite Clay Using Ultraviolet Light Spectroscopy.”</span> <em>Biomacromolecules</em> 10 (1): 105–12. <a href="https://doi.org/10.1021/bm800970v">https://doi.org/10.1021/bm800970v</a>.
</div>
<div id="ref-Buchwald2024-wd" class="csl-entry" role="listitem">
Buchwald, Stella Z, Ulrike Herzschuh, Dirk Nürnberg, Lars Harms, and Kathleen R Stoof-Leichsenring. 2024. <span>“Plankton Community Changes During the Last 124 000 Years in the Subarctic Bering Sea Derived from Sedimentary Ancient <span>DNA</span>.”</span> <em>The ISME Journal</em> 18 (1): wrad006. <a href="https://doi.org/10.1093/ismejo/wrad006">https://doi.org/10.1093/ismejo/wrad006</a>.
</div>
<div id="ref-Capo2023-og" class="csl-entry" role="listitem">
Capo, E, C Barouillet, and J P Smol, eds. 2023. <em>Tracking Environmental Change Using Lake Sediments</em>. Springer International Publishing. <a href="https://doi.org/10.1007/978-3-031-43799-1">https://doi.org/10.1007/978-3-031-43799-1</a>.
</div>
<div id="ref-Capo2021" class="csl-entry" role="listitem">
Capo, Eric, Charline Giguet-Covex, Alexandra Rouillard, Kevin Nota, Peter D Heintzman, Aurèle Vuillemin, Daniel Ariztegui, et al. 2021. <span>“Lake Sedimentary <span>DNA</span> Research on Past Terrestrial and Aquatic Biodiversity: Overview and Recommendations.”</span> <em>Quaternary</em> 4 (1): 6. <a href="https://doi.org/10.3390/quat4010006">https://doi.org/10.3390/quat4010006</a>.
</div>
<div id="ref-Capo2022-qv" class="csl-entry" role="listitem">
Capo, Eric, Marie-Eve Monchamp, Marco J L Coolen, Isabelle Domaizon, Linda Armbrecht, and Stefan Bertilsson. 2022. <span>“Environmental Paleomicrobiology: Using <span>DNA</span> Preserved in Aquatic Sediments to Its Full Potential.”</span> <em>Environmental Microbiology</em> 24 (5): 2201–9. <a href="https://doi.org/10.1111/1462-2920.15913">https://doi.org/10.1111/1462-2920.15913</a>.
</div>
<div id="ref-Coolen1998-qi" class="csl-entry" role="listitem">
Coolen, Marco J L, and Jörg Overmann. 1998. <span>“Analysis of Subfossil Molecular Remains of Purple Sulfur Bacteria in a Lake Sediment.”</span> <em>Applied and Environmental Microbiology</em>, November. <a href="https://doi.org/10.1128/AEM.64.11.4513-4521.1998">https://doi.org/10.1128/AEM.64.11.4513-4521.1998</a>.
</div>
<div id="ref-Cribdon2020" class="csl-entry" role="listitem">
Cribdon, Becky, Roselyn Ware, Oliver Smith, Vincent Gaffney, and Robin G Allaby. 2020. <span>“<span>PIA</span>: More Accurate Taxonomic Assignment of Metagenomic Data Demonstrated on <span class="nocase">sedaDNA</span> from the North Sea.”</span> <em>Frontiers in Ecology and Evolution</em> 8: 84. <a href="https://doi.org/10.3389/fevo.2020.00084">https://doi.org/10.3389/fevo.2020.00084</a>.
</div>
<div id="ref-Crump2021" class="csl-entry" role="listitem">
Crump, Sarah E. 2021. <span>“Sedimentary Ancient <span>DNA</span> as a Tool in Paleoecology.”</span> <em>Nature Reviews. Earth &amp; Environment</em> 2 (4): 229–29. <a href="https://doi.org/10.1038/s43017-021-00158-8">https://doi.org/10.1038/s43017-021-00158-8</a>.
</div>
<div id="ref-Epp2019-cz" class="csl-entry" role="listitem">
Epp, Laura S, Heike H Zimmermann, and Kathleen R Stoof-Leichsenring. 2019. <span>“Sampling and Extraction of Ancient <span>DNA</span> from Sediments.”</span> <em>Ancient DNA</em>, 31–44. <a href="https://doi.org/10.1007/978-1-4939-9176-1\_5">https://doi.org/10.1007/978-1-4939-9176-1\_5</a>.
</div>
<div id="ref-Garces2023" class="csl-entry" role="listitem">
Garcés-Pastor, Sandra, Kevin Nota, Dilli P Rijal, Sisi Liu, Weihan Jia, Maria Leunda, Christoph Schwörer, Sarah E Crump, Laura Parducci, and Inger G Alsos. 2023. <span>“Terrestrial Plant <span>DNA</span> from Lake Sediments.”</span> <em>Tracking Environmental Change Using Lake Sediments</em>, 275–98. <a href="https://doi.org/10.1007/978-3-031-43799-1\_10">https://doi.org/10.1007/978-3-031-43799-1\_10</a>.
</div>
<div id="ref-Giguet-Covex2023-qr" class="csl-entry" role="listitem">
Giguet-Covex, Charline, Stanislav Jelavić, Anthony Foucher, Marina A Morlock, Susanna A Wood, Femke Augustijns, Isabelle Domaizon, Ludovic Gielly, and Eric Capo. 2023. <span>“The Sources and Fates of Lake Sedimentary <span>DNA</span>.”</span> <em>Tracking Environmental Change Using Lake Sediments</em>, 9–52. <a href="https://doi.org/10.1007/978-3-031-43799-1\_2">https://doi.org/10.1007/978-3-031-43799-1\_2</a>.
</div>
<div id="ref-Haile2007-ch" class="csl-entry" role="listitem">
Haile, James, Richard Holdaway, Karen Oliver, Michael Bunce, M Thomas P Gilbert, Rasmus Nielsen, Kasper Munch, Simon Y W Ho, Beth Shapiro, and Eske Willerslev. 2007. <span>“Ancient <span>DNA</span> Chronology Within Sediment Deposits: Are Paleobiological Reconstructions Possible and Is <span>DNA</span> Leaching a Factor?”</span> <em>Molecular Biology and Evolution</em> 24 (4): 982–89. <a href="https://doi.org/10.1093/molbev/msm016">https://doi.org/10.1093/molbev/msm016</a>.
</div>
<div id="ref-Heintzman2023" class="csl-entry" role="listitem">
Heintzman, Peter D, Kevin Nota, Alexandra Rouillard, Youri Lammers, Tyler J Murchie, Linda Armbrecht, Sandra Garcés-Pastor, and Benjamin Vernot. 2023. <span>“The Sedimentary Ancient <span>DNA</span> Workflow.”</span> <em>Tracking Environmental Change Using Lake Sediments</em>, 53–84. <a href="https://doi.org/10.1007/978-3-031-43799-1\_3">https://doi.org/10.1007/978-3-031-43799-1\_3</a>.
</div>
<div id="ref-Jenkins2012-ac" class="csl-entry" role="listitem">
Jenkins, Dennis L, Loren G Davis, Thomas W Stafford Jr, Paula F Campos, Bryan Hockett, George T Jones, Linda Scott Cummings, et al. 2012. <span>“Clovis Age Western Stemmed Projectile Points and Human Coprolites at the Paisley Caves.”</span> <em>Science</em> 337 (6091): 223–28. <a href="https://doi.org/10.1126/science.1218443">https://doi.org/10.1126/science.1218443</a>.
</div>
<div id="ref-Jia2022-vv" class="csl-entry" role="listitem">
Jia, Weihan, Xingqi Liu, Kathleen R Stoof-Leichsenring, Sisi Liu, Kai Li, and Ulrike Herzschuh. 2022. <span>“Preservation of Sedimentary Plant <span>DNA</span> Is Related to Lake Water Chemistry.”</span> <em>Environmental DNA</em> 4 (2): 425–39. <a href="https://doi.org/10.1002/edn3.259">https://doi.org/10.1002/edn3.259</a>.
</div>
<div id="ref-Kanbar2020-fk" class="csl-entry" role="listitem">
Kanbar, Hussein Jaafar, Fredrik Olajos, Göran Englund, and Michael Holmboe. 2020. <span>“Geochemical Identification of Potential <span>DNA</span>-Hotspots and <span>DNA</span>-Infrared Fingerprints in Lake Sediments.”</span> <em>Applied Geochemistry: Journal of the International Association of Geochemistry and Cosmochemistry</em> 122 (104728): 104728. <a href="https://doi.org/10.1016/j.apgeochem.2020.104728">https://doi.org/10.1016/j.apgeochem.2020.104728</a>.
</div>
<div id="ref-Kjaer2022-bf" class="csl-entry" role="listitem">
Kjær, Kurt H, Mikkel Winther Pedersen, Bianca De Sanctis, Binia De Cahsan, Thorfinn S Korneliussen, Christian S Michelsen, Karina K Sand, et al. 2022. <span>“A 2-Million-Year-Old Ecosystem in Greenland Uncovered by Environmental <span>DNA</span>.”</span> <em>Nature</em> 612 (7939): 283–91. <a href="https://doi.org/10.1038/s41586-022-05453-y">https://doi.org/10.1038/s41586-022-05453-y</a>.
</div>
<div id="ref-Lorenz1981" class="csl-entry" role="listitem">
Lorenz, M G, B W Aardema, and W E Krumbein. 1981. <span>“Interaction of Marine Sediment with <span>DNA</span> and <span>DNA</span> Availability to Nucleases.”</span> <em>Marine Biology</em> 64 (2): 225–30. <a href="https://doi.org/10.1007/bf00397113">https://doi.org/10.1007/bf00397113</a>.
</div>
<div id="ref-Lorenz1987" class="csl-entry" role="listitem">
Lorenz, M G, and W Wackernagel. 1987. <span>“Adsorption of <span>DNA</span> to Sand and Variable Degradation Rates of Adsorbed <span>DNA</span>.”</span> <em>Applied and Environmental Microbiology</em> 53 (12): 2948–52. <a href="https://doi.org/10.1128/aem.53.12.2948-2952.1987">https://doi.org/10.1128/aem.53.12.2948-2952.1987</a>.
</div>
<div id="ref-Massilani2022-zc" class="csl-entry" role="listitem">
Massilani, Diyendo, Mike W Morley, Susan M Mentzer, Vera Aldeias, Benjamin Vernot, Christopher Miller, Mareike Stahlschmidt, et al. 2022. <span>“Microstratigraphic Preservation of Ancient Faunal and Hominin <span>DNA</span> in Pleistocene Cave Sediments.”</span> <em>Proceedings of the National Academy of Sciences of the United States of America</em> 119 (1). <a href="https://doi.org/10.1073/pnas.2113666118">https://doi.org/10.1073/pnas.2113666118</a>.
</div>
<div id="ref-Murchie2023" class="csl-entry" role="listitem">
Murchie, Tyler J, Charline Giguet-Covex, Peter D Heintzman, Viviane Slon, and Yucheng Wang. 2023. <span>“Terrestrial Fauna and Hominin <span>DNA</span> from Sedimentary Archives: Volume 6: Sedimentary <span>DNA</span>.”</span> In <em>Tracking Environmental Change Using Lake Sediments</em>, edited by Eric Capo, Cécilia Barouillet, and John P Smol, 21:299–378. Developments in Paleoenvironmental Research. Cham: Springer International Publishing. <a href="https://doi.org/10.1007/978-3-031-43799-1\_11">https://doi.org/10.1007/978-3-031-43799-1\_11</a>.
</div>
<div id="ref-Murchie2021" class="csl-entry" role="listitem">
Murchie, Tyler J, Melanie Kuch, Ana T Duggan, Marissa L Ledger, Kévin Roche, Jennifer Klunk, Emil Karpinski, et al. 2021. <span>“Optimizing Extraction and Targeted Capture of Ancient Environmental <span>DNA</span> for Reconstructing Past Environments Using the <span>PalaeoChip</span> Arctic-1.0 Bait-Set.”</span> <em>Quaternary Research</em> 99 (January): 305–28. <a href="https://doi.org/10.1017/qua.2020.59">https://doi.org/10.1017/qua.2020.59</a>.
</div>
<div id="ref-Orlando2021" class="csl-entry" role="listitem">
Orlando, Ludovic, Robin Allaby, Pontus Skoglund, Clio Der Sarkissian, Philipp W Stockhammer, María C Ávila-Arcos, Qiaomei Fu, et al. 2021. <span>“Ancient <span>DNA</span> Analysis.”</span> <em>Nature Reviews Methods Primers</em> 1 (1): 1–26. <a href="https://doi.org/10.1038/s43586-020-00011-0">https://doi.org/10.1038/s43586-020-00011-0</a>.
</div>
<div id="ref-Overballe-Petersen2013-kw" class="csl-entry" role="listitem">
Overballe-Petersen, Søren, Klaus Harms, Ludovic A A Orlando, J Victor Moreno Mayar, Simon Rasmussen, Tais W Dahl, Minik T Rosing, et al. 2013. <span>“Bacterial Natural Transformation by Highly Fragmented and Damaged <span>DNA</span>.”</span> <em>Proceedings of the National Academy of Sciences of the United States of America</em> 110 (49): 19860–65. <a href="https://doi.org/10.1073/pnas.1315278110">https://doi.org/10.1073/pnas.1315278110</a>.
</div>
<div id="ref-Pedersen2015" class="csl-entry" role="listitem">
Pedersen, Mikkel Winther, Søren Overballe-Petersen, Luca Ermini, Clio Der Sarkissian, James Haile, Micaela Hellstrom, Johan Spens, et al. 2015. <span>“Ancient and Modern Environmental <span>DNA</span>.”</span> <em>Philosophical Transactions of the Royal Society of London. Series B, Biological Sciences</em> 370 (1660): 20130383. <a href="https://doi.org/10.1098/rstb.2013.0383">https://doi.org/10.1098/rstb.2013.0383</a>.
</div>
<div id="ref-Perez2022" class="csl-entry" role="listitem">
Pérez, Vilma, Yichen Liu, Martha B Hengst, and Laura S Weyrich. 2022. <span>“A Case Study for the Recovery of Authentic Microbial Ancient <span>DNA</span> from Soil Samples.”</span> <em>Microorganisms</em> 10 (8): 1623. <a href="https://doi.org/10.3390/microorganisms10081623">https://doi.org/10.3390/microorganisms10081623</a>.
</div>
<div id="ref-Picard2024-gi" class="csl-entry" role="listitem">
Picard, Maïlys, Jordan Von Eggers, Katie A Brasell, Dongna Yan, Jonatan Klaminder, Inger G Alsos, Cécilia Barouillet, et al. 2024. <span>“Using <span>DNA</span> Archived in Lake Sediments to Reconstruct Past Ecosystems.”</span> In <em>Reference Module in Earth Systems and Environmental Sciences</em>. Elsevier. <a href="https://doi.org/10.1016/b978-0-323-99931-1.00171-9">https://doi.org/10.1016/b978-0-323-99931-1.00171-9</a>.
</div>
<div id="ref-Rawlence2014-ex" class="csl-entry" role="listitem">
Rawlence, Nicolas J, David J Lowe, Jamie R Wood, Jennifer M Young, G Jock Churchman, Yu-Tuan Huang, and Alan Cooper. 2014. <span>“Using Palaeoenvironmental <span>DNA</span> to Reconstruct Past Environments: Progress and Prospects: <span>USING</span> <span>PALAEOENVIRONMENTAL</span> <span>DNA</span> <span>TO</span> <span>RECONSTRUCT</span> <span>PAST</span> <span>ENVIRONMENTS</span>.”</span> <em>Journal of Quaternary Science</em> 29 (7): 610–26. <a href="https://doi.org/10.1002/jqs.2740">https://doi.org/10.1002/jqs.2740</a>.
</div>
<div id="ref-Sand2024" class="csl-entry" role="listitem">
Sand, K K, S Jelavić, K H Kjær, and A Prohaska. 2024. <span>“Importance of <span class="nocase">eDNA</span> Taphonomy and Sediment Provenance for Robust Ecological Inference: Insights from Interfacial Geochemistry.”</span> <em>Environmental DNA (Hoboken, N.J.)</em> 6 (2). <a href="https://doi.org/10.1002/edn3.519">https://doi.org/10.1002/edn3.519</a>.
</div>
<div id="ref-Schulte2021" class="csl-entry" role="listitem">
Schulte, Luise, Nadine Bernhardt, Kathleen Stoof-Leichsenring, Heike H Zimmermann, Luidmila A Pestryakova, Laura S Epp, and Ulrike Herzschuh. 2021. <span>“Hybridization Capture of Larch (Larix Mill.) Chloroplast Genomes from Sedimentary Ancient <span>DNA</span> Reveals Past Changes of Siberian Forest.”</span> <em>Molecular Ecology Resources</em> 21 (3): 801–15. <a href="https://doi.org/10.1111/1755-0998.13311">https://doi.org/10.1111/1755-0998.13311</a>.
</div>
<div id="ref-Slon2017" class="csl-entry" role="listitem">
Slon, Viviane, Charlotte Hopfe, Clemens L Weiß, Fabrizio Mafessoni, Marco de la Rasilla, Carles Lalueza-Fox, Antonio Rosas, et al. 2017. <span>“Neandertal and Denisovan <span>DNA</span> from Pleistocene Sediments.”</span> <em>Science</em> 356 (6338): 605–8. <a href="https://doi.org/10.1126/science.aam9695">https://doi.org/10.1126/science.aam9695</a>.
</div>
<div id="ref-Stotzky2000" class="csl-entry" role="listitem">
Stotzky, G. 2000. <span>“Persistence and Biological Activity in Soil of Insecticidal Proteins from Bacillus Thuringiensis and of Bacterial <span>DNA</span> Bound on Clays and Humic Acids.”</span> <em>Journal of Environmental Quality</em> 29 (3): 691–705. <a href="https://doi.org/10.2134/jeq2000.00472425002900030003x">https://doi.org/10.2134/jeq2000.00472425002900030003x</a>.
</div>
<div id="ref-Taberlet2018-xa" class="csl-entry" role="listitem">
Taberlet, Pierre, Aurelie Bonin, Lucie Zinger, and Eric Coissac. 2018. <em>Environmental <span>DNA</span>: For Biodiversity Research and Monitoring</em>. London, England: Oxford University Press. <a href="https://academic.oup.com/book/32663">https://academic.oup.com/book/32663</a>.
</div>
<div id="ref-Torti2015" class="csl-entry" role="listitem">
Torti, Andrea, Mark Alexander Lever, and Bo Barker Jørgensen. 2015. <span>“Origin, Dynamics, and Implications of Extracellular <span>DNA</span> Pools in Marine Sediments.”</span> <em>Marine Genomics</em> 24 Pt 3 (December): 185–96. <a href="https://doi.org/10.1016/j.margen.2015.08.007">https://doi.org/10.1016/j.margen.2015.08.007</a>.
</div>
<div id="ref-Vuillemin2023" class="csl-entry" role="listitem">
Vuillemin, Aurèle, Marco J L Coolen, Jens Kallmeyer, Susanne Liebner, and Stefan Bertilsson. 2023. <span>“Bacterial and Archaeal <span>DNA</span> from Lake Sediments.”</span> <em>Tracking Environmental Change Using Lake Sediments</em>, 85–151. <a href="https://doi.org/10.1007/978-3-031-43799-1\_4">https://doi.org/10.1007/978-3-031-43799-1\_4</a>.
</div>
<div id="ref-Willerslev2003-cv" class="csl-entry" role="listitem">
Willerslev, Eske, Anders J Hansen, Jonas Binladen, Tina B Brand, M Thomas P Gilbert, Beth Shapiro, Michael Bunce, Carsten Wiuf, David A Gilichinsky, and Alan Cooper. 2003. <span>“Diverse Plant and Animal Genetic Records from Holocene and Pleistocene Sediments.”</span> <em>Science (New York, N.Y.)</em> 300 (5620): 791–95. <a href="https://doi.org/10.1126/science.1084114">https://doi.org/10.1126/science.1084114</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/spaam-community\.github\.org\/intro-to-ancient-metagenomics-book");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./introduction-to-microbial-genomics.html" class="pagination-link" aria-label="Introduction to Microbial Genomics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./section-useful-skills.html" class="pagination-link" aria-label="Useful Skills">
        <span class="nav-page-text">Useful Skills</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2024 SPAAM Community &amp; Authors with ❤️. Available under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>