<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Clemens Schmid">

<title>Introduction to Ancient Metagenomics - 7&nbsp; Introduction to R and the Tidyverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./python-pandas.html" rel="next">
<link href="./bare-bones-bash.html" rel="prev">
<link href="./favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<meta property="og:title" content="Introduction to Ancient Metagenomics - 7&nbsp; Introduction to R and the Tidyverse">
<meta property="og:description" content="">
<meta property="og:site_name" content="Introduction to Ancient Metagenomics">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-useful-skills.html">Useful Skills</a></li><li class="breadcrumb-item"><a href="./r-tidyverse.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-evolutionary-biology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Evolutionary Biology</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication-decontamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Authentication and Decontamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Introduction to Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancient-metagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-template.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Chapter Template</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-working-environment" id="toc-the-working-environment" class="nav-link active" data-scroll-target="#the-working-environment"><span class="header-section-number">7.1</span> The working environment</a>
  <ul class="collapse">
  <li><a href="#r-rstudio-the-tidyverse-and-penguins" id="toc-r-rstudio-the-tidyverse-and-penguins" class="nav-link" data-scroll-target="#r-rstudio-the-tidyverse-and-penguins"><span class="header-section-number">7.1.1</span> R, RStudio, the tidyverse and penguins</a></li>
  </ul></li>
  <li><a href="#loading-data-into-tibbles" id="toc-loading-data-into-tibbles" class="nav-link" data-scroll-target="#loading-data-into-tibbles"><span class="header-section-number">7.2</span> Loading data into tibbles</a>
  <ul class="collapse">
  <li><a href="#reading-data-with-readr" id="toc-reading-data-with-readr" class="nav-link" data-scroll-target="#reading-data-with-readr"><span class="header-section-number">7.2.1</span> Reading data with readr</a></li>
  <li><a href="#how-does-the-interface-of-read_csv-work" id="toc-how-does-the-interface-of-read_csv-work" class="nav-link" data-scroll-target="#how-does-the-interface-of-read_csv-work"><span class="header-section-number">7.2.2</span> How does the interface of <code>read_csv</code> work?</a></li>
  <li><a href="#what-does-readr-produce-the-tibble" id="toc-what-does-readr-produce-the-tibble" class="nav-link" data-scroll-target="#what-does-readr-produce-the-tibble"><span class="header-section-number">7.2.3</span> What does <code>readr</code> produce? The <code>tibble</code>!</a></li>
  <li><a href="#how-to-look-at-a-tibble" id="toc-how-to-look-at-a-tibble" class="nav-link" data-scroll-target="#how-to-look-at-a-tibble"><span class="header-section-number">7.2.4</span> How to look at a <code>tibble</code>?</a></li>
  </ul></li>
  <li><a href="#plotting-data-in-tibbles" id="toc-plotting-data-in-tibbles" class="nav-link" data-scroll-target="#plotting-data-in-tibbles"><span class="header-section-number">7.3</span> Plotting data in <code>tibble</code>s</a>
  <ul class="collapse">
  <li><a href="#ggplot2-and-the-grammar-of-graphics" id="toc-ggplot2-and-the-grammar-of-graphics" class="nav-link" data-scroll-target="#ggplot2-and-the-grammar-of-graphics"><span class="header-section-number">7.3.1</span> <code>ggplot2</code> and the “grammar of graphics”</a></li>
  <li><a href="#scales-control-the-behaviour-of-visual-elements" id="toc-scales-control-the-behaviour-of-visual-elements" class="nav-link" data-scroll-target="#scales-control-the-behaviour-of-visual-elements"><span class="header-section-number">7.3.2</span> <code>scale</code>s control the behaviour of visual elements</a></li>
  <li><a href="#colour-scales" id="toc-colour-scales" class="nav-link" data-scroll-target="#colour-scales"><span class="header-section-number">7.3.3</span> Colour <code>scale</code>s</a></li>
  <li><a href="#more-variables-defining-plot-matrices-via-facets" id="toc-more-variables-defining-plot-matrices-via-facets" class="nav-link" data-scroll-target="#more-variables-defining-plot-matrices-via-facets"><span class="header-section-number">7.3.4</span> More variables! Defining plot matrices via <code>facet</code>s</a></li>
  <li><a href="#setting-purely-aesthetic-settings-with-theme" id="toc-setting-purely-aesthetic-settings-with-theme" class="nav-link" data-scroll-target="#setting-purely-aesthetic-settings-with-theme"><span class="header-section-number">7.3.5</span> Setting purely aesthetic settings with <code>theme</code></a></li>
  <li><a href="#ordering-elements-in-a-plot-with-factors" id="toc-ordering-elements-in-a-plot-with-factors" class="nav-link" data-scroll-target="#ordering-elements-in-a-plot-with-factors"><span class="header-section-number">7.3.6</span> Ordering elements in a plot with <code>factors</code></a></li>
  <li><a href="#exercise" id="toc-exercise" class="nav-link" data-scroll-target="#exercise"><span class="header-section-number">7.3.7</span> Exercise</a></li>
  </ul></li>
  <li><a href="#conditional-queries-on-tibbles" id="toc-conditional-queries-on-tibbles" class="nav-link" data-scroll-target="#conditional-queries-on-tibbles"><span class="header-section-number">7.4</span> Conditional queries on tibbles</a>
  <ul class="collapse">
  <li><a href="#selecting-columns-and-filtering-rows-with-select-and-filter" id="toc-selecting-columns-and-filtering-rows-with-select-and-filter" class="nav-link" data-scroll-target="#selecting-columns-and-filtering-rows-with-select-and-filter"><span class="header-section-number">7.4.1</span> Selecting columns and filtering rows with <code>select</code> and <code>filter</code></a></li>
  <li><a href="#chaining-functions-together-with-the-pipe" id="toc-chaining-functions-together-with-the-pipe" class="nav-link" data-scroll-target="#chaining-functions-together-with-the-pipe"><span class="header-section-number">7.4.2</span> Chaining functions together with the pipe <code>%&gt;%</code></a></li>
  <li><a href="#summary-statistics-in-base-r" id="toc-summary-statistics-in-base-r" class="nav-link" data-scroll-target="#summary-statistics-in-base-r"><span class="header-section-number">7.4.3</span> Summary statistics in <code>base</code> R</a></li>
  <li><a href="#group-wise-summaries-with-group_by-and-summarise" id="toc-group-wise-summaries-with-group_by-and-summarise" class="nav-link" data-scroll-target="#group-wise-summaries-with-group_by-and-summarise"><span class="header-section-number">7.4.4</span> Group-wise summaries with <code>group_by</code> and <code>summarise</code></a></li>
  <li><a href="#sorting-and-slicing-tibbles-with-arrange-and-slice" id="toc-sorting-and-slicing-tibbles-with-arrange-and-slice" class="nav-link" data-scroll-target="#sorting-and-slicing-tibbles-with-arrange-and-slice"><span class="header-section-number">7.4.5</span> Sorting and slicing tibbles with <code>arrange</code> and <code>slice</code></a></li>
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1"><span class="header-section-number">7.4.6</span> Exercise</a></li>
  </ul></li>
  <li><a href="#transforming-and-manipulating-tibbles" id="toc-transforming-and-manipulating-tibbles" class="nav-link" data-scroll-target="#transforming-and-manipulating-tibbles"><span class="header-section-number">7.5</span> Transforming and manipulating tibbles</a>
  <ul class="collapse">
  <li><a href="#renaming-and-reordering-columns-with-rename-and-relocate" id="toc-renaming-and-reordering-columns-with-rename-and-relocate" class="nav-link" data-scroll-target="#renaming-and-reordering-columns-with-rename-and-relocate"><span class="header-section-number">7.5.1</span> Renaming and reordering columns with <code>rename</code> and <code>relocate</code></a></li>
  <li><a href="#conditional-operations-with-ifelse-case_when-and-case_match" id="toc-conditional-operations-with-ifelse-case_when-and-case_match" class="nav-link" data-scroll-target="#conditional-operations-with-ifelse-case_when-and-case_match"><span class="header-section-number">7.5.2</span> Conditional operations with <code>ifelse</code>, <code>case_when</code> and <code>case_match</code></a></li>
  <li><a href="#switching-between-long-and-wide-data-with-pivot_longer-and-pivot_wider" id="toc-switching-between-long-and-wide-data-with-pivot_longer-and-pivot_wider" class="nav-link" data-scroll-target="#switching-between-long-and-wide-data-with-pivot_longer-and-pivot_wider"><span class="header-section-number">7.5.3</span> Switching between long and wide data with <code>pivot_longer</code> and <code>pivot_wider</code></a></li>
  <li><a href="#exercise-2" id="toc-exercise-2" class="nav-link" data-scroll-target="#exercise-2"><span class="header-section-number">7.5.4</span> Exercise</a></li>
  </ul></li>
  <li><a href="#combining-tibbles-with-join-operations" id="toc-combining-tibbles-with-join-operations" class="nav-link" data-scroll-target="#combining-tibbles-with-join-operations"><span class="header-section-number">7.6</span> Combining tibbles with join operations</a>
  <ul class="collapse">
  <li><a href="#types-of-joins" id="toc-types-of-joins" class="nav-link" data-scroll-target="#types-of-joins"><span class="header-section-number">7.6.1</span> Types of joins</a></li>
  <li><a href="#left-join-with-left_join" id="toc-left-join-with-left_join" class="nav-link" data-scroll-target="#left-join-with-left_join"><span class="header-section-number">7.6.2</span> Left join with <code>left_join</code></a></li>
  <li><a href="#right-join-with-right_join" id="toc-right-join-with-right_join" class="nav-link" data-scroll-target="#right-join-with-right_join"><span class="header-section-number">7.6.3</span> Right join with <code>right_join</code></a></li>
  <li><a href="#inner-join-with-inner_join" id="toc-inner-join-with-inner_join" class="nav-link" data-scroll-target="#inner-join-with-inner_join"><span class="header-section-number">7.6.4</span> Inner join with <code>inner_join</code></a></li>
  <li><a href="#full-join-with-full_join" id="toc-full-join-with-full_join" class="nav-link" data-scroll-target="#full-join-with-full_join"><span class="header-section-number">7.6.5</span> Full join with <code>full_join</code></a></li>
  <li><a href="#semi-join-with-semi_join" id="toc-semi-join-with-semi_join" class="nav-link" data-scroll-target="#semi-join-with-semi_join"><span class="header-section-number">7.6.6</span> Semi join with <code>semi_join</code></a></li>
  <li><a href="#anti-join-with-anti_join" id="toc-anti-join-with-anti_join" class="nav-link" data-scroll-target="#anti-join-with-anti_join"><span class="header-section-number">7.6.7</span> Anti join with <code>anti_join</code></a></li>
  <li><a href="#exercise-3" id="toc-exercise-3" class="nav-link" data-scroll-target="#exercise-3"><span class="header-section-number">7.6.8</span> Exercise</a></li>
  </ul></li>
  <li><a href="#optional-clean-up" id="toc-optional-clean-up" class="nav-link" data-scroll-target="#optional-clean-up"><span class="header-section-number">7.7</span> (Optional) clean-up</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">7.8</span> References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-useful-skills.html">Useful Skills</a></li><li class="breadcrumb-item"><a href="./r-tidyverse.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Clemens Schmid </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This session is typically ran held in parallel to the Introduction to Python and Pandas. Participants of the summer schools chose which to attend based on their prior experience. We recommend this session if you have no experience with neither R nor Python.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to download the chapter’s dataset, decompress the archive, and create and activate the conda environment.</p>
<p>Do this, use <code>wget</code> or right click and save to download this Zenodo archive: <a href="https://doi.org/10.5281/zenodo.8413026">10.5281/zenodo.8413026</a>, and unpack</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvf r-tidyverse.tar.gz </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> r-tidyverse/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can then create the subsequently activate environment with</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">-f</span> r-tidyverse.yml</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate r-tidyverse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="the-working-environment" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="the-working-environment"><span class="header-section-number">7.1</span> The working environment</h2>
<section id="r-rstudio-the-tidyverse-and-penguins" class="level3" data-number="7.1.1">
<h3 data-number="7.1.1" class="anchored" data-anchor-id="r-rstudio-the-tidyverse-and-penguins"><span class="header-section-number">7.1.1</span> R, RStudio, the tidyverse and penguins</h3>
<p>R <span class="citation" data-cites="RCoreTeam2023">(<a href="#ref-RCoreTeam2023" role="doc-biblioref">R Core Team 2023</a>)</span> is a fully featured programming language, but it excels as an environment for (statistical) data analysis (<a href="https://www.r-project.org" class="uri">https://www.r-project.org</a>) RStudio <span class="citation" data-cites="RstudioTeam">(<a href="#ref-RstudioTeam" role="doc-biblioref">RStudio Team 2020</a>)</span> is an integrated development environment (IDE) for R (and other languages) (<a href="https://www.rstudio.com/products/rstudio" class="uri">https://www.rstudio.com/products/rstudio</a>)</p>
<p>The tidyverse <span class="citation" data-cites="Wickham2019-ot">(<a href="#ref-Wickham2019-ot" role="doc-biblioref">Wickham et al. 2019</a>)</span> is a powerful collection of R packages with well-designed and consistent interfaces for the main steps of data analysis: loading, transforming and plotting data (<a href="https://www.tidyverse.org" class="uri">https://www.tidyverse.org</a>). This tutorial works with tidyverse ~v2.0. We will learn about the packages <code>readr</code>, <code>tibble</code>, <code>ggplot2</code>, <code>dplyr</code>, <code>magrittr</code> and <code>tidyr</code>. <code>forcats</code> will be briefly mentioned, but <code>purrr</code> and <code>stringr</code> are left out.</p>
<p>The <code>palmerpenguins</code> package <span class="citation" data-cites="Horst2020">(<a href="#ref-Horst2020" role="doc-biblioref">Horst, Hill, and Gorman 2020</a>)</span> provides a neat example dataset to learn data exploration and visualization in R (<a href="https://allisonhorst.github.io/palmerpenguins" class="uri">https://allisonhorst.github.io/palmerpenguins</a>)</p>
<div class="callout callout-style-default callout-note callout-titled" title="Instructions for running this chapter without conda">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Instructions for running this chapter without conda
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you wish to follow this chapter without installing conda, you can recreate the conda environment and data generation by running the following steps.</p>
<ol type="1">
<li>Install R and RStudio for your operating system according to one of the many instructions online (e.g.&nbsp;<a href="https://www.r-bloggers.com/2022/01/how-to-install-and-update-r-and-rstudio">here</a>)</li>
<li>Download the .qmd file underlying this webpage <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book/blob/main/r-tidyverse.qmd">here</a> and copy it to a new directory</li>
<li>Open RStudio and create a new project in this directory with <code>File</code> -&gt; <code>New Project...</code> -&gt; <code>Existing directory</code></li>
<li>Open this very file from the RStudio project, so that you can easily follow along</li>
<li>Run the <code>Environment preparation code</code> code to install necessary R package dependencies and prepare the required data used below</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># installing necessary R packages</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"tidyverse"</span>, <span class="st">"palmerpenguins"</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># preparing data</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># (at the end of this tutorial</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># you will understand this code)</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5678</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>peng_prepped <span class="ot">&lt;-</span> palmerpenguins<span class="sc">::</span>penguins <span class="sc">%&gt;%</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>        <span class="sc">!</span>dplyr<span class="sc">::</span><span class="fu">if_any</span>(</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">.cols =</span> tidyselect<span class="sc">::</span><span class="fu">everything</span>(),</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">.fns =</span> is.na</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">add_column</span>(., <span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(.), <span class="at">.before =</span> <span class="st">"species"</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>peng_prepped <span class="sc">%&gt;%</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">300</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(id) <span class="sc">%&gt;%</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span>bill_length_mm, <span class="sc">-</span>bill_depth_mm) <span class="sc">%&gt;%</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="st">"penguins.csv"</span>)</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>peng_prepped <span class="sc">%&gt;%</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">300</span>) <span class="sc">%&gt;%</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(id) <span class="sc">%&gt;%</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(id, bill_length_mm, bill_depth_mm) <span class="sc">%&gt;%</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    readr<span class="sc">::</span><span class="fu">write_csv</span>(<span class="st">"penguin_bills.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>To load RStudio from within your <code>conda</code> environment, please run</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">rstudio</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And the RStudio window will open.</p>
</section>
</section>
<section id="loading-data-into-tibbles" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="loading-data-into-tibbles"><span class="header-section-number">7.2</span> Loading data into tibbles</h2>
<section id="reading-data-with-readr" class="level3" data-number="7.2.1">
<h3 data-number="7.2.1" class="anchored" data-anchor-id="reading-data-with-readr"><span class="header-section-number">7.2.1</span> Reading data with readr</h3>
<p>With R we usually operate on data in our computer’s memory. The tidyverse provides the package <code>readr</code> to read data from text files into the memory. <code>readr</code> can read from our file system or the internet. It provides functions to read data in almost any (text) format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_csv</span>() <span class="co"># .csv files -&gt; see penguins.csv</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_tsv</span>() <span class="co"># .tsv files</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_delim</span>() <span class="co"># tabular files with an arbitrary separator</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_fwf</span>() <span class="co"># fixed width files</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_lines</span>() <span class="co"># read linewise to parse yourself</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="how-does-the-interface-of-read_csv-work" class="level3" data-number="7.2.2">
<h3 data-number="7.2.2" class="anchored" data-anchor-id="how-does-the-interface-of-read_csv-work"><span class="header-section-number">7.2.2</span> How does the interface of <code>read_csv</code> work?</h3>
<p>We can learn more about an R function with the <code>?</code> operator.</p>
<p>To open a help file for a specific function run <code>?&lt;function_name&gt;</code> (e.g.&nbsp;<code>?readr::read_csv</code>) in the R console. <code>readr::read_csv</code> has many options to specify how to read a text file.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_csv</span>(</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    file, <span class="co"># The path to the file we want to read</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_names =</span> <span class="cn">TRUE</span>, <span class="co"># Are there column names?</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_types =</span> <span class="cn">NULL</span>, <span class="co"># Which types do the columns have? NULL -&gt; auto</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">locale =</span> <span class="fu">default_locale</span>(), <span class="co"># How is information encoded in this file?</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">na =</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">"NA"</span>), <span class="co"># Which values mean "no data"</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">trim_ws =</span> <span class="cn">TRUE</span>, <span class="co"># Should superfluous white-spaces be removed?</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">skip =</span> <span class="dv">0</span>, <span class="co"># Skip X lines at the beginning of the file</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">n_max =</span> <span class="cn">Inf</span>, <span class="co"># Only read X lines</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">skip_empty_rows =</span> <span class="cn">TRUE</span>, <span class="co"># Should empty lines be ignored?</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">comment =</span> <span class="st">""</span>, <span class="co"># Should comment lines be ignored?</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">name_repair =</span> <span class="st">"unique"</span>, <span class="co"># How should "broken" column names be fixed</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>    ...</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="what-does-readr-produce-the-tibble" class="level3" data-number="7.2.3">
<h3 data-number="7.2.3" class="anchored" data-anchor-id="what-does-readr-produce-the-tibble"><span class="header-section-number">7.2.3</span> What does <code>readr</code> produce? The <code>tibble</code>!</h3>
<p>To read a .csv file (here <code>"penguins.csv"</code>) into a variable (here <code>peng_auto</code>) run.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>peng_auto <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">"penguins.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 300 Columns: 7
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): species, island, sex
dbl (4): id, flipper_length_mm, body_mass_g, year

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p><code>readr</code> first prints some information on the number and type of rows and columns it discovered in the file.</p>
<p>It automatically detects column types – but you can also define them manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>peng <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"penguins.csv"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">col_types =</span> <span class="st">"iccddcc"</span> <span class="co"># this string encodes the desired types for each column</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It then returns an in-memory representation of this data, a <code>tibble</code>.</p>
<p>A <code>tibble</code> is a “data frame”, a tabular data structure with rows and columns. Unlike a simple array, each column can have another data type.</p>
</section>
<section id="how-to-look-at-a-tibble" class="level3" data-number="7.2.4">
<h3 data-number="7.2.4" class="anchored" data-anchor-id="how-to-look-at-a-tibble"><span class="header-section-number">7.2.4</span> How to look at a <code>tibble</code>?</h3>
<p>Typing the name of any object into the R console will print an overview to the console.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>peng</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
      id species island    flipper_length_mm body_mass_g sex    year 
   &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
 1     1 Adelie  Torgersen               181        3750 male   2007 
 2     2 Adelie  Torgersen               186        3800 female 2007 
 3     4 Adelie  Torgersen               193        3450 female 2007 
 4     5 Adelie  Torgersen               190        3650 male   2007 
 5     6 Adelie  Torgersen               181        3625 female 2007 
 6     7 Adelie  Torgersen               195        4675 male   2007 
 7     9 Adelie  Torgersen               191        3800 male   2007 
 8    10 Adelie  Torgersen               198        4400 male   2007 
 9    11 Adelie  Torgersen               185        3700 female 2007 
10    12 Adelie  Torgersen               195        3450 female 2007 
# ℹ 290 more rows</code></pre>
</div>
</div>
<p>But there are various other ways to inspect the content of a <code>tibble</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(peng) <span class="co"># a structural overview of an R object</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(peng) <span class="co"># a human-readable summary of an R object</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(peng) <span class="co"># open RStudio's interactive data browser</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="plotting-data-in-tibbles" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="plotting-data-in-tibbles"><span class="header-section-number">7.3</span> Plotting data in <code>tibble</code>s</h2>
<section id="ggplot2-and-the-grammar-of-graphics" class="level3" data-number="7.3.1">
<h3 data-number="7.3.1" class="anchored" data-anchor-id="ggplot2-and-the-grammar-of-graphics"><span class="header-section-number">7.3.1</span> <code>ggplot2</code> and the “grammar of graphics”</h3>
<p>To understand and present data, we usually have to visualize it.</p>
<p><code>ggplot2</code> is an R package that offers an unusual, but powerful and logical interface for this task <span class="citation" data-cites="Wickham2016">(<a href="#ref-Wickham2016" role="doc-biblioref">Wickham 2016</a>)</span>. The following example describes a stacked bar chart.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co"># Loading a library to use its functions without ::</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>( <span class="co"># Every plot starts with a call to the ggplot() function</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">data =</span> peng <span class="co"># This function can also take the input tibble</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="co"># The plot consists of individual functions linked with "+"</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>( <span class="co"># "geoms" define the plot layers we want to draw</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">mapping =</span> <span class="fu">aes</span>( <span class="co"># The aes() function maps variables to visual properties</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">x =</span> island, <span class="co"># publication_year -&gt; x-axis</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">fill =</span> species <span class="co"># community_type   -&gt; fill color</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>A <code>geom_*</code> combines data (here <code>peng</code>), a geometry (here vertical, stacked bars) and a statistical transformation (here counting the number of penguins per island and species). <code>ggplot2</code> features many such geoms: A good overview is provided by this cheatsheet: <a href="https://rstudio.github.io/cheatsheets/html/data-visualization.html">https://rstudio.github.io/cheatsheets/html/data-visualization.html</a>.</p>
<p>Beyond <code>geom</code>s, a ggplot plot can be further specified with (among others) <code>scale</code>s, <code>facet</code>s and <code>theme</code>s.</p>
</section>
<section id="scales-control-the-behaviour-of-visual-elements" class="level3" data-number="7.3.2">
<h3 data-number="7.3.2" class="anchored" data-anchor-id="scales-control-the-behaviour-of-visual-elements"><span class="header-section-number">7.3.2</span> <code>scale</code>s control the behaviour of visual elements</h3>
<p>Here is another plot to demonstrate this: Boxplots of penguin weight per species.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Let’s assume we had some extreme outliers in this dataset. To simulate this, we replace some random weights with extreme values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234</span>) <span class="co"># we set a seed for reproducible randomness</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>peng_out <span class="ot">&lt;-</span> peng</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>peng_out<span class="sc">$</span>body_mass_g[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(peng_out), <span class="dv">10</span>)] <span class="ot">&lt;-</span> <span class="dv">50000</span> <span class="sc">+</span> <span class="dv">50000</span> <span class="sc">*</span> <span class="fu">runif</span>(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we plot the dataset with these “outliers”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng_out) <span class="sc">+</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This is not well readable, because the extreme outliers dictate the scale of the y-axis. A 50+kg penguin is a scary thought and we would probably remove these outliers, but let’s assume they are valid observation we want to include in the plot.</p>
<p>To mitigate this issue we can change the <strong>scale</strong> of different visual elements - e.g.&nbsp;the y-axis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng_out) <span class="sc">+</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g)) <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>() <span class="co"># adding the log-scale improves readability</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We will now go back to the normal dataset without the artificial outliers.</p>
</section>
<section id="colour-scales" class="level3" data-number="7.3.3">
<h3 data-number="7.3.3" class="anchored" data-anchor-id="colour-scales"><span class="header-section-number">7.3.3</span> Colour <code>scale</code>s</h3>
<p>(Fill) colour is a visual element of a plot and its scaling can be adjusted as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> species)) <span class="sc">+</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_viridis_d</span>(<span class="at">option =</span> <span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We use the <code>scale_*</code> function to select one of the visually appealing (and robust to colourblindness) viridis colour palettes (<a href="https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html">https://cran.r-project.org/web/packages/viridis/vignettes/intro-to-viridis.html</a>).</p>
</section>
<section id="more-variables-defining-plot-matrices-via-facets" class="level3" data-number="7.3.4">
<h3 data-number="7.3.4" class="anchored" data-anchor-id="more-variables-defining-plot-matrices-via-facets"><span class="header-section-number">7.3.4</span> More variables! Defining plot matrices via <code>facet</code>s</h3>
<p>In the previous example we didn’t add additional information with the fill colour, as the plot already distinguished by species on the x-axis.</p>
<p>We can instead use colour to encode more information, for example by mapping the variable sex to it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> sex))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Another way to visualize more variables in one plot is to split the plot by categories into <strong>facets</strong>, so sub-plots per category.</p>
<p>Here we split by sex, which is already mapped to the fill colour.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> sex)) <span class="sc">+</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The fill colour is therefore free again to show yet another variable, for example the year a given penguin was examined.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> year)) <span class="sc">+</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="setting-purely-aesthetic-settings-with-theme" class="level3" data-number="7.3.5">
<h3 data-number="7.3.5" class="anchored" data-anchor-id="setting-purely-aesthetic-settings-with-theme"><span class="header-section-number">7.3.5</span> Setting purely aesthetic settings with <code>theme</code></h3>
<p>Aesthetic changes can be applied as part of the <code>theme</code>, which allows for very detailed configuration (see <code>?theme</code>).</p>
<p>Here we rotate the x-axis labels by 45°, which often helps to resolve overplotting.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> species, <span class="at">y =</span> body_mass_g, <span class="at">fill =</span> year)) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span>sex) <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="ordering-elements-in-a-plot-with-factors" class="level3" data-number="7.3.6">
<h3 data-number="7.3.6" class="anchored" data-anchor-id="ordering-elements-in-a-plot-with-factors"><span class="header-section-number">7.3.6</span> Ordering elements in a plot with <code>factors</code></h3>
<p>R supports defining ordinal data with <code>factor</code>s. This can be used to set the order of elements in a plot, e.g.&nbsp;the order of bars in a bar chart.</p>
<p>We do not cover <code>factor</code>s beyond the following example here, although the tidyverse includes a package (<code>forcats</code>) specifically for that purpose.</p>
<p>Elements based on <code>character</code> columns are generally ordered alphabetically.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> species)) <span class="co"># bars are alphabetically ordered</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>With <code>forcats::fct_reorder</code> we can transform an input vector to a <code>factor</code>, ordered by a summary statistic (even based on another vector).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>peng2 <span class="ot">&lt;-</span> peng</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>peng2<span class="sc">$</span>species_ordered <span class="ot">&lt;-</span> forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    peng2<span class="sc">$</span>species,</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    peng2<span class="sc">$</span>species, length</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>With this change, the plot will be ordered according to the intrinsic order defined for <code>species_ordered</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(peng2) <span class="sc">+</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> species_ordered)) <span class="co"># bars are ordered by size</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="exercise" class="level3" data-number="7.3.7">
<h3 data-number="7.3.7" class="anchored" data-anchor-id="exercise"><span class="header-section-number">7.3.7</span> Exercise</h3>
<ol type="1">
<li><p>Look at the <code>mtcars</code> dataset and read up on the meaning of its variables with the help operator <code>?</code>. <code>mtcars</code> is a test dataset integrated in R and can be accessed by typing <code>mtcars</code> in the console</p></li>
<li><p>Visualize the relationship between <em>Gross horsepower</em> and <em>1/4 mile time</em></p></li>
</ol>
<ol start="3" type="1">
<li>Integrate the <em>Number of cylinders</em> into your plot as an additional variable</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled" title="Possible solutions">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Possible solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>?mtcars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mtcars) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> hp, <span class="at">y =</span> qsec))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mtcars) <span class="sc">+</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> hp, <span class="at">y =</span> qsec, <span class="at">color =</span> <span class="fu">as.factor</span>(cyl)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-tidyverse_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="conditional-queries-on-tibbles" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="conditional-queries-on-tibbles"><span class="header-section-number">7.4</span> Conditional queries on tibbles</h2>
<section id="selecting-columns-and-filtering-rows-with-select-and-filter" class="level3" data-number="7.4.1">
<h3 data-number="7.4.1" class="anchored" data-anchor-id="selecting-columns-and-filtering-rows-with-select-and-filter"><span class="header-section-number">7.4.1</span> Selecting columns and filtering rows with <code>select</code> and <code>filter</code></h3>
<p>The <code>dplyr</code> package includes powerful functions to subset data in tibbles based on conditions. <code>dplyr::select</code> allows to select columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(peng, id, flipper_length_mm) <span class="co"># reduce tibble to two columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 2
     id flipper_length_mm
  &lt;int&gt;             &lt;dbl&gt;
1     1               181
2     2               186
3     4               193
4     5               190
5     6               181
# ℹ 295 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(peng, <span class="sc">-</span>island, <span class="sc">-</span>flipper_length_mm) <span class="co"># remove two columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 5
     id species body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;         &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie         3750 male   2007 
2     2 Adelie         3800 female 2007 
3     4 Adelie         3450 female 2007 
4     5 Adelie         3650 male   2007 
5     6 Adelie         3625 female 2007 
# ℹ 295 more rows</code></pre>
</div>
</div>
<p><code>dplyr::filter</code> allows for conditional filtering of rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(peng, year <span class="sc">==</span> <span class="dv">2007</span>) <span class="co"># penguins examined in 2007</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 93 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie  Torgersen               181        3750 male   2007 
2     2 Adelie  Torgersen               186        3800 female 2007 
3     4 Adelie  Torgersen               193        3450 female 2007 
4     5 Adelie  Torgersen               190        3650 male   2007 
5     6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 88 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(peng, year <span class="sc">==</span> <span class="dv">2007</span> <span class="sc">|</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">==</span> <span class="dv">2009</span>) <span class="co"># penguins examined in 2007 OR 2009</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 198 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie  Torgersen               181        3750 male   2007 
2     2 Adelie  Torgersen               186        3800 female 2007 
3     4 Adelie  Torgersen               193        3450 female 2007 
4     5 Adelie  Torgersen               190        3650 male   2007 
5     6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 193 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    peng, <span class="co"># an alternative way to express</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    year <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">2007</span>, <span class="dv">2009</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>) <span class="co"># OR with the match operator "%in%"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 198 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie  Torgersen               181        3750 male   2007 
2     2 Adelie  Torgersen               186        3800 female 2007 
3     4 Adelie  Torgersen               193        3450 female 2007 
4     5 Adelie  Torgersen               190        3650 male   2007 
5     6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 193 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(peng, species <span class="sc">==</span> <span class="st">"Adelie"</span> <span class="sc">&amp;</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>    body_mass_g <span class="sc">&gt;=</span> <span class="dv">4000</span>) <span class="co"># Adelie penguins heavier than 4kg</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 29 × 7
     id species island    flipper_length_mm body_mass_g sex   year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;
1     7 Adelie  Torgersen               195        4675 male  2007 
2    10 Adelie  Torgersen               198        4400 male  2007 
3    13 Adelie  Torgersen               197        4500 male  2007 
4    31 Adelie  Dream                   196        4150 male  2007 
5    39 Adelie  Dream                   196        4400 male  2007 
# ℹ 24 more rows</code></pre>
</div>
</div>
</section>
<section id="chaining-functions-together-with-the-pipe" class="level3" data-number="7.4.2">
<h3 data-number="7.4.2" class="anchored" data-anchor-id="chaining-functions-together-with-the-pipe"><span class="header-section-number">7.4.2</span> Chaining functions together with the pipe <code>%&gt;%</code></h3>
<p>A core feature of the tidyverse is the pipe <code>%&gt;%</code> in the <code>magrittr</code> package. This ‘infix’ operator allows to chain data and operations for concise and clear data analysis syntax.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(year <span class="sc">==</span> <span class="dv">2007</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 93 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie  Torgersen               181        3750 male   2007 
2     2 Adelie  Torgersen               186        3800 female 2007 
3     4 Adelie  Torgersen               193        3450 female 2007 
4     5 Adelie  Torgersen               190        3650 male   2007 
5     6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 88 more rows</code></pre>
</div>
</div>
<p>It forwards the LHS (left-hand side) as the first argument of the function appearing on the RHS (right-hand side), which enables sequences of functions (“tidyverse style”).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(id, species, body_mass_g) <span class="sc">%&gt;%</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">"Adelie"</span> <span class="sc">&amp;</span> body_mass_g <span class="sc">&gt;=</span> <span class="dv">4000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">nrow</span>() <span class="co"># count the resulting rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 29</code></pre>
</div>
</div>
<p><code>magrittr</code> also offers some more operators, among which the extraction <code>%$%</code> is particularly useful to easily extract individual variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(island <span class="sc">==</span> <span class="st">"Biscoe"</span>) <span class="sc">%$%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>    species <span class="sc">%&gt;%</span> <span class="co"># extract the species column as a vector</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="co"># get the unique elements of said vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] "Adelie" "Gentoo"</code></pre>
</div>
</div>
<p>Here we already use the base R summary function <code>unique</code>.</p>
</section>
<section id="summary-statistics-in-base-r" class="level3" data-number="7.4.3">
<h3 data-number="7.4.3" class="anchored" data-anchor-id="summary-statistics-in-base-r"><span class="header-section-number">7.4.3</span> Summary statistics in <code>base</code> R</h3>
<p>Summarising and counting data is indispensable and R offers all basic operations you would expect in its <code>base</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>chinstraps_weights <span class="ot">&lt;-</span> peng <span class="sc">%&gt;%</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(species <span class="sc">==</span> <span class="st">"Chinstrap"</span>) <span class="sc">%$%</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>    body_mass_g</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(chinstraps_weights) <span class="co"># length/size of a vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 63</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(chinstraps_weights) <span class="co"># unique elements of a vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code> [1] 3500 3900 3650 3525 3725 3950 3250 3750 4150 3700 3800 3775 4050 3300 3450
[16] 4400 3400 2900 4550 3200 4300 3350 4100 3600 3850 4800 2700 4500 3550 3675
[31] 4450 3325 4000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(chinstraps_weights) <span class="co"># minimum</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 2700</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(chinstraps_weights) <span class="co"># maximum</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 4800</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(chinstraps_weights) <span class="co"># mean</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 3751.19</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(chinstraps_weights) <span class="co"># median</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 3725</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(chinstraps_weights) <span class="co"># variance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 153032.8</code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(chinstraps_weights) <span class="co"># standard deviation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 391.1941</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(chinstraps_weights, <span class="at">probs =</span> <span class="fl">0.75</span>) <span class="co"># quantiles for the given probabilities</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code> 75% 
3975 </code></pre>
</div>
</div>
<p>Many of these functions can ignore missing values with an option <code>na.rm = TRUE</code>.</p>
</section>
<section id="group-wise-summaries-with-group_by-and-summarise" class="level3" data-number="7.4.4">
<h3 data-number="7.4.4" class="anchored" data-anchor-id="group-wise-summaries-with-group_by-and-summarise"><span class="header-section-number">7.4.4</span> Group-wise summaries with <code>group_by</code> and <code>summarise</code></h3>
<p>These summary statistics are particular useful when applied to conditional subsets of a dataset.</p>
<p><code>dplyr</code> allows such summary operations with a combination of the functions <code>group_by</code> and <code>summarise</code>, where the former tags a <code>tibble</code> with categories based on its variables and the latter reduces it to these groups while simultanously creating new columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span> <span class="co"># group the tibble by the material column</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">min_weight =</span> <span class="fu">min</span>(body_mass_g), <span class="co"># new col: min weight for each group</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">median_weight =</span> <span class="fu">median</span>(body_mass_g), <span class="co"># new col: median weight for each group</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">max_weight =</span> <span class="fu">max</span>(body_mass_g) <span class="co"># new col: max weight for each group</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 3 × 4
  species   min_weight median_weight max_weight
  &lt;chr&gt;          &lt;dbl&gt;         &lt;dbl&gt;      &lt;dbl&gt;
1 Adelie          2850          3650       4775
2 Chinstrap       2700          3725       4800
3 Gentoo          3950          5050       6300</code></pre>
</div>
</div>
<p>Grouping can also be applied across multiple columns at once.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(species, year) <span class="sc">%&gt;%</span> <span class="co"># group by species and year</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">n =</span> dplyr<span class="sc">::</span><span class="fu">n</span>(), <span class="co"># a new column: number of penguins for each group</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">.groups =</span> <span class="st">"drop"</span> <span class="co"># drop the grouping after this summary operation</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 9 × 3
  species   year      n
  &lt;chr&gt;     &lt;chr&gt; &lt;int&gt;
1 Adelie    2007     38
2 Adelie    2008     45
3 Adelie    2009     43
4 Chinstrap 2007     23
5 Chinstrap 2008     18
# ℹ 4 more rows</code></pre>
</div>
</div>
<p>If we group by more than one variable, then <code>summarise</code> will not entirely remove the group tagging when generating the result dataset. We can force this with <code>.groups = "drop"</code> to avoid undesired behaviour with this dataset later on.</p>
</section>
<section id="sorting-and-slicing-tibbles-with-arrange-and-slice" class="level3" data-number="7.4.5">
<h3 data-number="7.4.5" class="anchored" data-anchor-id="sorting-and-slicing-tibbles-with-arrange-and-slice"><span class="header-section-number">7.4.5</span> Sorting and slicing tibbles with <code>arrange</code> and <code>slice</code></h3>
<p><code>dplyr</code> allows to <code>arrange</code> tibbles by one or multiple columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(sex) <span class="co"># sort by sex</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     2 Adelie  Torgersen               186        3800 female 2007 
2     4 Adelie  Torgersen               193        3450 female 2007 
3     6 Adelie  Torgersen               181        3625 female 2007 
4    11 Adelie  Torgersen               185        3700 female 2007 
5    12 Adelie  Torgersen               195        3450 female 2007 
# ℹ 295 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(sex, body_mass_g) <span class="co"># sort by sex and weight</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
     id species   island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1   304 Chinstrap Dream                   192        2700 female 2008 
2    53 Adelie    Biscoe                  181        2850 female 2008 
3    59 Adelie    Biscoe                  184        2850 female 2008 
4    49 Adelie    Biscoe                  187        2900 female 2008 
5   111 Adelie    Torgersen               188        2900 female 2009 
# ℹ 295 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(body_mass_g)) <span class="co"># sort descending</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
     id species island flipper_length_mm body_mass_g sex   year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;
1   164 Gentoo  Biscoe               221        6300 male  2007 
2   179 Gentoo  Biscoe               230        6050 male  2007 
3   260 Gentoo  Biscoe               222        6000 male  2009 
4   224 Gentoo  Biscoe               223        5950 male  2008 
5   160 Gentoo  Biscoe               213        5850 male  2007 
# ℹ 295 more rows</code></pre>
</div>
</div>
<p>Sorting also works within groups and can be paired with <code>slice</code> to extract extreme values per group.</p>
<p>Here we extract the heaviest individuals per species.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(species) <span class="sc">%&gt;%</span> <span class="co"># group by species</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(body_mass_g)) <span class="sc">%&gt;%</span> <span class="co"># sort by weight within (!) groups</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">3</span>) <span class="sc">%&gt;%</span> <span class="co"># keep the first three penguins per group</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="co"># remove the still lingering grouping</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 9 × 7
     id species   island    flipper_length_mm body_mass_g sex   year 
  &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;
1   104 Adelie    Biscoe                  197        4775 male  2009 
2    96 Adelie    Biscoe                  203        4725 male  2009 
3    76 Adelie    Torgersen               196        4700 male  2008 
4   303 Chinstrap Dream                   210        4800 male  2008 
5   295 Chinstrap Dream                   205        4550 male  2008 
# ℹ 4 more rows</code></pre>
</div>
</div>
<p>Slicing is also the relevant operation to take random samples from the observations in a <code>tibble</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 10 × 7
     id species island flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1    47 Adelie  Biscoe               190        3450 female 2008 
2   239 Gentoo  Biscoe               214        4850 female 2009 
3   221 Gentoo  Biscoe               209        4600 female 2008 
4   104 Adelie  Biscoe               197        4775 male   2009 
5   138 Adelie  Dream                190        3725 male   2009 
# ℹ 5 more rows</code></pre>
</div>
</div>
</section>
<section id="exercise-1" class="level3" data-number="7.4.6">
<h3 data-number="7.4.6" class="anchored" data-anchor-id="exercise-1"><span class="header-section-number">7.4.6</span> Exercise</h3>
<p>For this exercise we once more go back to the <code>mtcars</code> dataset. See <code>?mtcars</code> for details</p>
<ol type="1">
<li>Determine the number of cars with four <em>forward gears</em> (<code>gear</code>) in the <code>mtcars</code> dataset</li>
</ol>
<ol start="2" type="1">
<li>Determine the mean <em>1/4 mile time</em> (<code>qsec</code>) per <em>Number of cylinders</em> (<code>cyl</code>) group</li>
</ol>
<ol start="3" type="1">
<li>Identify the least efficient cars for both <em>transmission types</em> (<code>am</code>)</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled" title="Possible solutions">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Possible solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(gear <span class="sc">==</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] 12</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(cyl) <span class="sc">%&gt;%</span></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">qsec_mean =</span> <span class="fu">mean</span>(qsec)</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 3 × 2
    cyl qsec_mean
  &lt;dbl&gt;     &lt;dbl&gt;
1     4      19.1
2     6      18.0
3     8      16.8</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(mtcars, <span class="at">var =</span> <span class="st">"car"</span>)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="sc">%&gt;%</span></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(am) <span class="sc">%&gt;%</span></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(mpg) <span class="sc">%&gt;%</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">slice_head</span>() <span class="sc">%$%</span></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>    car</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code>[1] "Cadillac Fleetwood" "Maserati Bora"     </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="transforming-and-manipulating-tibbles" class="level2" data-number="7.5">
<h2 data-number="7.5" class="anchored" data-anchor-id="transforming-and-manipulating-tibbles"><span class="header-section-number">7.5</span> Transforming and manipulating tibbles</h2>
<section id="renaming-and-reordering-columns-with-rename-and-relocate" class="level3" data-number="7.5.1">
<h3 data-number="7.5.1" class="anchored" data-anchor-id="renaming-and-reordering-columns-with-rename-and-relocate"><span class="header-section-number">7.5.1</span> Renaming and reordering columns with <code>rename</code> and <code>relocate</code></h3>
<p>Columns in tibbles can be renamed with <code>dplyr::rename</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">penguin_name =</span> id) <span class="co"># rename a column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
  penguin_name species island    flipper_length_mm body_mass_g sex    year 
         &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1            1 Adelie  Torgersen               181        3750 male   2007 
2            2 Adelie  Torgersen               186        3800 female 2007 
3            4 Adelie  Torgersen               193        3450 female 2007 
4            5 Adelie  Torgersen               190        3650 male   2007 
5            6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 295 more rows</code></pre>
</div>
</div>
<p>And with <code>dplyr::relocate</code> they can be reordered.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">relocate</span>(year, <span class="at">.before =</span> species) <span class="co"># reorder columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 7
     id year  species island    flipper_length_mm body_mass_g sex   
  &lt;int&gt; &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; 
1     1 2007  Adelie  Torgersen               181        3750 male  
2     2 2007  Adelie  Torgersen               186        3800 female
3     4 2007  Adelie  Torgersen               193        3450 female
4     5 2007  Adelie  Torgersen               190        3650 male  
5     6 2007  Adelie  Torgersen               181        3625 female
# ℹ 295 more rows</code></pre>
</div>
</div>
<p>Adding columns to tibbles with <code>mutate</code> and <code>transmute</code>.</p>
<p>A common application of data manipulation is adding new, derived columns, that combine or modify the information in the already available columns. <code>dplyr</code> offers this core feature with the <code>mutate</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>( <span class="co"># add a column that</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">kg =</span> body_mass_g <span class="sc">/</span> <span class="dv">1000</span> <span class="co"># manipulates an existing column</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 8
     id species island    flipper_length_mm body_mass_g sex    year     kg
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;dbl&gt;
1     1 Adelie  Torgersen               181        3750 male   2007   3.75
2     2 Adelie  Torgersen               186        3800 female 2007   3.8 
3     4 Adelie  Torgersen               193        3450 female 2007   3.45
4     5 Adelie  Torgersen               190        3650 male   2007   3.65
5     6 Adelie  Torgersen               181        3625 female 2007   3.62
# ℹ 295 more rows</code></pre>
</div>
</div>
<p><code>dplyr::transmute</code> has the same interface as <code>dplyr::mutate</code>, but it removes all columns except for the newly created ones.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">transmute</span>(</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">id =</span> <span class="fu">paste</span>(<span class="st">"Penguin Nr."</span>, id), <span class="co"># overwrite this column</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>        flipper_length_mm <span class="co"># select this column</span></span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 2
  id            flipper_length_mm
  &lt;chr&gt;                     &lt;dbl&gt;
1 Penguin Nr. 1               181
2 Penguin Nr. 2               186
3 Penguin Nr. 4               193
4 Penguin Nr. 5               190
5 Penguin Nr. 6               181
# ℹ 295 more rows</code></pre>
</div>
</div>
<p><code>tibble::add_column</code> behaves as <code>dplyr::mutate</code>, but gives more control over column position.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">add_column</span>(</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">flipper_length_cm =</span> .<span class="sc">$</span>flipper_length_mm <span class="sc">/</span> <span class="dv">10</span>, <span class="co"># not the . representing the LHS of the pipe</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">.after =</span> <span class="st">"flipper_length_mm"</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 8
     id species island    flipper_length_mm flipper_length_cm body_mass_g sex   
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;             &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; 
1     1 Adelie  Torgersen               181              18.1        3750 male  
2     2 Adelie  Torgersen               186              18.6        3800 female
3     4 Adelie  Torgersen               193              19.3        3450 female
4     5 Adelie  Torgersen               190              19          3650 male  
5     6 Adelie  Torgersen               181              18.1        3625 female
# ℹ 295 more rows
# ℹ 1 more variable: year &lt;chr&gt;</code></pre>
</div>
</div>
<p><code>dplyr::mutate</code> can also be combined with <code>dplyr::group_by</code> (instead of <code>dplyr::summarise</code>) to add information on a group level. This is relevant, when a value for an individual entity should be put into context of a group-wise summary statistic.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(species, sex, year) <span class="sc">%&gt;%</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">mean_weight =</span> <span class="fu">mean</span>(body_mass_g, <span class="at">na.rm =</span> T),</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">relation_to_mean =</span> body_mass_g <span class="sc">/</span> mean_weight</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">select</span>(id, species, sex, year, relation_to_mean) <span class="sc">%&gt;%</span></span>
<span id="cb96-9"><a href="#cb96-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># mutate does not remove rows, unlike summarise, so we use select</span></span>
<span id="cb96-10"><a href="#cb96-10" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(relation_to_mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 5
     id species   sex    year  relation_to_mean
  &lt;int&gt; &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;
1   104 Adelie    male   2009              1.21
2    96 Adelie    male   2009              1.20
3   274 Chinstrap female 2007              1.17
4     7 Adelie    male   2007              1.17
5   303 Chinstrap male   2008              1.16
# ℹ 295 more rows</code></pre>
</div>
</div>
</section>
<section id="conditional-operations-with-ifelse-case_when-and-case_match" class="level3" data-number="7.5.2">
<h3 data-number="7.5.2" class="anchored" data-anchor-id="conditional-operations-with-ifelse-case_when-and-case_match"><span class="header-section-number">7.5.2</span> Conditional operations with <code>ifelse</code>, <code>case_when</code> and <code>case_match</code></h3>
<p><code>ifelse</code> allows to implement conditional <code>mutate</code> operations, that consider information from other columns.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">weight =</span> <span class="fu">ifelse</span>(</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">test =</span> body_mass_g <span class="sc">&gt;=</span> <span class="dv">4200</span>, <span class="co"># is weight below or above mean weight?</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">yes =</span> <span class="st">"above mean"</span>,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">no =</span> <span class="st">"below mean"</span></span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 8
     id species island    flipper_length_mm body_mass_g sex    year  weight    
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;     
1     1 Adelie  Torgersen               181        3750 male   2007  below mean
2     2 Adelie  Torgersen               186        3800 female 2007  below mean
3     4 Adelie  Torgersen               193        3450 female 2007  below mean
4     5 Adelie  Torgersen               190        3650 male   2007  below mean
5     6 Adelie  Torgersen               181        3625 female 2007  below mean
# ℹ 295 more rows</code></pre>
</div>
</div>
<p><code>ifelse</code> gets cumbersome easily. <code>dplyr::case_when</code> is more readable and scales much better for this application.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">weight =</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>        body_mass_g <span class="sc">&gt;=</span> <span class="dv">4200</span> <span class="sc">~</span> <span class="st">"above mean"</span>, <span class="co"># the number of conditions is arbitrary</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>        body_mass_g <span class="sc">&lt;</span> <span class="dv">4200</span> <span class="sc">~</span> <span class="st">"below mean"</span>,</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>        <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"unknown"</span> <span class="co"># TRUE catches all remaining cases</span></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 8
     id species island    flipper_length_mm body_mass_g sex    year  weight    
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;     
1     1 Adelie  Torgersen               181        3750 male   2007  below mean
2     2 Adelie  Torgersen               186        3800 female 2007  below mean
3     4 Adelie  Torgersen               193        3450 female 2007  below mean
4     5 Adelie  Torgersen               190        3650 male   2007  below mean
5     6 Adelie  Torgersen               181        3625 female 2007  below mean
# ℹ 295 more rows</code></pre>
</div>
</div>
<p><code>dplyr::case_match</code> is similar, but unlike <code>dplyr::case_when</code> it does not check logical expressions, but matches by value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>peng <span class="sc">%&gt;%</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">island_rating =</span> dplyr<span class="sc">::</span><span class="fu">case_match</span>(</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>            island,</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Torgersen"</span> <span class="sc">~</span> <span class="st">"My favourite island"</span>,</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Biscoe"</span> <span class="sc">~</span> <span class="st">"Overrated tourist trap"</span>,</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Dream"</span> <span class="sc">~</span> <span class="st">"Lost my wallet there. 4/10"</span></span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># here we use group_by+summarise only to show the result</span></span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(island, island_rating) <span class="sc">%&gt;%</span></span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">.groups =</span> <span class="st">"drop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 3 × 2
  island    island_rating             
  &lt;chr&gt;     &lt;chr&gt;                     
1 Biscoe    Overrated tourist trap    
2 Dream     Lost my wallet there. 4/10
3 Torgersen My favourite island       </code></pre>
</div>
</div>
</section>
<section id="switching-between-long-and-wide-data-with-pivot_longer-and-pivot_wider" class="level3" data-number="7.5.3">
<h3 data-number="7.5.3" class="anchored" data-anchor-id="switching-between-long-and-wide-data-with-pivot_longer-and-pivot_wider"><span class="header-section-number">7.5.3</span> Switching between long and wide data with <code>pivot_longer</code> and <code>pivot_wider</code></h3>
<p>To simplify certain analysis or plotting operations data often has to be transformed from a <strong>wide</strong> to a <strong>long</strong> format or vice versa (<a href="#fig-rtidyverse-longtowide" class="quarto-xref">Figure&nbsp;<span>7.1</span></a>).</p>
<div id="fig-rtidyverse-longtowide" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-longtowide-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/pivot_longer_wider.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-longtowide-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.1: Graphical representation of converting a table from a wide format (first column are categories, e.g.&nbsp;species, first row are also categories e.g., samples, and counts are in cells after first row/column. Conversion from wide to long makes the long-format so that the first columnn still has categories, but the second column becomes the column-categories from the first row, and third column has values. Converting long format back to wider shows again first row and column as categories and cells as values).
</figcaption>
</figure>
</div>
<ul>
<li>A table in <strong>wide</strong> format has N key columns and N value columns.</li>
<li>A table in <strong>long</strong> format has N key columns, one descriptor column and one value column.</li>
</ul>
<p>Here is an example of a wide dataset. It features information about the number of cars sold per year per brand at a dealership.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>carsales <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tribble</span>(</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="sc">~</span>brand, <span class="sc">~</span><span class="st">`</span><span class="at">2014</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2015</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2016</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2017</span><span class="st">`</span>,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"BMW"</span>, <span class="dv">20</span>, <span class="dv">25</span>, <span class="dv">30</span>, <span class="dv">45</span>,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"VW"</span>, <span class="dv">67</span>, <span class="dv">40</span>, <span class="dv">120</span>, <span class="dv">55</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>carsales</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 2 × 5
  brand `2014` `2015` `2016` `2017`
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 BMW       20     25     30     45
2 VW        67     40    120     55</code></pre>
</div>
</div>
<p>Wide format data becomes a problem, when the columns are semantically identical. This dataset is in wide format and we can not easily plot it. We generally prefer data in long format, although it is more verbose with more duplication.</p>
<p>To transform this dataset to a long format, we can apply <code>tidyr::pivot_longer</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>carsales_long <span class="ot">&lt;-</span> carsales <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> tidyselect<span class="sc">::</span><span class="fu">num_range</span>(</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>        <span class="st">""</span>,</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">range =</span> <span class="dv">2014</span><span class="sc">:</span><span class="dv">2017</span></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>    ), <span class="co"># define a set of columns to transform</span></span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"year"</span>, <span class="co"># the name of the descriptor column we want</span></span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_transform =</span> as.integer, <span class="co"># a transformation function to apply to the names</span></span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_to =</span> <span class="st">"sales"</span> <span class="co"># the name of the value column we want</span></span>
<span id="cb106-9"><a href="#cb106-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb106-10"><a href="#cb106-10" aria-hidden="true" tabindex="-1"></a>carsales_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 8 × 3
  brand  year sales
  &lt;chr&gt; &lt;int&gt; &lt;dbl&gt;
1 BMW    2014    20
2 BMW    2015    25
3 BMW    2016    30
4 BMW    2017    45
5 VW     2014    67
# ℹ 3 more rows</code></pre>
</div>
</div>
<p>Wide datasets are not always the wrong choice, for example for adjacency matrices to represent graphs, covariance matrices or other pairwise statistics. When data gets big, then wide formats can be significantly more efficient (e.g.&nbsp;for spatial data).</p>
<p>So transform data from long to wide, we can use <code>tidyr::pivot_wider</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>carsales_wide <span class="ot">&lt;-</span> carsales_long <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">id_cols =</span> <span class="st">"brand"</span>, <span class="co"># the set of id columns that should not be changed</span></span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_from =</span> year, <span class="co"># the descriptor column with the names of the new columns</span></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">values_from =</span> sales <span class="co"># the value column from which the values should be extracted</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>carsales_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 2 × 5
  brand `2014` `2015` `2016` `2017`
  &lt;chr&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1 BMW       20     25     30     45
2 VW        67     40    120     55</code></pre>
</div>
</div>
</section>
<section id="exercise-2" class="level3" data-number="7.5.4">
<h3 data-number="7.5.4" class="anchored" data-anchor-id="exercise-2"><span class="header-section-number">7.5.4</span> Exercise</h3>
<ol type="1">
<li>Move the column <code>gear</code> to the first position of the mtcars dataset</li>
</ol>
<ol start="2" type="1">
<li>Make a new dataset <code>mtcars2</code> with the column <code>mpg</code> and an additional column <code>am_v</code>, which encodes the <em>transmission type</em> (<code>am</code>) as either <code>"manual"</code> or <code>"automatic"</code></li>
</ol>
<ol start="3" type="1">
<li>Count the number of cars per <em>transmission type</em> (<code>am_v</code>) and <em>number of gears</em> (<code>gear</code>). Then transform the result to a wide format, with one column per <em>transmission type</em>.</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled" title="Possible solutions">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Possible solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(gear, <span class="at">.before =</span> mpg) <span class="sc">%&gt;%</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as_tibble</span>() <span class="co"># transforming the raw dataset for better printing</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 32 × 11
   gear   mpg   cyl  disp    hp  drat    wt  qsec    vs    am  carb
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     4  21       6   160   110  3.9   2.62  16.5     0     1     4
2     4  21       6   160   110  3.9   2.88  17.0     0     1     4
3     4  22.8     4   108    93  3.85  2.32  18.6     1     1     1
4     3  21.4     6   258   110  3.08  3.22  19.4     1     0     1
5     3  18.7     8   360   175  3.15  3.44  17.0     0     0     2
# ℹ 27 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="ot">&lt;-</span> mtcars <span class="sc">%&gt;%</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>        gear,</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">am_v =</span> dplyr<span class="sc">::</span><span class="fu">case_match</span>(</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>            am,</span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>            <span class="dv">0</span> <span class="sc">~</span> <span class="st">"automatic"</span>,</span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a>            <span class="dv">1</span> <span class="sc">~</span> <span class="st">"manual"</span></span>
<span id="cb112-8"><a href="#cb112-8" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb112-9"><a href="#cb112-9" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">%&gt;%</span></span>
<span id="cb112-10"><a href="#cb112-10" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as_tibble</span>()</span>
<span id="cb112-11"><a href="#cb112-11" aria-hidden="true" tabindex="-1"></a>mtcars2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 32 × 12
    mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb am_v     
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;    
1  21       6   160   110  3.9   2.62  16.5     0     1     4     4 manual   
2  21       6   160   110  3.9   2.88  17.0     0     1     4     4 manual   
3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1 manual   
4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1 automatic
5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2 automatic
# ℹ 27 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="sc">%&gt;%</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(am_v, gear) <span class="sc">%&gt;%</span></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">tally</span>() <span class="sc">%&gt;%</span> <span class="co"># dplyr::tally() is identical to</span></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># dplyr::summarise(n = dplyr::n())</span></span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># it counts the number of entities in a group</span></span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">names_from =</span> am_v,</span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">values_from =</span> n</span>
<span id="cb114-9"><a href="#cb114-9" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 3 × 3
   gear automatic manual
  &lt;dbl&gt;     &lt;int&gt;  &lt;int&gt;
1     3        15     NA
2     4         4      8
3     5        NA      5</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="combining-tibbles-with-join-operations" class="level2" data-number="7.6">
<h2 data-number="7.6" class="anchored" data-anchor-id="combining-tibbles-with-join-operations"><span class="header-section-number">7.6</span> Combining tibbles with join operations</h2>
<section id="types-of-joins" class="level3" data-number="7.6.1">
<h3 data-number="7.6.1" class="anchored" data-anchor-id="types-of-joins"><span class="header-section-number">7.6.1</span> Types of joins</h3>
<p>Joins combine two datasets x and y based on overlapping key columns.</p>
<p>Mutating joins add columns from one dataset to the other:</p>
<ul>
<li>Left join: Take observations from x and add fitting information from y.</li>
<li>Right join: Take observations from y and add fitting information from x.</li>
<li>Inner join: Join the overlapping observations from x and y.</li>
<li>Full join: Join all observations from x and y, even if information is missing.</li>
</ul>
<p>Filtering joins remove observations from x based on their presence in y.</p>
<p>Types of filtering consist of:</p>
<ul>
<li>Semi join: Keep every observation in x that is in y.</li>
<li>Anti join: Keep every observation in x that is not in y.</li>
</ul>
<p>The following sections will introduce each join with an example.</p>
<p>To experiment with joins, we need a second dataset with complementary information. This new dataset contains additional variables for a subset of the penguins in our first dataset – both datasets feature 300 penguins, but only with a partial overlap in individuals.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>bills <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_csv</span>(<span class="st">"penguin_bills.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 3
     id bill_length_mm bill_depth_mm
  &lt;dbl&gt;          &lt;dbl&gt;         &lt;dbl&gt;
1     1           39.1          18.7
2     2           39.5          17.4
3     3           40.3          18  
4     4           36.7          19.3
5     5           39.3          20.6
# ℹ 295 more rows</code></pre>
</div>
</div>
</section>
<section id="left-join-with-left_join" class="level3" data-number="7.6.2">
<h3 data-number="7.6.2" class="anchored" data-anchor-id="left-join-with-left_join"><span class="header-section-number">7.6.2</span> Left join with <code>left_join</code></h3>
<p>Take observations from x and add fitting information from y (<a href="#fig-rtidyverse-leftjoin" class="quarto-xref">Figure&nbsp;<span>7.2</span></a>).</p>
<div id="fig-rtidyverse-leftjoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-leftjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/left_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-leftjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.2: Graphical representation of left join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have columns A B C D. As A and B have a one to one match of values, this remains the same in the joined table. The B column between the two have a different value on the third row, and thus is lost from the second table, retaining row 3 of the first table. Column D (from the second table) has an empty value on row three, as this row was not in row 3 of the second table that column D was derived from.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span> <span class="co"># the key column by which to join</span></span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 9
     id species island  flipper_length_mm body_mass_g sex   year  bill_length_mm
  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;               &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1     1 Adelie  Torger…               181        3750 male  2007            39.1
2     2 Adelie  Torger…               186        3800 fema… 2007            39.5
3     4 Adelie  Torger…               193        3450 fema… 2007            36.7
4     5 Adelie  Torger…               190        3650 male  2007            39.3
5     6 Adelie  Torger…               181        3625 fema… 2007            38.9
# ℹ 295 more rows
# ℹ 1 more variable: bill_depth_mm &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Left joins are the most common join operation: Add information from y to the main dataset x.</p>
</section>
<section id="right-join-with-right_join" class="level3" data-number="7.6.3">
<h3 data-number="7.6.3" class="anchored" data-anchor-id="right-join-with-right_join"><span class="header-section-number">7.6.3</span> Right join with <code>right_join</code></h3>
<p>Take observations from y and add fitting information from x (<a href="#fig-rtidyverse-rightjoin" class="quarto-xref">Figure&nbsp;<span>7.3</span></a>).</p>
<div id="fig-rtidyverse-rightjoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-rightjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/right_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-rightjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.3: Graphical representation of right join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have columns A B C D. As A and B have a one to one match of values, this remains the same in the joined table. The B column between the two have a different value on the third row, and thus is lost from the first table, retaining row 3 of the second table. Column C (from the first table) has an empty value on row three, as this row was not in row 3 of the first table that column C was derived from.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">right_join</span>(</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span></span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="co"># we arrange by id to highlight the missing</span></span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">arrange</span>(id) <span class="co"># observation in the peng dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 300 × 9
     id species island  flipper_length_mm body_mass_g sex   year  bill_length_mm
  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;               &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1     1 Adelie  Torger…               181        3750 male  2007            39.1
2     2 Adelie  Torger…               186        3800 fema… 2007            39.5
3     3 &lt;NA&gt;    &lt;NA&gt;                   NA          NA &lt;NA&gt;  &lt;NA&gt;            40.3
4     4 Adelie  Torger…               193        3450 fema… 2007            36.7
5     5 Adelie  Torger…               190        3650 male  2007            39.3
# ℹ 295 more rows
# ℹ 1 more variable: bill_depth_mm &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Right joins are almost identical to left joins – only x and y have reversed roles.</p>
</section>
<section id="inner-join-with-inner_join" class="level3" data-number="7.6.4">
<h3 data-number="7.6.4" class="anchored" data-anchor-id="inner-join-with-inner_join"><span class="header-section-number">7.6.4</span> Inner join with <code>inner_join</code></h3>
<p>Join the overlapping observations from x and y (<a href="#fig-rtidyverse-innerjoin" class="quarto-xref">Figure&nbsp;<span>7.4</span></a>).</p>
<div id="fig-rtidyverse-innerjoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-innerjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/inner_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-innerjoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.4: Graphical representation of inner join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have columns A B C D. Only rows from both table that have exact matches on columns A and B are retained. The third rows from both tables that had a different value in column B are lost.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">inner_join</span>(</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span></span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 275 × 9
     id species island  flipper_length_mm body_mass_g sex   year  bill_length_mm
  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;               &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1     1 Adelie  Torger…               181        3750 male  2007            39.1
2     2 Adelie  Torger…               186        3800 fema… 2007            39.5
3     4 Adelie  Torger…               193        3450 fema… 2007            36.7
4     5 Adelie  Torger…               190        3650 male  2007            39.3
5     6 Adelie  Torger…               181        3625 fema… 2007            38.9
# ℹ 270 more rows
# ℹ 1 more variable: bill_depth_mm &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Inner joins are a fast and easy way to check, to which degree two dataset overlap.</p>
</section>
<section id="full-join-with-full_join" class="level3" data-number="7.6.5">
<h3 data-number="7.6.5" class="anchored" data-anchor-id="full-join-with-full_join"><span class="header-section-number">7.6.5</span> Full join with <code>full_join</code></h3>
<p>Join all observations from x and y, even if information is missing (<a href="#fig-rtidyverse-fulljoin" class="quarto-xref">Figure&nbsp;<span>7.5</span></a>).</p>
<div id="fig-rtidyverse-fulljoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-fulljoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/inner_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-fulljoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.5: Graphical representation of full join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have columns A B C D. All rows from both tables are retained, even though they do not share the same value on column B on both tables. The missing values for the two third rows (i.e., column C from the second table, and column D from the first table) are are filled with an empty cell indicated with <code>-</code>.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">full_join</span>(</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span></span>
<span id="cb124-5"><a href="#cb124-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(id)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 325 × 9
     id species island  flipper_length_mm body_mass_g sex   year  bill_length_mm
  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;               &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;          &lt;dbl&gt;
1     1 Adelie  Torger…               181        3750 male  2007            39.1
2     2 Adelie  Torger…               186        3800 fema… 2007            39.5
3     3 &lt;NA&gt;    &lt;NA&gt;                   NA          NA &lt;NA&gt;  &lt;NA&gt;            40.3
4     4 Adelie  Torger…               193        3450 fema… 2007            36.7
5     5 Adelie  Torger…               190        3650 male  2007            39.3
# ℹ 320 more rows
# ℹ 1 more variable: bill_depth_mm &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Full joins allow to preserve every bit of information.</p>
</section>
<section id="semi-join-with-semi_join" class="level3" data-number="7.6.6">
<h3 data-number="7.6.6" class="anchored" data-anchor-id="semi-join-with-semi_join"><span class="header-section-number">7.6.6</span> Semi join with <code>semi_join</code></h3>
<p>Keep every observation in x that is in y (<a href="#fig-rtidyverse-semijoin" class="quarto-xref">Figure&nbsp;<span>7.6</span></a>).</p>
<div id="fig-rtidyverse-semijoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-semijoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/semi_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-semijoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.6: Graphical representation of semi join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have only columns A B C. Only columns A B and C are retained in the joined table. Row three of both tables are not included as the column values in columns A and B do not match.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">semi_join</span>(</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb126-4"><a href="#cb126-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span></span>
<span id="cb126-5"><a href="#cb126-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 275 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1     1 Adelie  Torgersen               181        3750 male   2007 
2     2 Adelie  Torgersen               186        3800 female 2007 
3     4 Adelie  Torgersen               193        3450 female 2007 
4     5 Adelie  Torgersen               190        3650 male   2007 
5     6 Adelie  Torgersen               181        3625 female 2007 
# ℹ 270 more rows</code></pre>
</div>
</div>
<p>Semi joins are underused (!) operations to filter datasets.</p>
</section>
<section id="anti-join-with-anti_join" class="level3" data-number="7.6.7">
<h3 data-number="7.6.7" class="anchored" data-anchor-id="anti-join-with-anti_join"><span class="header-section-number">7.6.7</span> Anti join with <code>anti_join</code></h3>
<p>Keep every observation in x that is not in y (<a href="#fig-rtidyverse-antijoin" class="quarto-xref">Figure&nbsp;<span>7.7</span></a>).</p>
<div id="fig-rtidyverse-antijoin" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-rtidyverse-antijoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/r-tidyverse/anti_join.png" height="150" class="figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-rtidyverse-antijoin-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7.7: Graphical representation of anti join operation. Two tables with a shared first column (A B C, and A B D respectively) are merged together to have only columns A B C, and of that only row 3 of the first table. Only row three is retained from the first table as this is the only unique row present only in the first table.
</figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">anti_join</span>(</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> peng, <span class="co"># 300 observations</span></span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> bills, <span class="co"># 300 observations</span></span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">by =</span> <span class="st">"id"</span></span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 25 × 7
     id species island    flipper_length_mm body_mass_g sex    year 
  &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;
1    22 Adelie  Biscoe                  183        3550 male   2007 
2    34 Adelie  Dream                   181        3300 female 2007 
3    74 Adelie  Torgersen               195        4000 male   2008 
4    92 Adelie  Dream                   196        4350 male   2008 
5    99 Adelie  Biscoe                  193        2925 female 2009 
# ℹ 20 more rows</code></pre>
</div>
</div>
<p>Anti joins allow to quickly determine what information is missing in a dataset compared to an other one.</p>
</section>
<section id="exercise-3" class="level3" data-number="7.6.8">
<h3 data-number="7.6.8" class="anchored" data-anchor-id="exercise-3"><span class="header-section-number">7.6.8</span> Exercise</h3>
<p>Consider the following additional dataset with my opinions on cars with a specific number of gears:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>gear_opinions <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">gear =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">5</span>),</span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">opinion =</span> <span class="fu">c</span>(<span class="st">"boring"</span>, <span class="st">"wow"</span>)</span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol type="1">
<li>Add my opinions about gears to the <code>mtcars</code> dataset</li>
</ol>
<ol start="2" type="1">
<li>Remove all cars from the dataset for which I do not have an opinion</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled" title="Possible solutions">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Possible solutions
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">left_join</span>(mtcars, gear_opinions, <span class="at">by =</span> <span class="st">"gear"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as_tibble</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 32 × 12
    mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb opinion
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;  
1  21       6   160   110  3.9   2.62  16.5     0     1     4     4 &lt;NA&gt;   
2  21       6   160   110  3.9   2.88  17.0     0     1     4     4 &lt;NA&gt;   
3  22.8     4   108    93  3.85  2.32  18.6     1     1     4     1 &lt;NA&gt;   
4  21.4     6   258   110  3.08  3.22  19.4     1     0     3     1 boring 
5  18.7     8   360   175  3.15  3.44  17.0     0     0     3     2 boring 
# ℹ 27 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">anti_join</span>(mtcars, gear_opinions, <span class="at">by =</span> <span class="st">"gear"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">as_tibble</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre style="border: 1px; border-style: solid; margin-left: 10px; margin-right: 10px;"><code># A tibble: 12 × 11
    mpg   cyl  disp    hp  drat    wt  qsec    vs    am  gear  carb
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1  21       6  160    110  3.9   2.62  16.5     0     1     4     4
2  21       6  160    110  3.9   2.88  17.0     0     1     4     4
3  22.8     4  108     93  3.85  2.32  18.6     1     1     4     1
4  24.4     4  147.    62  3.69  3.19  20       1     0     4     2
5  22.8     4  141.    95  3.92  3.15  22.9     1     0     4     2
# ℹ 7 more rows</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="optional-clean-up" class="level2" data-number="7.7">
<h2 data-number="7.7" class="anchored" data-anchor-id="optional-clean-up"><span class="header-section-number">7.7</span> (Optional) clean-up</h2>
<p>Let’s clean up your working directory by removing all the data and output from this chapter.</p>
<p>When closing <code>rstudio</code>, say no to saving any additional files.</p>
<p>The command below will remove the <code>/&lt;PATH&gt;/&lt;TO&gt;/r-tidyverse</code> directory <strong>as well as all of its contents</strong>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pro Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Always be VERY careful when using <code>rm -r</code>. Check 3x that the path you are specifying is exactly what you want to delete and nothing more before pressing ENTER!</p>
</div>
</div>
<div class="sourceCode" id="cb135"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> /<span class="op">&lt;</span>PATH<span class="op">&gt;</span>/<span class="op">&lt;</span>TO<span class="op">&gt;</span>/r-tidyverse<span class="pp">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once deleted you can move elsewhere (e.g.&nbsp;<code>cd ~</code>).</p>
<p>We can also get out of the <code>conda</code> environment with</p>
<div class="sourceCode" id="cb136"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To delete the conda environment</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> remove <span class="at">--name</span> r-tidyverse <span class="at">--all</span> <span class="at">-y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="references" class="level2" data-number="7.8">
<h2 data-number="7.8" class="anchored" data-anchor-id="references"><span class="header-section-number">7.8</span> References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Horst2020" class="csl-entry" role="listitem">
Horst, Allison Marie, Alison Presmanes Hill, and Kristen B Gorman. 2020. <em>Palmerpenguins: Palmer Archipelago (Antarctica) Penguin Data</em>. <a href="https://doi.org/10.5281/zenodo.3960218">https://doi.org/10.5281/zenodo.3960218</a>.
</div>
<div id="ref-RCoreTeam2023" class="csl-entry" role="listitem">
R Core Team. 2023. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="https://www.R-project.org">https://www.R-project.org</a>.
</div>
<div id="ref-RstudioTeam" class="csl-entry" role="listitem">
RStudio Team. 2020. <em>RStudio: Integrated Development Environment for r</em>. Boston, MA: RStudio, PBC. <a href="http://www.rstudio.com/">http://www.rstudio.com/</a>.
</div>
<div id="ref-Wickham2016" class="csl-entry" role="listitem">
Wickham, Hadley. 2016. <em>Ggplot2: Elegant Graphics for Data Analysis</em>. Springer-Verlag New York. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-Wickham2019-ot" class="csl-entry" role="listitem">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy McGowan, Romain François, Garrett Grolemund, et al. 2019. <span>“Welcome to the Tidyverse.”</span> <em>Journal of Open Source Software</em> 4 (43): 1686. <a href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/spaam-community\.github\.org\/intro-to-ancient-metagenomics-book");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
          // target, if specified
          link.setAttribute("target", "_blank");
          if (link.getAttribute("rel") === null) {
            link.setAttribute("rel", "noopener");
          }
          // default icon
          link.classList.add("external");
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./bare-bones-bash.html" class="pagination-link" aria-label="Introduction to the Command Line">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./python-pandas.html" class="pagination-link" aria-label="Introduction to Python and Pandas">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2023 SPAAM Community &amp; Authors with ❤️. Available under <a href="http://creativecommons.org/licenses/by-sa/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>