<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Maxime Borry">

<title>10&nbsp; Taxonomic Profiling, OTU Tables and Visualisation – Introduction to Ancient Metagenomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./denovo-assembly.html" rel="next">
<link href="./section-ancient-metagenomics.html" rel="prev">
<link href="./favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<meta property="og:title" content="10&nbsp; Taxonomic Profiling, OTU Tables and Visualisation – Introduction to Ancient Metagenomics">
<meta property="og:description" content="">
<meta property="og:site_name" content="Introduction to Ancient Metagenomics">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-ancient-metagenomics.html">Ancient Metagenomics</a></li><li class="breadcrumb-item"><a href="./taxonomic-profiling.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./foreword.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Foreword</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-edna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Environmental DNA</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title"><em>De novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Authentication</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./contamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Contamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancient-metagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./deprecated-chapters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Deprecated Chapters</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-writing-guidelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Chapter Writing Guidelines</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-template.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Chapter Page Template</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./chapter-checklist.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Chapter Checklist</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">10.1</span> Introduction</a></li>
  <li><a href="#chapter-overview" id="toc-chapter-overview" class="nav-link" data-scroll-target="#chapter-overview"><span class="header-section-number">10.2</span> Chapter Overview</a></li>
  <li><a href="#working-in-a-jupyter-environment" id="toc-working-in-a-jupyter-environment" class="nav-link" data-scroll-target="#working-in-a-jupyter-environment"><span class="header-section-number">10.3</span> Working in a jupyter environment</a></li>
  <li><a href="#data-pre-processing" id="toc-data-pre-processing" class="nav-link" data-scroll-target="#data-pre-processing"><span class="header-section-number">10.4</span> Data pre-processing</a></li>
  <li><a href="#adapter-sequence-trimming-and-low-quality-bases-trimming" id="toc-adapter-sequence-trimming-and-low-quality-bases-trimming" class="nav-link" data-scroll-target="#adapter-sequence-trimming-and-low-quality-bases-trimming"><span class="header-section-number">10.5</span> Adapter sequence trimming and low-quality bases trimming</a></li>
  <li><a href="#taxonomic-profiling-with-metaphlan" id="toc-taxonomic-profiling-with-metaphlan" class="nav-link" data-scroll-target="#taxonomic-profiling-with-metaphlan"><span class="header-section-number">10.6</span> Taxonomic profiling with Metaphlan</a></li>
  <li><a href="#visualizing-the-taxonomic-profile" id="toc-visualizing-the-taxonomic-profile" class="nav-link" data-scroll-target="#visualizing-the-taxonomic-profile"><span class="header-section-number">10.7</span> Visualizing the taxonomic profile</a>
  <ul class="collapse">
  <li><a href="#visualizing-metaphlan-taxonomic-profile-with-pavian" id="toc-visualizing-metaphlan-taxonomic-profile-with-pavian" class="nav-link" data-scroll-target="#visualizing-metaphlan-taxonomic-profile-with-pavian"><span class="header-section-number">10.7.1</span> Visualizing metaphlan taxonomic profile with Pavian</a></li>
  <li><a href="#visualizing-metaphlan-taxonomic-profile-with-krona" id="toc-visualizing-metaphlan-taxonomic-profile-with-krona" class="nav-link" data-scroll-target="#visualizing-metaphlan-taxonomic-profile-with-krona"><span class="header-section-number">10.7.2</span> Visualizing metaphlan taxonomic profile with Krona</a></li>
  </ul></li>
  <li><a href="#getting-modern-comparative-reference-data" id="toc-getting-modern-comparative-reference-data" class="nav-link" data-scroll-target="#getting-modern-comparative-reference-data"><span class="header-section-number">10.8</span> Getting modern comparative reference data</a></li>
  <li><a href="#loading-the-ancient-sample-taxonomic-profile" id="toc-loading-the-ancient-sample-taxonomic-profile" class="nav-link" data-scroll-target="#loading-the-ancient-sample-taxonomic-profile"><span class="header-section-number">10.9</span> Loading the ancient sample taxonomic profile</a></li>
  <li><a href="#bringing-together-ancient-and-modern-samples" id="toc-bringing-together-ancient-and-modern-samples" class="nav-link" data-scroll-target="#bringing-together-ancient-and-modern-samples"><span class="header-section-number">10.10</span> Bringing together ancient and modern samples</a>
  <ul class="collapse">
  <li><a href="#time-to-merge" id="toc-time-to-merge" class="nav-link" data-scroll-target="#time-to-merge"><span class="header-section-number">10.10.1</span> Time to merge !</a></li>
  </ul></li>
  <li><a href="#comparing-ancient-and-modern-samples" id="toc-comparing-ancient-and-modern-samples" class="nav-link" data-scroll-target="#comparing-ancient-and-modern-samples"><span class="header-section-number">10.11</span> Comparing ancient and modern samples</a>
  <ul class="collapse">
  <li><a href="#taxonomic-composition" id="toc-taxonomic-composition" class="nav-link" data-scroll-target="#taxonomic-composition"><span class="header-section-number">10.11.1</span> Taxonomic composition</a></li>
  </ul></li>
  <li><a href="#lets-make-some-plots" id="toc-lets-make-some-plots" class="nav-link" data-scroll-target="#lets-make-some-plots"><span class="header-section-number">10.12</span> Let’s make some plots</a></li>
  <li><a href="#ecological-diversity" id="toc-ecological-diversity" class="nav-link" data-scroll-target="#ecological-diversity"><span class="header-section-number">10.13</span> Ecological diversity</a>
  <ul class="collapse">
  <li><a href="#alpha-diversity" id="toc-alpha-diversity" class="nav-link" data-scroll-target="#alpha-diversity"><span class="header-section-number">10.13.1</span> Alpha diversity</a></li>
  <li><a href="#beta-diversity" id="toc-beta-diversity" class="nav-link" data-scroll-target="#beta-diversity"><span class="header-section-number">10.13.2</span> Beta diversity</a></li>
  </ul></li>
  <li><a href="#optional-clean-up" id="toc-optional-clean-up" class="nav-link" data-scroll-target="#optional-clean-up"><span class="header-section-number">10.14</span> (Optional) clean-up</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">10.15</span> Summary</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">10.16</span> References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-ancient-metagenomics.html">Ancient Metagenomics</a></li><li class="breadcrumb-item"><a href="./taxonomic-profiling.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></a></li></ol></nav>
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Maxime Borry </p>
          </div>
  </div>
    
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-note callout-titled" title="Self guided: chapter environment setup">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Self guided: chapter environment setup
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to download the chapter’s dataset, decompress the archive, and create and activate the conda environment.</p>
<p>Do this, use <code>wget</code> or right click and save to download this Zenodo archive: <a href="https://doi.org/10.5281/zenodo.8413138">10.5281/zenodo.8413138</a>, and unpack</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> xvf taxonomic-profiling.tar.gz </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> taxonomic-profiling/</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can then create the subsequently activate environment with</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">-f</span> taxonomic-profiling.yml</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate taxonomic-profiling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="introduction" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">10.1</span> Introduction</h2>
<p>In this chapter, we’re going to look at taxonomic profiling, or in other words, how to get the microbial composition of a sample from the DNA sequencing data.</p>
<p>Though there are many algorithms, and even more different tools available to perform taxonomic profiling, the general idea remains the same (<a href="#fig-taxonomicprofiling-overview" class="quarto-xref">Figure&nbsp;<span>10.1</span></a>).</p>
<p>After cleaning up the sequencing data, generally saved as FASTQ files, a <strong>taxonomic profiler</strong> is used to compare the sequenced DNA to a reference database of sequences from known organisms, in order to generate a taxonomic profile of all organisms identified in a sample (<a href="#fig-taxonomicprofiling-overview" class="quarto-xref">Figure&nbsp;<span>10.1</span></a>)</p>
<div id="fig-taxonomicprofiling-overview" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/how_to_analyze.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-overview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.1: General overview of a taxonomic profiling analysis workflow
</figcaption>
</figure>
</div>
<p>If you prefer text instead of pictograms, the workflow we’re going to cover today is outlined in <a href="#fig-taxonomicprofiling-workflow" class="quarto-xref">Figure&nbsp;<span>10.2</span></a>, adapted from <span class="citation" data-cites="sharptonIntroductionAnalysisShotgun2014">Sharpton (<a href="#ref-sharptonIntroductionAnalysisShotgun2014" role="doc-biblioref">2014</a>)</span></p>
<div id="fig-taxonomicprofiling-workflow" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_workflow.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-workflow-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.2: A typical metagenomics analysis workflow, adapted from <span class="citation" data-cites="sharptonIntroductionAnalysisShotgun2014">Sharpton (<a href="#ref-sharptonIntroductionAnalysisShotgun2014" role="doc-biblioref">2014</a>)</span>
</figcaption>
</figure>
</div>
<p>Because different organisms can possess the same DNA, especially when looking at shorter sequences, taxonomic profilers need to have a way to resolve the ambiguity in the taxonomic assignation (<a href="#fig-taxonomicprofiling-ambiguity" class="quarto-xref">Figure&nbsp;<span>10.3</span></a>).</p>
<div id="fig-taxonomicprofiling-ambiguity" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-ambiguity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/ambiguity.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-ambiguity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.3: Different species can share the same DNA sequence
</figcaption>
</figure>
</div>
<p>By leveraging an algorithm known as the <strong>Lowest Common Ancestor (LCA)</strong>, and the taxonomic tree of all known species, ambiguities are going to be resolved by assigning a higher, less precise, taxonomic rank to ambiguous matches (<a href="#fig-taxonomicprofiling-lca" class="quarto-xref">Figure&nbsp;<span>10.4</span></a>).</p>
<div id="fig-taxonomicprofiling-lca" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-lca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/lca.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-lca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.4: A diagram of the LCA algorithm, a way to resolve these ambiguities
</figcaption>
</figure>
</div>
</section>
<section id="chapter-overview" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="chapter-overview"><span class="header-section-number">10.2</span> Chapter Overview</h2>
<p>Today, we’re going to use the following tools:</p>
<ul>
<li>fastp <span class="citation" data-cites="chenFastpUltrafastAllinone2018">(<a href="#ref-chenFastpUltrafastAllinone2018" role="doc-biblioref">Chen et al. 2018</a>)</span> for sequencing data cleaning</li>
<li>MetaPhlAn <span class="citation" data-cites="segata2012metagenomic">(<a href="#ref-segata2012metagenomic" role="doc-biblioref">Segata et al. 2012</a>)</span>, for taxonomic profiling</li>
<li>Krona <span class="citation" data-cites="ondovInteractiveMetagenomicVisualization2011">(<a href="#ref-ondovInteractiveMetagenomicVisualization2011" role="doc-biblioref">Brian D. Ondov, Bergman, and Phillippy 2011a</a>)</span> and Pavian <span class="citation" data-cites="breitwieserPavianInteractiveAnalysis2016">(<a href="#ref-breitwieserPavianInteractiveAnalysis2016" role="doc-biblioref">Florian P. Breitwieser and Salzberg 2016</a>)</span> for the interactive exploration of the taxonomic profiles</li>
<li>curatedMetagenomicData <span class="citation" data-cites="pasolliAccessibleCuratedMetagenomic2017">(<a href="#ref-pasolliAccessibleCuratedMetagenomic2017" role="doc-biblioref">Pasolli et al. 2017</a>)</span> for retrieving modern comparison data</li>
<li>Python, pandas <span class="citation" data-cites="rebackPandasdevPandasPandas2022">(<a href="#ref-rebackPandasdevPandasPandas2022" role="doc-biblioref">Reback et al. 2022</a>)</span>, plotnine (<a href="https://plotnine.readthedocs.io/en/stable/">https://plotnine.readthedocs.io/en/stable/</a>), and scikit-bio <span class="citation" data-cites="scikit-bioScikitbioBioinformaticsLibrary2022">(<a href="#ref-scikit-bioScikitbioBioinformaticsLibrary2022" role="doc-biblioref">scikit-bio 2022</a>)</span> to perform exploratory data analysis and a bit of microbial ecology</li>
</ul>
<p>to explore a toy dataset that has already been prepared for you.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Preamble: what has been done to generate this toy dataset
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="download-and-subsample" class="level3" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="download-and-subsample"><span class="header-section-number">10.2.1</span> Download and Subsample</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For this tutorial, we will be using the ERR5766177 library from the sample 2612 published by <span class="citation" data-cites="maixner2021hallstatt">(<a href="#ref-maixner2021hallstatt" role="doc-biblioref">Maixner et al. 2021</a>)</span></p>
<p><img src="assets/images/chapters/taxonomic-profiling/1-s2.0-S0960982221012719-fx1.jpg" class="img-fluid"></p>
<p><img src="assets/images/chapters/taxonomic-profiling/1-s2.0-S0960982221012719-gr1.jpg" class="img-fluid">`</p>
</section>
<section id="subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial" class="level3" data-number="10.2.2">
<h3 data-number="10.2.2" class="anchored" data-anchor-id="subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial"><span class="header-section-number">10.2.2</span> Subsampling the sequencing files to make the analysis quicker for this tutorial</h3>
<p>This Python code defines a function called subsample that takes in a FASTQ file name, an output directory, and a depth value (defaulting to 1000000). The function uses the seqtk command-line tool to subsample the input FASTQ file to the desired depth and saves the output to a new file in the specified output directory. The function prints the constructed command string to the console for debugging purposes.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> subsample(filename, outdir, depth<span class="op">=</span><span class="dv">1000000</span>):</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    basename <span class="op">=</span> Path(filename).stem</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="ss">f"seqtk sample -s42 </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>depth<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>outdir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>basename<span class="sc">}</span><span class="ss">_subsample_</span><span class="sc">{</span>depth<span class="sc">}</span><span class="ss">.fastq"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(cmd)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    subprocess.check_output(cmd, shell<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This Python code uses a for loop to iterate over all the files in the <code>../data/raw/</code> directory that match the pattern <code>*</code>, and calls the <code>subsample</code> (defined above) function on each file in the directory <code>../data/subsampled</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> f <span class="kw">in</span> glob.glob(<span class="st">"../data/raw/*"</span>):</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    outdir <span class="op">=</span> <span class="st">"../data/subsampled"</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    subsample(f, outdir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb6"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>seqtk sample -s42 ../data/raw/ERR5766177_PE.mapped.hostremoved.fwd.fq.gz 1000000 &gt;</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>seqtk sample -s42 ../data/raw/ERR5766177_PE.mapped.hostremoved.rev.fq.gz 1000000 &gt;</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>../data/subsampled/ERR5766177_PE.mapped.hostremoved.rev.fq_subsample_1000000.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Finally, we compress all files to gzip format</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gzip</span> <span class="at">-f</span> ../data/subsampled/<span class="pp">*</span>.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</div>
</div>
</div>
</section>
<section id="working-in-a-jupyter-environment" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="working-in-a-jupyter-environment"><span class="header-section-number">10.3</span> Working in a jupyter environment</h2>
<p>This tutorial run-through is using a Jupyter Notebook (<a href="https://jupyter.org">https://jupyter.org</a>) for writing &amp; executing Python code and for annotating.</p>
<p>Jupyter notebooks are convenient and have two types of cells: <em>Markdown</em> and <em>Code</em>. The <em>markup cell</em> syntax is very similar to <em>R markdown</em>. The markdown cells are used for annotating, which is important for sharing code with collaborators, reproducibility, and documentation.</p>
<p>To load, please run the following command from within the chapter’s directory.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> notebooks/</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="ex">jupyter</span> notebook analysis.ipynb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can then follow that notebook, which should mirror the contents of this chapter! Otherwise try making a new notebook within Jupyter <code>File &gt; New &gt; Notebook</code>!</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you wish to run all commands manually (i.e., without the notebook), you must make sure you run all commands while within the <code>notebook</code> directory of this chapter.</p>
</div>
</div>
</section>
<section id="data-pre-processing" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="data-pre-processing"><span class="header-section-number">10.4</span> Data pre-processing</h2>
<p>Before starting to analyze our data, we will need to pre-process them to remove reads mapping to the host genome, here, <em>Homo sapiens</em>.</p>
<p>To do so, I’ve used the first steps of the nf-core/eager <span class="citation" data-cites="Fellows-Yates2020-iy">(<a href="#ref-Fellows-Yates2020-iy" role="doc-biblioref">Fellows Yates et al. 2020</a>)</span> pipeline, more information of which can be found in the <a href="./ancient-metagenomic-pipelines.html">Ancient Metagenomic Pipelines</a> chapter.</p>
<p>I’ve already done some pre-processed the data, and the resulting cleaned files are available in the <code>data/eager_cleaned/</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Self-guided: onstructions for manual pre-preparation of data
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you wish to re-pre-prepare the data yourself, the basic eager command to do so is below, running on the output of the previous block in <a href="#chapter-overview">chapter overview</a>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">nextflow</span> run nf-core/eager <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>-r 2.4.7 <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>-profile <span class="op">&lt;</span>docker/singularity/podman/conda/institute<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>--input <span class="st">'*_R{1,2}.fastq.gz'</span> <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>--fasta <span class="st">'human_genome.fasta'</span> <span class="dt">\</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>--hostremoval_input_fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="adapter-sequence-trimming-and-low-quality-bases-trimming" class="level2" data-number="10.5">
<h2 data-number="10.5" class="anchored" data-anchor-id="adapter-sequence-trimming-and-low-quality-bases-trimming"><span class="header-section-number">10.5</span> Adapter sequence trimming and low-quality bases trimming</h2>
<p>Sequencing adapters are small DNA sequences adding prior to DNA sequencing to allow the DNA fragments to attach to the sequencing flow cells (see <a href="./introduction-to-ngs-sequencing.html">Introduction to NGS Sequencing</a>). Because these adapters could interfere with downstream analyses, we need to remove them before proceeding any further. Furthermore, because the quality of the sequencing is not always optimal, we need to remove bases of lower sequencing quality to might lead to spurious results in downstream analyses.</p>
<p>To perform both of these tasks, we’ll use the program fastp (<a href="https://github.com/OpenGene/fastp">https://github.com/OpenGene/fastp</a>) by Chen et al. <span class="citation" data-cites="chenFastpUltrafastAllinone2018">(<a href="#ref-chenFastpUltrafastAllinone2018" role="doc-biblioref">2018</a>)</span>.</p>
<p>The following command gets you the help of fastp (the <code>--help</code> option is a common option in command-line tools that displays a list of available options and their descriptions).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>option needs value: --html
usage: fastp [options] ...
options:
  -i, --in1                            read1 input file name (string [=])
  -o, --out1                           read1 output file name (string [=])
  -I, --in2                            read2 input file name (string [=])
  -O, --out2                           read2 output file name (string [=])
      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1.
                                       Default is to discard it. (string [=])
      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2.
                                       If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be
                                       written to this same file. (string [=])
      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched
                                       base. (string [=])
      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])
  -m, --merge                          for paired-end input, merge each pair of reads into a single read if they are
                                       overlapped. The merged reads will be written
                                       to the file given by --merged_out, the unmerged reads will be written to the
                                       files specified by --out1 and --out2. The merging mode is disabled by default.
      --merged_out                     in the merging mode, specify the file name to store merged output, or specify
                                       --stdout to stream the merged output (string [=])
      --include_unmerged               in the merging mode, write the unmerged or unpaired reads to the file specified
                                       by --merge. Disabled by default.
  -6, --phred64                        indicate the input is using phred64 scoring (it'll be converted to phred33,
                                       so the output will still be phred33)
  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])
      --stdin                          input from STDIN. If the STDIN is interleaved paired-end FASTQ, please also add --interleaved_in.
      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved
                                       FASTQ output for paired-end output. Disabled by default.
      --interleaved_in                 indicate that &lt;in1&gt; is an interleaved FASTQ which contains both read1 and read2.
                                       Disabled by default.
      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])
      --dont_overwrite                 don't overwrite existing files. Overwritting is allowed by default.
      --fix_mgi_id                     the MGI FASTQ ID format is not compatible with many BAM operation tools, enable this option to fix it.
  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).
  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled
  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected.
                                       For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])
      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped.
                                       If not specified, it will be the same as &lt;adapter_sequence&gt; (string [=auto])
      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])
      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this
                                    option to enable it for PE data.
  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])
  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])
  -b, --max_len1                       if read1 is longer than max_len1, then trim read1 at its tail to make it as
                                       long as max_len1. Default 0 means no limitation (int [=0])
  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])
  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])
  -B, --max_len2                       if read2 is longer than max_len2, then trim read2 at its tail to make it as long as max_len2.
                                       Default 0 means no limitation. If it's not specified, it will follow read1's settings (int [=0])
  -D, --dedup                          enable deduplication to drop the duplicated reads/pairs
      --dup_calc_accuracy              accuracy level to calculate duplication (1~6), higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G).
                                       Default 1 for no-dedup mode, and 3 for dedup mode. (int [=0])
      --dont_eval_duplication          don't evaluate duplication rate to save time and use less memory.
  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data
      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])
  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data
  -x, --trim_poly_x                    enable polyX trimming in 3' ends.
      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])
  -5, --cut_front                      move a sliding window from front (5') to tail, drop the bases in the window if
                                       its mean quality &lt; threshold, stop otherwise.
  -3, --cut_tail                       move a sliding window from tail (3') to front, drop the bases in the window if
                                       its mean quality &lt; threshold, stop otherwise.
  -r, --cut_right                      move a sliding window from front to tail, if meet one window with mean quality
                                       &lt; threshold, drop the bases in the window and the right part, and then stop.
  -W, --cut_window_size                the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])
  -M, --cut_mean_quality               the mean quality requirement option shared by cut_front, cut_tail or cut_sliding.
                                       Range: 1~36 default: 20 (Q20) (int [=20])
      --cut_front_window_size          the window size option of cut_front, default to cut_window_size if not specified (int [=4])
      --cut_front_mean_quality         the mean quality requirement option for cut_front, default to cut_mean_quality if not specified (int [=20])
      --cut_tail_window_size           the window size option of cut_tail, default to cut_window_size if not specified (int [=4])
      --cut_tail_mean_quality          the mean quality requirement option for cut_tail, default to cut_mean_quality if not specified (int [=20])
      --cut_right_window_size          the window size option of cut_right, default to cut_window_size if not specified (int [=4])
      --cut_right_mean_quality         the mean quality requirement option for cut_right, default to cut_mean_quality if not specified (int [=20])
  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled
  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])
  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])
  -n, --n_base_limit                   if one read's number of N base is &gt;n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])
  -e, --average_qual                   if one read's average quality score &lt;avg_qual, then this read/pair is discarded.
                                       Default 0 means no requirement (int [=0])
  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled
  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])
      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])
  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base
                                       that is different from its next base (base[i] != base[i+1]).
  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])
      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])
      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])
      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])
  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled
      --overlap_len_require            the minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge,
                                       adapter trimming and correction. 30 by default. (int [=30])
      --overlap_diff_limit             the maximum number of mismatched bases to detect overlapped region of PE reads.
                                       This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default. (int [=5])
      --overlap_diff_percent_limit     the maximum percentage of mismatched bases to detect overlapped region of PE reads.
                                       This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%. (int [=20])
  -U, --umi                            enable unique molecular identifier (UMI) preprocessing
      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])
      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])
      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e.
                                       prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])
      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])
  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.
  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation
                                       analysis (1~10000), smaller is slower, default is 20. (int [=20])
  -j, --json                           the json format report file name (string [=fastp.json])
  -h, --html                           the html format report file name (string [=fastp.html])
  -R, --report_title                   should be quoted with ' or ", default is "fastp report" (string [=fastp report])
  -w, --thread                         worker thread number, default is 3 (int [=3])
  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix
                                       will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])
  -S, --split_by_lines                 split output by limiting lines of each file with this option(&gt;=1000), a sequential number prefix will be
                                       added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])
  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as
                                       0001.xxx, 0 to disable padding (int [=4])
      --cut_by_quality5                DEPRECATED, use --cut_front instead.
      --cut_by_quality3                DEPRECATED, use --cut_tail instead.
      --cut_by_quality_aggressive      DEPRECATED, use --cut_right instead.
      --discard_unmerged               DEPRECATED, no effect now, see the introduction for merging.
  -?, --help                           print this message</code></pre>
</div>
</div>
</div>
<p>Here we use fastp to preprocess a pair of FASTQ files. The code specifies the input files, merges the paired-end reads on their overlaps, removes duplicate reads, and generates JSON and HTML reports. The output files are saved in the <code>../results/fastp/ directory</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--in1</span> ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--in2</span> ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--merge</span> <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--merged_out</span> ../results/fastp/ERR5766177.merged.fastq.gz <span class="dt">\</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--include_unmerged</span> <span class="dt">\</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--dedup</span> <span class="dt">\</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--json</span> ../results/fastp/ERR5766177.fastp.json <span class="dt">\</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--html</span> ../results/fastp/ERR5766177.fastp.html <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>Read1 before filtering:
total reads: 1000000
total bases: 101000000
Q20 bases: 99440729(98.4562%)
Q30 bases: 94683150(93.7457%)

Read2 before filtering:
total reads: 1000000
total bases: 101000000
Q20 bases: 99440729(98.4562%)
Q30 bases: 94683150(93.7457%)

Merged and filtered:
total reads: 1994070
total bases: 201397311
Q20 bases: 198330392(98.4772%)
Q30 bases: 188843169(93.7665%)

Filtering result:
reads passed filter: 1999252
reads failed due to low quality: 728
reads failed due to too many N: 20
reads failed due to too short: 0
reads with adapter trimmed: 282
bases trimmed due to adapters: 18654
reads corrected by overlap analysis: 0
bases corrected by overlap analysis: 0

Duplication rate: 0.2479%

Insert size peak (evaluated by paired-end reads): 31

Read pairs merged: 228
% of original read pairs: 0.0228%
% in reads after filtering: 0.0114339%


JSON report: ../results/fastp/ERR5766177.fastp.json
HTML report: ../results/fastp/ERR5766177.fastp.html

fastp --in1 ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz \
--in2 ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz --merge \
--merged_out ../results/fastp/ERR5766177.merged.fastq.gz --include_unmerged --dedup \
--json ../results/fastp/ERR5766177.fastp.json --html ../results/fastp/ERR5766177.fastp.html
fastp v0.23.2, time used: 11 seconds</code></pre>
</div>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What do you think of the number of read pairs that were merged ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Here, only 228 read pairs were merged. This is due to the length of the reads of 100bp, and length of the DNA fragments. If you would use fewer cycles, and have shorter DNA fragments, you would expect this number to go up.</p>
</div>
</div>
</div>
</section>
<section id="taxonomic-profiling-with-metaphlan" class="level2" data-number="10.6">
<h2 data-number="10.6" class="anchored" data-anchor-id="taxonomic-profiling-with-metaphlan"><span class="header-section-number">10.6</span> Taxonomic profiling with Metaphlan</h2>
<p>MetaPhlAn is a computational tool for profiling the composition of microbial communities from metagenomic shotgun sequencing data.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">metaphlan</span>  <span class="at">--help</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>usage: metaphlan --input_type {fastq,fasta,bowtie2out,sam} [--force]
                 [--bowtie2db METAPHLAN_BOWTIE2_DB] [-x INDEX]
                 [--bt2_ps BowTie2 presets] [--bowtie2_exe BOWTIE2_EXE]
                 [--bowtie2_build BOWTIE2_BUILD] [--bowtie2out FILE_NAME]
                 [--min_mapq_val MIN_MAPQ_VAL] [--no_map] [--tmp_dir]
                 [--tax_lev TAXONOMIC_LEVEL] [--min_cu_len]
                 [--min_alignment_len] [--add_viruses] [--ignore_eukaryotes]
                 [--ignore_bacteria] [--ignore_archaea] [--stat_q]
                 [--perc_nonzero] [--ignore_markers IGNORE_MARKERS]
                 [--avoid_disqm] [--stat] [-t ANALYSIS TYPE]
                 [--nreads NUMBER_OF_READS] [--pres_th PRESENCE_THRESHOLD]
                 [--clade] [--min_ab] [-o output file] [--sample_id_key name]
                 [--use_group_representative] [--sample_id value]
                 [-s sam_output_file] [--legacy-output] [--CAMI_format_output]
                 [--unknown_estimation] [--biom biom_output] [--mdelim mdelim]
                 [--nproc N] [--install] [--force_download]
                 [--read_min_len READ_MIN_LEN] [-v] [-h]
                 [INPUT_FILE] [OUTPUT_FILE]

DESCRIPTION
 MetaPhlAn version 3.1.0 (25 Jul 2022):
 METAgenomic PHyLogenetic ANalysis for metagenomic taxonomic profiling.

AUTHORS: Francesco Beghini (francesco.beghini@unitn.it),Nicola Segata (nicola.segata@unitn.it), Duy Tin Truong,
Francesco Asnicar (f.asnicar@unitn.it), Aitor Blanco Miguez (aitor.blancomiguez@unitn.it)

COMMON COMMANDS

 We assume here that MetaPhlAn is installed using the several options available (pip, conda, PyPi)
 Also BowTie2 should be in the system path with execution and read permissions, and Perl should be installed)

========== MetaPhlAn clade-abundance estimation =================

The basic usage of MetaPhlAn consists in the identification of the clades (from phyla to species )
present in the metagenome obtained from a microbiome sample and their
relative abundance. This correspond to the default analysis type (-t rel_ab).

*  Profiling a metagenome from raw reads:
$ metaphlan metagenome.fastq --input_type fastq -o profiled_metagenome.txt

*  You can take advantage of multiple CPUs and save the intermediate BowTie2 output for re-running
   MetaPhlAn extremely quickly:
$ metaphlan metagenome.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq -o profiled_metagenome.txt

*  If you already mapped your metagenome against the marker DB (using a previous MetaPhlAn run), you
   can obtain the results in few seconds by using the previously saved --bowtie2out file and
   specifying the input (--input_type bowtie2out):
$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out -o profiled_metagenome.txt

*  bowtie2out files generated with MetaPhlAn versions below 3 are not compatibile.
   Starting from MetaPhlAn 3.0, the BowTie2 ouput now includes the size of the profiled metagenome and the average read length.
   If you want to re-run MetaPhlAn using these file you should provide the metagenome s via --nreads:
$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out --nreads 520000 -o profiled_metagenome.txt

*  You can also provide an externally BowTie2-mapped SAM if you specify this format with
   --input_type. Two steps: first apply BowTie2 and then feed MetaPhlAn with the obtained sam:
$ bowtie2 --sam-no-hd --sam-no-sq --no-unal --very-sensitive -S metagenome.sam -x \
  ${mpa_dir}/metaphlan_databases/mpa_v30_CHOCOPhlAn_201901 -U metagenome.fastq
$ metaphlan metagenome.sam --input_type sam -o profiled_metagenome.txt

*  We can also natively handle paired-end metagenomes, and, more generally, metagenomes stored in
  multiple files (but you need to specify the --bowtie2out parameter):
$ metaphlan metagenome_1.fastq,metagenome_2.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq

-------------------------------------------------------------------


========== Marker level analysis ============================

MetaPhlAn introduces the capability of characterizing organisms at the strain level using non
aggregated marker information. Such capability comes with several slightly different flavours and
are a way to perform strain tracking and comparison across multiple samples.
Usually, MetaPhlAn is first ran with the default -t to profile the species present in
the community, and then a strain-level profiling can be performed to zoom-in into specific species
of interest. This operation can be performed quickly as it exploits the --bowtie2out intermediate
file saved during the execution of the default analysis type.

*  The following command will output the abundance of each marker with a RPK (reads per kilo-base)
   higher 0.0. (we are assuming that metagenome_outfmt.bz2 has been generated before as
   shown above).
$ metaphlan -t marker_ab_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt
   The obtained RPK can be optionally normald by the total number of reads in the metagenome
   to guarantee fair comparisons of abundances across samples. The number of reads in the metagenome
   needs to be passed with the '--nreads' argument

*  The list of markers present in the sample can be obtained with '-t marker_pres_table'
$ metaphlan -t marker_pres_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt
   The --pres_th argument (default 1.0) set the minimum RPK value to consider a marker present

*  The list '-t clade_profiles' analysis type reports the same information of '-t marker_ab_table'
   but the markers are reported on a clade-by-clade basis.
$ metaphlan -t clade_profiles metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt

*  Finally, to obtain all markers present for a specific clade and all its subclades, the
   '-t clade_specific_strain_tracker' should be used. For example, the following command
   is reporting the presence/absence of the markers for the B. fragilis species and its strains
   the optional argument --min_ab specifies the minimum clade abundance for reporting the markers

$ metaphlan -t clade_specific_strain_tracker --clade s__Bacteroides_fragilis metagenome_outfmt.bz2 --input_typ
  bowtie2out -o marker_abundance_table.txt

-------------------------------------------------------------------

positional arguments:
  INPUT_FILE            the input file can be:
                        * a fastq file containing metagenomic reads
                        OR
                        * a BowTie2 produced SAM file.
                        OR
                        * an intermediary mapping file of the metagenome generated by a previous MetaPhlAn run
                        If the input file is missing, the script assumes that the input is provided using the standard
                        input, or named pipes.
                        IMPORTANT: the type of input needs to be specified with --input_type
  OUTPUT_FILE           the tab-separated output file of the predicted taxon relative abundances
                        [stdout if not present]

Required arguments:
  --input_type {fastq,fasta,bowtie2out,sam}
                        set whether the input is the FASTA file of metagenomic reads or
                        the SAM file of the mapping of the reads against the MetaPhlAn db.

Mapping arguments:
  --force               Force profiling of the input file by removing the bowtie2out file
  --bowtie2db METAPHLAN_BOWTIE2_DB
                        Folder containing the MetaPhlAn database. You can specify the location by exporting the
                        DEFAULT_DB_FOLDER variable in the shell.
                        [default /Users/maxime/mambaforge/envs/summer_school_microbiome/lib/python3.9/site-packages/metaphlan/metaphlan_databases]
  -x INDEX, --index INDEX
                        Specify the id of the database version to use. If "latest", MetaPhlAn will get the latest version.
                        If an index name is provided, MetaPhlAn will try to use it, if available, and skip the online check.
                        If the database files are not found on the local MetaPhlAn installation they
                        will be automatically downloaded [default latest]
  --bt2_ps BowTie2 presets
                        Presets options for BowTie2 (applied only when a FASTA file is provided)
                        The choices enabled in MetaPhlAn are:
                         * sensitive
                         * very-sensitive
                         * sensitive-local
                         * very-sensitive-local
                        [default very-sensitive]
  --bowtie2_exe BOWTIE2_EXE
                        Full path and name of the BowTie2 executable. This option allowsMetaPhlAn to reach the
                        executable even when it is not in the system PATH or the system PATH is unreachable
  --bowtie2_build BOWTIE2_BUILD
                        Full path to the bowtie2-build command to use, deafult assumes that 'bowtie2-build is present in the system path
  --bowtie2out FILE_NAME
                        The file for saving the output of BowTie2
  --min_mapq_val MIN_MAPQ_VAL
                        Minimum mapping quality value (MAPQ) [default 5]
  --no_map              Avoid storing the --bowtie2out map file
  --tmp_dir             The folder used to store temporary files [default is the OS dependent tmp dir]

Post-mapping arguments:
  --tax_lev TAXONOMIC_LEVEL
                        The taxonomic level for the relative abundance output:
                        'a' : all taxonomic levels
                        'k' : kingdoms
                        'p' : phyla only
                        'c' : classes only
                        'o' : orders only
                        'f' : families only
                        'g' : genera only
                        's' : species only
                        [default 'a']
  --min_cu_len          minimum total nucleotide length for the markers in a clade for
                        estimating the abundance without considering sub-clade abundances
                        [default 2000]
  --min_alignment_len   The sam records for aligned reads with the longest subalignment
                        length smaller than this threshold will be discarded.
                        [default None]
  --add_viruses         Allow the profiling of viral organisms
  --ignore_eukaryotes   Do not profile eukaryotic organisms
  --ignore_bacteria     Do not profile bacterial organisms
  --ignore_archaea      Do not profile archeal organisms
  --stat_q              Quantile value for the robust average
                        [default 0.2]
  --perc_nonzero        Percentage of markers with a non zero relative abundance for misidentify a species
                        [default 0.33]
  --ignore_markers IGNORE_MARKERS
                        File containing a list of markers to ignore.
  --avoid_disqm         Deactivate the procedure of disambiguating the quasi-markers based on the
                        marker abundance pattern found in the sample. It is generally recommended
                        to keep the disambiguation procedure in order to minimize false positives
  --stat                Statistical approach for converting marker abundances into clade abundances
                        'avg_g'  : clade global (i.e. normalizing all markers together) average
                        'avg_l'  : average of length-normalized marker counts
                        'tavg_g' : truncated clade global average at --stat_q quantile
                        'tavg_l' : truncated average of length-normalized marker counts (at --stat_q)
                        'wavg_g' : winsorized clade global average (at --stat_q)
                        'wavg_l' : winsorized average of length-normalized marker counts (at --stat_q)
                        'med'    : median of length-normalized marker counts
                        [default tavg_g]

Additional analysis types and arguments:
  -t ANALYSIS TYPE      Type of analysis to perform:
                         * rel_ab: profiling a metagenomes in terms of relative abundances
                         * rel_ab_w_read_stats: profiling a metagenomes in terms of relative abundances and estimate
                                                the number of reads coming from each clade.
                         * reads_map: mapping from reads to clades (only reads hitting a marker)
                         * clade_profiles: normalized marker counts for clades with at least a non-null marker
                         * marker_ab_table: normalized marker counts (only when &gt; 0.0 and normalized by metagenome size if --nreads is specified)
                         * marker_counts: non-normalized marker counts [use with extreme caution]
                         * marker_pres_table: list of markers present in the sample (threshold at 1.0 if not differently specified with --pres_th
                         * clade_specific_strain_tracker: list of markers present for a specific clade, specified with --clade, and all its subclades
                        [default 'rel_ab']
  --nreads NUMBER_OF_READS
                        The total number of reads in the original metagenome. It is used only when
                        -t marker_table is specified for normalizing the length-normalized counts
                        with the metagenome size as well. No normalization applied if --nreads is not
                        specified
  --pres_th PRESENCE_THRESHOLD
                        Threshold for calling a marker present by the -t marker_pres_table option
  --clade               The clade for clade_specific_strain_tracker analysis
  --min_ab              The minimum percentage abundance for the clade in the clade_specific_strain_tracker analysis

Output arguments:
  -o output file, --output_file output file
                        The output file (if not specified as positional argument)
  --sample_id_key name  Specify the sample ID key for this analysis. Defaults to 'SampleID'.
  --use_group_representative
                        Use a species as representative for species groups.
  --sample_id value     Specify the sample ID for this analysis. Defaults to 'Metaphlan_Analysis'.
  -s sam_output_file, --samout sam_output_file
                        The sam output file
  --legacy-output       Old MetaPhlAn2 two columns output
  --CAMI_format_output  Report the profiling using the CAMI output format
  --unknown_estimation  Scale relative abundances to the number of reads mapping to known clades in order to estimate unknowness
  --biom biom_output, --biom_output_file biom_output
                        If requesting biom file output: The name of the output file in biom format
  --mdelim mdelim, --metadata_delimiter_char mdelim
                        Delimiter for bug metadata: - defaults to pipe. e.g. the pipe in k__Bacteria|p__Proteobacteria

Other arguments:
  --nproc N             The number of CPUs to use for parallelizing the mapping [default 4]
  --install             Only checks if the MetaPhlAn DB is installed and installs it if not. All other parameters are ignored.
  --force_download      Force the re-download of the latest MetaPhlAn database.
  --read_min_len READ_MIN_LEN
                        Specify the minimum length of the reads to be considered when parsing the input file with
                        'read_fastx.py' script, default value is 70
  -v, --version         Prints the current MetaPhlAn version and exit
  -h, --help            show this help message and exit</code></pre>
</div>
</div>
</div>
<p>The following command uses MetaPhlAn to profile the taxonomic composition of the ERR5766177 metagenomic sample. The input file is specified as a merged FASTQ file, and the output is saved as a text file containing the taxonomic profile. The <code>--bowtie2out</code> option is used to specify the output file for the Bowtie2 alignment, and the –nproc option is used to specify the number of CPUs to use for the analysis.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">metaphlan</span> ../results/fastp/ERR5766177.merged.fastq.gz  <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input_type</span> fastq <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--bowtie2out</span> ../results/metaphlan/ERR5766177.bt2.out  <span class="dt">\</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--nproc</span> 4 <span class="dt">\</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">&gt;</span> ../results/metaphlan/ERR5766177.metaphlan_profile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The main results files that we’re interested in is located at <code>../results/metaphlan/ERR5766177.metaphlan_profile.txt</code></p>
<p>It’s a tab separated file, with taxons in rows, with their relative abundance in the sample</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span> ../results/metaphlan/ERR5766177.metaphlan_profile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>#mpa_v30_CHOCOPhlAn_201901
#/home/maxime_borry/.conda/envs/maxime/envs/summer_school_microbiome/bin/metaphlan ../results/fastp/ERR5766177.merged.fastq.gz \
--input_type fastq --bowtie2out ../results/metaphlan/ERR5766177.bt2.out --nproc 8
#SampleID   Metaphlan_Analysis
#clade_name NCBI_tax_id relative_abundance  additional_species
k__Bacteria 2   82.23198
k__Archaea  2157    17.76802
k__Bacteria|p__Firmicutes   2|1239  33.47957
k__Bacteria|p__Bacteroidetes    2|976   28.4209
k__Bacteria|p__Actinobacteria   2|201174    20.33151
k__Archaea|p__Euryarchaeota 2157|28890  17.76802</code></pre>
</div>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the letter preceding the <code>__</code> before every clade name ? Eg. <code>k__Bacteria</code></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>This letter corresponds to the rank of the clade. - k for Kingdom - p for Phylum - […] - f for Family - g for Genus - s for Species</p>
</div>
</div>
</div>
</section>
<section id="visualizing-the-taxonomic-profile" class="level2" data-number="10.7">
<h2 data-number="10.7" class="anchored" data-anchor-id="visualizing-the-taxonomic-profile"><span class="header-section-number">10.7</span> Visualizing the taxonomic profile</h2>
<section id="visualizing-metaphlan-taxonomic-profile-with-pavian" class="level3" data-number="10.7.1">
<h3 data-number="10.7.1" class="anchored" data-anchor-id="visualizing-metaphlan-taxonomic-profile-with-pavian"><span class="header-section-number">10.7.1</span> Visualizing metaphlan taxonomic profile with Pavian</h3>
<p>Pavian (<a href="https://github.com/fbreitwieser/pavian">https://github.com/fbreitwieser/pavian</a>) by Breitweiser et al. <span class="citation" data-cites="Breitwieser2020-zw">(<a href="#ref-Breitwieser2020-zw" role="doc-biblioref">2020</a>)</span> is a web-based tool for interactive visualization and analysis of metagenomics data. It provides a user-friendly interface for exploring taxonomic and functional profiles of microbial communities, and allows users to generate interactive plots and tables that can be customised and shared (<a href="#fig-taxonomicprofiling-pavianinterface" class="quarto-xref">Figure&nbsp;<span>10.5</span></a>).</p>
<p>You can open Pavian in your browser by visiting <a href="https://fbreitwieser.shinyapps.io/pavian/.">https://fbreitwieser.shinyapps.io/pavian</a>.</p>
<div id="fig-taxonomicprofiling-pavianinterface" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-pavianinterface-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/pavian.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-pavianinterface-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.5: Screenshot of the pavian metagenomics visualisation interface, with menus on the left, a select sample and filter taxa search bar at the top, and a Sankey visualisation of the example metagenome sample
</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Expand for instructions for running yourself">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand for instructions for running yourself
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are different ways to run it:</p>
<ul>
<li><p>If you have docker (<a href="https://www.docker.com/">https://www.docker.com/</a>) installed</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull <span class="st">'florianbw/pavian'</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-p</span> 5000:80 florianbw/pavian</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
<p>Then open your browser and visit <a href="localhost:5000">localhost:5000</a></p>
<ul>
<li><p>If you are familiar with R (<a href="https://www.r-project.org/">https://www.r-project.org/</a>)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(remotes)) {</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"remotes"</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"fbreitwieser/pavian"</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>pavian<span class="sc">::</span><span class="fu">runApp</span>(<span class="at">port =</span> <span class="dv">5000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></li>
</ul>
<p>Then open your browser and visit <a href="localhost:5000">localhost:5000</a></p>
</div>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the relative abundance of the phylum Firmicutes ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>In this example, the relative abundance of the Firmicutes (officially renamed Bacillota since 2021) is 33.5%. You can verify this number either in the Sankey plot (section “Sample”), or in the comparison table (section “Comparison”) by selecting the Phylum tab.</p>
</div>
</div>
</div>
</section>
<section id="visualizing-metaphlan-taxonomic-profile-with-krona" class="level3" data-number="10.7.2">
<h3 data-number="10.7.2" class="anchored" data-anchor-id="visualizing-metaphlan-taxonomic-profile-with-krona"><span class="header-section-number">10.7.2</span> Visualizing metaphlan taxonomic profile with Krona</h3>
<p>Krona (<a href="https://github.com/marbl/Krona/wiki">https://github.com/marbl/Krona/wiki</a>) by Ondov et al. <span class="citation" data-cites="Ondov2011-se">(<a href="#ref-Ondov2011-se" role="doc-biblioref">Brian D. Ondov, Bergman, and Phillippy 2011b</a>)</span> is a software tool for interactive visualization of hierarchical data, such as taxonomic profiles generated by metagenomics tools like MetaPhlAn. Krona allows users to explore the taxonomic composition of microbial communities in a hierarchical manner, from the highest taxonomic level down to the species level.</p>
<p>The <code>metaphlan2krona.py</code> script is used to convert the MetaPhlAn taxonomic profile output to a format that can be visualised by Krona. The output of the script is a text file that contains the taxonomic profile in a hierarchical format that can be read by Krona. The <code>ktImportText</code> command is then used to generate an interactive HTML file that displays the taxonomic profile in a hierarchical manner using Krona.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> ../scripts/metaphlan2krona.py <span class="at">-p</span> ../results/metaphlan/ERR5766177.metaphlan_profile.txt <span class="at">-k</span> ../results/krona/ERR5766177_krona.out</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="ex">ktImportText</span> <span class="at">-o</span> ../results/krona/ERR5766177_krona.html ../results/krona/ERR5766177_krona.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>Writing ../results/krona/ERR5766177_krona.html...</code></pre>
</div>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which proportion of the Firmicutes is made of Clostridiales ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The Clostridiales represent 63% of the Firmicutes in this sample. You can verify this by clicking in the Clostridiales and looking at the pie chart for Firmicutes on the top right of the screen.</p>
</div>
</div>
</div>
</section>
</section>
<section id="getting-modern-comparative-reference-data" class="level2" data-number="10.8">
<h2 data-number="10.8" class="anchored" data-anchor-id="getting-modern-comparative-reference-data"><span class="header-section-number">10.8</span> Getting modern comparative reference data</h2>
<p>In order to compare our sample with modern reference samples, I used the curatedMetagenomicsData package, which provides both curated metadata, and pre-computed metaphlan taxonomic profiles for published modern human samples.</p>
<p>The full R code to get these data is available in <code>curatedMetagenomics/get_sources.Rmd</code>.</p>
<p>I pre-selected 200 gut microbiome samples from non-westernised (100) and westernised (100) from healthy, non-antibiotic users donors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedMetagenomicData)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter samples based on specific criteria</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>sampleMetadata <span class="sc">%&gt;%</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(body_site <span class="sc">==</span> <span class="st">"stool"</span> <span class="sc">&amp;</span> antibiotics_current_use <span class="sc">==</span> <span class="st">"no"</span> <span class="sc">&amp;</span> disease <span class="sc">==</span> <span class="st">"healthy"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(non_westernized) <span class="sc">%&gt;%</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sample_n</span>(<span class="dv">100</span>) <span class="sc">%&gt;%</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ungroup</span>() <span class="ot">-&gt;</span> selected_samples</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract relative abundance data for selected samples</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>selected_samples <span class="sc">%&gt;%</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">returnSamples</span>(<span class="st">"relative_abundance"</span>) <span class="ot">-&gt;</span> rel_ab</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Split relative abundance data by taxonomic rank and write to CSV files</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>data_ranks <span class="ot">&lt;-</span> <span class="fu">splitByRanks</span>(rel_ab)</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="fu">names</span>(data_ranks)) {</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Print taxonomic rank and output file name</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(r)</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>    output_file <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"../../data/curated_metagenomics/modern_sources_"</span>, <span class="fu">tolower</span>(r), <span class="st">".csv"</span>)</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(output_file)</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Write relative abundance data to CSV file</span></span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>    assay_rank <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">assay</span>(data_ranks[[r]]))</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">write.csv</span>(assay_rank, output_file)</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>The resulting pre-computed metaphlan taxonomic profiles (split by taxonomic ranks) are available in <code>data/curated_metagenomics</code></li>
<li>The associated metadata is available at <code>data/metadata/curated_metagenomics_modern_sources.csv</code></li>
</ul>
</section>
<section id="loading-the-ancient-sample-taxonomic-profile" class="level2" data-number="10.9">
<h2 data-number="10.9" class="anchored" data-anchor-id="loading-the-ancient-sample-taxonomic-profile"><span class="header-section-number">10.9</span> Loading the ancient sample taxonomic profile</h2>
<p>This is the moment where we will use the Pandas (<a href="https://pandas.pydata.org">https://pandas.pydata.org</a>) Python library (<a href="https://www.python.org/">https://www.python.org/</a>) to perform some data manipulation. We will also use the Taxopy (<a href="https://github.com/apcamargo/taxopy">https://github.com/apcamargo/taxopy</a>) library to work with taxonomic information.</p>
<p>In python we need to import necessary libraries, i.e.&nbsp;pandas and taxopy, and a couple of other utility libraries.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> taxopy</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gzip</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And we then create an instance of the taxopy taxonomy database. This will take a few seconds/minutes, as it needs to download the entire NCBI taxonomy before storing in a local database.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>taxdb <span class="op">=</span> taxopy.TaxDb()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s read the metaphlan profile table with pandas (a python package with a similar concept to tidyverse dyplyr, tidyr packa). It’s a tab separated file, so we need to specify the delimiter as <code>\t</code>, and skip the comment lines of the files that start with <code>#</code>.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> pd.read_csv(<span class="st">"../results/metaphlan/ERR5766177.metaphlan_profile.txt"</span>,</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                            comment<span class="op">=</span><span class="st">"#"</span>,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                            delimiter<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                            names<span class="op">=</span>[<span class="st">'clade_name'</span>,<span class="st">'NCBI_tax_id'</span>,<span class="st">'relative_abundance'</span>,<span class="st">'additional_species'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To look at the head of a dataframe (<a href="#tbl-taxonomicprofiling-mp3raw" class="quarto-xref">Table&nbsp;<span>10.1</span></a>) with pandas</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-mp3raw" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-mp3raw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.1: Top few lines of a MetaPhlAn taxonomic profile
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-mp3raw-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 3%">
<col style="width: 36%">
<col style="width: 14%">
<col style="width: 22%">
<col style="width: 22%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>clade_name</th>
<th>NCBI_tax_id</th>
<th>relative_abundance</th>
<th>additional_species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>k__Bacteria</td>
<td>2</td>
<td>82.23198</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>1</td>
<td>k__Archaea</td>
<td>2157</td>
<td>17.76802</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td>2</td>
<td>k__Bacteria|p__Firmicutes</td>
<td>2|1239</td>
<td>33.47957</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>3</td>
<td>k__Bacteria|p__Bacteroidetes</td>
<td>2|976</td>
<td>28.42090</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td>4</td>
<td>k__Bacteria|p__Actinobacteria</td>
<td>2|201174</td>
<td>20.33151</td>
<td>NaN</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>We can also specify more rows by randomly picking 10 rows to display (<a href="#tbl-taxonomicprofiling-mp3sampled" class="quarto-xref">Table&nbsp;<span>10.2</span></a>).</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ancient_data.sample(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-mp3sampled" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-mp3sampled-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.2: Ten randomly selected lines of a MetaPhlAn taxonomic profile
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-mp3sampled-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 2%">
<col style="width: 30%">
<col style="width: 26%">
<col style="width: 11%">
<col style="width: 30%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>clade_name</th>
<th>NCBI_tax_id</th>
<th>relative_abundance</th>
<th>additional_species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>k__Archaea</td>
<td>2157</td>
<td>17.76802</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>46</td>
<td>k__Bacteria|p__Bacteroidetes|c_<em>Bacteroidia|o</em>…</td>
<td>2|976|200643|171549|171552|838|165179</td>
<td>25.75544</td>
<td>k__Bacteria|p__Bacteroidetes|c_<em>Bacteroidia|o</em>…</td>
</tr>
<tr class="odd">
<td>55</td>
<td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…</td>
<td>2|1239|186801|186802|186803|189330|88431</td>
<td>0.91178</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>18</td>
<td>k__Archaea|p__Euryarchaeota|c_<em>Halobacteria|o</em>…</td>
<td>2157|28890|183963|2235</td>
<td>0.71177</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td>36</td>
<td>k__Bacteria|p__Actinobacteria|c__Actinobacteri…</td>
<td>2|201174|1760|85004|31953|1678</td>
<td>9.39377</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>65</td>
<td>k__Bacteria|p__Actinobacteria|c__Actinobacteri…</td>
<td>2|201174|1760|85004|31953|1678|216816</td>
<td>0.05447</td>
<td>k__Bacteria|p__Actinobacteria|c__Actinobacteri…</td>
</tr>
<tr class="odd">
<td>37</td>
<td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…</td>
<td>2|1239|186801|186802|186803|</td>
<td>2.16125</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>38</td>
<td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…</td>
<td>2|1239|186801|186802|541000|216851</td>
<td>1.24537</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td>26</td>
<td>k__Bacteria|p__Actinobacteria|c__Actinobacteri…</td>
<td>2|201174|1760|85004|31953</td>
<td>9.39377</td>
<td>NaN</td>
</tr>
<tr class="even">
<td>48</td>
<td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…</td>
<td>2|1239|186801|186802|541000|1263|40518</td>
<td>14.96816</td>
<td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>Because for this analysis, we’re only going to look at the relative abundance, we’ll only use this column, and the Taxonomic ID (TAXID) (<a href="https://www.ncbi.nlm.nih.gov/taxonomy">https://www.ncbi.nlm.nih.gov/taxonomy</a>) information, so we can <code>drop</code> (get rid of) the unnecessary columns.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> (</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    ancient_data</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'NCBI_tax_id'</span>: <span class="st">'TAXID'</span>})</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    .drop([<span class="st">'clade_name'</span>,<span class="st">'additional_species'</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Always investigate your data at first !</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>ancient_data.relative_abundance.<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>700.00007</code></pre>
</div>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What do you think of a 700% relative abundance ?</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Let’s proceed further and try to understand what’s happening (<a href="#tbl-taxonomicprofiling-taxidtable" class="quarto-xref">Table&nbsp;<span>10.3</span></a>).</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtable" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtable-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.3: A two column table of TAXIDs and the organisms corresponding relative abundance
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtable-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th>TAXID</th>
<th>relative_abundance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>2</td>
<td>82.23198</td>
</tr>
<tr class="even">
<td>1</td>
<td>2157</td>
<td>17.76802</td>
</tr>
<tr class="odd">
<td>2</td>
<td>2|1239</td>
<td>33.47957</td>
</tr>
<tr class="even">
<td>3</td>
<td>2|976</td>
<td>28.42090</td>
</tr>
<tr class="odd">
<td>4</td>
<td>2|201174</td>
<td>20.33151</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>To make sense of the TAXID, we will use taxopy to get all the taxonomic related informations such as (<a href="#tbl-taxonomicprofiling-taxidtablewithpath" class="quarto-xref">Table&nbsp;<span>10.4</span></a>):</p>
<ul>
<li>Name of the taxon</li>
<li>Rank of the taxon</li>
<li>Lineage of the taxon</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">## This function is here to help us get the taxon information</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co">## from the metaphlan taxonomic ID lineage, of the following form</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">## 2|976|200643|171549|171552|838|165179</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> to_taxopy(taxid_entry, taxo_db):</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Returns a taxopy taxon object</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="co">        taxid_entry(str): metaphlan TAXID taxonomic lineage</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co">        taxo_db(taxopy database)</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="co">        (bool): Returns a taxopy taxon object</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>    taxid <span class="op">=</span> taxid_entry.split(<span class="st">"|"</span>)[<span class="op">-</span><span class="dv">1</span>] <span class="co"># get the last element</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(taxid) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> taxopy.Taxon(<span class="bu">int</span>(taxid), taxo_db) <span class="co"># if it's not empty, get the taxon corresponding to the taxid</span></span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> taxopy.Taxon(<span class="dv">12908</span>, taxo_db) <span class="co"># otherwise, return the taxon associated with unclassified sequences</span></span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> taxopy.exceptions.TaxidError <span class="im">as</span> e:</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> taxopy.Taxon(<span class="dv">12908</span>, taxo_db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>ancient_data[<span class="st">'taxopy'</span>] <span class="op">=</span> ancient_data[<span class="st">'TAXID'</span>].<span class="bu">apply</span>(to_taxopy, taxo_db<span class="op">=</span>taxo_db)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtablewithpath" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtablewithpath-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.4: A three column table of TAXIDs and the organisms corresponding relative abundance, and the attached taxonomic path associated with the TAXID
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtablewithpath-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 23%">
<col style="width: 60%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>TAXID</th>
<th>relative_abundance</th>
<th>taxopy</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>2</td>
<td>82.23198</td>
<td>s__Bacteria</td>
</tr>
<tr class="even">
<td>1</td>
<td>2157</td>
<td>17.76802</td>
<td>s__Archaea</td>
</tr>
<tr class="odd">
<td>2</td>
<td>2|1239</td>
<td>33.47957</td>
<td>s__Bacteria;c__Terrabacteria group;p__Firmicutes</td>
</tr>
<tr class="even">
<td>3</td>
<td>2|976</td>
<td>28.42090</td>
<td>s__Bacteria;c__FCB group;p__Bacteroidetes</td>
</tr>
<tr class="odd">
<td>4</td>
<td>2|201174</td>
<td>20.33151</td>
<td>s__Bacteria;c__Terrabacteria group;p__Actinoba…</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<div class="sourceCode" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> ancient_data.assign(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>    rank <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.rank),</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>    name <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.name),</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    lineage <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.name_lineage),</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>ancient_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtablewithpathsep" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtablewithpathsep-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.5: A five column table of TAXIDs and the organisms corresponding relative abundance, and the attached taxonomic path associated with the TAXID, but also the rank and name of the particular taxonomic ID
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtablewithpathsep-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="bivtfxazkm" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#bivtfxazkm table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#bivtfxazkm thead, #bivtfxazkm tbody, #bivtfxazkm tfoot, #bivtfxazkm tr, #bivtfxazkm td, #bivtfxazkm th {
  border-style: none;
}

#bivtfxazkm p {
  margin: 0;
  padding: 0;
}

#bivtfxazkm .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#bivtfxazkm .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#bivtfxazkm .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#bivtfxazkm .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#bivtfxazkm .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bivtfxazkm .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bivtfxazkm .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#bivtfxazkm .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#bivtfxazkm .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#bivtfxazkm .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#bivtfxazkm .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#bivtfxazkm .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#bivtfxazkm .gt_spanner_row {
  border-bottom-style: hidden;
}

#bivtfxazkm .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#bivtfxazkm .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#bivtfxazkm .gt_from_md > :first-child {
  margin-top: 0;
}

#bivtfxazkm .gt_from_md > :last-child {
  margin-bottom: 0;
}

#bivtfxazkm .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#bivtfxazkm .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#bivtfxazkm .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#bivtfxazkm .gt_row_group_first td {
  border-top-width: 2px;
}

#bivtfxazkm .gt_row_group_first th {
  border-top-width: 2px;
}

#bivtfxazkm .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bivtfxazkm .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#bivtfxazkm .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#bivtfxazkm .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bivtfxazkm .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#bivtfxazkm .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#bivtfxazkm .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#bivtfxazkm .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#bivtfxazkm .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#bivtfxazkm .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bivtfxazkm .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bivtfxazkm .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#bivtfxazkm .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#bivtfxazkm .gt_left {
  text-align: left;
}

#bivtfxazkm .gt_center {
  text-align: center;
}

#bivtfxazkm .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#bivtfxazkm .gt_font_normal {
  font-weight: normal;
}

#bivtfxazkm .gt_font_bold {
  font-weight: bold;
}

#bivtfxazkm .gt_font_italic {
  font-style: italic;
}

#bivtfxazkm .gt_super {
  font-size: 65%;
}

#bivtfxazkm .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#bivtfxazkm .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#bivtfxazkm .gt_indent_1 {
  text-indent: 5px;
}

#bivtfxazkm .gt_indent_2 {
  text-indent: 10px;
}

#bivtfxazkm .gt_indent_3 {
  text-indent: 15px;
}

#bivtfxazkm .gt_indent_4 {
  text-indent: 20px;
}

#bivtfxazkm .gt_indent_5 {
  text-indent: 25px;
}

#bivtfxazkm .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#bivtfxazkm div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="TAXID" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">TAXID</th>
<th id="relative_abundance" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">relative_abundance</th>
<th id="taxopy" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">taxopy</th>
<th id="rank" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">rank</th>
<th id="name" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">name</th>
<th id="lineage" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">lineage</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="...1">0</td>
<td class="gt_row gt_left" headers="TAXID">2</td>
<td class="gt_row gt_right" headers="relative_abundance">82.23198</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria</td>
<td class="gt_row gt_left" headers="rank">superkingdom</td>
<td class="gt_row gt_left" headers="name">Bacteria</td>
<td class="gt_row gt_left" headers="lineage">[Bacteria, cellular organisms, root]</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">1</td>
<td class="gt_row gt_left" headers="TAXID">2157</td>
<td class="gt_row gt_right" headers="relative_abundance">17.76802</td>
<td class="gt_row gt_left" headers="taxopy">s__Archaea</td>
<td class="gt_row gt_left" headers="rank">superkingdom</td>
<td class="gt_row gt_left" headers="name">Archaea</td>
<td class="gt_row gt_left" headers="lineage">[Archaea, cellular organisms, root]</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">2</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239</td>
<td class="gt_row gt_right" headers="relative_abundance">33.47957</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicutes</td>
<td class="gt_row gt_left" headers="rank">phylum</td>
<td class="gt_row gt_left" headers="name">Firmicutes</td>
<td class="gt_row gt_left" headers="lineage">[Firmicutes, Terrabacteria group, Bacteria, ce...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">3</td>
<td class="gt_row gt_left" headers="TAXID">2\|976</td>
<td class="gt_row gt_right" headers="relative_abundance">28.4209</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__FCB group;p__Bacteroidetes</td>
<td class="gt_row gt_left" headers="rank">phylum</td>
<td class="gt_row gt_left" headers="name">Bacteroidetes</td>
<td class="gt_row gt_left" headers="lineage">[Bacteroidetes, Bacteroidetes/Chlorobi group, ...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">4</td>
<td class="gt_row gt_left" headers="TAXID">2\|201174</td>
<td class="gt_row gt_right" headers="relative_abundance">20.33151</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Actinoba...</td>
<td class="gt_row gt_left" headers="rank">phylum</td>
<td class="gt_row gt_left" headers="name">Actinobacteria</td>
<td class="gt_row gt_left" headers="lineage">[Actinobacteria, Terrabacteria group, Bacteria...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">...</td>
<td class="gt_row gt_left" headers="TAXID">...</td>
<td class="gt_row gt_right" headers="relative_abundance">...</td>
<td class="gt_row gt_left" headers="taxopy">...</td>
<td class="gt_row gt_left" headers="rank">...</td>
<td class="gt_row gt_left" headers="name">...</td>
<td class="gt_row gt_left" headers="lineage">...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">62</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239\|186801\|186802\|186803\|572511\|33039</td>
<td class="gt_row gt_right" headers="relative_abundance">0.2491</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicut...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">[Ruminococcus] torques</td>
<td class="gt_row gt_left" headers="lineage">[[Ruminococcus] torques, Mediterraneibacter, L...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">63</td>
<td class="gt_row gt_left" headers="TAXID">2\|201174\|84998\|84999\|84107\|1472762\|1232426</td>
<td class="gt_row gt_right" headers="relative_abundance">0.17084</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Actinoba...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">[Collinsella] massiliensis</td>
<td class="gt_row gt_left" headers="lineage">[[Collinsella] massiliensis, Enorma, Coriobact...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">64</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239\|186801\|186802\|186803\|189330\|39486</td>
<td class="gt_row gt_right" headers="relative_abundance">0.0769</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicut...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Dorea formicigenerans</td>
<td class="gt_row gt_left" headers="lineage">[Dorea formicigenerans, Dorea, Lachnospiraceae...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">65</td>
<td class="gt_row gt_left" headers="TAXID">2\|201174\|1760\|85004\|31953\|1678\|216816</td>
<td class="gt_row gt_right" headers="relative_abundance">0.05447</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Actinoba...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Bifidobacterium longum</td>
<td class="gt_row gt_left" headers="lineage">[Bifidobacterium longum, Bifidobacterium, Bifi...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">66</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239\|186801\|186802\|541000\|1263\|1262959</td>
<td class="gt_row gt_right" headers="relative_abundance">0.0144</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicut...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Ruminococcus sp. CAG:488</td>
<td class="gt_row gt_left" headers="lineage">[Ruminococcus sp. CAG:488, environmental sampl...</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<p>Because our modern data are split by ranks, we’ll first split our ancient sample by rank</p>
<p>Which of the entries are at the <code>species rank</code> level?</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>ancient_species <span class="op">=</span> ancient_data.query(<span class="st">"rank == 'species'"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>ancient_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtablewithpathspecies" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtablewithpathspecies-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.6: A five column table of TAXIDs and the organisms corresponding relative abundance, and the attached taxonomic path associated with the TAXID, but also the rank and name of the particular taxonomic ID, filtered to only species
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtablewithpathspecies-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="lqauwaqodm" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#lqauwaqodm table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#lqauwaqodm thead, #lqauwaqodm tbody, #lqauwaqodm tfoot, #lqauwaqodm tr, #lqauwaqodm td, #lqauwaqodm th {
  border-style: none;
}

#lqauwaqodm p {
  margin: 0;
  padding: 0;
}

#lqauwaqodm .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#lqauwaqodm .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#lqauwaqodm .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#lqauwaqodm .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#lqauwaqodm .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#lqauwaqodm .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#lqauwaqodm .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#lqauwaqodm .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#lqauwaqodm .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#lqauwaqodm .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#lqauwaqodm .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#lqauwaqodm .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#lqauwaqodm .gt_spanner_row {
  border-bottom-style: hidden;
}

#lqauwaqodm .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#lqauwaqodm .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#lqauwaqodm .gt_from_md > :first-child {
  margin-top: 0;
}

#lqauwaqodm .gt_from_md > :last-child {
  margin-bottom: 0;
}

#lqauwaqodm .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#lqauwaqodm .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#lqauwaqodm .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#lqauwaqodm .gt_row_group_first td {
  border-top-width: 2px;
}

#lqauwaqodm .gt_row_group_first th {
  border-top-width: 2px;
}

#lqauwaqodm .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#lqauwaqodm .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#lqauwaqodm .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#lqauwaqodm .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#lqauwaqodm .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#lqauwaqodm .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#lqauwaqodm .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#lqauwaqodm .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#lqauwaqodm .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#lqauwaqodm .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#lqauwaqodm .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#lqauwaqodm .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#lqauwaqodm .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#lqauwaqodm .gt_left {
  text-align: left;
}

#lqauwaqodm .gt_center {
  text-align: center;
}

#lqauwaqodm .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#lqauwaqodm .gt_font_normal {
  font-weight: normal;
}

#lqauwaqodm .gt_font_bold {
  font-weight: bold;
}

#lqauwaqodm .gt_font_italic {
  font-style: italic;
}

#lqauwaqodm .gt_super {
  font-size: 65%;
}

#lqauwaqodm .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#lqauwaqodm .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#lqauwaqodm .gt_indent_1 {
  text-indent: 5px;
}

#lqauwaqodm .gt_indent_2 {
  text-indent: 10px;
}

#lqauwaqodm .gt_indent_3 {
  text-indent: 15px;
}

#lqauwaqodm .gt_indent_4 {
  text-indent: 20px;
}

#lqauwaqodm .gt_indent_5 {
  text-indent: 25px;
}

#lqauwaqodm .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#lqauwaqodm div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="TAXID" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">TAXID</th>
<th id="relative_abundance" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">relative_abundance</th>
<th id="taxopy" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">taxopy</th>
<th id="rank" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">rank</th>
<th id="name" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">name</th>
<th id="lineage" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">lineage</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="...1">46</td>
<td class="gt_row gt_left" headers="TAXID">2\|976\|200643\|171549\|171552\|838\|165179</td>
<td class="gt_row gt_right" headers="relative_abundance">25.75544</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__FCB group;p__Bacteroidetes;c__B...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Prevotella copri</td>
<td class="gt_row gt_left" headers="lineage">[Prevotella copri, Prevotella, Prevotellaceae,...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">47</td>
<td class="gt_row gt_left" headers="TAXID">2157\|28890\|183925\|2158\|2159\|2172\|2173</td>
<td class="gt_row gt_right" headers="relative_abundance">17.05626</td>
<td class="gt_row gt_left" headers="taxopy">s__Archaea;p__Euryarchaeota;c__Methanomada gro...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Methanobrevibacter smithii</td>
<td class="gt_row gt_left" headers="lineage">[Methanobrevibacter smithii, Methanobrevibacte...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">48</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239\|186801\|186802\|541000\|1263\|40518</td>
<td class="gt_row gt_right" headers="relative_abundance">14.96816</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicut...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Ruminococcus bromii</td>
<td class="gt_row gt_left" headers="lineage">[Ruminococcus bromii, Ruminococcus, Oscillospi...</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">49</td>
<td class="gt_row gt_left" headers="TAXID">2\|1239\|186801\|186802\|186803\|841\|301302</td>
<td class="gt_row gt_right" headers="relative_abundance">13.57908</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Firmicut...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Roseburia faecis</td>
<td class="gt_row gt_left" headers="lineage">[Roseburia faecis, Roseburia, Lachnospiraceae,...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">50</td>
<td class="gt_row gt_left" headers="TAXID">2\|201174\|84998\|84999\|84107\|102106\|74426</td>
<td class="gt_row gt_right" headers="relative_abundance">9.49165</td>
<td class="gt_row gt_left" headers="taxopy">s__Bacteria;c__Terrabacteria group;p__Actinoba...</td>
<td class="gt_row gt_left" headers="rank">species</td>
<td class="gt_row gt_left" headers="name">Collinsella aerofaciens</td>
<td class="gt_row gt_left" headers="lineage">[Collinsella aerofaciens, Collinsella, Corioba...</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<p>Let’s do a bit of renaming to prepare for what’s coming next</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ancient_species <span class="op">=</span> ancient_species[[<span class="st">'relative_abundance'</span>,<span class="st">'name'</span>]].set_index(<span class="st">'name'</span>).rename(columns<span class="op">=</span>{<span class="st">'relative_abundance'</span>:<span class="st">'ERR5766177'</span>})</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>ancient_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtablewithname" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtablewithname-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.7: Reconstruction of the first two column taxonomic profile but with species-level organism names rather than TAXIDs
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtablewithname-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th>name</th>
<th>ERR5766177</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Prevotella copri</td>
<td>25.75544</td>
</tr>
<tr class="even">
<td>Methanobrevibacter smithii</td>
<td>17.05626</td>
</tr>
<tr class="odd">
<td>Ruminococcus bromii</td>
<td>14.96816</td>
</tr>
<tr class="even">
<td>Roseburia faecis</td>
<td>13.57908</td>
</tr>
<tr class="odd">
<td>Collinsella aerofaciens</td>
<td>9.49165</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>ancient_phylums <span class="op">=</span> ancient_data.query(<span class="st">"rank == 'phylum'"</span>)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>ancient_phylums <span class="op">=</span> ancient_phylums[[<span class="st">'relative_abundance'</span>,<span class="st">'name'</span>]].set_index(<span class="st">'name'</span>).rename(columns<span class="op">=</span>{<span class="st">'relative_abundance'</span>:<span class="st">'ERR5766177'</span>})</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>ancient_phylums</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-taxidtablewithname" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-taxidtablewithname-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.8: Reconstruction of the first two column taxonomic profile but with phylum-level organism names rather than TAXIDs
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-taxidtablewithname-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th>name</th>
<th>ERR5766177</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Firmicutes</td>
<td>33.47957</td>
</tr>
<tr class="even">
<td>Bacteroidetes</td>
<td>28.42090</td>
</tr>
<tr class="odd">
<td>Actinobacteria</td>
<td>20.33151</td>
</tr>
<tr class="even">
<td>Euryarchaeota</td>
<td>17.76802</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>Now, let’s go back to the 700% relative abundance issue…</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>ancient_data.groupby(<span class="st">'rank'</span>)[<span class="st">'relative_abundance'</span>].<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>    rank
    class            99.72648
    family           83.49854
    genus            97.56524
    no rank          19.48331
    order            99.72648
    phylum          100.00000
    species         100.00002
    superkingdom    100.00000
    Name: relative_abundance, dtype: float64</code></pre>
</div>
</div>
</div>
<p>Seems better, right ?</p>
<p><em>Pause and think: why don’t we get exactly 100%</em> ?</p>
</div>
</div>
</div>
</section>
<section id="bringing-together-ancient-and-modern-samples" class="level2" data-number="10.10">
<h2 data-number="10.10" class="anchored" data-anchor-id="bringing-together-ancient-and-modern-samples"><span class="header-section-number">10.10</span> Bringing together ancient and modern samples</h2>
<p>Now let’s load our modern reference samples</p>
<p>First at the phylum level (<a href="#tbl-taxonomicprofiling-tblmoderntaxtablephylum" class="quarto-xref">Table&nbsp;<span>10.9</span></a>)</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>modern_phylums <span class="op">=</span> pd.read_csv(<span class="st">"../data/curated_metagenomics/modern_sources_phylum.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>modern_phylums.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-tblmoderntaxtablephylum" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-tblmoderntaxtablephylum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.9: Taxonomic profiles at phylum level of multiple modern samples
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-tblmoderntaxtablephylum-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="xuuryfgiql" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#xuuryfgiql table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#xuuryfgiql thead, #xuuryfgiql tbody, #xuuryfgiql tfoot, #xuuryfgiql tr, #xuuryfgiql td, #xuuryfgiql th {
  border-style: none;
}

#xuuryfgiql p {
  margin: 0;
  padding: 0;
}

#xuuryfgiql .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#xuuryfgiql .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#xuuryfgiql .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#xuuryfgiql .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#xuuryfgiql .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#xuuryfgiql .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xuuryfgiql .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#xuuryfgiql .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#xuuryfgiql .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#xuuryfgiql .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#xuuryfgiql .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#xuuryfgiql .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#xuuryfgiql .gt_spanner_row {
  border-bottom-style: hidden;
}

#xuuryfgiql .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#xuuryfgiql .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#xuuryfgiql .gt_from_md > :first-child {
  margin-top: 0;
}

#xuuryfgiql .gt_from_md > :last-child {
  margin-bottom: 0;
}

#xuuryfgiql .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#xuuryfgiql .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#xuuryfgiql .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#xuuryfgiql .gt_row_group_first td {
  border-top-width: 2px;
}

#xuuryfgiql .gt_row_group_first th {
  border-top-width: 2px;
}

#xuuryfgiql .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#xuuryfgiql .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#xuuryfgiql .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#xuuryfgiql .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xuuryfgiql .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#xuuryfgiql .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#xuuryfgiql .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#xuuryfgiql .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#xuuryfgiql .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#xuuryfgiql .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#xuuryfgiql .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#xuuryfgiql .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#xuuryfgiql .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#xuuryfgiql .gt_left {
  text-align: left;
}

#xuuryfgiql .gt_center {
  text-align: center;
}

#xuuryfgiql .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#xuuryfgiql .gt_font_normal {
  font-weight: normal;
}

#xuuryfgiql .gt_font_bold {
  font-weight: bold;
}

#xuuryfgiql .gt_font_italic {
  font-style: italic;
}

#xuuryfgiql .gt_super {
  font-size: 65%;
}

#xuuryfgiql .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#xuuryfgiql .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#xuuryfgiql .gt_indent_1 {
  text-indent: 5px;
}

#xuuryfgiql .gt_indent_2 {
  text-indent: 10px;
}

#xuuryfgiql .gt_indent_3 {
  text-indent: 15px;
}

#xuuryfgiql .gt_indent_4 {
  text-indent: 20px;
}

#xuuryfgiql .gt_indent_5 {
  text-indent: 25px;
}

#xuuryfgiql .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#xuuryfgiql div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="de028ad4-7ae6-11e9-a106-68b59976a384" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">de028ad4-7ae6-11e9-a106-68b59976a384</th>
<th id="PNP_Main_283" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Main_283</th>
<th id="PNP_Validation_55" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Validation_55</th>
<th id="G80275" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">G80275</th>
<th id="PNP_Main_363" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Main_363</th>
<th id="SAMEA7045572" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SAMEA7045572</th>
<th id="SAMEA7045355" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SAMEA7045355</th>
<th id="HD-13" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">HD-13</th>
<th id="EGAR00001420773_9002000001423910" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">EGAR00001420773_9002000001423910</th>
<th id="SID5428-4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SID5428-4</th>
<th id="...12" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...12</th>
<th id="A46_02_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A46_02_1FE</th>
<th id="TZ_87532" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">TZ_87532</th>
<th id="A94_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A94_01_1FE</th>
<th id="KHG_7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_7</th>
<th id="LDK_4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">LDK_4</th>
<th id="KHG_9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_9</th>
<th id="A48_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A48_01_1FE</th>
<th id="KHG_1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_1</th>
<th id="TZ_81781" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">TZ_81781</th>
<th id="A09_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A09_01_1FE</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Bacteroidetes</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0.00000</td>
<td class="gt_row gt_right" headers="PNP_Main_283">17.44332</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">82.86400</td>
<td class="gt_row gt_right" headers="G80275">69.99087</td>
<td class="gt_row gt_right" headers="PNP_Main_363">31.93081</td>
<td class="gt_row gt_right" headers="SAMEA7045572">51.76204</td>
<td class="gt_row gt_right" headers="SAMEA7045355">53.32801</td>
<td class="gt_row gt_right" headers="HD-13">74.59667</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">8.81074</td>
<td class="gt_row gt_right" headers="SID5428-4">26.39694</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">1.97760</td>
<td class="gt_row gt_right" headers="TZ_87532">1.49601</td>
<td class="gt_row gt_right" headers="A94_01_1FE">67.21410</td>
<td class="gt_row gt_right" headers="KHG_7">4.29848</td>
<td class="gt_row gt_right" headers="LDK_4">68.16890</td>
<td class="gt_row gt_right" headers="KHG_9">38.59709</td>
<td class="gt_row gt_right" headers="A48_01_1FE">14.81828</td>
<td class="gt_row gt_right" headers="KHG_1">10.13908</td>
<td class="gt_row gt_right" headers="TZ_81781">57.14031</td>
<td class="gt_row gt_right" headers="A09_01_1FE">11.61544</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">Firmicutes</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">95.24231</td>
<td class="gt_row gt_right" headers="PNP_Main_283">60.47031</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">16.53946</td>
<td class="gt_row gt_right" headers="G80275">22.81977</td>
<td class="gt_row gt_right" headers="PNP_Main_363">65.23075</td>
<td class="gt_row gt_right" headers="SAMEA7045572">41.96928</td>
<td class="gt_row gt_right" headers="SAMEA7045355">45.77661</td>
<td class="gt_row gt_right" headers="HD-13">23.51065</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">54.35341</td>
<td class="gt_row gt_right" headers="SID5428-4">62.23094</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">76.68499</td>
<td class="gt_row gt_right" headers="TZ_87532">78.13269</td>
<td class="gt_row gt_right" headers="A94_01_1FE">29.72394</td>
<td class="gt_row gt_right" headers="KHG_7">33.51772</td>
<td class="gt_row gt_right" headers="LDK_4">19.11149</td>
<td class="gt_row gt_right" headers="KHG_9">46.87139</td>
<td class="gt_row gt_right" headers="A48_01_1FE">72.68136</td>
<td class="gt_row gt_right" headers="KHG_1">35.43789</td>
<td class="gt_row gt_right" headers="TZ_81781">40.57101</td>
<td class="gt_row gt_right" headers="A09_01_1FE">24.72113</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Proteobacteria</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">4.49959</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.77098</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.05697</td>
<td class="gt_row gt_right" headers="G80275">4.07757</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.27316</td>
<td class="gt_row gt_right" headers="SAMEA7045572">3.33972</td>
<td class="gt_row gt_right" headers="SAMEA7045355">0.02001</td>
<td class="gt_row gt_right" headers="HD-13">1.72865</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">0.00000</td>
<td class="gt_row gt_right" headers="SID5428-4">1.81016</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">16.57250</td>
<td class="gt_row gt_right" headers="TZ_87532">0.76159</td>
<td class="gt_row gt_right" headers="A94_01_1FE">2.35058</td>
<td class="gt_row gt_right" headers="KHG_7">9.83772</td>
<td class="gt_row gt_right" headers="LDK_4">5.32392</td>
<td class="gt_row gt_right" headers="KHG_9">0.19699</td>
<td class="gt_row gt_right" headers="A48_01_1FE">3.64655</td>
<td class="gt_row gt_right" headers="KHG_1">17.64151</td>
<td class="gt_row gt_right" headers="TZ_81781">0.30580</td>
<td class="gt_row gt_right" headers="A09_01_1FE">56.20177</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">Actinobacteria</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0.25809</td>
<td class="gt_row gt_right" headers="PNP_Main_283">10.27631</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.45187</td>
<td class="gt_row gt_right" headers="G80275">1.11902</td>
<td class="gt_row gt_right" headers="PNP_Main_363">2.31075</td>
<td class="gt_row gt_right" headers="SAMEA7045572">2.92715</td>
<td class="gt_row gt_right" headers="SAMEA7045355">0.77667</td>
<td class="gt_row gt_right" headers="HD-13">0.16403</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">36.55138</td>
<td class="gt_row gt_right" headers="SID5428-4">1.19951</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">3.01814</td>
<td class="gt_row gt_right" headers="TZ_87532">19.20468</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.69913</td>
<td class="gt_row gt_right" headers="KHG_7">46.99479</td>
<td class="gt_row gt_right" headers="LDK_4">7.39093</td>
<td class="gt_row gt_right" headers="KHG_9">14.26365</td>
<td class="gt_row gt_right" headers="A48_01_1FE">5.47750</td>
<td class="gt_row gt_right" headers="KHG_1">36.77145</td>
<td class="gt_row gt_right" headers="TZ_81781">1.16426</td>
<td class="gt_row gt_right" headers="A09_01_1FE">7.40894</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Verrucomicrobia</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0.00000</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.00784</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.00000</td>
<td class="gt_row gt_right" headers="G80275">1.99276</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.25451</td>
<td class="gt_row gt_right" headers="SAMEA7045572">0.00000</td>
<td class="gt_row gt_right" headers="SAMEA7045355">0.00000</td>
<td class="gt_row gt_right" headers="HD-13">0.00000</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">0.09940</td>
<td class="gt_row gt_right" headers="SID5428-4">3.29795</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.05011</td>
<td class="gt_row gt_right" headers="TZ_87532">0.00000</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_7">0.00000</td>
<td class="gt_row gt_right" headers="LDK_4">0.00000</td>
<td class="gt_row gt_right" headers="KHG_9">0.00000</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_1">0.00000</td>
<td class="gt_row gt_right" headers="TZ_81781">0.00000</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.00000</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<p>Then at the species level</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>modern_species <span class="op">=</span> pd.read_csv(<span class="st">"../data/curated_metagenomics/modern_sources_species.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>As usual, we always check if our data has been loaded correctly (<a href="#tbl-taxonomicprofiling-tblmoderntaxtablespecies" class="quarto-xref">Table&nbsp;<span>10.10</span></a>)</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>modern_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-tblmoderntaxtablespecies" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-tblmoderntaxtablespecies-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.10: Taxonomic profiles at species level of multiple modern samples
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-tblmoderntaxtablespecies-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="tskurpqzib" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#tskurpqzib table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#tskurpqzib thead, #tskurpqzib tbody, #tskurpqzib tfoot, #tskurpqzib tr, #tskurpqzib td, #tskurpqzib th {
  border-style: none;
}

#tskurpqzib p {
  margin: 0;
  padding: 0;
}

#tskurpqzib .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#tskurpqzib .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#tskurpqzib .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#tskurpqzib .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#tskurpqzib .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#tskurpqzib .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#tskurpqzib .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#tskurpqzib .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#tskurpqzib .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#tskurpqzib .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#tskurpqzib .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#tskurpqzib .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#tskurpqzib .gt_spanner_row {
  border-bottom-style: hidden;
}

#tskurpqzib .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#tskurpqzib .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#tskurpqzib .gt_from_md > :first-child {
  margin-top: 0;
}

#tskurpqzib .gt_from_md > :last-child {
  margin-bottom: 0;
}

#tskurpqzib .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#tskurpqzib .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#tskurpqzib .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#tskurpqzib .gt_row_group_first td {
  border-top-width: 2px;
}

#tskurpqzib .gt_row_group_first th {
  border-top-width: 2px;
}

#tskurpqzib .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#tskurpqzib .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#tskurpqzib .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#tskurpqzib .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#tskurpqzib .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#tskurpqzib .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#tskurpqzib .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#tskurpqzib .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#tskurpqzib .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#tskurpqzib .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#tskurpqzib .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#tskurpqzib .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#tskurpqzib .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#tskurpqzib .gt_left {
  text-align: left;
}

#tskurpqzib .gt_center {
  text-align: center;
}

#tskurpqzib .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#tskurpqzib .gt_font_normal {
  font-weight: normal;
}

#tskurpqzib .gt_font_bold {
  font-weight: bold;
}

#tskurpqzib .gt_font_italic {
  font-style: italic;
}

#tskurpqzib .gt_super {
  font-size: 65%;
}

#tskurpqzib .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#tskurpqzib .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#tskurpqzib .gt_indent_1 {
  text-indent: 5px;
}

#tskurpqzib .gt_indent_2 {
  text-indent: 10px;
}

#tskurpqzib .gt_indent_3 {
  text-indent: 15px;
}

#tskurpqzib .gt_indent_4 {
  text-indent: 20px;
}

#tskurpqzib .gt_indent_5 {
  text-indent: 25px;
}

#tskurpqzib .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#tskurpqzib div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="de028ad4-7ae6-11e9-a106-68b59976a384" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">de028ad4-7ae6-11e9-a106-68b59976a384</th>
<th id="PNP_Main_283" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Main_283</th>
<th id="PNP_Validation_55" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Validation_55</th>
<th id="G80275" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">G80275</th>
<th id="PNP_Main_363" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PNP_Main_363</th>
<th id="SAMEA7045572" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SAMEA7045572</th>
<th id="SAMEA7045355" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SAMEA7045355</th>
<th id="HD-13" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">HD-13</th>
<th id="EGAR00001420773_9002000001423910" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">EGAR00001420773_9002000001423910</th>
<th id="SID5428-4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">SID5428-4</th>
<th id="...12" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...12</th>
<th id="A46_02_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A46_02_1FE</th>
<th id="TZ_87532" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">TZ_87532</th>
<th id="A94_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A94_01_1FE</th>
<th id="KHG_7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_7</th>
<th id="LDK_4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">LDK_4</th>
<th id="KHG_9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_9</th>
<th id="A48_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A48_01_1FE</th>
<th id="KHG_1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">KHG_1</th>
<th id="TZ_81781" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">TZ_81781</th>
<th id="A09_01_1FE" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">A09_01_1FE</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Bacteroides vulgatus</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.60446</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">1.59911</td>
<td class="gt_row gt_right" headers="G80275">4.39085</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.04494</td>
<td class="gt_row gt_right" headers="SAMEA7045572">4.66505</td>
<td class="gt_row gt_right" headers="SAMEA7045355">2.99431</td>
<td class="gt_row gt_right" headers="HD-13">29.30325</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">1.48560</td>
<td class="gt_row gt_right" headers="SID5428-4">0.98818</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.20717</td>
<td class="gt_row gt_right" headers="TZ_87532">0</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.00309</td>
<td class="gt_row gt_right" headers="KHG_7">0.48891</td>
<td class="gt_row gt_right" headers="LDK_4">0.00000</td>
<td class="gt_row gt_right" headers="KHG_9">0.02230</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_1">0.15112</td>
<td class="gt_row gt_right" headers="TZ_81781">0</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.00836</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">Bacteroides stercoris</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.00546</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.00000</td>
<td class="gt_row gt_right" headers="G80275">0.00000</td>
<td class="gt_row gt_right" headers="PNP_Main_363">2.50789</td>
<td class="gt_row gt_right" headers="SAMEA7045572">0.00000</td>
<td class="gt_row gt_right" headers="SAMEA7045355">20.57498</td>
<td class="gt_row gt_right" headers="HD-13">8.28443</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">1.23261</td>
<td class="gt_row gt_right" headers="SID5428-4">0.00000</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.00000</td>
<td class="gt_row gt_right" headers="TZ_87532">0</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_7">0.00693</td>
<td class="gt_row gt_right" headers="LDK_4">0.00000</td>
<td class="gt_row gt_right" headers="KHG_9">0.02603</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_1">0.19318</td>
<td class="gt_row gt_right" headers="TZ_81781">0</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.00000</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Acidaminococcus intestini</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.00000</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.00000</td>
<td class="gt_row gt_right" headers="G80275">0.00000</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.00000</td>
<td class="gt_row gt_right" headers="SAMEA7045572">0.00000</td>
<td class="gt_row gt_right" headers="SAMEA7045355">0.00000</td>
<td class="gt_row gt_right" headers="HD-13">0.00000</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">0.32822</td>
<td class="gt_row gt_right" headers="SID5428-4">0.00000</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.00000</td>
<td class="gt_row gt_right" headers="TZ_87532">0</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_7">0.00000</td>
<td class="gt_row gt_right" headers="LDK_4">0.00000</td>
<td class="gt_row gt_right" headers="KHG_9">0.00000</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_1">0.00000</td>
<td class="gt_row gt_right" headers="TZ_81781">0</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.00000</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">Eubacterium sp CAG 38</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0</td>
<td class="gt_row gt_right" headers="PNP_Main_283">0.06712</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">0.81149</td>
<td class="gt_row gt_right" headers="G80275">0.05247</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.26027</td>
<td class="gt_row gt_right" headers="SAMEA7045572">0.00000</td>
<td class="gt_row gt_right" headers="SAMEA7045355">0.00000</td>
<td class="gt_row gt_right" headers="HD-13">2.62415</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">0.46585</td>
<td class="gt_row gt_right" headers="SID5428-4">0.23372</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.78140</td>
<td class="gt_row gt_right" headers="TZ_87532">0</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_7">0.00499</td>
<td class="gt_row gt_right" headers="LDK_4">0.00000</td>
<td class="gt_row gt_right" headers="KHG_9">0.02446</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.00000</td>
<td class="gt_row gt_right" headers="KHG_1">0.00000</td>
<td class="gt_row gt_right" headers="TZ_81781">0</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.00000</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">Parabacteroides distasonis</td>
<td class="gt_row gt_right" headers="de028ad4-7ae6-11e9-a106-68b59976a384">0</td>
<td class="gt_row gt_right" headers="PNP_Main_283">1.34931</td>
<td class="gt_row gt_right" headers="PNP_Validation_55">2.00672</td>
<td class="gt_row gt_right" headers="G80275">5.85067</td>
<td class="gt_row gt_right" headers="PNP_Main_363">0.59019</td>
<td class="gt_row gt_right" headers="SAMEA7045572">7.00027</td>
<td class="gt_row gt_right" headers="SAMEA7045355">1.28075</td>
<td class="gt_row gt_right" headers="HD-13">0.61758</td>
<td class="gt_row gt_right" headers="EGAR00001420773_9002000001423910">0.07383</td>
<td class="gt_row gt_right" headers="SID5428-4">2.80355</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="A46_02_1FE">0.11423</td>
<td class="gt_row gt_right" headers="TZ_87532">0</td>
<td class="gt_row gt_right" headers="A94_01_1FE">0.01181</td>
<td class="gt_row gt_right" headers="KHG_7">0.01386</td>
<td class="gt_row gt_right" headers="LDK_4">0.03111</td>
<td class="gt_row gt_right" headers="KHG_9">0.07463</td>
<td class="gt_row gt_right" headers="A48_01_1FE">0.15597</td>
<td class="gt_row gt_right" headers="KHG_1">0.07541</td>
<td class="gt_row gt_right" headers="TZ_81781">0</td>
<td class="gt_row gt_right" headers="A09_01_1FE">0.01932</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>How would we get a random sample of your species table ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>modern_species.sample()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="time-to-merge" class="level3" data-number="10.10.1">
<h3 data-number="10.10.1" class="anchored" data-anchor-id="time-to-merge"><span class="header-section-number">10.10.1</span> Time to merge !</h3>
<p>Now, let’s merge our ancient sample with the modern data in one single table. For that, we’ll use the pandas <code>merge</code> function which will merge the two tables together, using the index as the merge key.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>all_species <span class="op">=</span> ancient_species.merge(modern_species, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>, how<span class="op">=</span><span class="st">'outer'</span>).fillna(<span class="dv">0</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>all_phylums <span class="op">=</span> ancient_phylums.merge(modern_phylums, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>, how<span class="op">=</span><span class="st">'outer'</span>).fillna(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, let’s load the metadata, which contains the information about the modern samples (<a href="#tbl-taxonomicprofiling-tblsamplemetadata" class="quarto-xref">Table&nbsp;<span>10.11</span></a>).</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> pd.read_csv(<span class="st">"../data/metadata/curated_metagenomics_modern_sources.csv"</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>metadata.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-tblsamplemetadata" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-tblsamplemetadata-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.11: Various metadata information about the samples in this example
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-tblsamplemetadata-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="afxkyiruhl" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#afxkyiruhl table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#afxkyiruhl thead, #afxkyiruhl tbody, #afxkyiruhl tfoot, #afxkyiruhl tr, #afxkyiruhl td, #afxkyiruhl th {
  border-style: none;
}

#afxkyiruhl p {
  margin: 0;
  padding: 0;
}

#afxkyiruhl .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#afxkyiruhl .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#afxkyiruhl .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#afxkyiruhl .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#afxkyiruhl .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#afxkyiruhl .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#afxkyiruhl .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#afxkyiruhl .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#afxkyiruhl .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#afxkyiruhl .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#afxkyiruhl .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#afxkyiruhl .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#afxkyiruhl .gt_spanner_row {
  border-bottom-style: hidden;
}

#afxkyiruhl .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#afxkyiruhl .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#afxkyiruhl .gt_from_md > :first-child {
  margin-top: 0;
}

#afxkyiruhl .gt_from_md > :last-child {
  margin-bottom: 0;
}

#afxkyiruhl .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#afxkyiruhl .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#afxkyiruhl .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#afxkyiruhl .gt_row_group_first td {
  border-top-width: 2px;
}

#afxkyiruhl .gt_row_group_first th {
  border-top-width: 2px;
}

#afxkyiruhl .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#afxkyiruhl .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#afxkyiruhl .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#afxkyiruhl .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#afxkyiruhl .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#afxkyiruhl .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#afxkyiruhl .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#afxkyiruhl .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#afxkyiruhl .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#afxkyiruhl .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#afxkyiruhl .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#afxkyiruhl .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#afxkyiruhl .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#afxkyiruhl .gt_left {
  text-align: left;
}

#afxkyiruhl .gt_center {
  text-align: center;
}

#afxkyiruhl .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#afxkyiruhl .gt_font_normal {
  font-weight: normal;
}

#afxkyiruhl .gt_font_bold {
  font-weight: bold;
}

#afxkyiruhl .gt_font_italic {
  font-style: italic;
}

#afxkyiruhl .gt_super {
  font-size: 65%;
}

#afxkyiruhl .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#afxkyiruhl .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#afxkyiruhl .gt_indent_1 {
  text-indent: 5px;
}

#afxkyiruhl .gt_indent_2 {
  text-indent: 10px;
}

#afxkyiruhl .gt_indent_3 {
  text-indent: 15px;
}

#afxkyiruhl .gt_indent_4 {
  text-indent: 20px;
}

#afxkyiruhl .gt_indent_5 {
  text-indent: 25px;
}

#afxkyiruhl .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#afxkyiruhl div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="study_name" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">study_name</th>
<th id="sample_id" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">sample_id</th>
<th id="subject_id" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">subject_id</th>
<th id="body_site" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">body_site</th>
<th id="antibiotics_current_use" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">antibiotics_current_use</th>
<th id="study_condition" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">study_condition</th>
<th id="disease" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">disease</th>
<th id="age" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">age</th>
<th id="infant_age" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">infant_age</th>
<th id="age_category" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">age_category</th>
<th id="...12" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...12</th>
<th id="hla_drb11" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">hla_drb11</th>
<th id="birth_order" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">birth_order</th>
<th id="age_twins_started_to_live_apart" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">age_twins_started_to_live_apart</th>
<th id="zigosity" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">zigosity</th>
<th id="brinkman_index" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">brinkman_index</th>
<th id="alcohol_numeric" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">alcohol_numeric</th>
<th id="breastfeeding_duration" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">breastfeeding_duration</th>
<th id="formula_first_day" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">formula_first_day</th>
<th id="ALT" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">ALT</th>
<th id="eGFR" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">eGFR</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="...1">0</td>
<td class="gt_row gt_left" headers="study_name">ShaoY_2019</td>
<td class="gt_row gt_left" headers="sample_id">de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td class="gt_row gt_left" headers="subject_id">C01528_ba</td>
<td class="gt_row gt_left" headers="body_site">stool</td>
<td class="gt_row gt_left" headers="antibiotics_current_use">no</td>
<td class="gt_row gt_left" headers="study_condition">control</td>
<td class="gt_row gt_left" headers="disease">healthy</td>
<td class="gt_row gt_right" headers="age">0</td>
<td class="gt_row gt_right" headers="infant_age">4</td>
<td class="gt_row gt_left" headers="age_category">newborn</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="hla_drb11">NaN</td>
<td class="gt_row gt_right" headers="birth_order">NaN</td>
<td class="gt_row gt_right" headers="age_twins_started_to_live_apart">NaN</td>
<td class="gt_row gt_right" headers="zigosity">NaN</td>
<td class="gt_row gt_right" headers="brinkman_index">NaN</td>
<td class="gt_row gt_right" headers="alcohol_numeric">NaN</td>
<td class="gt_row gt_right" headers="breastfeeding_duration">NaN</td>
<td class="gt_row gt_right" headers="formula_first_day">NaN</td>
<td class="gt_row gt_right" headers="ALT">NaN</td>
<td class="gt_row gt_right" headers="eGFR">NaN</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">1</td>
<td class="gt_row gt_left" headers="study_name">ZeeviD_2015</td>
<td class="gt_row gt_left" headers="sample_id">PNP_Main_283</td>
<td class="gt_row gt_left" headers="subject_id">PNP_Main_283</td>
<td class="gt_row gt_left" headers="body_site">stool</td>
<td class="gt_row gt_left" headers="antibiotics_current_use">no</td>
<td class="gt_row gt_left" headers="study_condition">control</td>
<td class="gt_row gt_left" headers="disease">healthy</td>
<td class="gt_row gt_right" headers="age">NaN</td>
<td class="gt_row gt_right" headers="infant_age">NaN</td>
<td class="gt_row gt_left" headers="age_category">adult</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="hla_drb11">NaN</td>
<td class="gt_row gt_right" headers="birth_order">NaN</td>
<td class="gt_row gt_right" headers="age_twins_started_to_live_apart">NaN</td>
<td class="gt_row gt_right" headers="zigosity">NaN</td>
<td class="gt_row gt_right" headers="brinkman_index">NaN</td>
<td class="gt_row gt_right" headers="alcohol_numeric">NaN</td>
<td class="gt_row gt_right" headers="breastfeeding_duration">NaN</td>
<td class="gt_row gt_right" headers="formula_first_day">NaN</td>
<td class="gt_row gt_right" headers="ALT">NaN</td>
<td class="gt_row gt_right" headers="eGFR">NaN</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">2</td>
<td class="gt_row gt_left" headers="study_name">ZeeviD_2015</td>
<td class="gt_row gt_left" headers="sample_id">PNP_Validation_55</td>
<td class="gt_row gt_left" headers="subject_id">PNP_Validation_55</td>
<td class="gt_row gt_left" headers="body_site">stool</td>
<td class="gt_row gt_left" headers="antibiotics_current_use">no</td>
<td class="gt_row gt_left" headers="study_condition">control</td>
<td class="gt_row gt_left" headers="disease">healthy</td>
<td class="gt_row gt_right" headers="age">NaN</td>
<td class="gt_row gt_right" headers="infant_age">NaN</td>
<td class="gt_row gt_left" headers="age_category">adult</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="hla_drb11">NaN</td>
<td class="gt_row gt_right" headers="birth_order">NaN</td>
<td class="gt_row gt_right" headers="age_twins_started_to_live_apart">NaN</td>
<td class="gt_row gt_right" headers="zigosity">NaN</td>
<td class="gt_row gt_right" headers="brinkman_index">NaN</td>
<td class="gt_row gt_right" headers="alcohol_numeric">NaN</td>
<td class="gt_row gt_right" headers="breastfeeding_duration">NaN</td>
<td class="gt_row gt_right" headers="formula_first_day">NaN</td>
<td class="gt_row gt_right" headers="ALT">NaN</td>
<td class="gt_row gt_right" headers="eGFR">NaN</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">3</td>
<td class="gt_row gt_left" headers="study_name">VatanenT_2016</td>
<td class="gt_row gt_left" headers="sample_id">G80275</td>
<td class="gt_row gt_left" headers="subject_id">T014806</td>
<td class="gt_row gt_left" headers="body_site">stool</td>
<td class="gt_row gt_left" headers="antibiotics_current_use">no</td>
<td class="gt_row gt_left" headers="study_condition">control</td>
<td class="gt_row gt_left" headers="disease">healthy</td>
<td class="gt_row gt_right" headers="age">1</td>
<td class="gt_row gt_right" headers="infant_age">NaN</td>
<td class="gt_row gt_left" headers="age_category">child</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="hla_drb11">NaN</td>
<td class="gt_row gt_right" headers="birth_order">NaN</td>
<td class="gt_row gt_right" headers="age_twins_started_to_live_apart">NaN</td>
<td class="gt_row gt_right" headers="zigosity">NaN</td>
<td class="gt_row gt_right" headers="brinkman_index">NaN</td>
<td class="gt_row gt_right" headers="alcohol_numeric">NaN</td>
<td class="gt_row gt_right" headers="breastfeeding_duration">NaN</td>
<td class="gt_row gt_right" headers="formula_first_day">NaN</td>
<td class="gt_row gt_right" headers="ALT">NaN</td>
<td class="gt_row gt_right" headers="eGFR">NaN</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">4</td>
<td class="gt_row gt_left" headers="study_name">ZeeviD_2015</td>
<td class="gt_row gt_left" headers="sample_id">PNP_Main_363</td>
<td class="gt_row gt_left" headers="subject_id">PNP_Main_363</td>
<td class="gt_row gt_left" headers="body_site">stool</td>
<td class="gt_row gt_left" headers="antibiotics_current_use">no</td>
<td class="gt_row gt_left" headers="study_condition">control</td>
<td class="gt_row gt_left" headers="disease">healthy</td>
<td class="gt_row gt_right" headers="age">NaN</td>
<td class="gt_row gt_right" headers="infant_age">NaN</td>
<td class="gt_row gt_left" headers="age_category">adult</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="hla_drb11">NaN</td>
<td class="gt_row gt_right" headers="birth_order">NaN</td>
<td class="gt_row gt_right" headers="age_twins_started_to_live_apart">NaN</td>
<td class="gt_row gt_right" headers="zigosity">NaN</td>
<td class="gt_row gt_right" headers="brinkman_index">NaN</td>
<td class="gt_row gt_right" headers="alcohol_numeric">NaN</td>
<td class="gt_row gt_right" headers="breastfeeding_duration">NaN</td>
<td class="gt_row gt_right" headers="formula_first_day">NaN</td>
<td class="gt_row gt_right" headers="ALT">NaN</td>
<td class="gt_row gt_right" headers="eGFR">NaN</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why did we use an “outer” join when merging the ancient and modern taxonomic profiles ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>An outer join will include entries from both left (ancient) and right (modern) table, even if they’re not present in both. This ensures that taxons present in the ancient sample are not excluded when merging with the modern samples, and vice versa. Missing entries in one table will be replaced with <code>NaN</code> which we then replaced by zeros (with <code>.fillna(0)</code>).</p>
</div>
</div>
</div>
</section>
</section>
<section id="comparing-ancient-and-modern-samples" class="level2" data-number="10.11">
<h2 data-number="10.11" class="anchored" data-anchor-id="comparing-ancient-and-modern-samples"><span class="header-section-number">10.11</span> Comparing ancient and modern samples</h2>
<section id="taxonomic-composition" class="level3" data-number="10.11.1">
<h3 data-number="10.11.1" class="anchored" data-anchor-id="taxonomic-composition"><span class="header-section-number">10.11.1</span> Taxonomic composition</h3>
<p>One common plot in microbiome papers in a stacked barplot, often at the phylum or family level.</p>
<p>First, we’ll do some renaming, to make the value of the metadata variables a bit easier to understand (<a href="#tbl-taxonomicsampling-samplegroups" class="quarto-xref">Table&nbsp;<span>10.12</span></a>)</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>group_info <span class="op">=</span> pd.concat(</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>        (</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>        metadata[<span class="st">'non_westernized'</span>]</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>        .<span class="bu">map</span>({<span class="st">'no'</span>:<span class="st">'westernized'</span>,<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>}) <span class="co"># for the non_westernized in the modern sample metadata, rename the value levels</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>        .to_frame(name<span class="op">=</span><span class="st">'group'</span>).set_index(metadata[<span class="st">'sample_id'</span>]) <span class="co"># rename the column to group</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>        .reset_index()</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>        ),</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>        (</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        pd.Series({<span class="st">'sample_id'</span>:<span class="st">'ERR5766177'</span>, <span class="st">'group'</span>:<span class="st">'ancient'</span>}).to_frame().transpose()</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    axis<span class="op">=</span><span class="dv">0</span>, ignore_index<span class="op">=</span><span class="va">True</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>group_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-28-contents" aria-controls="callout-28" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-28" class="callout-28-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb48"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>    /var/folders/1c/l1qb09f15jddsh65f6xv1n_r0000gp/T/ipykernel_40830/27419655.py:2:</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>    FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version.</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>    Use pandas.concat instead.</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>      metadata['non_westernized']</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div id="tbl-taxonomicsampling-samplegroups" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicsampling-samplegroups-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.12: Table of samples and their group
</figcaption>
<div aria-describedby="tbl-taxonomicsampling-samplegroups-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th>sample_id</th>
<th>group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td>westernized</td>
</tr>
<tr class="even">
<td>1</td>
<td>PNP_Main_283</td>
<td>westernized</td>
</tr>
<tr class="odd">
<td>2</td>
<td>PNP_Validation_55</td>
<td>westernized</td>
</tr>
<tr class="even">
<td>3</td>
<td>G80275</td>
<td>westernized</td>
</tr>
<tr class="odd">
<td>4</td>
<td>PNP_Main_363</td>
<td>westernized</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="odd">
<td>196</td>
<td>A48_01_1FE</td>
<td>non_westernized</td>
</tr>
<tr class="even">
<td>197</td>
<td>KHG_1</td>
<td>non_westernized</td>
</tr>
<tr class="odd">
<td>198</td>
<td>TZ_81781</td>
<td>non_westernized</td>
</tr>
<tr class="even">
<td>199</td>
<td>A09_01_1FE</td>
<td>non_westernized</td>
</tr>
<tr class="odd">
<td>200</td>
<td>ERR5766177</td>
<td>ancient</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>We need transform our data in ‘tidy format’ (<a href="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html">https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html</a>) to plot with plotnine (<a href="https://plotnine.readthedocs.io/en/stable/">https://plotnine.readthedocs.io/en/stable/</a>), a python clone of ggplot (<a href="https://ggplot2.tidyverse.org/index.html">https://ggplot2.tidyverse.org/index.html</a>).</p>
<div id="tbl-taxonomicprofiling-tblrawmultisampletable" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-tblrawmultisampletable-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.13: Table of the raw multi-sample taxonomic table
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-tblrawmultisampletable-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="drwzomnwjm" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#drwzomnwjm table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#drwzomnwjm thead, #drwzomnwjm tbody, #drwzomnwjm tfoot, #drwzomnwjm tr, #drwzomnwjm td, #drwzomnwjm th {
  border-style: none;
}

#drwzomnwjm p {
  margin: 0;
  padding: 0;
}

#drwzomnwjm .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#drwzomnwjm .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#drwzomnwjm .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#drwzomnwjm .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#drwzomnwjm .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#drwzomnwjm .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#drwzomnwjm .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#drwzomnwjm .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#drwzomnwjm .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#drwzomnwjm .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#drwzomnwjm .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#drwzomnwjm .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#drwzomnwjm .gt_spanner_row {
  border-bottom-style: hidden;
}

#drwzomnwjm .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#drwzomnwjm .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#drwzomnwjm .gt_from_md > :first-child {
  margin-top: 0;
}

#drwzomnwjm .gt_from_md > :last-child {
  margin-bottom: 0;
}

#drwzomnwjm .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#drwzomnwjm .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#drwzomnwjm .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#drwzomnwjm .gt_row_group_first td {
  border-top-width: 2px;
}

#drwzomnwjm .gt_row_group_first th {
  border-top-width: 2px;
}

#drwzomnwjm .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#drwzomnwjm .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#drwzomnwjm .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#drwzomnwjm .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#drwzomnwjm .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#drwzomnwjm .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#drwzomnwjm .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#drwzomnwjm .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#drwzomnwjm .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#drwzomnwjm .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#drwzomnwjm .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#drwzomnwjm .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#drwzomnwjm .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#drwzomnwjm .gt_left {
  text-align: left;
}

#drwzomnwjm .gt_center {
  text-align: center;
}

#drwzomnwjm .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#drwzomnwjm .gt_font_normal {
  font-weight: normal;
}

#drwzomnwjm .gt_font_bold {
  font-weight: bold;
}

#drwzomnwjm .gt_font_italic {
  font-style: italic;
}

#drwzomnwjm .gt_super {
  font-size: 65%;
}

#drwzomnwjm .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#drwzomnwjm .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#drwzomnwjm .gt_indent_1 {
  text-indent: 5px;
}

#drwzomnwjm .gt_indent_2 {
  text-indent: 10px;
}

#drwzomnwjm .gt_indent_3 {
  text-indent: 15px;
}

#drwzomnwjm .gt_indent_4 {
  text-indent: 20px;
}

#drwzomnwjm .gt_indent_5 {
  text-indent: 25px;
}

#drwzomnwjm .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#drwzomnwjm div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="Actinobacteria" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Actinobacteria</th>
<th id="Apicomplexa" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Apicomplexa</th>
<th id="Ascomycota" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Ascomycota</th>
<th id="Bacteroidetes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Bacteroidetes</th>
<th id="Basidiomycota" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Basidiomycota</th>
<th id="Candidatus Melainabacteria" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Candidatus Melainabacteria</th>
<th id="Chlamydiae" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Chlamydiae</th>
<th id="Chloroflexi" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Chloroflexi</th>
<th id="Cyanobacteria" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Cyanobacteria</th>
<th id="Deferribacteres" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Deferribacteres</th>
<th id="...12" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...12</th>
<th id="Fusobacteria" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Fusobacteria</th>
<th id="Lentisphaerae" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Lentisphaerae</th>
<th id="Planctomycetes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Planctomycetes</th>
<th id="Proteobacteria" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Proteobacteria</th>
<th id="Spirochaetes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Spirochaetes</th>
<th id="Synergistetes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Synergistetes</th>
<th id="Tenericutes" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Tenericutes</th>
<th id="Verrucomicrobia" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Verrucomicrobia</th>
<th id="sample_id" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">sample_id</th>
<th id="group" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">group</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_right" headers="...1">200</td>
<td class="gt_row gt_right" headers="Actinobacteria">20.33151</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">28.4209</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">0</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">ERR5766177</td>
<td class="gt_row gt_left" headers="group">ancient</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">0</td>
<td class="gt_row gt_right" headers="Actinobacteria">0.25809</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">0</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">4.49959</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td class="gt_row gt_left" headers="group">westernized</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">1</td>
<td class="gt_row gt_right" headers="Actinobacteria">10.27631</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">17.44332</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0.01486</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">0.77098</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0.00784</td>
<td class="gt_row gt_left" headers="sample_id">PNP_Main_283</td>
<td class="gt_row gt_left" headers="group">westernized</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">2</td>
<td class="gt_row gt_right" headers="Actinobacteria">0.45187</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">82.864</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">0.05697</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">PNP_Validation_55</td>
<td class="gt_row gt_left" headers="group">westernized</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">3</td>
<td class="gt_row gt_right" headers="Actinobacteria">1.11902</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">69.99087</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">4.07757</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">1.99276</td>
<td class="gt_row gt_left" headers="sample_id">G80275</td>
<td class="gt_row gt_left" headers="group">westernized</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">...</td>
<td class="gt_row gt_right" headers="Actinobacteria">...</td>
<td class="gt_row gt_right" headers="Apicomplexa">...</td>
<td class="gt_row gt_right" headers="Ascomycota">...</td>
<td class="gt_row gt_right" headers="Bacteroidetes">...</td>
<td class="gt_row gt_right" headers="Basidiomycota">...</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">...</td>
<td class="gt_row gt_right" headers="Chlamydiae">...</td>
<td class="gt_row gt_right" headers="Chloroflexi">...</td>
<td class="gt_row gt_right" headers="Cyanobacteria">...</td>
<td class="gt_row gt_right" headers="Deferribacteres">...</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">...</td>
<td class="gt_row gt_right" headers="Lentisphaerae">...</td>
<td class="gt_row gt_right" headers="Planctomycetes">...</td>
<td class="gt_row gt_right" headers="Proteobacteria">...</td>
<td class="gt_row gt_right" headers="Spirochaetes">...</td>
<td class="gt_row gt_right" headers="Synergistetes">...</td>
<td class="gt_row gt_right" headers="Tenericutes">...</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">...</td>
<td class="gt_row gt_left" headers="sample_id">...</td>
<td class="gt_row gt_left" headers="group">...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">195</td>
<td class="gt_row gt_right" headers="Actinobacteria">14.26365</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">38.59709</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">0.19699</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">KHG_9</td>
<td class="gt_row gt_left" headers="group">non_westernized</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">196</td>
<td class="gt_row gt_right" headers="Actinobacteria">5.4775</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">14.81828</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">3.64655</td>
<td class="gt_row gt_right" headers="Spirochaetes">0.09964</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">A48_01_1FE</td>
<td class="gt_row gt_left" headers="group">non_westernized</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">197</td>
<td class="gt_row gt_right" headers="Actinobacteria">36.77145</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">10.13908</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">17.64151</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">KHG_1</td>
<td class="gt_row gt_left" headers="group">non_westernized</td>
</tr>
<tr class="even">
<td class="gt_row gt_right" headers="...1">198</td>
<td class="gt_row gt_right" headers="Actinobacteria">1.16426</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">57.14031</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">0.3058</td>
<td class="gt_row gt_right" headers="Spirochaetes">0.70467</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">TZ_81781</td>
<td class="gt_row gt_left" headers="group">non_westernized</td>
</tr>
<tr class="odd">
<td class="gt_row gt_right" headers="...1">199</td>
<td class="gt_row gt_right" headers="Actinobacteria">7.40894</td>
<td class="gt_row gt_right" headers="Apicomplexa">0</td>
<td class="gt_row gt_right" headers="Ascomycota">0</td>
<td class="gt_row gt_right" headers="Bacteroidetes">11.61544</td>
<td class="gt_row gt_right" headers="Basidiomycota">0</td>
<td class="gt_row gt_right" headers="Candidatus Melainabacteria">0</td>
<td class="gt_row gt_right" headers="Chlamydiae">0</td>
<td class="gt_row gt_right" headers="Chloroflexi">0</td>
<td class="gt_row gt_right" headers="Cyanobacteria">0</td>
<td class="gt_row gt_right" headers="Deferribacteres">0</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="Fusobacteria">0</td>
<td class="gt_row gt_right" headers="Lentisphaerae">0</td>
<td class="gt_row gt_right" headers="Planctomycetes">0</td>
<td class="gt_row gt_right" headers="Proteobacteria">56.20177</td>
<td class="gt_row gt_right" headers="Spirochaetes">0</td>
<td class="gt_row gt_right" headers="Synergistetes">0</td>
<td class="gt_row gt_right" headers="Tenericutes">0</td>
<td class="gt_row gt_right" headers="Verrucomicrobia">0</td>
<td class="gt_row gt_left" headers="sample_id">A09_01_1FE</td>
<td class="gt_row gt_left" headers="group">non_westernized</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<p>Now, we need transform this (<a href="#tbl-taxonomicprofiling-tblrawmultisampletable" class="quarto-xref">Table&nbsp;<span>10.13</span></a>) in the tidy format, with the <code>melt</code> function.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums <span class="op">=</span> (</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>    all_phylums</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>    .transpose()</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>    .merge(group_info, left_index<span class="op">=</span><span class="va">True</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>)</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    .melt(id_vars<span class="op">=</span>[<span class="st">'sample_id'</span>, <span class="st">'group'</span>], value_name<span class="op">=</span><span class="st">'relative_abundance'</span>, var_name<span class="op">=</span><span class="st">'Phylum'</span>, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, we only want to keep the mean relative abundance for each phylum. To do so, we will compute the mean relative abundance, for each phylum, for each group (<code>ancient</code>, <code>westernized</code>, and <code>non_westernized</code>).</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums <span class="op">=</span> tidy_phylums.groupby([<span class="st">'group'</span>, <span class="st">'Phylum'</span>]).mean().reset_index()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We then verify that the sum of the mean relative abundance is still ~100%, as an extra sanity check.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums.groupby(<span class="st">'group'</span>)[<span class="st">'relative_abundance'</span>].<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>group
ancient            100.000000
non_westernized     99.710255
westernized         99.905089
Name: relative_abundance, dtype: float64</code></pre>
</section>
</section>
<section id="lets-make-some-plots" class="level2" data-number="10.12">
<h2 data-number="10.12" class="anchored" data-anchor-id="lets-make-some-plots"><span class="header-section-number">10.12</span> Let’s make some plots</h2>
<p>We first import plotnine</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> plotnine <span class="im">import</span> <span class="op">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>And then run plotnine to a barplot of the mean abundance per group (<a href="#fig-taxonomicprofiling-plotninephylumabundance" class="quarto-xref">Figure&nbsp;<span>10.6</span></a>).</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>ggplot(tidy_phylums, aes(x<span class="op">=</span><span class="st">'group'</span>, y<span class="op">=</span><span class="st">'relative_abundance'</span>, fill<span class="op">=</span><span class="st">'Phylum'</span>)) <span class="op">\</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_bar(position<span class="op">=</span><span class="st">'stack'</span>, stat<span class="op">=</span><span class="st">'identity'</span>) <span class="op">\</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> ylab(<span class="st">'mean abundance'</span>) <span class="op">\</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> xlab(<span class="st">""</span>) <span class="op">\</span></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-plotninephylumabundance" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-plotninephylumabundance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_80_0.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-plotninephylumabundance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.6: Stacked bar chart of ancient, non-westernised, and westernised sample groups on the X axis columns, and mean abundance percentage on the Y-axis. The legend and stacks of the bar represent different phyla each with a different colour
</figcaption>
</figure>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the most abundant Phylum in all samples</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-30-contents" aria-controls="callout-30" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-30" class="callout-30-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Whether ancient, non westernised, or westernised gut microbiome sample, the phylum Firmicutes (officially renamed Bacillota since 2021) is the most abundant.</p>
</div>
</div>
</div>
</section>
<section id="ecological-diversity" class="level2" data-number="10.13">
<h2 data-number="10.13" class="anchored" data-anchor-id="ecological-diversity"><span class="header-section-number">10.13</span> Ecological diversity</h2>
<section id="alpha-diversity" class="level3" data-number="10.13.1">
<h3 data-number="10.13.1" class="anchored" data-anchor-id="alpha-diversity"><span class="header-section-number">10.13.1</span> Alpha diversity</h3>
<p>Alpha diversity is the measure of diversity withing each sample. It is used to estimate how many species are present in a sample, and how diverse they are. We’ll use the python library scikit-bio (<a href="http://scikit-bio.org/">http://scikit-bio.org/</a>) to compute it, and the plotnine (<a href="https://plotnine.readthedocs.io/">https://plotnine.readthedocs.io/</a>) library (a python port of ggplot2 (<a href="https://ggplot2.tidyverse.org/reference/ggplot.html">https://ggplot2.tidyverse.org/reference/ggplot.html</a>) to visualise the results).</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> skbio</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s compute the <strong>species richness</strong>, the <strong>Shannon index</strong>, and <strong>Simpson index of diversity</strong> (<a href="#tbl-taxonomicprofiling-alphadiv" class="quarto-xref">Table&nbsp;<span>10.14</span></a>)</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>shannon <span class="op">=</span> skbio.diversity.alpha_diversity(</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>    metric<span class="op">=</span><span class="st">"shannon"</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>simpson <span class="op">=</span> skbio.diversity.alpha_diversity(</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>    metric<span class="op">=</span><span class="st">"simpson"</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>richness <span class="op">=</span> (all_species <span class="op">!=</span> <span class="dv">0</span>).astype(<span class="bu">int</span>).<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> (</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>    shannon.to_frame(name<span class="op">=</span><span class="st">"shannon"</span>)</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>    .merge(simpson.to_frame(name<span class="op">=</span><span class="st">"simpson"</span>), left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>    .merge(richness.to_frame(name<span class="op">=</span><span class="st">"richness"</span>), left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-alphadiv" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-alphadiv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.14: Table of the shannon, simpson, and richness alpha diversity indicies for a subset of samples
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-alphadiv-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 55%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>shannon</th>
<th>simpson</th>
<th>richness</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ERR5766177</td>
<td>3.032945</td>
<td>0.844769</td>
<td>21</td>
</tr>
<tr class="even">
<td>de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td>0.798112</td>
<td>0.251280</td>
<td>11</td>
</tr>
<tr class="odd">
<td>PNP_Main_283</td>
<td>5.092878</td>
<td>0.954159</td>
<td>118</td>
</tr>
<tr class="even">
<td>PNP_Validation_55</td>
<td>3.670162</td>
<td>0.812438</td>
<td>72</td>
</tr>
<tr class="odd">
<td>G80275</td>
<td>3.831358</td>
<td>0.876712</td>
<td>66</td>
</tr>
<tr class="even">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="odd">
<td>KHG_9</td>
<td>3.884285</td>
<td>0.861683</td>
<td>87</td>
</tr>
<tr class="even">
<td>A48_01_1FE</td>
<td>4.377755</td>
<td>0.930024</td>
<td>53</td>
</tr>
<tr class="odd">
<td>KHG_1</td>
<td>3.733834</td>
<td>0.875335</td>
<td>108</td>
</tr>
<tr class="even">
<td>TZ_81781</td>
<td>2.881856</td>
<td>0.719491</td>
<td>44</td>
</tr>
<tr class="odd">
<td>A09_01_1FE</td>
<td>2.982322</td>
<td>0.719962</td>
<td>75</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>Let’s load the group information from the metadata. To do so, we merge the alpha diversity dataframe that we compute previously, with the metadata dataframe, using the <code>sample_id</code> as a merge key. Finally, we do a bit of renaming to re-encode <code>yes</code>/<code>no</code> as <code>non_westernized</code>/<code>westernized</code>.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> (</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>    alpha_diversity</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>    .merge(metadata[[<span class="st">'sample_id'</span>, <span class="st">'non_westernized'</span>]], left_index<span class="op">=</span><span class="va">True</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>, how<span class="op">=</span><span class="st">'outer'</span>)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>    .set_index(<span class="st">'sample_id'</span>)</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'non_westernized'</span>:<span class="st">'group'</span>})</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>alpha_diversity[<span class="st">'group'</span>] <span class="op">=</span> alpha_diversity[<span class="st">'group'</span>].replace({<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>,<span class="st">'no'</span>:<span class="st">'westernized'</span>, pd.NA:<span class="st">'ERR5766177'</span>})</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-alphadivgroup" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-alphadivgroup-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.15: Table of the shannon, simpson, and richness alpha diversity indicies for a subset of samples but with the group metadata
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-alphadivgroup-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 44%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>shannon</th>
<th>simpson</th>
<th>richness</th>
<th>group</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sample_id</td>
<td></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>ERR5766177</td>
<td>3.032945</td>
<td>0.844769</td>
<td>21</td>
<td>ERR5766177</td>
</tr>
<tr class="odd">
<td>de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td>0.798112</td>
<td>0.251280</td>
<td>11</td>
<td>westernized</td>
</tr>
<tr class="even">
<td>PNP_Main_283</td>
<td>5.092878</td>
<td>0.954159</td>
<td>118</td>
<td>westernized</td>
</tr>
<tr class="odd">
<td>PNP_Validation_55</td>
<td>3.670162</td>
<td>0.812438</td>
<td>72</td>
<td>westernized</td>
</tr>
<tr class="even">
<td>G80275</td>
<td>3.831358</td>
<td>0.876712</td>
<td>66</td>
<td>westernized</td>
</tr>
<tr class="odd">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="even">
<td>KHG_9</td>
<td>3.884285</td>
<td>0.861683</td>
<td>87</td>
<td>non_westernized</td>
</tr>
<tr class="odd">
<td>A48_01_1FE</td>
<td>4.377755</td>
<td>0.930024</td>
<td>53</td>
<td>non_westernized</td>
</tr>
<tr class="even">
<td>KHG_1</td>
<td>3.733834</td>
<td>0.875335</td>
<td>108</td>
<td>non_westernized</td>
</tr>
<tr class="odd">
<td>TZ_81781</td>
<td>2.881856</td>
<td>0.719491</td>
<td>44</td>
<td>non_westernized</td>
</tr>
<tr class="even">
<td>A09_01_1FE</td>
<td>2.982322</td>
<td>0.719962</td>
<td>75</td>
<td>non_westernized</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>And as always, we need it in tidy format (<a href="#tbl-taxonomicprofiling-alphadivgrouplong" class="quarto-xref">Table&nbsp;<span>10.16</span></a>) for plotnine.</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> alpha_diversity.melt(id_vars<span class="op">=</span><span class="st">'group'</span>, value_name<span class="op">=</span><span class="st">'alpha diversity'</span>, var_name<span class="op">=</span><span class="st">'diversity_index'</span>, ignore_index<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-alphadivgrouplong" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-alphadivgrouplong-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.16: Table of the shannon, simpson, and richness alpha diversity indicies for a subset of samples but with the group metadata but in long-form tidy format
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-alphadivgrouplong-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<colgroup>
<col style="width: 42%">
<col style="width: 19%">
<col style="width: 19%">
<col style="width: 19%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>group</th>
<th>diversity_index</th>
<th>alpha diversity</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>sample_id</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>ERR5766177</td>
<td>ERR5766177</td>
<td>shannon</td>
<td>3.032945</td>
</tr>
<tr class="odd">
<td>de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td>westernized</td>
<td>shannon</td>
<td>0.798112</td>
</tr>
<tr class="even">
<td>PNP_Main_283</td>
<td>westernized</td>
<td>shannon</td>
<td>5.092878</td>
</tr>
<tr class="odd">
<td>PNP_Validation_55</td>
<td>westernized</td>
<td>shannon</td>
<td>3.670162</td>
</tr>
<tr class="even">
<td>G80275</td>
<td>westernized</td>
<td>shannon</td>
<td>3.831358</td>
</tr>
<tr class="odd">
<td>…</td>
<td>…</td>
<td>…</td>
<td>…</td>
</tr>
<tr class="even">
<td>KHG_9</td>
<td>non_westernized</td>
<td>richness</td>
<td>87.000000</td>
</tr>
<tr class="odd">
<td>A48_01_1FE</td>
<td>non_westernized</td>
<td>richness</td>
<td>53.000000</td>
</tr>
<tr class="even">
<td>KHG_1</td>
<td>non_westernized</td>
<td>richness</td>
<td>108.000000</td>
</tr>
<tr class="odd">
<td>TZ_81781</td>
<td>non_westernized</td>
<td>richness</td>
<td>44.000000</td>
</tr>
<tr class="even">
<td>A09_01_1FE</td>
<td>non_westernized</td>
<td>richness</td>
<td>75.000000</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p>We now make a violin plot to compare the alpha diversity for each group, faceted by the type of alpha diversity index (<a href="#fig-taxonomicprofiling-diversityviolinplots" class="quarto-xref">Figure&nbsp;<span>10.7</span></a>).</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> ggplot(alpha_diversity, aes(x<span class="op">=</span><span class="st">'group'</span>, y<span class="op">=</span><span class="st">'alpha diversity'</span>, color<span class="op">=</span><span class="st">'group'</span>))</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> geom_violin()</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> geom_jitter()</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme_classic()</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> facet_wrap(<span class="st">'~diversity_index'</span>, scales <span class="op">=</span> <span class="st">'free'</span>)</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme(axis_text_x<span class="op">=</span>element_text(rotation<span class="op">=</span><span class="dv">45</span>, hjust<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme(subplots_adjust<span class="op">=</span>{<span class="st">'wspace'</span>: <span class="fl">0.15</span>})</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-diversityviolinplots" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-diversityviolinplots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_91_1.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-diversityviolinplots-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.7: Three groups of violin plots of an ancient sample, westernised samples and non-westernised samples (x-axis) of the alpha diversity (y-axis) calculated for richness, shannon and simpson alpha indicies
</figcaption>
</figure>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why do we observe a smaller species richness and diversity in our sample ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-32-contents" aria-controls="callout-32" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-32" class="callout-32-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There are different possibilies: - First, our ancient sample might genuilely have a lower diversity. - Second, our ancient sample diversity might have been lost through the degradation of the sample. - And third, the most likely explanation here: to make this tutorial fast enough, the reads were downsampled by quite a bit, which artifically lowered the diversity of this ancient sample.</p>
</div>
</div>
</div>
</section>
<section id="beta-diversity" class="level3" data-number="10.13.2">
<h3 data-number="10.13.2" class="anchored" data-anchor-id="beta-diversity"><span class="header-section-number">10.13.2</span> Beta diversity</h3>
<p>The Beta diversity is the measure of diversity between a pair of samples. It is used to compare the diversity between samples and see how they relate.</p>
<p>We will compute the beta diversity using the <strong>bray-curtis</strong> dissimilarity</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>beta_diversity <span class="op">=</span> skbio.diversity.beta_diversity(metric<span class="op">=</span><span class="st">'braycurtis'</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns, validate<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We get a distance matrix</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(beta_diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-33-contents" aria-controls="callout-33" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-33" class="callout-33-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<pre><code>    201x201 distance matrix
    IDs:
    'ERR5766177', 'de028ad4-7ae6-11e9-a106-68b59976a384', 'PNP_Main_283', ...
    Data:
    [[0.         1.         0.81508134 ... 0.85716612 0.69790092 0.8303726 ]
     [1.         0.         0.99988327 ... 0.99853413 0.994116   0.99877258]
     [0.81508134 0.99988327 0.         ... 0.82311942 0.87202543 0.91363156]
     ...
     [0.85716612 0.99853413 0.82311942 ... 0.         0.84253376 0.76616679]
     [0.69790092 0.994116   0.87202543 ... 0.84253376 0.         0.82409272]
     [0.8303726  0.99877258 0.91363156 ... 0.76616679 0.82409272 0.        ]]</code></pre>
</div>
</div>
</div>
<p>To visualise this distance matrix in a lower dimensional space, we’ll use a <strong>PCoA</strong>, which is is a method very similar to a PCA, but taking a distance matrix as input.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>pcoa <span class="op">=</span> skbio.stats.ordination.pcoa(beta_diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see command output">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-34-contents" aria-controls="callout-34" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see command output
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-34" class="callout-34-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb64"><pre class="sourceCode default code-with-copy"><code class="sourceCode default"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>    /Users/maxime/mambaforge/envs/summer_school_microbiome/lib/python3.9/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:143: RuntimeWarning:</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>    The result contains negative eigenvalues. Please compare their magnitude with the magnitude of some of the largest positive eigenvalues.</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>    If the negative ones are smaller, it's probably safe to ignore them, but if they are large in magnitude, the results won't be useful.</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    See the Notes section for more details. The smallest eigenvalue is -0.25334842745723996 and the largest is 10.204440747987945.</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="sourceCode" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>pcoa.samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-taxonomicprofiling-tblpca" class="cell quarto-float quarto-figure quarto-figure-center anchored" data-tbl-cap-location="top">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-taxonomicprofiling-tblpca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;10.17: Table of principal coordinates (columns) for each of the samples (rows)
</figcaption>
<div aria-describedby="tbl-taxonomicprofiling-tblpca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div id="ftzjndefyw" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:scroll;overflow-y:auto;width:auto;height:auto;">
<style>#ftzjndefyw table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ftzjndefyw thead, #ftzjndefyw tbody, #ftzjndefyw tfoot, #ftzjndefyw tr, #ftzjndefyw td, #ftzjndefyw th {
  border-style: none;
}

#ftzjndefyw p {
  margin: 0;
  padding: 0;
}

#ftzjndefyw .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: 100%;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ftzjndefyw .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ftzjndefyw .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ftzjndefyw .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ftzjndefyw .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ftzjndefyw .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ftzjndefyw .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ftzjndefyw .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ftzjndefyw .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ftzjndefyw .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ftzjndefyw .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ftzjndefyw .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ftzjndefyw .gt_spanner_row {
  border-bottom-style: hidden;
}

#ftzjndefyw .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ftzjndefyw .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ftzjndefyw .gt_from_md > :first-child {
  margin-top: 0;
}

#ftzjndefyw .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ftzjndefyw .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ftzjndefyw .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ftzjndefyw .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ftzjndefyw .gt_row_group_first td {
  border-top-width: 2px;
}

#ftzjndefyw .gt_row_group_first th {
  border-top-width: 2px;
}

#ftzjndefyw .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ftzjndefyw .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ftzjndefyw .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ftzjndefyw .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ftzjndefyw .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ftzjndefyw .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ftzjndefyw .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ftzjndefyw .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ftzjndefyw .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ftzjndefyw .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ftzjndefyw .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ftzjndefyw .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ftzjndefyw .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ftzjndefyw .gt_left {
  text-align: left;
}

#ftzjndefyw .gt_center {
  text-align: center;
}

#ftzjndefyw .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ftzjndefyw .gt_font_normal {
  font-weight: normal;
}

#ftzjndefyw .gt_font_bold {
  font-weight: bold;
}

#ftzjndefyw .gt_font_italic {
  font-style: italic;
}

#ftzjndefyw .gt_super {
  font-size: 65%;
}

#ftzjndefyw .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ftzjndefyw .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ftzjndefyw .gt_indent_1 {
  text-indent: 5px;
}

#ftzjndefyw .gt_indent_2 {
  text-indent: 10px;
}

#ftzjndefyw .gt_indent_3 {
  text-indent: 15px;
}

#ftzjndefyw .gt_indent_4 {
  text-indent: 20px;
}

#ftzjndefyw .gt_indent_5 {
  text-indent: 25px;
}

#ftzjndefyw .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#ftzjndefyw div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="...1" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">...1</th>
<th id="PC1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC1</th>
<th id="PC2" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC2</th>
<th id="PC3" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC3</th>
<th id="PC4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC4</th>
<th id="PC5" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC5</th>
<th id="PC6" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC6</th>
<th id="PC7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC7</th>
<th id="PC8" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC8</th>
<th id="PC9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC9</th>
<th id="PC10" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC10</th>
<th id="...12" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">...12</th>
<th id="PC192" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC192</th>
<th id="PC193" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC193</th>
<th id="PC194" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC194</th>
<th id="PC195" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC195</th>
<th id="PC196" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC196</th>
<th id="PC197" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC197</th>
<th id="PC198" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC198</th>
<th id="PC199" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC199</th>
<th id="PC200" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC200</th>
<th id="PC201" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">PC201</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_left" headers="...1">ERR5766177</td>
<td class="gt_row gt_right" headers="PC1">0.216901</td>
<td class="gt_row gt_right" headers="PC2">-0.039778</td>
<td class="gt_row gt_right" headers="PC3">0.107412</td>
<td class="gt_row gt_right" headers="PC4">0.273272</td>
<td class="gt_row gt_right" headers="PC5">0.02054</td>
<td class="gt_row gt_right" headers="PC6">0.114876</td>
<td class="gt_row gt_right" headers="PC7">-0.256332</td>
<td class="gt_row gt_right" headers="PC8">-0.151069</td>
<td class="gt_row gt_right" headers="PC9">0.097451</td>
<td class="gt_row gt_right" headers="PC10">0.060211</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">de028ad4-7ae6-11e9-a106-68b59976a384</td>
<td class="gt_row gt_right" headers="PC1">-0.099355</td>
<td class="gt_row gt_right" headers="PC2">0.145224</td>
<td class="gt_row gt_right" headers="PC3">-0.191676</td>
<td class="gt_row gt_right" headers="PC4">0.127626</td>
<td class="gt_row gt_right" headers="PC5">0.119754</td>
<td class="gt_row gt_right" headers="PC6">-0.132209</td>
<td class="gt_row gt_right" headers="PC7">-0.097382</td>
<td class="gt_row gt_right" headers="PC8">0.036728</td>
<td class="gt_row gt_right" headers="PC9">0.081294</td>
<td class="gt_row gt_right" headers="PC10">-0.056686</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">PNP_Main_283</td>
<td class="gt_row gt_right" headers="PC1">-0.214108</td>
<td class="gt_row gt_right" headers="PC2">-0.147466</td>
<td class="gt_row gt_right" headers="PC3">0.116027</td>
<td class="gt_row gt_right" headers="PC4">0.090059</td>
<td class="gt_row gt_right" headers="PC5">0.076644</td>
<td class="gt_row gt_right" headers="PC6">0.111536</td>
<td class="gt_row gt_right" headers="PC7">0.092115</td>
<td class="gt_row gt_right" headers="PC8">0.026477</td>
<td class="gt_row gt_right" headers="PC9">-0.00646</td>
<td class="gt_row gt_right" headers="PC10">-0.018592</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">PNP_Validation_55</td>
<td class="gt_row gt_right" headers="PC1">0.244827</td>
<td class="gt_row gt_right" headers="PC2">-0.173996</td>
<td class="gt_row gt_right" headers="PC3">-0.311197</td>
<td class="gt_row gt_right" headers="PC4">-0.012836</td>
<td class="gt_row gt_right" headers="PC5">0.031759</td>
<td class="gt_row gt_right" headers="PC6">0.117548</td>
<td class="gt_row gt_right" headers="PC7">0.148715</td>
<td class="gt_row gt_right" headers="PC8">-0.135641</td>
<td class="gt_row gt_right" headers="PC9">0.03473</td>
<td class="gt_row gt_right" headers="PC10">-0.009395</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">G80275</td>
<td class="gt_row gt_right" headers="PC1">-0.261358</td>
<td class="gt_row gt_right" headers="PC2">-0.077147</td>
<td class="gt_row gt_right" headers="PC3">-0.254374</td>
<td class="gt_row gt_right" headers="PC4">-0.065932</td>
<td class="gt_row gt_right" headers="PC5">0.088538</td>
<td class="gt_row gt_right" headers="PC6">0.16597</td>
<td class="gt_row gt_right" headers="PC7">-0.00526</td>
<td class="gt_row gt_right" headers="PC8">-0.028739</td>
<td class="gt_row gt_right" headers="PC9">-0.002016</td>
<td class="gt_row gt_right" headers="PC10">0.015719</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">...</td>
<td class="gt_row gt_right" headers="PC1">...</td>
<td class="gt_row gt_right" headers="PC2">...</td>
<td class="gt_row gt_right" headers="PC3">...</td>
<td class="gt_row gt_right" headers="PC4">...</td>
<td class="gt_row gt_right" headers="PC5">...</td>
<td class="gt_row gt_right" headers="PC6">...</td>
<td class="gt_row gt_right" headers="PC7">...</td>
<td class="gt_row gt_right" headers="PC8">...</td>
<td class="gt_row gt_right" headers="PC9">...</td>
<td class="gt_row gt_right" headers="PC10">...</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">...</td>
<td class="gt_row gt_right" headers="PC193">...</td>
<td class="gt_row gt_right" headers="PC194">...</td>
<td class="gt_row gt_right" headers="PC195">...</td>
<td class="gt_row gt_right" headers="PC196">...</td>
<td class="gt_row gt_right" headers="PC197">...</td>
<td class="gt_row gt_right" headers="PC198">...</td>
<td class="gt_row gt_right" headers="PC199">...</td>
<td class="gt_row gt_right" headers="PC200">...</td>
<td class="gt_row gt_right" headers="PC201">...</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">KHG_9</td>
<td class="gt_row gt_right" headers="PC1">0.296057</td>
<td class="gt_row gt_right" headers="PC2">-0.1503</td>
<td class="gt_row gt_right" headers="PC3">0.013941</td>
<td class="gt_row gt_right" headers="PC4">0.032649</td>
<td class="gt_row gt_right" headers="PC5">-0.147692</td>
<td class="gt_row gt_right" headers="PC6">0.019663</td>
<td class="gt_row gt_right" headers="PC7">-0.06312</td>
<td class="gt_row gt_right" headers="PC8">-0.034453</td>
<td class="gt_row gt_right" headers="PC9">-0.073514</td>
<td class="gt_row gt_right" headers="PC10">0.070085</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">A48_01_1FE</td>
<td class="gt_row gt_right" headers="PC1">0.110621</td>
<td class="gt_row gt_right" headers="PC2">0.030971</td>
<td class="gt_row gt_right" headers="PC3">0.154231</td>
<td class="gt_row gt_right" headers="PC4">-0.185961</td>
<td class="gt_row gt_right" headers="PC5">-0.008512</td>
<td class="gt_row gt_right" headers="PC6">-0.10342</td>
<td class="gt_row gt_right" headers="PC7">0.028169</td>
<td class="gt_row gt_right" headers="PC8">-0.04453</td>
<td class="gt_row gt_right" headers="PC9">0.041902</td>
<td class="gt_row gt_right" headers="PC10">0.068597</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">KHG_1</td>
<td class="gt_row gt_right" headers="PC1">-0.100009</td>
<td class="gt_row gt_right" headers="PC2">0.167885</td>
<td class="gt_row gt_right" headers="PC3">0.009915</td>
<td class="gt_row gt_right" headers="PC4">0.076842</td>
<td class="gt_row gt_right" headers="PC5">-0.405582</td>
<td class="gt_row gt_right" headers="PC6">-0.039111</td>
<td class="gt_row gt_right" headers="PC7">-0.006421</td>
<td class="gt_row gt_right" headers="PC8">-0.009774</td>
<td class="gt_row gt_right" headers="PC9">-0.072252</td>
<td class="gt_row gt_right" headers="PC10">0.15</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="...1">TZ_81781</td>
<td class="gt_row gt_right" headers="PC1">0.405716</td>
<td class="gt_row gt_right" headers="PC2">-0.139297</td>
<td class="gt_row gt_right" headers="PC3">-0.075026</td>
<td class="gt_row gt_right" headers="PC4">-0.079716</td>
<td class="gt_row gt_right" headers="PC5">-0.053264</td>
<td class="gt_row gt_right" headers="PC6">-0.119271</td>
<td class="gt_row gt_right" headers="PC7">0.068261</td>
<td class="gt_row gt_right" headers="PC8">-0.018821</td>
<td class="gt_row gt_right" headers="PC9">0.198152</td>
<td class="gt_row gt_right" headers="PC10">-0.012792</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="...1">A09_01_1FE</td>
<td class="gt_row gt_right" headers="PC1">0.089101</td>
<td class="gt_row gt_right" headers="PC2">0.471135</td>
<td class="gt_row gt_right" headers="PC3">0.069629</td>
<td class="gt_row gt_right" headers="PC4">-0.125644</td>
<td class="gt_row gt_right" headers="PC5">-0.036793</td>
<td class="gt_row gt_right" headers="PC6">0.115151</td>
<td class="gt_row gt_right" headers="PC7">0.060507</td>
<td class="gt_row gt_right" headers="PC8">-0.000912</td>
<td class="gt_row gt_right" headers="PC9">-0.027239</td>
<td class="gt_row gt_right" headers="PC10">-0.138436</td>
<td class="gt_row gt_right" headers="...12">...</td>
<td class="gt_row gt_right" headers="PC192">0</td>
<td class="gt_row gt_right" headers="PC193">0</td>
<td class="gt_row gt_right" headers="PC194">0</td>
<td class="gt_row gt_right" headers="PC195">0</td>
<td class="gt_row gt_right" headers="PC196">0</td>
<td class="gt_row gt_right" headers="PC197">0</td>
<td class="gt_row gt_right" headers="PC198">0</td>
<td class="gt_row gt_right" headers="PC199">0</td>
<td class="gt_row gt_right" headers="PC200">0</td>
<td class="gt_row gt_right" headers="PC201">0</td>
</tr>
</tbody>
</table>

</div>
</div>
</figure>
</div>
<p>Let’s look at the variance explained by the first axes by using a scree plot (<a href="#fig-taxonomicprofiling-screeplot" class="quarto-xref">Figure&nbsp;<span>10.8</span></a>).</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>var_explained <span class="op">=</span> pcoa.proportion_explained[:<span class="dv">9</span>].to_frame(name<span class="op">=</span><span class="st">'variance explained'</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'PC'</span>})</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>ggplot(var_explained, aes(x<span class="op">=</span><span class="st">'PC'</span>, y<span class="op">=</span><span class="st">'variance explained'</span>, group<span class="op">=</span><span class="dv">1</span>)) <span class="op">\</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_point() <span class="op">\</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_line() <span class="op">\</span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-screeplot" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-screeplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_102_0.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-screeplot-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.8: Scree plot describing the variance explained (Y-axis), for each Principal Componanent (X-axis), with a curved line from PC1 having highest variance to lowest on PC9.
</figcaption>
</figure>
</div>
<p>In this scree plot, we’re looking for the “elbow”, where there is a drop in the slope. Here, it seems that most of the variance is captures by the 3 first principal components</p>
<div class="sourceCode" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>pcoa_embed <span class="op">=</span> pcoa.samples[[<span class="st">'PC1'</span>,<span class="st">'PC2'</span>,<span class="st">'PC3'</span>]].rename_axis(<span class="st">'sample'</span>).reset_index()</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>pcoa_embed <span class="op">=</span> (</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>    pcoa_embed</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>    .merge(metadata[[<span class="st">'sample_id'</span>, <span class="st">'non_westernized'</span>]], left_on<span class="op">=</span><span class="st">'sample'</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>, how<span class="op">=</span><span class="st">'outer'</span>)</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>    .drop(<span class="st">'sample_id'</span>, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'non_westernized'</span>:<span class="st">'group'</span>})</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>pcoa_embed[<span class="st">'group'</span>] <span class="op">=</span> pcoa_embed[<span class="st">'group'</span>].replace({<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>,<span class="st">'no'</span>:<span class="st">'westernized'</span>, pd.NA:<span class="st">'ERR5766177'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s first look at these components with 2D plots (<a href="#fig-taxonomicprofiling-poca1" class="quarto-xref">Figure&nbsp;<span>10.9</span></a>, <a href="#fig-taxonomicprofiling-poca2" class="quarto-xref">Figure&nbsp;<span>10.10</span></a>)</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>ggplot(pcoa_embed, aes(x<span class="op">=</span><span class="st">'PC1'</span>, y<span class="op">=</span><span class="st">'PC2'</span>, color<span class="op">=</span><span class="st">'group'</span>)) <span class="op">\</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_point() <span class="op">\</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic() <span class="op">\</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-poca1" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-poca1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_107_0.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-poca1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.9: Principal Coordinate Analysis plot of PC1 (X-axis) and PC2 (Y-axis), with three groups of points in the scatter plot - blue circles of westernised data points in the bottom left, overlapping with green circles of non-westernised datapoints in the top right, and the single ancient sample as a red circle falling in between the two on the right of the overlap
</figcaption>
</figure>
</div>
<div class="sourceCode" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>ggplot(pcoa_embed, aes(x<span class="op">=</span><span class="st">'PC1'</span>, y<span class="op">=</span><span class="st">'PC3'</span>, color<span class="op">=</span><span class="st">'group'</span>)) <span class="op">+</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>geom_point() <span class="op">+</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>theme_classic() <span class="op">+</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-poca2" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-poca2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_108_0.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-poca2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.10: Principal Coordinate Analysis plot of PC1 (X-axis) and PC3 (Y-axis), a similar overlap between westernised/non-westernised individuals and position of the ancient sample as in the PC1-PC2 PCoA, however this time in a horseshoe shape from bottom left for the westernised data points, curving up to the top of PC3 at a peak, and then falling again at the top of PC1
</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Expand to see additional visualisations">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-35-contents" aria-controls="callout-35" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Expand to see additional visualisations
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-35" class="callout-35-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>You can also plot the data above as a with a 3d plot if you were to run the following command</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.express <span class="im">as</span> px</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> px.scatter_3d(pcoa_embed, x<span class="op">=</span><span class="st">"PC1"</span>, y<span class="op">=</span><span class="st">"PC2"</span>, z<span class="op">=</span><span class="st">"PC3"</span>,</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>                  color <span class="op">=</span> <span class="st">"group"</span>,</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>                  color_discrete_map<span class="op">=</span>{<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>},</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>                  hover_name<span class="op">=</span><span class="st">"sample"</span>)</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>fig.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Finally, we can also visualise this distance matrix using a clustered heatmap, where pairs of sample with a small beta diversity are clustered together (<a href="#fig-taxonomicprofiling-heatmap" class="quarto-xref">Figure&nbsp;<span>10.11</span></a>).</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.spatial <span class="im">as</span> sp, scipy.cluster.hierarchy <span class="im">as</span> hc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We set the color in seaborn to match the color palette we’ve used so far.</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>pcoa_embed[<span class="st">'colour'</span>] <span class="op">=</span> pcoa_embed[<span class="st">'group'</span>].<span class="bu">map</span>({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>linkage <span class="op">=</span> hc.linkage(sp.distance.squareform(beta_diversity.to_data_frame()), method<span class="op">=</span><span class="st">'average'</span>)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>sns.clustermap(</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>    beta_diversity.to_data_frame(),</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>    row_linkage<span class="op">=</span>linkage,</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>    col_linkage<span class="op">=</span>linkage,</span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>    row_colors <span class="op">=</span> pcoa_embed[<span class="st">'colour'</span>].to_list()</span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-taxonomicprofiling-heatmap" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-taxonomicprofiling-heatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="assets/images/chapters/taxonomic-profiling/analysis_115_1.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-taxonomicprofiling-heatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10.11: Sample-by-sample clustered heatmap, with tree representation of the clustering on the left and top of the heatmap
</figcaption>
</figure>
</div>
<div class="callout callout-style-simple callout-tip callout-titled" title="Question">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Based on the PCoA and heatmap, how you think our ancient sample relates to modern ones ?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled" title="Answer">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-37-contents" aria-controls="callout-37" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-37" class="callout-37-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>The sample we looked at in this tutorial is actually coming from the <span class="citation" data-cites="maixner2021hallstatt">(<a href="#ref-maixner2021hallstatt" role="doc-biblioref">Maixner et al. 2021</a>)</span> article, looking at a human gut microbiome found in a salt mine in nowadays Austria. This sample is actually for the Baroque period (150 years BP). Reassuringly for our analysis, the title of this article is <em>Hallstatt miners consumed blue cheese and beer during the Iron Age and retained a non-Westernized gut microbiome until the Baroque period</em>.</p>
</div>
</div>
</div>
</section>
</section>
<section id="optional-clean-up" class="level2" data-number="10.14">
<h2 data-number="10.14" class="anchored" data-anchor-id="optional-clean-up"><span class="header-section-number">10.14</span> (Optional) clean-up</h2>
<p>Let’s clean up your working directory by removing all the data and output from this chapter.</p>
<p>When closing your <code>jupyter</code> notebook(s), say no to saving any additional files.</p>
<p>Press <kbd>ctrl</kbd> + <kbd>c</kbd> on your terminal, and type <kbd>y</kbd> when requested. Once completed, the command below will remove the <code>/&lt;PATH&gt;/&lt;TO&gt;/taxonomic-profiling</code> directory <em>as well as all of its contents</em>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Pro Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Always be VERY careful when using <code>rm -r</code>. Check 3x that the path you are specifying is exactly what you want to delete and nothing more before pressing ENTER!</p>
</div>
</div>
<div class="sourceCode" id="cb73"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span> <span class="at">-r</span> /<span class="op">&lt;</span>PATH<span class="op">&gt;</span>/<span class="op">&lt;</span>TO<span class="op">&gt;</span>/taxonomic-profiling<span class="pp">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Once deleted you can move elsewhere (e.g.&nbsp;<code>cd ~</code>).</p>
<p>We can also get out of the <code>conda</code> environment with</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To delete the conda environment</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> remove <span class="at">--name</span> taxonomic-profiling <span class="at">--all</span> <span class="at">-y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="summary" class="level2" data-number="10.15">
<h2 data-number="10.15" class="anchored" data-anchor-id="summary"><span class="header-section-number">10.15</span> Summary</h2>
<p>In this practical session we</p>
<ul>
<li>Looked at how to process the raw sequencing data to focus only on the non-human reads</li>
<li>Profiled these reads with the Metaphlan3 taxonomic profiler</li>
<li>Looked at phylum composition, and alpha, and beta diversity with the help of PCoA and abundance heatmaps</li>
</ul>
<div class="callout callout-style-default callout-caution callout-titled" title="Cautionary note - sequencing depth">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Cautionary note - sequencing depth
</div>
</div>
<div class="callout-body-container callout-body">
<p>We worked on relative abundances provided by metaphlan. But beware that not all taxonomic profilers will output relative abundances. To perform quantitative analysis of taxonomic profiles, sequencing data need to be normalised for sequencing depth. There are many different ways to normalise sequencing data, but this out of the scope of this session. Just to name a few, the most commonly used are RLE, TSS, rarefaction, CLR, or GMPR.</p>
</div>
</div>
</section>
<section id="references" class="level2" data-number="10.16">
<h2 data-number="10.16" class="anchored" data-anchor-id="references"><span class="header-section-number">10.16</span> References</h2>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-breitwieserPavianInteractiveAnalysis2016" class="csl-entry" role="listitem">
Breitwieser, Florian P., and Steven L. Salzberg. 2016. <span>“Pavian: <span>Interactive</span> Analysis of Metagenomics Data for Microbiomics and Pathogen Identification.”</span> <em>bioRxiv</em>, October, 084715.
</div>
<div id="ref-Breitwieser2020-zw" class="csl-entry" role="listitem">
Breitwieser, Florian P, and Steven L Salzberg. 2020. <span>“Pavian: Interactive Analysis of Metagenomics Data for Microbiome Studies and Pathogen Identification.”</span> <em>Bioinformatics (Oxford, England)</em> 36 (4): 1303–4. <a href="https://doi.org/10.1093/bioinformatics/btz715">https://doi.org/10.1093/bioinformatics/btz715</a>.
</div>
<div id="ref-chenFastpUltrafastAllinone2018" class="csl-entry" role="listitem">
Chen, Shifu, Yanqing Zhou, Yaru Chen, and Jia Gu. 2018. <span>“Fastp: An Ultra-Fast All-in-One <span>FASTQ</span> Preprocessor.”</span> <em>Bioinformatics</em> 34 (17): i884–90.
</div>
<div id="ref-Fellows-Yates2020-iy" class="csl-entry" role="listitem">
Fellows Yates, James A, T C Lamnidis, M Borry, A A Valtueña, Zandra Fagernäs, S Clayton, M Garcia, Judith Neukamm, and Alexander Peltzer. 2020. <span>“Reproducible, Portable, and Efficient Ancient Genome Reconstruction with Nf-Core/Eager.”</span> <em>bioRxiv</em> 9 (June): e10947. <a href="https://doi.org/10.7717/peerj.10947">https://doi.org/10.7717/peerj.10947</a>.
</div>
<div id="ref-maixner2021hallstatt" class="csl-entry" role="listitem">
Maixner, Frank, Mohamed S Sarhan, Kun D Huang, Adrian Tett, Alexander Schoenafinger, Stefania Zingale, Aitor Blanco-Mı́guez, et al. 2021. <span>“Hallstatt Miners Consumed Blue Cheese and Beer During the Iron Age and Retained a Non-Westernized Gut Microbiome Until the Baroque Period.”</span> <em>Current Biology</em> 31 (23): 5149–62.
</div>
<div id="ref-ondovInteractiveMetagenomicVisualization2011" class="csl-entry" role="listitem">
Ondov, Brian D., Nicholas H. Bergman, and Adam M. Phillippy. 2011a. <span>“Interactive Metagenomic Visualization in a <span>Web</span> Browser.”</span> <em>BMC Bioinformatics</em> 12 (1): 385.
</div>
<div id="ref-Ondov2011-se" class="csl-entry" role="listitem">
Ondov, Brian D, Nicholas H Bergman, and Adam M Phillippy. 2011b. <span>“Interactive Metagenomic Visualization in a Web Browser.”</span> <em>BMC Bioinformatics</em> 12 (1): 385. <a href="https://doi.org/10.1186/1471-2105-12-385">https://doi.org/10.1186/1471-2105-12-385</a>.
</div>
<div id="ref-pasolliAccessibleCuratedMetagenomic2017" class="csl-entry" role="listitem">
Pasolli, Edoardo, Lucas Schiffer, Paolo Manghi, Audrey Renson, Valerie Obenchain, Duy Tin Truong, Francesco Beghini, et al. 2017. <span>“Accessible, Curated Metagenomic Data Through <span>ExperimentHub</span>.”</span> <em>Nature Methods</em> 14 (11): 1023–24.
</div>
<div id="ref-rebackPandasdevPandasPandas2022" class="csl-entry" role="listitem">
Reback, Jeff, jbrockmendel, Wes McKinney, Joris Van den Bossche, Tom Augspurger, Matthew Roeschke, Simon Hawkins, et al. 2022. <span>“Pandas-Dev/Pandas: <span>Pandas</span> 1.4.2.”</span> Zenodo.
</div>
<div id="ref-scikit-bioScikitbioBioinformaticsLibrary2022" class="csl-entry" role="listitem">
scikit-bio, developers. 2022. <span>“Scikit-Bio: <span>A Bioinformatics Library</span> for <span>Data Scientists</span>, <span>Students</span>, and <span>Developers</span>.”</span>
</div>
<div id="ref-segata2012metagenomic" class="csl-entry" role="listitem">
Segata, Nicola, Levi Waldron, Annalisa Ballarini, Vagheesh Narasimhan, Olivier Jousson, and Curtis Huttenhower. 2012. <span>“Metagenomic Microbial Community Profiling Using Unique Clade-Specific Marker Genes.”</span> <em>Nature Methods</em> 9 (8): 811–14.
</div>
<div id="ref-sharptonIntroductionAnalysisShotgun2014" class="csl-entry" role="listitem">
Sharpton, Thomas J. 2014. <span>“<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4059276">An Introduction to the Analysis of Shotgun Metagenomic Data</a>.”</span> <em>Frontiers in Plant Science</em> 5 (June).
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/spaam-community\.github\.org\/intro-to-ancient-metagenomics-book");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./section-ancient-metagenomics.html" class="pagination-link" aria-label="Ancient Metagenomics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Ancient Metagenomics</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./denovo-assembly.html" class="pagination-link" aria-label="_De novo_ Genome Assembly">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title"><em>De novo</em> Genome Assembly</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2024 SPAAM Community &amp; Authors with ❤️. Available under <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>