<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Robin Warner, Kevin Nota, and Maxime Borry">

<title>Introduction to Ancient Metagenomics - 8&nbsp; Introduction to Python and Pandas</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./git-github.html" rel="next">
<link href="./r-tidyverse.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-useful-skills.html">Useful Skills</a></li><li class="breadcrumb-item"><a href="./python-pandas.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main tools-wide">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-evolutionary-biology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Evolutionary Biology</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication-decontamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Authentication and Decontamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Introduction to Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancientmetagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction-to-data-manipulation-in-python-with-pandas-and-visulization" id="toc-introduction-to-data-manipulation-in-python-with-pandas-and-visulization" class="nav-link active" data-scroll-target="#introduction-to-data-manipulation-in-python-with-pandas-and-visulization"><span class="header-section-number">8.1</span> Introduction to data manipulation in Python with Pandas and visulization</a></li>
  <li><a href="#table-of-content" id="toc-table-of-content" class="nav-link" data-scroll-target="#table-of-content"><span class="header-section-number">8.2</span> Table of content:</a></li>
  <li><a href="#working-in-a-jupyter-environment" id="toc-working-in-a-jupyter-environment" class="nav-link" data-scroll-target="#working-in-a-jupyter-environment"><span class="header-section-number">8.3</span> Working in a jupyter environment</a></li>
  <li><a href="#pandas" id="toc-pandas" class="nav-link" data-scroll-target="#pandas"><span class="header-section-number">8.4</span> Pandas</a>
  <ul class="collapse">
  <li><a href="#getting-started" id="toc-getting-started" class="nav-link" data-scroll-target="#getting-started"><span class="header-section-number">8.4.1</span> Getting started</a></li>
  <li><a href="#pandas-data-structures" id="toc-pandas-data-structures" class="nav-link" data-scroll-target="#pandas-data-structures"><span class="header-section-number">8.4.2</span> Pandas data structures</a></li>
  <li><a href="#reading-in-data" id="toc-reading-in-data" class="nav-link" data-scroll-target="#reading-in-data"><span class="header-section-number">8.4.3</span> Reading in data</a></li>
  </ul></li>
  <li><a href="#reading-data-with-pandas" id="toc-reading-data-with-pandas" class="nav-link" data-scroll-target="#reading-data-with-pandas"><span class="header-section-number">8.5</span> 3 - Reading data with Pandas</a>
  <ul class="collapse">
  <li><a href="#listing-the-columns-of-the-sample-dataframe" id="toc-listing-the-columns-of-the-sample-dataframe" class="nav-link" data-scroll-target="#listing-the-columns-of-the-sample-dataframe"><span class="header-section-number">8.5.1</span> Listing the columns of the sample dataframe</a></li>
  <li><a href="#looking-at-the-data-type-of-the-sample-dataframe" id="toc-looking-at-the-data-type-of-the-sample-dataframe" class="nav-link" data-scroll-target="#looking-at-the-data-type-of-the-sample-dataframe"><span class="header-section-number">8.5.2</span> Looking at the data type of the sample dataframe</a></li>
  <li><a href="#lets-inspect-our-data" id="toc-lets-inspect-our-data" class="nav-link" data-scroll-target="#lets-inspect-our-data"><span class="header-section-number">8.5.3</span> Let’s inspect our data</a></li>
  </ul></li>
  <li><a href="#computing-basic-statistics" id="toc-computing-basic-statistics" class="nav-link" data-scroll-target="#computing-basic-statistics"><span class="header-section-number">8.6</span> 5 - Computing basic statistics</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering"><span class="header-section-number">8.7</span> 6 - Filtering</a></li>
  <li><a href="#groupby-operations-and-computing-statistics-on-grouped-values" id="toc-groupby-operations-and-computing-statistics-on-grouped-values" class="nav-link" data-scroll-target="#groupby-operations-and-computing-statistics-on-grouped-values"><span class="header-section-number">8.8</span> 7 - GroupBy operations, and computing statistics on grouped values</a></li>
  <li><a href="#reshaping-data-from-wide-to-long-and-back" id="toc-reshaping-data-from-wide-to-long-and-back" class="nav-link" data-scroll-target="#reshaping-data-from-wide-to-long-and-back"><span class="header-section-number">8.9</span> 8 - Reshaping data, from wide to long and back</a>
  <ul class="collapse">
  <li><a href="#from-wide-to-longtidy" id="toc-from-wide-to-longtidy" class="nav-link" data-scroll-target="#from-wide-to-longtidy"><span class="header-section-number">8.9.1</span> From wide to long/tidy</a></li>
  <li><a href="#from-longtidy-to-wide-format-using-the-.pivot-function." id="toc-from-longtidy-to-wide-format-using-the-.pivot-function." class="nav-link" data-scroll-target="#from-longtidy-to-wide-format-using-the-.pivot-function."><span class="header-section-number">8.9.2</span> From long/tidy to wide format using the <code>.pivot()</code> function.</a></li>
  </ul></li>
  <li><a href="#joining-two-different-tables" id="toc-joining-two-different-tables" class="nav-link" data-scroll-target="#joining-two-different-tables"><span class="header-section-number">8.10</span> 9 - Joining two different tables</a></li>
  <li><a href="#visualizing-some-of-the-results-with-plotnine" id="toc-visualizing-some-of-the-results-with-plotnine" class="nav-link" data-scroll-target="#visualizing-some-of-the-results-with-plotnine"><span class="header-section-number">8.11</span> 10 - Visualizing some of the results with Plotnine</a></li>
  <li><a href="#bonus-dealing-with-ill-formatted-columns" id="toc-bonus-dealing-with-ill-formatted-columns" class="nav-link" data-scroll-target="#bonus-dealing-with-ill-formatted-columns"><span class="header-section-number">8.12</span> 11 - Bonus, dealing with ill-formatted columns</a></li>
  <li><a href="#lecture-from-2022" id="toc-lecture-from-2022" class="nav-link" data-scroll-target="#lecture-from-2022"><span class="header-section-number">8.13</span> Lecture from 2022</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Robin Warner, Kevin Nota, and Maxime Borry </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This session is typically ran held in parallel to the Introduction to R and Tidyverse. Participants of the summer schools chose which to attend based on their prior experience. We recommend the <a href="./r-tidyverse.html">introduction to R session</a> if you have no experience with neither R nor Python.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to create the <a href="./before-you-start.html#creating-a-conda-environment">conda environment</a> from the <code>yml</code> file in the following <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book/raw/main/assets/envs/python-pandas.yml">link</a> (right click and save as to download), and once created, activate the environment with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate python-pandas</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="introduction-to-data-manipulation-in-python-with-pandas-and-visulization" class="level2" data-number="8.1">
<h2 data-number="8.1" class="anchored" data-anchor-id="introduction-to-data-manipulation-in-python-with-pandas-and-visulization"><span class="header-section-number">8.1</span> Introduction to data manipulation in Python with Pandas and visulization</h2>
<p>SPAAM Summer School 2023</p>
<p>Over the last few years, Python has gained popularity thanks to the numerous libraries (packages with prewritten functions) in the field of machine learning, statistical data analysis, and bioinformatics. While a few years ago, it was often necessary to go to R for performing routine data manipulation and analysis tasks, nowadays Python has a vast ecosystem of libraries. Libraries exist for many different file formats that you might encounter in metagenomics, such as fasta, fastq, sam, bam, etc.</p>
<p>This tutorial/walkthrough will provide a short introduction to the most popular libraries for data analysis <a href="https://pandas.pydata.org/">pandas</a>. This library has functions for reading and manipulating tabular data similar to the data.frame() in R together with some basic data plotting codes.</p>
<p>The aim of this walkthrough is to first: get familiar with the Phyton code syntax and use Jupiter Notebook for executing codes and secondly get a kickstart to utilising the endless possibilities of data analysis in Python that can be applied to your data.</p>
</section>
<section id="table-of-content" class="level2" data-number="8.2">
<h2 data-number="8.2" class="anchored" data-anchor-id="table-of-content"><span class="header-section-number">8.2</span> Table of content:</h2>
<ul>
<li>8.3 <a href="#working-in-a-jupyter-environment">Working in a jupyter environment</a></li>
<li>8.4 <a href="#pandas">Pandas</a></li>
<li>2 - Foreword on Pandas</li>
<li>3 - Reading data with Pandas</li>
<li>4 - Dealing with missing data</li>
<li>5 - Computing basic statistics</li>
<li>6 - Filtering</li>
<li>8 - GroupBy operations</li>
<li>9 - Joining different tables</li>
<li>10 - Visualization with Plotnine</li>
</ul>
</section>
<section id="working-in-a-jupyter-environment" class="level2" data-number="8.3">
<h2 data-number="8.3" class="anchored" data-anchor-id="working-in-a-jupyter-environment"><span class="header-section-number">8.3</span> Working in a jupyter environment</h2>
<p>This tutorial run-through is using a <a href="https://jupyter.org">Jupyter Notebook</a> for writing &amp; executing Python code and for annotating.</p>
<p>Jupyter notebooks are convenient and have two types of cells: <strong>Markdown</strong> and <strong>Code</strong>. The <strong>markup cell</strong> syntax is very similar to <strong>R markdown</strong>. The markdown cells are used for annotating, which is important for sharing code with collaborators, reproducibility, and documentation.</p>
<p>A few examples are shown below. For a full list of possible syntax <a href="https://www.ibm.com/docs/en/watson-studio-local/1.2.3?topic=notebooks-markdown-jupyter-cheatsheet">click here</a> for a Jupyter Notebook cheatsheet.</p>
<p>list of <strong>markdown cell</strong> examples:</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>In many cases, there are multiple possible syntaxes which give the same result. We present only one way in this run-through.</p>
</div>
</div>
<p>Text</p>
<ul>
<li><code>**bold**</code> : <strong>bold</strong></li>
<li><code>*italics*</code> : <em>italics</em></li>
</ul>
<p>Code</p>
<ul>
<li>`inline code` : <code>inline code</code></li>
</ul>
<p>LaTeX math</p>
<ul>
<li><code>$ x = \frac{\pi}{42} $</code> : <span class="math display">\[ x = \frac{\pi}{42} \]</span></li>
</ul>
<p>url links</p>
<ul>
<li><code>[link](https://www.python.org/)</code><a href="https://www.python.org/">link</a></li>
</ul>
<p>Images</p>
<ul>
<li><code>![](https://www.spaam-community.org/assets/media/SPAAM-Logo-Full-Colour_ShortName.svg)</code> : <img src="https://www.spaam-community.org/assets/media/SPAAM-Logo-Full-Colour_ShortName.svg" class="img-fluid"></li>
</ul>
<p><br>
The <strong>code cells</strong> can interpret many different coding languages including Python and Bash. The syntax of the code cells is the same as the syntax of the coding languages, in our case python.</p>
<p>Below are some examples of Python <strong>code cells</strong> with some useful basic python functions:</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Python function
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>print()</code> is a python fuction for printing lines in the terminal <code>print() == echo</code> in bash</p>
</div>
</div>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Hello World from Python!"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>out -</em> <code>Hello World from Python!</code></p>
<p>But it can also, for example, run bash commands by adding a <em>!</em> at the start of the line.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> echo <span class="st">"Hello World from bash!"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>out -</em> <code>Hello World from bash!</code></p>
<p>Stings or numbers can be stored as a variable by using the <em>=</em> sign</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>i <span class="op">=</span> <span class="dv">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Ones a variable is set in one <strong>code cell</strong> they are stored and can be accessed in other downstream <strong>code celles</strong>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(i)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You can also print multiple things together in one <code>print</code> statement such as a number and a string:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"The number is"</span>, i, <span class="st">"Wow!"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>out -</em> <code>The number is 0 Wow!</code></p>
</section>
<section id="pandas" class="level2" data-number="8.4">
<h2 data-number="8.4" class="anchored" data-anchor-id="pandas"><span class="header-section-number">8.4</span> Pandas</h2>
<section id="getting-started" class="level3" data-number="8.4.1">
<h3 data-number="8.4.1" class="anchored" data-anchor-id="getting-started"><span class="header-section-number">8.4.1</span> Getting started</h3>
<p>Pandas is a Python library used for data manipulation and analysis.</p>
<p>We can import the library like this:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>We set “pandas” to the alias “pd” because we are lazy and don’t want to write the full word too many times.</p>
</div>
</div>
<p>Now, we can print the current version:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pd.__version__</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><em>out -</em> <code>'2.0.1'</code></p>
</section>
<section id="pandas-data-structures" class="level3" data-number="8.4.2">
<h3 data-number="8.4.2" class="anchored" data-anchor-id="pandas-data-structures"><span class="header-section-number">8.4.2</span> Pandas data structures</h3>
<p>The primary data structures in Pandas are <code>Series</code> and <code>DataFrame</code>.</p>
<p>A <code>DataFrame</code> is a table with <strong>columns</strong> and <strong>rows</strong>.</p>
<p>Each <strong>column</strong> has a <em>column name</em> and each <strong>row</strong> has an <em>index</em>.</p>
<p><img src="assets/images/chapters/python-pandas/01_table_dataframe.svg" class="img-fluid"></p>
<p>A single row or column (1 dimensional data) is a <code>Series</code>.</p>
<p><img src="assets/images/chapters/python-pandas/01_table_series.svg" class="img-fluid"></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>For a more in detail pandas getting started tutorial <a href="https://pandas.pydata.org/docs/getting_started/index.html#">click here</a></p>
</div>
</div>
</section>
<section id="reading-in-data" class="level3" data-number="8.4.3">
<h3 data-number="8.4.3" class="anchored" data-anchor-id="reading-in-data"><span class="header-section-number">8.4.3</span> Reading in data</h3>
<p>Pandas can read in <strong>csv</strong> (comma separated values) files, which are tables in text format.</p>
<p>It’s called <strong>c</strong>sv becase each value is separated from the others via a comma, like this:</p>
<pre class="{text}"><code>A,B
5,6
8,4</code></pre>
<p>another common tabluar seperater are <strong>tsv</strong>, where each value is seperated by a <strong>tab</strong> <code>\t</code></p>
<pre class="{text}"><code>A\tB
5\t6
8\t4</code></pre>
<p><br>
<br>
<br>
<br>
<br>
<br>
</p>
</section>
</section>
<section id="reading-data-with-pandas" class="level2" data-number="8.5">
<h2 data-number="8.5" class="anchored" data-anchor-id="reading-data-with-pandas"><span class="header-section-number">8.5</span> 3 - Reading data with Pandas</h2>
<div class="sourceCode" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sample_table_url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/SPAAM-community/AncientMetagenomeDir/b187df6ebd23dfeb42935fd5020cb615ead3f164/</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="er">ancientmetagenome-hostassociated/samples/ancientmetagenome-hostassociated_samples.tsv"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>library_table_url <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/SPAAM-community/AncientMetagenomeDir/b187df6ebd23dfeb42935fd5020cb615ead3f164/</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="er">ancientmetagenome-hostassociated/libraries/ancientmetagenome-hostassociated_libraries.tsv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Getting help in Python</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="bu">help</span>(pd.read_csv)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]',
sep=&lt;no_default&gt;, delimiter=None, header='infer', names=&lt;no_default&gt;, index_col=None,
usecols=None, squeeze=None, prefix=&lt;no_default&gt;, mangle_dupe_cols=True,
dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters=None,
true_values=None, false_values=None, skipinitialspace=False, skiprows=None, skipfooter=0,
nrows=None, na_values=None, keep_default_na=True, na_filter=True, verbose=False,
skip_blank_lines=True, parse_dates=None, infer_datetime_format=False, keep_date_col=False,
 date_parser=None, dayfirst=False, cache_dates=True, iterator=False, chunksize=None,
 compression: 'CompressionOptions' = 'infer', thousands=None, decimal: 'str' = '.',
 lineterminator=None, quotechar='"', quoting=0, doublequote=True, escapechar=None,
 comment=None, encoding=None, encoding_errors: 'str | None' = 'strict', dialect=None,
 error_bad_lines=None, warn_bad_lines=None, on_bad_lines=None, delim_whitespace=False,
 low_memory=True, memory_map=False, float_precision=None, storage_options: 'StorageOptions' = None)
    Read a comma-separated values (csv) file into DataFrame.

    Also supports optionally iterating or breaking of the file
    into chunks.

    Additional help can be found in the online docs for
    `IO Tools &lt;https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html&gt;`_.

    Parameters
    ----------
    filepath_or_buffer : str, path object or file-like object
        Any valid string path is acceptable. The string could be a URL. Valid
        URL schemes include http, ftp, s3, gs, and file. For file URLs, a host is
        expected. A local file could be: file://localhost/path/to/table.csv.

        If you want to pass in a path object, pandas accepts any ``os.PathLike``.

        By file-like object, we refer to objects with a ``read()`` method, such as
        a file handle (e.g. via builtin ``open`` function) or ``StringIO``.
    sep : str, default ','
        Delimiter to use. If sep is None, the C engine cannot automatically detect
        the separator, but the Python parsing engine can, meaning the latter will
        be used and automatically detect the separator by Python's builtin sniffer
        tool, ``csv.Sniffer``. In addition, separators longer than 1 character and
        different from ``'\s+'`` will be interpreted as regular expressions and
        will also force the use of the Python parsing engine. Note that regex
        delimiters are prone to ignoring quoted data. Regex example: ``'\r\t'``.
    delimiter : str, default ``None``
        Alias for sep.
    header : int, list of int, None, default 'infer'
        Row number(s) to use as the column names, and the start of the
        data.  Default behavior is to infer the column names: if no names
        are passed the behavior is identical to ``header=0`` and column
        names are inferred from the first line of the file, if column
        names are passed explicitly then the behavior is identical to
        ``header=None``. Explicitly pass ``header=0`` to be able to
        replace existing names. The header can be a list of integers that
        specify row locations for a multi-index on the columns
        e.g. [0,1,3]. Intervening rows that are not specified will be
        skipped (e.g. 2 in this example is skipped). Note that this
        parameter ignores commented lines and empty lines if
        ``skip_blank_lines=True``, so ``header=0`` denotes the first line of
        data rather than the first line of the file.
    names : array-like, optional
        List of column names to use. If the file contains a header row,
        then you should explicitly pass ``header=0`` to override the column names.
        Duplicates in this list are not allowed.
    index_col : int, str, sequence of int / str, or False, optional, default ``None``
      Column(s) to use as the row labels of the ``DataFrame``, either given as
      string name or column index. If a sequence of int / str is given, a
      MultiIndex is used.

      Note: ``index_col=False`` can be used to force pandas to *not* use the first
      column as the index, e.g. when you have a malformed file with delimiters at
      the end of each line.
    usecols : list-like or callable, optional
        Return a subset of the columns. If list-like, all elements must either
        be positional (i.e. integer indices into the document columns) or strings
        that correspond to column names provided either by the user in `names` or
        inferred from the document header row(s). If ``names`` are given, the document
        header row(s) are not taken into account. For example, a valid list-like
        `usecols` parameter would be ``[0, 1, 2]`` or ``['foo', 'bar', 'baz']``.
        Element order is ignored, so ``usecols=[0, 1]`` is the same as ``[1, 0]``.
        To instantiate a DataFrame from ``data`` with element order preserved use
        ``pd.read_csv(data, usecols=['foo', 'bar'])[['foo', 'bar']]`` for columns
        in ``['foo', 'bar']`` order or
        ``pd.read_csv(data, usecols=['foo', 'bar'])[['bar', 'foo']]``
        for ``['bar', 'foo']`` order.

        If callable, the callable function will be evaluated against the column
        names, returning names where the callable function evaluates to True. An
        example of a valid callable argument would be ``lambda x: x.upper() in
        ['AAA', 'BBB', 'DDD']``. Using this parameter results in much faster
        parsing time and lower memory usage.
    squeeze : bool, default False
        If the parsed data only contains one column then return a Series.

        .. deprecated:: 1.4.0
            Append ``.squeeze("columns")`` to the call to ``read_csv`` to squeeze
            the data.
    prefix : str, optional
        Prefix to add to column numbers when no header, e.g. 'X' for X0, X1, ...

        .. deprecated:: 1.4.0
           Use a list comprehension on the DataFrame's columns after calling ``read_csv``.
    mangle_dupe_cols : bool, default True
        Duplicate columns will be specified as 'X', 'X.1', ...'X.N', rather than
        'X'...'X'. Passing in False will cause data to be overwritten if there
        are duplicate names in the columns.
    dtype : Type name or dict of column -&gt; type, optional
        Data type for data or columns. E.g. {'a': np.float64, 'b': np.int32,
        'c': 'Int64'}
        Use `str` or `object` together with suitable `na_values` settings
        to preserve and not interpret dtype.
        If converters are specified, they will be applied INSTEAD
        of dtype conversion.
    engine : {'c', 'python', 'pyarrow'}, optional
        Parser engine to use. The C and pyarrow engines are faster, while the python engine
        is currently more feature-complete. Multithreading is currently only supported by
        the pyarrow engine.

        .. versionadded:: 1.4.0

            The "pyarrow" engine was added as an *experimental* engine, and some features
            are unsupported, or may not work correctly, with this engine.
    converters : dict, optional
        Dict of functions for converting values in certain columns. Keys can either
        be integers or column labels.
    true_values : list, optional
        Values to consider as True.
    false_values : list, optional
        Values to consider as False.
    skipinitialspace : bool, default False
        Skip spaces after delimiter.
    skiprows : list-like, int or callable, optional
        Line numbers to skip (0-indexed) or number of lines to skip (int)
        at the start of the file.

        If callable, the callable function will be evaluated against the row
        indices, returning True if the row should be skipped and False otherwise.
        An example of a valid callable argument would be ``lambda x: x in [0, 2]``.
    skipfooter : int, default 0
        Number of lines at bottom of file to skip (Unsupported with engine='c').
    nrows : int, optional
        Number of rows of file to read. Useful for reading pieces of large files.
    na_values : scalar, str, list-like, or dict, optional
        Additional strings to recognize as NA/NaN. If dict passed, specific
        per-column NA values.  By default the following values are interpreted as
        NaN: '', '#N/A', '#N/A N/A', '#NA', '-1.#IND', '-1.#QNAN', '-NaN', '-nan',
        '1.#IND', '1.#QNAN', '&lt;NA&gt;', 'N/A', 'NA', 'NULL', 'NaN', 'n/a',
        'nan', 'null'.
    keep_default_na : bool, default True
        Whether or not to include the default NaN values when parsing the data.
        Depending on whether `na_values` is passed in, the behavior is as follows:

        * If `keep_default_na` is True, and `na_values` are specified, `na_values`
          is appended to the default NaN values used for parsing.
        * If `keep_default_na` is True, and `na_values` are not specified, only
          the default NaN values are used for parsing.
        * If `keep_default_na` is False, and `na_values` are specified, only
          the NaN values specified `na_values` are used for parsing.
        * If `keep_default_na` is False, and `na_values` are not specified, no
          strings will be parsed as NaN.

        Note that if `na_filter` is passed in as False, the `keep_default_na` and
        `na_values` parameters will be ignored.
    na_filter : bool, default True
        Detect missing value markers (empty strings and the value of na_values). In
        data without any NAs, passing na_filter=False can improve the performance
        of reading a large file.
    verbose : bool, default False
        Indicate number of NA values placed in non-numeric columns.
    skip_blank_lines : bool, default True
        If True, skip over blank lines rather than interpreting as NaN values.
    parse_dates : bool or list of int or names or list of lists or dict, default False
        The behavior is as follows:

        * boolean. If True -&gt; try parsing the index.
        * list of int or names. e.g. If [1, 2, 3] -&gt; try parsing columns 1, 2, 3
          each as a separate date column.
        * list of lists. e.g.  If [[1, 3]] -&gt; combine columns 1 and 3 and parse as
          a single date column.
        * dict, e.g. {'foo' : [1, 3]} -&gt; parse columns 1, 3 as date and call
          result 'foo'

        If a column or index cannot be represented as an array of datetimes,
        say because of an unparsable value or a mixture of timezones, the column
        or index will be returned unaltered as an object data type. For
        non-standard datetime parsing, use ``pd.to_datetime`` after
        ``pd.read_csv``. To parse an index or column with a mixture of timezones,
        specify ``date_parser`` to be a partially-applied
        :func:`pandas.to_datetime` with ``utc=True``. See
        :ref:`io.csv.mixed_timezones` for more.

        Note: A fast-path exists for iso8601-formatted dates.
    infer_datetime_format : bool, default False
        If True and `parse_dates` is enabled, pandas will attempt to infer the
        format of the datetime strings in the columns, and if it can be inferred,
        switch to a faster method of parsing them. In some cases this can increase
        the parsing speed by 5-10x.
    keep_date_col : bool, default False
        If True and `parse_dates` specifies combining multiple columns then
        keep the original columns.
    date_parser : function, optional
        Function to use for converting a sequence of string columns to an array of
        datetime instances. The default uses ``dateutil.parser.parser`` to do the
        conversion. Pandas will try to call `date_parser` in three different ways,
        advancing to the next if an exception occurs: 1) Pass one or more arrays
        (as defined by `parse_dates`) as arguments; 2) concatenate (row-wise) the
        string values from the columns defined by `parse_dates` into a single array
        and pass that; and 3) call `date_parser` once for each row using one or
        more strings (corresponding to the columns defined by `parse_dates`) as
        arguments.
    dayfirst : bool, default False
        DD/MM format dates, international and European format.
    cache_dates : bool, default True
        If True, use a cache of unique, converted dates to apply the datetime
        conversion. May produce significant speed-up when parsing duplicate
        date strings, especially ones with timezone offsets.

        .. versionadded:: 0.25.0
    iterator : bool, default False
        Return TextFileReader object for iteration or getting chunks with
        ``get_chunk()``.

        .. versionchanged:: 1.2

           ``TextFileReader`` is a context manager.
    chunksize : int, optional
        Return TextFileReader object for iteration.
        See the `IO Tools docs
        &lt;https://pandas.pydata.org/pandas-docs/stable/io.html#io-chunking&gt;`_
        for more information on ``iterator`` and ``chunksize``.

        .. versionchanged:: 1.2

           ``TextFileReader`` is a context manager.
    compression : str or dict, default 'infer'
        For on-the-fly decompression of on-disk data. If 'infer' and '%s' is
        path-like, then detect compression from the following extensions: '.gz',
        '.bz2', '.zip', '.xz', or '.zst' (otherwise no compression). If using
        'zip', the ZIP file must contain only one data file to be read in. Set to
        ``None`` for no decompression. Can also be a dict with key ``'method'`` set
        to one of {``'zip'``, ``'gzip'``, ``'bz2'``, ``'zstd'``} and other
        key-value pairs are forwarded to ``zipfile.ZipFile``, ``gzip.GzipFile``,
        ``bz2.BZ2File``, or ``zstandard.ZstdDecompressor``, respectively. As an
        example, the following could be passed for Zstandard decompression using a
        custom compression dictionary:
        ``compression={'method': 'zstd', 'dict_data': my_compression_dict}``.

        .. versionchanged:: 1.4.0 Zstandard support.

    thousands : str, optional
        Thousands separator.
    decimal : str, default '.'
        Character to recognize as decimal point (e.g. use ',' for European data).
    lineterminator : str (length 1), optional
        Character to break file into lines. Only valid with C parser.
    quotechar : str (length 1), optional
        The character used to denote the start and end of a quoted item. Quoted
        items can include the delimiter and it will be ignored.
    quoting : int or csv.QUOTE_* instance, default 0
        Control field quoting behavior per ``csv.QUOTE_*`` constants. Use one of
        QUOTE_MINIMAL (0), QUOTE_ALL (1), QUOTE_NONNUMERIC (2) or QUOTE_NONE (3).
    doublequote : bool, default ``True``
       When quotechar is specified and quoting is not ``QUOTE_NONE``, indicate
       whether or not to interpret two consecutive quotechar elements INSIDE a
       field as a single ``quotechar`` element.
    escapechar : str (length 1), optional
        One-character string used to escape other characters.
    comment : str, optional
        Indicates remainder of line should not be parsed. If found at the beginning
        of a line, the line will be ignored altogether. This parameter must be a
        single character. Like empty lines (as long as ``skip_blank_lines=True``),
        fully commented lines are ignored by the parameter `header` but not by
        `skiprows`. For example, if ``comment='#'``, parsing
        ``#empty\na,b,c\n1,2,3`` with ``header=0`` will result in 'a,b,c' being
        treated as the header.
    encoding : str, optional
        Encoding to use for UTF when reading/writing (ex. 'utf-8'). `List of Python
        standard encodings
        &lt;https://docs.python.org/3/library/codecs.html#standard-encodings&gt;`_ .

        .. versionchanged:: 1.2

           When ``encoding`` is ``None``, ``errors="replace"`` is passed to
           ``open()``. Otherwise, ``errors="strict"`` is passed to ``open()``.
           This behavior was previously only the case for ``engine="python"``.

        .. versionchanged:: 1.3.0

           ``encoding_errors`` is a new argument. ``encoding`` has no longer an
           influence on how encoding errors are handled.

    encoding_errors : str, optional, default "strict"
        How encoding errors are treated. `List of possible values
        &lt;https://docs.python.org/3/library/codecs.html#error-handlers&gt;`_ .

        .. versionadded:: 1.3.0

    dialect : str or csv.Dialect, optional
        If provided, this parameter will override values (default or not) for the
        following parameters: `delimiter`, `doublequote`, `escapechar`,
        `skipinitialspace`, `quotechar`, and `quoting`. If it is necessary to
        override values, a ParserWarning will be issued. See csv.Dialect
        documentation for more details.
    error_bad_lines : bool, optional, default ``None``
        Lines with too many fields (e.g. a csv line with too many commas) will by
        default cause an exception to be raised, and no DataFrame will be returned.
        If False, then these "bad lines" will be dropped from the DataFrame that is
        returned.

        .. deprecated:: 1.3.0
           The ``on_bad_lines`` parameter should be used instead to specify behavior upon
           encountering a bad line instead.
    warn_bad_lines : bool, optional, default ``None``
        If error_bad_lines is False, and warn_bad_lines is True, a warning for each
        "bad line" will be output.

        .. deprecated:: 1.3.0
           The ``on_bad_lines`` parameter should be used instead to specify behavior upon
           encountering a bad line instead.
    on_bad_lines : {'error', 'warn', 'skip'} or callable, default 'error'
        Specifies what to do upon encountering a bad line (a line with too many fields).
        Allowed values are :

            - 'error', raise an Exception when a bad line is encountered.
            - 'warn', raise a warning when a bad line is encountered and skip that line.
            - 'skip', skip bad lines without raising or warning when they are encountered.

        .. versionadded:: 1.3.0

            - callable, function with signature
              ``(bad_line: list[str]) -&gt; list[str] | None`` that will process a single
              bad line. ``bad_line`` is a list of strings split by the ``sep``.
              If the function returns ``None``, the bad line will be ignored.
              If the function returns a new list of strings with more elements than
              expected, a ``ParserWarning`` will be emitted while dropping extra elements.
              Only supported when ``engine="python"``

        .. versionadded:: 1.4.0

    delim_whitespace : bool, default False
        Specifies whether or not whitespace (e.g. ``' '`` or ``'    '``) will be
        used as the sep. Equivalent to setting ``sep='\s+'``. If this option
        is set to True, nothing should be passed in for the ``delimiter``
        parameter.
    low_memory : bool, default True
        Internally process the file in chunks, resulting in lower memory use
        while parsing, but possibly mixed type inference.  To ensure no mixed
        types either set False, or specify the type with the `dtype` parameter.
        Note that the entire file is read into a single DataFrame regardless,
        use the `chunksize` or `iterator` parameter to return the data in chunks.
        (Only valid with C parser).
    memory_map : bool, default False
        If a filepath is provided for `filepath_or_buffer`, map the file object
        directly onto memory and access the data directly from there. Using this
        option can improve performance because there is no longer any I/O overhead.
    float_precision : str, optional
        Specifies which converter the C engine should use for floating-point
        values. The options are ``None`` or 'high' for the ordinary converter,
        'legacy' for the original lower precision pandas converter, and
        'round_trip' for the round-trip converter.

        .. versionchanged:: 1.2

    storage_options : dict, optional
        Extra options that make sense for a particular storage connection, e.g.
        host, port, username, password, etc. For HTTP(S) URLs the key-value pairs
        are forwarded to ``urllib`` as header options. For other URLs (e.g.
        starting with "s3://", and "gcs://") the key-value pairs are forwarded to
        ``fsspec``. Please see ``fsspec`` and ``urllib`` for more details.

        .. versionadded:: 1.2

    Returns
    -------
    DataFrame or TextParser
        A comma-separated values (csv) file is returned as two-dimensional
        data structure with labeled axes.

    See Also
    --------
    DataFrame.to_csv : Write DataFrame to a comma-separated values (csv) file.
    read_csv : Read a comma-separated values (csv) file into DataFrame.
    read_fwf : Read a table of fixed-width formatted lines into DataFrame.

    Examples
    --------
    &gt;&gt;&gt; pd.read_csv('data.csv')  # doctest: +SKIP</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sample_df <span class="op">=</span> pd.read_csv(sample_table_url, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>library_df <span class="op">=</span> pd.read_csv(library_table_url, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sample_df.project_name.nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>45</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>library_df.project_name.nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>43</code></pre>
<section id="listing-the-columns-of-the-sample-dataframe" class="level3" data-number="8.5.1">
<h3 data-number="8.5.1" class="anchored" data-anchor-id="listing-the-columns-of-the-sample-dataframe"><span class="header-section-number">8.5.1</span> Listing the columns of the sample dataframe</h3>
<div class="sourceCode" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sample_df.columns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Index(['project_name', 'publication_year', 'publication_doi', 'site_name',
       'latitude', 'longitude', 'geo_loc_name', 'sample_name', 'sample_host',
       'sample_age', 'sample_age_doi', 'community_type', 'material', 'archive',
       'archive_project', 'archive_accession'],
      dtype='object')</code></pre>
</section>
<section id="looking-at-the-data-type-of-the-sample-dataframe" class="level3" data-number="8.5.2">
<h3 data-number="8.5.2" class="anchored" data-anchor-id="looking-at-the-data-type-of-the-sample-dataframe"><span class="header-section-number">8.5.2</span> Looking at the data type of the sample dataframe</h3>
<div class="sourceCode" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sample_df.dtypes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>project_name          object
publication_year       int64
publication_doi       object
site_name             object
latitude             float64
longitude            float64
geo_loc_name          object
sample_name           object
sample_host           object
sample_age             int64
sample_age_doi        object
community_type        object
material              object
archive               object
archive_project       object
archive_accession     object
dtype: object</code></pre>
<ul>
<li><code>int64</code> is for integers</li>
<li><code>floating64</code> is for floating point precision numbers, also known as double in some other programing languages</li>
<li><code>object</code> is a general type in pandas for everything that is not a number, interval, categorical, or date</li>
</ul>
</section>
<section id="lets-inspect-our-data" class="level3" data-number="8.5.3">
<h3 data-number="8.5.3" class="anchored" data-anchor-id="lets-inspect-our-data"><span class="header-section-number">8.5.3</span> Let’s inspect our data</h3>
<p>What is the size of our dataframe ?</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sample_df.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>(1060, 16)</code></pre>
<p>This dataframe has <strong>1060</strong> rows, and <strong>16</strong> columns</p>
<p>Let’s look at the first 5 rows</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sample_df.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
2
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Gola Forest
</td>
<td>
7.657
</td>
<td>
-10.841
</td>
<td>
Sierra Leone
</td>
<td>
Chimp
</td>
<td>
Pan troglodytes
</td>
<td>
100
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890499
</td>
</tr>
<tr>
<th>
3
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.328
</td>
<td>
Spain
</td>
<td>
ElSidron1
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890498
</td>
</tr>
<tr>
<th>
4
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.329
</td>
<td>
Spain
</td>
<td>
ElSidron2
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890496
</td>
</tr>
</tbody>

</table>
</div>
<p><strong>Unlike R, Python is <code>0</code> based language, meaning the first element is of index <code>0</code>, not like R where it is <code>1</code>.</strong></p>
<p>Let’s look at the last 5 rows</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sample_df.tail()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
1055
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T2
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283094,ERS7283095
</td>
</tr>
<tr>
<th>
1056
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283096,ERS7283097
</td>
</tr>
<tr>
<th>
1057
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283098,ERS7283099
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
Dom Square, Riga
</td>
<td>
56.949
</td>
<td>
24.104
</td>
<td>
Latvia
</td>
<td>
TZA3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283100,ERS7283101
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Peter’s Church, Riga
</td>
<td>
56.947
</td>
<td>
24.109
</td>
<td>
Latvia
</td>
<td>
TZA4
</td>
<td>
Homo sapiens
</td>
<td>
500
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283102,ERS7283103
</td>
</tr>
</tbody>

</table>
</div>
<p>Let’s randomly inspect 5 rows</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>sample_df.sample(n<span class="op">=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
413
</th>
<td>
Neukamm2020
</td>
<td>
2020
</td>
<td>
10.1186/s12915-020-00839-8
</td>
<td>
Abusir el-Meleq
</td>
<td>
29.240
</td>
<td>
31.100
</td>
<td>
Egypt
</td>
<td>
Abusir1576
</td>
<td>
Homo sapiens
</td>
<td>
2200
</td>
<td>
10.1038/ncomms15694
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
ENA
</td>
<td>
PRJEB33848
</td>
<td>
ERS3635981
</td>
</tr>
<tr>
<th>
754
</th>
<td>
Rampelli2021
</td>
<td>
2021
</td>
<td>
10.1038/s42003-021-01689-y
</td>
<td>
El Salt
</td>
<td>
38.687
</td>
<td>
-0.508
</td>
<td>
Spain
</td>
<td>
V3
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
44700
</td>
<td>
10.1038/s42003-021-01689-y
</td>
<td>
gut
</td>
<td>
sediment
</td>
<td>
ENA
</td>
<td>
PRJEB41665
</td>
<td>
ERS5428042
</td>
</tr>
<tr>
<th>
436
</th>
<td>
Neukamm2020
</td>
<td>
2020
</td>
<td>
10.1186/s12915-020-00839-8
</td>
<td>
Abusir el-Meleq
</td>
<td>
29.240
</td>
<td>
31.100
</td>
<td>
Egypt
</td>
<td>
Abusir1606
</td>
<td>
Homo sapiens
</td>
<td>
2600
</td>
<td>
10.1186/s12915-020-00839-8
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
ENA
</td>
<td>
PRJEB33848
</td>
<td>
ERS3635928
</td>
</tr>
<tr>
<th>
474
</th>
<td>
Neukamm2020
</td>
<td>
2020
</td>
<td>
10.1186/s12915-020-00839-8
</td>
<td>
Abusir el-Meleq
</td>
<td>
29.240
</td>
<td>
31.100
</td>
<td>
Egypt
</td>
<td>
Abusir1654
</td>
<td>
Homo sapiens
</td>
<td>
2300
</td>
<td>
10.1038/ncomms15694
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB33848
</td>
<td>
ERS3635960
</td>
</tr>
<tr>
<th>
573
</th>
<td>
Philips2017
</td>
<td>
2017
</td>
<td>
10.1186/s12864-020-06810-9
</td>
<td>
Kowalewko
</td>
<td>
52.699
</td>
<td>
17.605
</td>
<td>
Poland
</td>
<td>
PCA0040
</td>
<td>
Homo sapiens
</td>
<td>
1900
</td>
<td>
10.1186/s12864-020-06810-9
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
SRA
</td>
<td>
PRJNA354503
</td>
<td>
SRS1815407
</td>
</tr>
</tbody>

</table>
</div>
<section id="accessing-the-data-by-indexcolumns" class="level6" data-number="8.5.3.0.0.1">
<h6 data-number="8.5.3.0.0.1" class="anchored" data-anchor-id="accessing-the-data-by-indexcolumns"><span class="header-section-number">8.5.3.0.0.1</span> Accessing the data by index/columns</h6>
<p>The are different way of selecting of subset of a dataframe</p>
<p>Selecting by the row index</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># selecting the 10th row, and all columns</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>sample_df.iloc[<span class="dv">9</span>, :]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>project_name                    Weyrich2017
publication_year                       2017
publication_doi         10.1038/nature21674
site_name            Stuttgart-Mühlhausen I
latitude                             48.839
longitude                             9.227
geo_loc_name                        Germany
sample_name                        EuroLBK1
sample_host                    Homo sapiens
sample_age                             7400
sample_age_doi          10.1038/nature21674
community_type                         oral
material                    dental calculus
archive                                 SRA
archive_project                 PRJNA685265
archive_accession                SRS7890488
Name: 9, dtype: object</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># selecting the 10th to 12th row, and all columns</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>sample_df.iloc[<span class="dv">9</span>:<span class="dv">12</span>, :]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
9
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
<td>
48.839
</td>
<td>
9.227
</td>
<td>
Germany
</td>
<td>
EuroLBK1
</td>
<td>
Homo sapiens
</td>
<td>
7400
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890488
</td>
</tr>
<tr>
<th>
10
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
<td>
48.839
</td>
<td>
9.227
</td>
<td>
Germany
</td>
<td>
EuroLBK2
</td>
<td>
Homo sapiens
</td>
<td>
7400
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890485
</td>
</tr>
<tr>
<th>
11
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
<td>
48.839
</td>
<td>
9.227
</td>
<td>
Germany
</td>
<td>
EuroLBK3
</td>
<td>
Homo sapiens
</td>
<td>
7400
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890490
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># selecting the 10th to 12th row, and the first to the 4th column</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>sample_df.iloc[<span class="dv">9</span>:<span class="dv">12</span>, <span class="dv">0</span>:<span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
9
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
</tr>
<tr>
<th>
10
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
</tr>
<tr>
<th>
11
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Stuttgart-Mühlhausen I
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># selecting the column site_name</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'site_name'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>0                           Dalheim
1                           Dalheim
2                       Gola Forest
3                    El Sidrón Cave
4                    El Sidrón Cave
                   ...
1055    St. Gertrude’s Church, Riga
1056    St. Gertrude’s Church, Riga
1057    St. Gertrude’s Church, Riga
1058               Dom Square, Riga
1059       St. Peter’s Church, Riga
Name: site_name, Length: 1060, dtype: object</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Also valid, but less preferred</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>sample_df.site_name</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>0                           Dalheim
1                           Dalheim
2                       Gola Forest
3                    El Sidrón Cave
4                    El Sidrón Cave
                   ...
1055    St. Gertrude’s Church, Riga
1056    St. Gertrude’s Church, Riga
1057    St. Gertrude’s Church, Riga
1058               Dom Square, Riga
1059       St. Peter’s Church, Riga
Name: site_name, Length: 1060, dtype: object</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing a row</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sample_df.drop(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
2
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Gola Forest
</td>
<td>
7.657
</td>
<td>
-10.841
</td>
<td>
Sierra Leone
</td>
<td>
Chimp
</td>
<td>
Pan troglodytes
</td>
<td>
100
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890499
</td>
</tr>
<tr>
<th>
3
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.328
</td>
<td>
Spain
</td>
<td>
ElSidron1
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890498
</td>
</tr>
<tr>
<th>
4
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.329
</td>
<td>
Spain
</td>
<td>
ElSidron2
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890496
</td>
</tr>
<tr>
<th>
5
</th>
<td>
Weyrich2017
</td>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Spy Cave
</td>
<td>
50.480
</td>
<td>
4.674
</td>
<td>
Belgium
</td>
<td>
Spy1
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
35800
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890491
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
1055
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T2
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283094,ERS7283095
</td>
</tr>
<tr>
<th>
1056
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283096,ERS7283097
</td>
</tr>
<tr>
<th>
1057
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283098,ERS7283099
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
Dom Square, Riga
</td>
<td>
56.949
</td>
<td>
24.104
</td>
<td>
Latvia
</td>
<td>
TZA3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283100,ERS7283101
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Peter’s Church, Riga
</td>
<td>
56.947
</td>
<td>
24.109
</td>
<td>
Latvia
</td>
<td>
TZA4
</td>
<td>
Homo sapiens
</td>
<td>
500
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283102,ERS7283103
</td>
</tr>
</tbody>

</table>
<p>
1059 rows × 16 columns
</p>
</div>
<div class="sourceCode" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Removing a columm</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>sample_df.drop(<span class="st">'project_name'</span>, axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
2
</th>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
Gola Forest
</td>
<td>
7.657
</td>
<td>
-10.841
</td>
<td>
Sierra Leone
</td>
<td>
Chimp
</td>
<td>
Pan troglodytes
</td>
<td>
100
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890499
</td>
</tr>
<tr>
<th>
3
</th>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.328
</td>
<td>
Spain
</td>
<td>
ElSidron1
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890498
</td>
</tr>
<tr>
<th>
4
</th>
<td>
2017
</td>
<td>
10.1038/nature21674
</td>
<td>
El Sidrón Cave
</td>
<td>
43.386
</td>
<td>
-5.329
</td>
<td>
Spain
</td>
<td>
ElSidron2
</td>
<td>
Homo sapiens neanderthalensis
</td>
<td>
49000
</td>
<td>
10.1038/nature21674
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA685265
</td>
<td>
SRS7890496
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
1055
</th>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T2
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283094,ERS7283095
</td>
</tr>
<tr>
<th>
1056
</th>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283096,ERS7283097
</td>
</tr>
<tr>
<th>
1057
</th>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283098,ERS7283099
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
Dom Square, Riga
</td>
<td>
56.949
</td>
<td>
24.104
</td>
<td>
Latvia
</td>
<td>
TZA3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283100,ERS7283101
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Peter’s Church, Riga
</td>
<td>
56.947
</td>
<td>
24.109
</td>
<td>
Latvia
</td>
<td>
TZA4
</td>
<td>
Homo sapiens
</td>
<td>
500
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283102,ERS7283103
</td>
</tr>
</tbody>

</table>
<p>
1060 rows × 15 columns
</p>
</div>
</section>
<section id="dealing-with-missing-data" class="level5" data-number="8.5.3.0.1">
<h5 data-number="8.5.3.0.1" class="anchored" data-anchor-id="dealing-with-missing-data"><span class="header-section-number">8.5.3.0.1</span> 4 - Dealing with missing data</h5>
<p>Checking is some entries if the table have missing data (NA or NaN)</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sample_df.isna()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
1
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
2
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
3
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
4
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
1055
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
1056
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
1057
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
<td>
False
</td>
</tr>
</tbody>

</table>
<p>
1060 rows × 16 columns
</p>
</div>
<div class="sourceCode" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># making the sum by row - axis=1</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>sample_df.isna().<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>0       0
1       0
2       0
3       0
4       0
       ..
1055    0
1056    0
1057    0
1058    0
1059    0
Length: 1060, dtype: int64</code></pre>
<p>Sorting by decreasing order to check which rows have missing values</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>sample_df.isna().<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>).sort_values(ascending<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>800     2
962     2
992     2
801     2
802     2
       ..
362     0
363     0
364     0
365     0
1059    0
Length: 1060, dtype: int64</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sample_df.iloc[<span class="dv">800</span>,:]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>project_name                         FellowsYates2021
publication_year                                 2021
publication_doi               10.1073/pnas.2021655118
site_name                               Not specified
latitude                                          NaN
longitude                                         NaN
geo_loc_name         Democratic Republic of the Congo
sample_name                                  GDC002.A
sample_host                   Gorilla gorilla gorilla
sample_age                                        200
sample_age_doi                10.1073/pnas.2021655118
community_type                                   oral
material                              dental calculus
archive                                           ENA
archive_project                            PRJEB34569
archive_accession                          ERS3774403
Name: 800, dtype: object</code></pre>
<p>What to do now ? The ideal scenario would be to correct or impute the data.<br>
However, sometimes, the only thing we can do is remove the row with missing data, with the <code>.dropna() function</code>.<br>
Here, we’re just going to ignore them, and deal with it individually if necessary</p>
</section>
</section>
</section>
<section id="computing-basic-statistics" class="level2" data-number="8.6">
<h2 data-number="8.6" class="anchored" data-anchor-id="computing-basic-statistics"><span class="header-section-number">8.6</span> 5 - Computing basic statistics</h2>
<p>TLDR: use the <code>describe()</code> function, the equivalent of <code>summarize</code> in R</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>sample_df.describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
publication_year
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
sample_age
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
count
</th>
<td>
1060.000000
</td>
<td>
1021.000000
</td>
<td>
1021.000000
</td>
<td>
1060.000000
</td>
</tr>
<tr>
<th>
mean
</th>
<td>
2019.377358
</td>
<td>
40.600493
</td>
<td>
3.749624
</td>
<td>
3588.443396
</td>
</tr>
<tr>
<th>
std
</th>
<td>
1.633877
</td>
<td>
18.469421
</td>
<td>
43.790316
</td>
<td>
9862.416855
</td>
</tr>
<tr>
<th>
min
</th>
<td>
2014.000000
</td>
<td>
-34.030000
</td>
<td>
-121.800000
</td>
<td>
100.000000
</td>
</tr>
<tr>
<th>
25%
</th>
<td>
2018.000000
</td>
<td>
29.240000
</td>
<td>
-1.257000
</td>
<td>
200.000000
</td>
</tr>
<tr>
<th>
50%
</th>
<td>
2020.000000
</td>
<td>
45.450000
</td>
<td>
14.381000
</td>
<td>
1000.000000
</td>
</tr>
<tr>
<th>
75%
</th>
<td>
2021.000000
</td>
<td>
52.699000
</td>
<td>
23.892000
</td>
<td>
2200.000000
</td>
</tr>
<tr>
<th>
max
</th>
<td>
2021.000000
</td>
<td>
79.000000
</td>
<td>
159.346000
</td>
<td>
102000.000000
</td>
</tr>
</tbody>

</table>
</div>
<p>Let’s look at various individual summary statistics We can run them on the whole dataframe (for <code>int</code> or <code>float</code> columns), or on a subset of columns</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>sample_df.mean()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/var/folders/1c/l1qb09f15jddsh65f6xv1n_r0000gp/T/ipykernel_69168/2260452167.py:1:
FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None')
is deprecated; in a future version this will raise TypeError.  Select only valid columns
before calling the reduction.





publication_year    2019.377358
latitude              40.600493
longitude              3.749624
sample_age          3588.443396
dtype: float64</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].describe()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>count    1060.000000
mean     2019.377358
std         1.633877
min      2014.000000
25%      2018.000000
50%      2020.000000
75%      2021.000000
max      2021.000000
Name: publication_year, dtype: float64</code></pre>
<div class="sourceCode" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The average publication year</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].mean()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>2019.377358490566</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The median publication year</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].median()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>2020.0</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The minimum, or oldest publication year</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].<span class="bu">min</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>2014</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The maximum, or most recent publication year</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].<span class="bu">max</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>2021</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The number of sites</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'site_name'</span>].nunique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>246</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The number of samples from the different hosts</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'sample_host'</span>].value_counts()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Homo sapiens                      741
Ursus arctos                       85
Ambrosia artemisiifolia            46
Arabidopsis thaliana               34
Homo sapiens neanderthalensis      32
Pan troglodytes schweinfurthii     26
Gorilla beringei beringei          15
Canis lupus                        12
Gorilla gorilla gorilla             8
Mammuthus primigenius               8
Pan troglodytes verus               7
Rangifer tarandus                   6
Gorilla beringei graueri            6
Pan troglodytes ellioti             6
Papio hamadryas                     5
Alouatta palliata                   5
Conepatus chinga                    4
Gerbilliscus boehmi                 4
Strigocuscus celebensis             4
Papio anubis                        2
Gorilla beringei                    2
Papio sp.                           1
Pan troglodytes                     1
Name: sample_host, dtype: int64</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The quantile of the publication years</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].quantile(np.arange(<span class="dv">0</span>,<span class="dv">1</span>,<span class="fl">0.1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>0.0    2014.0
0.1    2017.0
0.2    2018.0
0.3    2018.0
0.4    2020.0
0.5    2020.0
0.6    2020.0
0.7    2021.0
0.8    2021.0
0.9    2021.0
Name: publication_year, dtype: float64</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can also visualize it with built-in plot functions of pandas</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>sample_df[<span class="st">'publication_year'</span>].plot.hist()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>&lt;AxesSubplot:ylabel='Frequency'&gt;</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/python-pandas/tutorial_74_1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
</section>
<section id="filtering" class="level2" data-number="8.7">
<h2 data-number="8.7" class="anchored" data-anchor-id="filtering"><span class="header-section-number">8.7</span> 6 - Filtering</h2>
<p>There are different ways of filtering data with Pandas:</p>
<ul>
<li>The <strong>classic</strong> method with bracket indexing/subsetting</li>
<li>The <code>query()</code> method</li>
</ul>
<p>The classic method</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Getting all the publications before 2015</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>sample_df[sample_df[<span class="st">'publication_year'</span>]  <span class="op">&lt;</span> <span class="dv">2015</span> ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
272
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP4
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428959
</td>
</tr>
<tr>
<th>
273
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP10
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428961
</td>
</tr>
<tr>
<th>
274
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP18
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428962
</td>
</tr>
<tr>
<th>
275
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP37
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428963
</td>
</tr>
<tr>
<th>
276
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428960
</td>
</tr>
<tr>
<th>
277
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP48
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428964
</td>
</tr>
<tr>
<th>
278
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP02,TP10,TP15,TP26
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428958
</td>
</tr>
<tr>
<th>
279
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP32,TP42,TP45,TP48
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428972
</td>
</tr>
<tr>
<th>
500
</th>
<td>
Appelt2014
</td>
<td>
2014
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
Place d’Armes, Namur
</td>
<td>
50.460
</td>
<td>
4.86
</td>
<td>
Belgium
</td>
<td>
4.453
</td>
<td>
Homo sapiens
</td>
<td>
600
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
gut
</td>
<td>
palaeofaeces
</td>
<td>
SRA
</td>
<td>
PRJNA230469
</td>
<td>
SRS510175
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Getting all the publications before 2015, only in the Northern hemisphere</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>sample_df[(sample_df[<span class="st">'publication_year'</span>]  <span class="op">&lt;</span> <span class="dv">2015</span>) <span class="op">&amp;</span> (sample_df[<span class="st">'longitude'</span>] <span class="op">&gt;</span> <span class="dv">0</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
500
</th>
<td>
Appelt2014
</td>
<td>
2014
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
Place d’Armes, Namur
</td>
<td>
50.460
</td>
<td>
4.86
</td>
<td>
Belgium
</td>
<td>
4.453
</td>
<td>
Homo sapiens
</td>
<td>
600
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
gut
</td>
<td>
palaeofaeces
</td>
<td>
SRA
</td>
<td>
PRJNA230469
</td>
<td>
SRS510175
</td>
</tr>
</tbody>

</table>
</div>
<p>This syntax can rapidly become quite cumbersome, which is why we can also use the <code>query()</code> method</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Getting all the publications before 2015</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>sample_df.query(<span class="st">"publication_year &lt; 2015"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
272
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP4
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428959
</td>
</tr>
<tr>
<th>
273
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP10
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428961
</td>
</tr>
<tr>
<th>
274
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP18
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428962
</td>
</tr>
<tr>
<th>
275
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP37
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428963
</td>
</tr>
<tr>
<th>
276
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428960
</td>
</tr>
<tr>
<th>
277
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP48
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428964
</td>
</tr>
<tr>
<th>
278
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP02,TP10,TP15,TP26
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428958
</td>
</tr>
<tr>
<th>
279
</th>
<td>
Campana2014
</td>
<td>
2014
</td>
<td>
10.1186/1756-0500-7-111
</td>
<td>
Teposcolula Yucundaa
</td>
<td>
17.490
</td>
<td>
-97.46
</td>
<td>
Mexico
</td>
<td>
TP32,TP42,TP45,TP48
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.7183/1045-6635.23.4.467
</td>
<td>
skeletal tissue
</td>
<td>
bone
</td>
<td>
SRA
</td>
<td>
PRJNA205039
</td>
<td>
SRS428972
</td>
</tr>
<tr>
<th>
500
</th>
<td>
Appelt2014
</td>
<td>
2014
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
Place d’Armes, Namur
</td>
<td>
50.460
</td>
<td>
4.86
</td>
<td>
Belgium
</td>
<td>
4.453
</td>
<td>
Homo sapiens
</td>
<td>
600
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
gut
</td>
<td>
palaeofaeces
</td>
<td>
SRA
</td>
<td>
PRJNA230469
</td>
<td>
SRS510175
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Getting all the publications before 2015, only the Northern hemisphere</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>sample_df.query(<span class="st">"publication_year &lt; 2015 and longitude &gt; 0 "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742,SRS473743,SRS473744,SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747,SRS473746,SRS473748,SRS473749,SRS473750
</td>
</tr>
<tr>
<th>
500
</th>
<td>
Appelt2014
</td>
<td>
2014
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
Place d’Armes, Namur
</td>
<td>
50.460
</td>
<td>
4.86
</td>
<td>
Belgium
</td>
<td>
4.453
</td>
<td>
Homo sapiens
</td>
<td>
600
</td>
<td>
10.1128/AEM.03242-13
</td>
<td>
gut
</td>
<td>
palaeofaeces
</td>
<td>
SRA
</td>
<td>
PRJNA230469
</td>
<td>
SRS510175
</td>
</tr>
</tbody>

</table>
</div>
</section>
<section id="groupby-operations-and-computing-statistics-on-grouped-values" class="level2" data-number="8.8">
<h2 data-number="8.8" class="anchored" data-anchor-id="groupby-operations-and-computing-statistics-on-grouped-values"><span class="header-section-number">8.8</span> 7 - GroupBy operations, and computing statistics on grouped values</h2>
<p>The “groupBy” operation, as the name suggests, allows us to group values by a grouping key, and perform a groupwise operation.<br>
For example, we can group by the <code>sample_host</code> and get the age of the <strong>youngest</strong> sample in each group</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>sample_df.groupby(<span class="st">"sample_host"</span>)[<span class="st">'sample_age'</span>].<span class="bu">min</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>sample_host
Alouatta palliata                   200
Ambrosia artemisiifolia             100
Arabidopsis thaliana                100
Canis lupus                         400
Conepatus chinga                    100
Gerbilliscus boehmi                 100
Gorilla beringei                    100
Gorilla beringei beringei           200
Gorilla beringei graueri            200
Gorilla gorilla gorilla             200
Homo sapiens                        100
Homo sapiens neanderthalensis     35800
Mammuthus primigenius             41800
Pan troglodytes                     100
Pan troglodytes ellioti             200
Pan troglodytes schweinfurthii      100
Pan troglodytes verus               200
Papio anubis                        100
Papio hamadryas                     100
Papio sp.                           100
Rangifer tarandus                   100
Strigocuscus celebensis             100
Ursus arctos                        100
Name: sample_age, dtype: int64</code></pre>
<p>Here <code>min()</code> is a so-called aggregation function</p>
<p>Notice that <code>.value_counts()</code> is actually a special case of <code>.groupby()</code></p>
<div class="sourceCode" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>sample_df.groupby(<span class="st">"sample_host"</span>)[<span class="st">"sample_host"</span>].count()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>sample_host
Alouatta palliata                   5
Ambrosia artemisiifolia            46
Arabidopsis thaliana               34
Canis lupus                        12
Conepatus chinga                    4
Gerbilliscus boehmi                 4
Gorilla beringei                    2
Gorilla beringei beringei          15
Gorilla beringei graueri            6
Gorilla gorilla gorilla             8
Homo sapiens                      741
Homo sapiens neanderthalensis      32
Mammuthus primigenius               8
Pan troglodytes                     1
Pan troglodytes ellioti             6
Pan troglodytes schweinfurthii     26
Pan troglodytes verus               7
Papio anubis                        2
Papio hamadryas                     5
Papio sp.                           1
Rangifer tarandus                   6
Strigocuscus celebensis             4
Ursus arctos                       85
Name: sample_host, dtype: int64</code></pre>
</section>
<section id="reshaping-data-from-wide-to-long-and-back" class="level2" data-number="8.9">
<h2 data-number="8.9" class="anchored" data-anchor-id="reshaping-data-from-wide-to-long-and-back"><span class="header-section-number">8.9</span> 8 - Reshaping data, from wide to long and back</h2>
<p><img src="assets/images/chapters/python-pandas/The-wide-versus-tidy-data-format-In-the-wide-spreadsheet-like-data-format-each-column.png" class="img-fluid"></p>
<section id="from-wide-to-longtidy" class="level3" data-number="8.9.1">
<h3 data-number="8.9.1" class="anchored" data-anchor-id="from-wide-to-longtidy"><span class="header-section-number">8.9.1</span> From wide to long/tidy</h3>
<p>The tidy format, or long format idea is that one column = one kind of data.<br>
Unfortunately for this tutorial, the AncientMetagenomeDir tables are already in the tidy format (good), so we’ll see an example or the wide format just below</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>wide_df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">150</span>,<span class="dv">155</span>,<span class="dv">157</span>,<span class="dv">160</span>],</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">149</span>,<span class="dv">153</span>,<span class="dv">154</span>,<span class="dv">155</span>]</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>    , index <span class="op">=</span> [<span class="st">'John'</span>,<span class="st">'Jack'</span>]</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>    , columns <span class="op">=</span> [<span class="dv">1991</span>,<span class="dv">1992</span>,<span class="dv">1993</span>, <span class="dv">1994</span>]</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>).rename_axis(<span class="st">'individual'</span>).reset_index()</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>wide_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
individual
</th>
<th>
1991
</th>
<th>
1992
</th>
<th>
1993
</th>
<th>
1994
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
John
</td>
<td>
150
</td>
<td>
155
</td>
<td>
157
</td>
<td>
160
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Jack
</td>
<td>
149
</td>
<td>
153
</td>
<td>
154
</td>
<td>
155
</td>
</tr>
</tbody>

</table>
</div>
<p>In this hypothetic dataframe, we have the years as column, the individual as index, and their height as value.<br>
We’ll reformat to the tidy/long format using the <code>.melt()</code> function</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>tidy_df <span class="op">=</span> wide_df.melt(id_vars<span class="op">=</span><span class="st">'individual'</span>, var_name<span class="op">=</span><span class="st">'birthyear'</span>, value_name<span class="op">=</span><span class="st">'height'</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>tidy_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
individual
</th>
<th>
birthyear
</th>
<th>
height
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
John
</td>
<td>
1991
</td>
<td>
150
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Jack
</td>
<td>
1991
</td>
<td>
149
</td>
</tr>
<tr>
<th>
2
</th>
<td>
John
</td>
<td>
1992
</td>
<td>
155
</td>
</tr>
<tr>
<th>
3
</th>
<td>
Jack
</td>
<td>
1992
</td>
<td>
153
</td>
</tr>
<tr>
<th>
4
</th>
<td>
John
</td>
<td>
1993
</td>
<td>
157
</td>
</tr>
<tr>
<th>
5
</th>
<td>
Jack
</td>
<td>
1993
</td>
<td>
154
</td>
</tr>
<tr>
<th>
6
</th>
<td>
John
</td>
<td>
1994
</td>
<td>
160
</td>
</tr>
<tr>
<th>
7
</th>
<td>
Jack
</td>
<td>
1994
</td>
<td>
155
</td>
</tr>
</tbody>

</table>
</div>
<blockquote class="blockquote">
<p>Bonus: How to deal with a dataframe with the kind of data indicated in the column name, typically like so</p>
</blockquote>
<div class="sourceCode" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>wide_df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">150</span>,<span class="dv">155</span>,<span class="dv">157</span>,<span class="dv">160</span>],</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">149</span>,<span class="dv">153</span>,<span class="dv">154</span>,<span class="dv">155</span>]</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>    , index <span class="op">=</span> [<span class="st">'John'</span>,<span class="st">'Jack'</span>]</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>    , columns <span class="op">=</span> [<span class="st">"year-1991"</span>,<span class="st">"year-1992"</span>,<span class="st">"year-1993"</span>, <span class="st">"year-1994"</span>]</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a>).rename_axis(<span class="st">'individual'</span>).reset_index()</span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>wide_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
individual
</th>
<th>
year-1991
</th>
<th>
year-1992
</th>
<th>
year-1993
</th>
<th>
year-1994
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
John
</td>
<td>
150
</td>
<td>
155
</td>
<td>
157
</td>
<td>
160
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Jack
</td>
<td>
149
</td>
<td>
153
</td>
<td>
154
</td>
<td>
155
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>pd.wide_to_long(wide_df, [<span class="st">'year'</span>], i<span class="op">=</span><span class="st">'individual'</span>, j<span class="op">=</span><span class="st">'birthyear'</span>, sep<span class="op">=</span><span class="st">"-"</span>).rename(columns<span class="op">=</span>{<span class="st">'year'</span>:<span class="st">'height'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
</th>
<th>
height
</th>
</tr>
<tr>
<th>
individual
</th>
<th>
birthyear
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
John
</th>
<th>
1991
</th>
<td>
150
</td>
</tr>
<tr>
<th>
Jack
</th>
<th>
1991
</th>
<td>
149
</td>
</tr>
<tr>
<th>
John
</th>
<th>
1992
</th>
<td>
155
</td>
</tr>
<tr>
<th>
Jack
</th>
<th>
1992
</th>
<td>
153
</td>
</tr>
<tr>
<th>
John
</th>
<th>
1993
</th>
<td>
157
</td>
</tr>
<tr>
<th>
Jack
</th>
<th>
1993
</th>
<td>
154
</td>
</tr>
<tr>
<th>
John
</th>
<th>
1994
</th>
<td>
160
</td>
</tr>
<tr>
<th>
Jack
</th>
<th>
1994
</th>
<td>
155
</td>
</tr>
</tbody>

</table>
</div>
</section>
<section id="from-longtidy-to-wide-format-using-the-.pivot-function." class="level3" data-number="8.9.2">
<h3 data-number="8.9.2" class="anchored" data-anchor-id="from-longtidy-to-wide-format-using-the-.pivot-function."><span class="header-section-number">8.9.2</span> From long/tidy to wide format using the <code>.pivot()</code> function.</h3>
<div class="sourceCode" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>tidy_df.pivot(index<span class="op">=</span><span class="st">'individual'</span>, columns<span class="op">=</span><span class="st">'birthyear'</span>, values<span class="op">=</span><span class="st">'height'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/Users/maxime/mambaforge/envs/intro-data/lib/python3.10/site-packages/pandas/core/algorithms.py:798: FutureWarning: In a future version, the Index constructor will not infer numeric dtypes when passed object-dtype sequences (matching Series behavior)</code></pre>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
birthyear
</th>
<th>
1991
</th>
<th>
1992
</th>
<th>
1993
</th>
<th>
1994
</th>
</tr>
<tr>
<th>
individual
</th>
<th>
</th>
<th>
</th>
<th>
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
Jack
</th>
<td>
149
</td>
<td>
153
</td>
<td>
154
</td>
<td>
155
</td>
</tr>
<tr>
<th>
John
</th>
<td>
150
</td>
<td>
155
</td>
<td>
157
</td>
<td>
160
</td>
</tr>
</tbody>

</table>
</div>
</section>
</section>
<section id="joining-two-different-tables" class="level2" data-number="8.10">
<h2 data-number="8.10" class="anchored" data-anchor-id="joining-two-different-tables"><span class="header-section-number">8.10</span> 9 - Joining two different tables</h2>
<p>In AncientMetagenomeDir, the information about each sample is located in sample table, and about the library in the library table.<br>
To match these two together, we need to join the tables together.</p>
<p>To do so, we need a column in common between the two tables, the so-called <strong>joining key</strong> (this key can be the index)</p>
<p><img src="assets/images/chapters/python-pandas/08_merge_left.svg" class="img-fluid"></p>
<p>For the samples and libraries dataframe, the joining key is the column <code>sample_name</code></p>
<div class="sourceCode" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>sample_df.merge(library_df, on<span class="op">=</span><span class="st">'sample_name'</span>).columns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Index(['project_name_x', 'publication_year_x', 'publication_doi', 'site_name',
       'latitude', 'longitude', 'geo_loc_name', 'sample_name', 'sample_host',
       'sample_age', 'sample_age_doi', 'community_type', 'material',
       'archive_x', 'archive_project_x', 'archive_accession', 'project_name_y',
       'publication_year_y', 'data_publication_doi', 'archive_y',
       'archive_project_y', 'archive_sample_accession', 'library_name',
       'strand_type', 'library_polymerase', 'library_treatment',
       'library_concentration', 'instrument_model', 'library_layout',
       'library_strategy', 'read_count', 'archive_data_accession',
       'download_links', 'download_md5s', 'download_sizes'],
      dtype='object')</code></pre>
<p>We have some duplicate columns that we can get rid of:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>merged_df <span class="op">=</span> sample_df.merge(library_df.drop([<span class="st">'project_name'</span>, <span class="st">'publication_year'</span>, <span class="st">'archive_project'</span>, <span class="st">'archive'</span>], axis<span class="op">=</span><span class="dv">1</span>), on<span class="op">=</span><span class="st">'sample_name'</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>merged_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
…
</th>
<th>
library_treatment
</th>
<th>
library_concentration
</th>
<th>
instrument_model
</th>
<th>
library_layout
</th>
<th>
library_strategy
</th>
<th>
read_count
</th>
<th>
archive_data_accession
</th>
<th>
download_links
</th>
<th>
download_md5s
</th>
<th>
download_sizes
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2000
</td>
<td>
SINGLE
</td>
<td>
WGS
</td>
<td>
13228381
</td>
<td>
SRR957738
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957738/…
</td>
<td>
9c40c43b5d455e760ae8db924347f0b2
</td>
<td>
953396663
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2000
</td>
<td>
SINGLE
</td>
<td>
WGS
</td>
<td>
13260566
</td>
<td>
SRR957739
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957739/…
</td>
<td>
dec1507f742de109529638bf00e0732f
</td>
<td>
1026825795
</td>
</tr>
<tr>
<th>
2
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2000
</td>
<td>
SINGLE
</td>
<td>
WGS
</td>
<td>
8869866
</td>
<td>
SRR957740
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957740/…
</td>
<td>
bc49c59f489b4009206f8abcb737d55d
</td>
<td>
661500786
</td>
</tr>
<tr>
<th>
3
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2000
</td>
<td>
SINGLE
</td>
<td>
WGS
</td>
<td>
11275013
</td>
<td>
SRR957741
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957741/…
</td>
<td>
e02e3549ddd3ba6dc278a7f573c07321
</td>
<td>
877360302
</td>
</tr>
<tr>
<th>
4
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.84
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2000
</td>
<td>
SINGLE
</td>
<td>
WGS
</td>
<td>
8978974
</td>
<td>
SRR957742
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/SRR957/SRR957742/…
</td>
<td>
b7c620b8ee165c08bee204529341ca5b
</td>
<td>
690614774
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
1802
</th>
<td>
Maixner2021
</td>
<td>
2021
</td>
<td>
10.1016/j.cub.2021.09.031
</td>
<td>
Edlersbergwerk - oben, Hallstatt
</td>
<td>
47.560
</td>
<td>
13.63
</td>
<td>
Austria
</td>
<td>
2612
</td>
<td>
Homo sapiens
</td>
<td>
150
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina MiSeq
</td>
<td>
PAIRED
</td>
<td>
WGS
</td>
<td>
1858404
</td>
<td>
ERR5766179
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/ERR576/009/ERR576…
</td>
<td>
542787c645b0aeebe15c66cc926d3f69;0bc58d56be3c3…
</td>
<td>
86783041;98100690
</td>
</tr>
<tr>
<th>
1803
</th>
<td>
Maixner2021
</td>
<td>
2021
</td>
<td>
10.1016/j.cub.2021.09.031
</td>
<td>
Edlersbergwerk - oben, Hallstatt
</td>
<td>
47.560
</td>
<td>
13.63
</td>
<td>
Austria
</td>
<td>
2612
</td>
<td>
Homo sapiens
</td>
<td>
150
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina MiSeq
</td>
<td>
PAIRED
</td>
<td>
WGS
</td>
<td>
1603064
</td>
<td>
ERR5766180
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/ERR576/000/ERR576…
</td>
<td>
022bb28da460e66590e974b4135bdd2e;f88acec67b648…
</td>
<td>
74375931;77621627
</td>
</tr>
<tr>
<th>
1804
</th>
<td>
Maixner2021
</td>
<td>
2021
</td>
<td>
10.1016/j.cub.2021.09.031
</td>
<td>
Edlersbergwerk - oben, Hallstatt
</td>
<td>
47.560
</td>
<td>
13.63
</td>
<td>
Austria
</td>
<td>
2612
</td>
<td>
Homo sapiens
</td>
<td>
150
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina MiSeq
</td>
<td>
PAIRED
</td>
<td>
WGS
</td>
<td>
1075088
</td>
<td>
ERR5766181
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/ERR576/001/ERR576…
</td>
<td>
57fc575d32db14f1d5c1ed7f6a106e91;4f57b9d978b53…
</td>
<td>
51852071;56288763
</td>
</tr>
<tr>
<th>
1805
</th>
<td>
Maixner2021
</td>
<td>
2021
</td>
<td>
10.1016/j.cub.2021.09.031
</td>
<td>
Edlersbergwerk - oben, Hallstatt
</td>
<td>
47.560
</td>
<td>
13.63
</td>
<td>
Austria
</td>
<td>
2612
</td>
<td>
Homo sapiens
</td>
<td>
150
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
Illumina HiSeq 2500
</td>
<td>
PAIRED
</td>
<td>
WGS
</td>
<td>
138836358
</td>
<td>
ERR5766182
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/ERR576/002/ERR576…
</td>
<td>
64e63df8da7542957d1d9eb08e764d38;3fc6cba02c74d…
</td>
<td>
4332353625;4420486328
</td>
</tr>
<tr>
<th>
1806
</th>
<td>
Maixner2021
</td>
<td>
2021
</td>
<td>
10.1016/j.cub.2021.09.031
</td>
<td>
Edlersbergwerk - oben, Hallstatt
</td>
<td>
47.560
</td>
<td>
13.63
</td>
<td>
Austria
</td>
<td>
2612
</td>
<td>
Homo sapiens
</td>
<td>
150
</td>
<td>
…
</td>
<td>
none
</td>
<td>
NaN
</td>
<td>
HiSeq X Ten
</td>
<td>
PAIRED
</td>
<td>
WGS
</td>
<td>
84192332
</td>
<td>
ERR5766183
</td>
<td>
ftp.sra.ebi.ac.uk/vol1/fastq/ERR576/003/ERR576…
</td>
<td>
43ac661c4e211ed6ee2940dcab8b13cb;88de66a85df92…
</td>
<td>
3128863954;3460789287
</td>
</tr>
</tbody>

</table>
<p>
1807 rows × 31 columns
</p>
</div>
</section>
<section id="visualizing-some-of-the-results-with-plotnine" class="level2" data-number="8.11">
<h2 data-number="8.11" class="anchored" data-anchor-id="visualizing-some-of-the-results-with-plotnine"><span class="header-section-number">8.11</span> 10 - Visualizing some of the results with Plotnine</h2>
<p>Plotnine is the Python clone of ggplot2, the syntax is identical, which makes it great if you’re working with data in tidy/long format, and are already familiar with the ggplot2 syntax</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>ggplot(merged_df, aes(x<span class="op">=</span><span class="st">'publication_year'</span>)) <span class="op">+</span> geom_histogram() <span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/Users/maxime/mambaforge/envs/intro-data/lib/python3.10/
site-packages/plotnine/stats/stat_bin.py:95:
PlotnineWarning: 'stat_bin()' using 'bins = 15'. Pick better value with 'binwidth'.</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/python-pandas/tutorial_111_1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (366051178)&gt;</code></pre>
<p>We can start to ask some questions, for example, is the sequencing depth increasing with the years ?</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>merged_df[<span class="st">'publication_year'</span>] <span class="op">=</span> merged_df[<span class="st">'publication_year'</span>].astype(<span class="st">'category'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>ggplot(merged_df, aes(x<span class="op">=</span><span class="st">'publication_year'</span>, y<span class="op">=</span><span class="st">'np.log10(read_count)'</span>, fill<span class="op">=</span><span class="st">'publication_year'</span>)) <span class="op">+</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>geom_jitter(alpha<span class="op">=</span><span class="fl">0.1</span>) <span class="op">+</span> geom_boxplot(alpha<span class="op">=</span><span class="fl">0.8</span>) <span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/python-pandas/tutorial_114_0.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (366112582)&gt;</code></pre>
<p>We could ask the same question, but first grouping the samples by publication year</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>avg_read_count_by_year <span class="op">=</span> merged_df.groupby(<span class="st">'publication_year'</span>)[<span class="st">'read_count'</span>].mean().to_frame().reset_index()</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>avg_read_count_by_year</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
publication_year
</th>
<th>
read_count
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
2014
</td>
<td>
1.437173e+07
</td>
</tr>
<tr>
<th>
1
</th>
<td>
2016
</td>
<td>
3.653450e+04
</td>
</tr>
<tr>
<th>
2
</th>
<td>
2017
</td>
<td>
5.712598e+06
</td>
</tr>
<tr>
<th>
3
</th>
<td>
2018
</td>
<td>
9.273287e+06
</td>
</tr>
<tr>
<th>
4
</th>
<td>
2019
</td>
<td>
2.211632e+07
</td>
</tr>
<tr>
<th>
5
</th>
<td>
2020
</td>
<td>
1.111819e+07
</td>
</tr>
<tr>
<th>
6
</th>
<td>
2021
</td>
<td>
2.547655e+07
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>ggplot(avg_read_count_by_year, aes(x<span class="op">=</span><span class="st">'publication_year'</span>, y<span class="op">=</span><span class="st">'np.log10(read_count)'</span>, fill<span class="op">=</span><span class="st">'publication_year'</span>)) <span class="op">+</span> geom_point()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/python-pandas/tutorial_117_0.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (366206706)&gt;</code></pre>
<p><strong>Your turn ! Make a plot to investigate the relation between the type of library treatment throughout the publication years</strong></p>
</section>
<section id="bonus-dealing-with-ill-formatted-columns" class="level2" data-number="8.12">
<h2 data-number="8.12" class="anchored" data-anchor-id="bonus-dealing-with-ill-formatted-columns"><span class="header-section-number">8.12</span> 11 - Bonus, dealing with ill-formatted columns</h2>
<p>Sometimes, colums can contains entries which could be split in multiple columns, typically values separated by a comma. In AncientMetagenomeDir, this is the case with the archive accession column.</p>
<p>Here is how we would solve it with pandas</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>sample_df.assign(archive_accession <span class="op">=</span> sample_df.archive_accession.<span class="bu">str</span>.split(<span class="st">","</span>)).explode(<span class="st">'archive_accession'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
project_name
</th>
<th>
publication_year
</th>
<th>
publication_doi
</th>
<th>
site_name
</th>
<th>
latitude
</th>
<th>
longitude
</th>
<th>
geo_loc_name
</th>
<th>
sample_name
</th>
<th>
sample_host
</th>
<th>
sample_age
</th>
<th>
sample_age_doi
</th>
<th>
community_type
</th>
<th>
material
</th>
<th>
archive
</th>
<th>
archive_project
</th>
<th>
archive_accession
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473742
</td>
</tr>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473743
</td>
</tr>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473744
</td>
</tr>
<tr>
<th>
0
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
B61
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473745
</td>
</tr>
<tr>
<th>
1
</th>
<td>
Warinner2014
</td>
<td>
2014
</td>
<td>
10.1038/ng.2906
</td>
<td>
Dalheim
</td>
<td>
51.565
</td>
<td>
8.840
</td>
<td>
Germany
</td>
<td>
G12
</td>
<td>
Homo sapiens
</td>
<td>
900
</td>
<td>
10.1038/ng.2906
</td>
<td>
oral
</td>
<td>
dental calculus
</td>
<td>
SRA
</td>
<td>
PRJNA216965
</td>
<td>
SRS473747
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
1057
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Gertrude’s Church, Riga
</td>
<td>
56.958
</td>
<td>
24.121
</td>
<td>
Latvia
</td>
<td>
T9
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283099
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
Dom Square, Riga
</td>
<td>
56.949
</td>
<td>
24.104
</td>
<td>
Latvia
</td>
<td>
TZA3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283100
</td>
</tr>
<tr>
<th>
1058
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
Dom Square, Riga
</td>
<td>
56.949
</td>
<td>
24.104
</td>
<td>
Latvia
</td>
<td>
TZA3
</td>
<td>
Homo sapiens
</td>
<td>
400
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283101
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Peter’s Church, Riga
</td>
<td>
56.947
</td>
<td>
24.109
</td>
<td>
Latvia
</td>
<td>
TZA4
</td>
<td>
Homo sapiens
</td>
<td>
500
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283102
</td>
</tr>
<tr>
<th>
1059
</th>
<td>
Kazarina2021b
</td>
<td>
2021
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
St.&nbsp;Peter’s Church, Riga
</td>
<td>
56.947
</td>
<td>
24.109
</td>
<td>
Latvia
</td>
<td>
TZA4
</td>
<td>
Homo sapiens
</td>
<td>
500
</td>
<td>
10.1016/j.jasrep.2021.103213
</td>
<td>
oral
</td>
<td>
tooth
</td>
<td>
ENA
</td>
<td>
PRJEB47251
</td>
<td>
ERS7283103
</td>
</tr>
</tbody>

</table>
<p>
1262 rows × 16 columns
</p>
</div>
</section>
<section id="lecture-from-2022" class="level2" data-number="8.13">
<h2 data-number="8.13" class="anchored" data-anchor-id="lecture-from-2022"><span class="header-section-number">8.13</span> Lecture from 2022</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here are the lecture from the previous course, as well as the links to the material from which the current practical was derived.</p>
</div>
</div>
<p>Lecture slides and video from the <a href="https://www.spaam-community.org/wss-summer-school/#/2022/README">2022 edition of the summer school</a>.</p>
<p>PDF version of these slides can be downloaded from <a href="https://github.com/SPAAM-community/wss-summer-school/raw/main/docs/assets/slides/2022/3b2-python-pandas/SPAAM%20Summer%20School%202022%20-%203B2%20-%20Intro%20to%20Python%20and%20Pandas.pdf">here</a>.</p>
<p>This session is run using a Jupyter notebook. This can be found <a href="https://github.com/maxibor/intro-to-pandas-plotnine">here</a>. However, it will already be installed on compute nodes during the summer school.</p>
<iframe width="100%" height="400px" src="https://www.youtube.com/embed/V_EKs_zlO1w" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="">
</iframe>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>We highly recommend viewing this walkthrough via the Jupyter notebook above! The output of commands on the website for this walkthrough are displayed in their own code blocks - be wary of what you copy-paste!</p>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./r-tidyverse.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./git-github.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">© 2023 SPAAM Community &amp; Authors with ❤️. Available under <a href="http://creativecommons.org/licenses/by-sa/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>