<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Clemens Schmid">

<title>Introduction to Ancient Metagenomics - 7&nbsp; Introduction to R and the Tidyverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./python-pandas.html" rel="next">
<link href="./bare-bones-bash.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-useful-skills.html">Useful Skills</a></li><li class="breadcrumb-item"><a href="./r-tidyverse.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main tools-wide">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./citing-this-book.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Citing this book</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-evolutionary-biology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Evolutionary Biology</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-tidyverse.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./python-pandas.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./git-github.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authentication-decontamination.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Authentication and Decontamination</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Introduction to Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancientmetagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Accessing Ancient Metagenomic Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lecture" id="toc-lecture" class="nav-link active" data-scroll-target="#lecture"><span class="header-section-number">7.1</span> Lecture</a></li>
  <li><a href="#the-working-environment" id="toc-the-working-environment" class="nav-link" data-scroll-target="#the-working-environment"><span class="header-section-number">7.2</span> The working environment</a>
  <ul class="collapse">
  <li><a href="#r-rstudio-and-the-tidyverse" id="toc-r-rstudio-and-the-tidyverse" class="nav-link" data-scroll-target="#r-rstudio-and-the-tidyverse"><span class="header-section-number">7.2.1</span> R, RStudio and the tidyverse</a></li>
  </ul></li>
  <li><a href="#loading-data-into-tibbles" id="toc-loading-data-into-tibbles" class="nav-link" data-scroll-target="#loading-data-into-tibbles"><span class="header-section-number">7.3</span> Loading data into tibbles</a>
  <ul class="collapse">
  <li><a href="#reading-data-with-readr" id="toc-reading-data-with-readr" class="nav-link" data-scroll-target="#reading-data-with-readr"><span class="header-section-number">7.3.1</span> Reading data with readr</a></li>
  <li><a href="#how-does-the-interface-of-read_csv-work" id="toc-how-does-the-interface-of-read_csv-work" class="nav-link" data-scroll-target="#how-does-the-interface-of-read_csv-work"><span class="header-section-number">7.3.2</span> How does the interface of <code>read_csv</code> work</a></li>
  <li><a href="#what-does-readr-produce-the-tibble" id="toc-what-does-readr-produce-the-tibble" class="nav-link" data-scroll-target="#what-does-readr-produce-the-tibble"><span class="header-section-number">7.3.3</span> What does <code>readr</code> produce? The <code>tibble</code></a></li>
  <li><a href="#how-to-look-at-a-tibble" id="toc-how-to-look-at-a-tibble" class="nav-link" data-scroll-target="#how-to-look-at-a-tibble"><span class="header-section-number">7.3.4</span> How to look at a <code>tibble</code></a></li>
  </ul></li>
  <li><a href="#plotting-data-in-tibbles" id="toc-plotting-data-in-tibbles" class="nav-link" data-scroll-target="#plotting-data-in-tibbles"><span class="header-section-number">7.4</span> Plotting data in <code>tibble</code>s</a>
  <ul class="collapse">
  <li><a href="#ggplot2-and-the-grammar-of-graphics" id="toc-ggplot2-and-the-grammar-of-graphics" class="nav-link" data-scroll-target="#ggplot2-and-the-grammar-of-graphics"><span class="header-section-number">7.4.1</span> <code>ggplot2</code> and the “grammar of graphics”</a></li>
  <li><a href="#ggplot2-and-the-grammar-of-graphics-1" id="toc-ggplot2-and-the-grammar-of-graphics-1" class="nav-link" data-scroll-target="#ggplot2-and-the-grammar-of-graphics-1"><span class="header-section-number">7.4.2</span> <code>ggplot2</code> and the “grammar of graphics”</a></li>
  <li><a href="#ggplot2-features-many-geoms" id="toc-ggplot2-features-many-geoms" class="nav-link" data-scroll-target="#ggplot2-features-many-geoms"><span class="header-section-number">7.4.3</span> <code>ggplot2</code> features many geoms</a></li>
  <li><a href="#scales-control-the-behaviour-of-visual-elements" id="toc-scales-control-the-behaviour-of-visual-elements" class="nav-link" data-scroll-target="#scales-control-the-behaviour-of-visual-elements"><span class="header-section-number">7.4.4</span> <code>scale</code>s control the behaviour of visual elements</a></li>
  <li><a href="#scales-control-the-behaviour-of-visual-elements-1" id="toc-scales-control-the-behaviour-of-visual-elements-1" class="nav-link" data-scroll-target="#scales-control-the-behaviour-of-visual-elements-1"><span class="header-section-number">7.4.5</span> <code>scale</code>s control the behaviour of visual elements</a></li>
  <li><a href="#scales-control-the-behaviour-of-visual-elements-2" id="toc-scales-control-the-behaviour-of-visual-elements-2" class="nav-link" data-scroll-target="#scales-control-the-behaviour-of-visual-elements-2"><span class="header-section-number">7.4.6</span> <code>scale</code>s control the behaviour of visual elements</a></li>
  <li><a href="#defining-plot-matrices-via-facets" id="toc-defining-plot-matrices-via-facets" class="nav-link" data-scroll-target="#defining-plot-matrices-via-facets"><span class="header-section-number">7.4.7</span> Defining plot matrices via <code>facet</code>s</a></li>
  <li><a href="#setting-purely-aesthetic-settings-with-theme" id="toc-setting-purely-aesthetic-settings-with-theme" class="nav-link" data-scroll-target="#setting-purely-aesthetic-settings-with-theme"><span class="header-section-number">7.4.8</span> Setting purely aesthetic settings with <code>theme</code></a></li>
  <li><a href="#exercise-1" id="toc-exercise-1" class="nav-link" data-scroll-target="#exercise-1"><span class="header-section-number">7.4.9</span> Exercise 1</a></li>
  <li><a href="#possible-solutions-1" id="toc-possible-solutions-1" class="nav-link" data-scroll-target="#possible-solutions-1"><span class="header-section-number">7.4.10</span> Possible solutions 1</a></li>
  </ul></li>
  <li><a href="#conditional-queries-on-tibbles" id="toc-conditional-queries-on-tibbles" class="nav-link" data-scroll-target="#conditional-queries-on-tibbles"><span class="header-section-number">7.5</span> Conditional queries on tibbles</a>
  <ul class="collapse">
  <li><a href="#selecting-columns-and-filtering-rows-with-select-and-filter" id="toc-selecting-columns-and-filtering-rows-with-select-and-filter" class="nav-link" data-scroll-target="#selecting-columns-and-filtering-rows-with-select-and-filter"><span class="header-section-number">7.5.1</span> Selecting columns and filtering rows with <code>select</code> and <code>filter</code></a></li>
  <li><a href="#chaining-functions-together-with-the-pipe" id="toc-chaining-functions-together-with-the-pipe" class="nav-link" data-scroll-target="#chaining-functions-together-with-the-pipe"><span class="header-section-number">7.5.2</span> Chaining functions together with the pipe <code>%&gt;%</code></a></li>
  <li><a href="#summary-statistics-in-base-r" id="toc-summary-statistics-in-base-r" class="nav-link" data-scroll-target="#summary-statistics-in-base-r"><span class="header-section-number">7.5.3</span> Summary statistics in <code>base</code> R</a></li>
  <li><a href="#group-wise-summaries-with-group_by-and-summarise" id="toc-group-wise-summaries-with-group_by-and-summarise" class="nav-link" data-scroll-target="#group-wise-summaries-with-group_by-and-summarise"><span class="header-section-number">7.5.4</span> Group-wise summaries with <code>group_by</code> and <code>summarise</code></a></li>
  <li><a href="#sorting-and-slicing-tibbles-with-arrange-and-slice" id="toc-sorting-and-slicing-tibbles-with-arrange-and-slice" class="nav-link" data-scroll-target="#sorting-and-slicing-tibbles-with-arrange-and-slice"><span class="header-section-number">7.5.5</span> Sorting and slicing tibbles with <code>arrange</code> and <code>slice</code></a></li>
  <li><a href="#exercise-2" id="toc-exercise-2" class="nav-link" data-scroll-target="#exercise-2"><span class="header-section-number">7.5.6</span> Exercise 2</a></li>
  <li><a href="#possible-solutions-2" id="toc-possible-solutions-2" class="nav-link" data-scroll-target="#possible-solutions-2"><span class="header-section-number">7.5.7</span> Possible solutions 2</a></li>
  </ul></li>
  <li><a href="#transforming-and-manipulating-tibbles" id="toc-transforming-and-manipulating-tibbles" class="nav-link" data-scroll-target="#transforming-and-manipulating-tibbles"><span class="header-section-number">7.6</span> Transforming and manipulating tibbles</a>
  <ul class="collapse">
  <li><a href="#renaming-and-reordering-columns-and-values-with-rename-relocate-and-recode" id="toc-renaming-and-reordering-columns-and-values-with-rename-relocate-and-recode" class="nav-link" data-scroll-target="#renaming-and-reordering-columns-and-values-with-rename-relocate-and-recode"><span class="header-section-number">7.6.1</span> Renaming and reordering columns and values with <code>rename</code>, <code>relocate</code> and <code>recode</code></a></li>
  <li><a href="#adding-columns-to-tibbles-with-mutate-and-transmute" id="toc-adding-columns-to-tibbles-with-mutate-and-transmute" class="nav-link" data-scroll-target="#adding-columns-to-tibbles-with-mutate-and-transmute"><span class="header-section-number">7.6.2</span> Adding columns to tibbles with <code>mutate</code> and <code>transmute</code></a></li>
  <li><a href="#conditional-operations-with-ifelse-and-case_when" id="toc-conditional-operations-with-ifelse-and-case_when" class="nav-link" data-scroll-target="#conditional-operations-with-ifelse-and-case_when"><span class="header-section-number">7.6.3</span> Conditional operations with <code>ifelse</code> and <code>case_when</code></a></li>
  <li><a href="#long-and-wide-data-formats" id="toc-long-and-wide-data-formats" class="nav-link" data-scroll-target="#long-and-wide-data-formats"><span class="header-section-number">7.6.4</span> Long and wide data formats</a></li>
  <li><a href="#a-wide-dataset" id="toc-a-wide-dataset" class="nav-link" data-scroll-target="#a-wide-dataset"><span class="header-section-number">7.6.5</span> A wide dataset</a></li>
  <li><a href="#making-a-wide-dataset-long-with-pivot_longer" id="toc-making-a-wide-dataset-long-with-pivot_longer" class="nav-link" data-scroll-target="#making-a-wide-dataset-long-with-pivot_longer"><span class="header-section-number">7.6.6</span> Making a wide dataset long with <code>pivot_longer</code></a></li>
  <li><a href="#making-a-long-dataset-wide-with-pivot_wider" id="toc-making-a-long-dataset-wide-with-pivot_wider" class="nav-link" data-scroll-target="#making-a-long-dataset-wide-with-pivot_wider"><span class="header-section-number">7.6.7</span> Making a long dataset wide with <code>pivot_wider</code></a></li>
  <li><a href="#exercise-3" id="toc-exercise-3" class="nav-link" data-scroll-target="#exercise-3"><span class="header-section-number">7.6.8</span> Exercise 3</a></li>
  <li><a href="#possible-solutions-3" id="toc-possible-solutions-3" class="nav-link" data-scroll-target="#possible-solutions-3"><span class="header-section-number">7.6.9</span> Possible solutions 3</a></li>
  </ul></li>
  <li><a href="#combining-tibbles-with-join-operations" id="toc-combining-tibbles-with-join-operations" class="nav-link" data-scroll-target="#combining-tibbles-with-join-operations"><span class="header-section-number">7.7</span> Combining tibbles with join operations</a>
  <ul class="collapse">
  <li><a href="#types-of-joins" id="toc-types-of-joins" class="nav-link" data-scroll-target="#types-of-joins"><span class="header-section-number">7.7.1</span> Types of joins</a></li>
  <li><a href="#a-second-dataset" id="toc-a-second-dataset" class="nav-link" data-scroll-target="#a-second-dataset"><span class="header-section-number">7.7.2</span> A second dataset</a></li>
  <li><a href="#meaningful-subsets" id="toc-meaningful-subsets" class="nav-link" data-scroll-target="#meaningful-subsets"><span class="header-section-number">7.7.3</span> Meaningful subsets</a></li>
  <li><a href="#left-join" id="toc-left-join" class="nav-link" data-scroll-target="#left-join"><span class="header-section-number">7.7.4</span> Left join</a></li>
  <li><a href="#right-join" id="toc-right-join" class="nav-link" data-scroll-target="#right-join"><span class="header-section-number">7.7.5</span> Right join</a></li>
  <li><a href="#inner-join" id="toc-inner-join" class="nav-link" data-scroll-target="#inner-join"><span class="header-section-number">7.7.6</span> Inner join</a></li>
  <li><a href="#full-join" id="toc-full-join" class="nav-link" data-scroll-target="#full-join"><span class="header-section-number">7.7.7</span> Full join</a></li>
  <li><a href="#semi-join" id="toc-semi-join" class="nav-link" data-scroll-target="#semi-join"><span class="header-section-number">7.7.8</span> Semi join</a></li>
  <li><a href="#anti-join" id="toc-anti-join" class="nav-link" data-scroll-target="#anti-join"><span class="header-section-number">7.7.9</span> Anti join</a></li>
  <li><a href="#exercise-4" id="toc-exercise-4" class="nav-link" data-scroll-target="#exercise-4"><span class="header-section-number">7.7.10</span> Exercise 4</a></li>
  <li><a href="#possible-solutions-4" id="toc-possible-solutions-4" class="nav-link" data-scroll-target="#possible-solutions-4"><span class="header-section-number">7.7.11</span> Possible Solutions 4</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Clemens Schmid </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This session is typically ran held in parallel to the Introduction to Python and Pandas. Participants of the summer schools chose which to attend based on their prior experience. We recommend this session if you have no experience with neither R nor Python.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to create the <a href="./before-you-start.html#creating-a-conda-environment">conda environment</a> from the <code>yml</code> file in the following <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book/raw/main/assets/envs/r-tidyverse.yml">link</a> (right click and save as to download), and once created, activate the environment with:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate r-tidyverse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="lecture" class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="lecture"><span class="header-section-number">7.1</span> Lecture</h2>
<p>Lecture slides and video from the <a href="https://www.spaam-community.org/wss-summer-school/#/2022/README">2022 edition of the summer school</a>.</p>
<iframe src="https://docs.google.com/viewer?url=https://raw.githubusercontent.com/nevrome/spaam_r_tidyverse_intro_2h/main/presentation.pdf&amp;embedded=true" width="100%" height="400px">
</iframe>
<p>PDF version of these slides can be downloaded from <a href="https://github.com/nevrome/spaam_r_tidyverse_intro_2h/raw/main/presentation.pdf">here</a>.</p>
<iframe width="100%" height="400px" src="https://www.youtube.com/embed/b32FRYWwSrg" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="">
</iframe>
</section>
<section id="the-working-environment" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="the-working-environment"><span class="header-section-number">7.2</span> The working environment</h2>
<section id="r-rstudio-and-the-tidyverse" class="level3" data-number="7.2.1">
<h3 data-number="7.2.1" class="anchored" data-anchor-id="r-rstudio-and-the-tidyverse"><span class="header-section-number">7.2.1</span> R, RStudio and the tidyverse</h3>
<ul>
<li><p>R is a fully featured programming language, but it excels as an environment for (statistical) data analysis (<a href="https://www.r-project.org">https://www.r-project.org</a>)</p></li>
<li><p>RStudio is an integrated development environment (IDE) for R (and other languages): (<a href="https://www.rstudio.com/products/rstudio">https://www.rstudio.com/products/rstudio</a>)</p></li>
<li><p>The tidyverse is a collection of R packages with well-designed and consistent interfaces for the main steps of data analysis: loading, transforming and plotting data (<a href="https://www.tidyverse.org">https://www.tidyverse.org</a>)</p>
<ul>
<li>This introduction works with tidyverse ~v1.3.0</li>
<li>We will learn about <code>readr</code>, <code>tibble</code>, <code>ggplot2</code>, <code>dplyr</code>, <code>magrittr</code> and <code>tidyr</code></li>
<li><code>forcats</code> will be briefly mentioned</li>
<li><code>purrr</code> and <code>stringr</code> are left out</li>
</ul></li>
</ul>
</section>
</section>
<section id="loading-data-into-tibbles" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="loading-data-into-tibbles"><span class="header-section-number">7.3</span> Loading data into tibbles</h2>
<section id="reading-data-with-readr" class="level3" data-number="7.3.1">
<h3 data-number="7.3.1" class="anchored" data-anchor-id="reading-data-with-readr"><span class="header-section-number">7.3.1</span> Reading data with readr</h3>
<ul>
<li><p>With R we usually operate on data in our computer’s memory</p></li>
<li><p>The tidyverse provides the package <code>readr</code> to read data from text files into the memory</p></li>
<li><p><code>readr</code> can read from our file system or the internet</p></li>
<li><p>It provides functions to read data in almost any (text) format:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_csv</span>()   <span class="co"># .csv files</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_tsv</span>()   <span class="co"># .tsv files</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_delim</span>() <span class="co"># tabular files with an arbitrary separator</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_fwf</span>()   <span class="co"># fixed width files</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>readr<span class="sc">::</span><span class="fu">read_lines</span>() <span class="co"># read linewise to parse yourself</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>readr</code> automatically detects column types – but you can also define them manually</p></li>
</ul>
</section>
<section id="how-does-the-interface-of-read_csv-work" class="level3" data-number="7.3.2">
<h3 data-number="7.3.2" class="anchored" data-anchor-id="how-does-the-interface-of-read_csv-work"><span class="header-section-number">7.3.2</span> How does the interface of <code>read_csv</code> work</h3>
<ul>
<li><p>We can learn more about a function with <code>?</code>. To open a help file: <code>?readr::read_csv</code></p></li>
<li><p><code>readr::read_csv</code> has many options to specify how to read a text file</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">read_csv</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>file,                      <span class="co"># The path to the file we want to read</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="at">col_names =</span> <span class="cn">TRUE</span>,          <span class="co"># Are there column names?</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="at">col_types =</span> <span class="cn">NULL</span>,          <span class="co"># Which types do the columns have? NULL -&gt; auto</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="at">locale =</span> <span class="fu">default_locale</span>(), <span class="co"># How is information encoded in this file?</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="at">na =</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">"NA"</span>),          <span class="co"># Which values mean "no data"</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="at">trim_ws =</span> <span class="cn">TRUE</span>,            <span class="co"># Should superfluous white-spaces be removed?</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="at">skip =</span> <span class="dv">0</span>,                  <span class="co"># Skip X lines at the beginning of the file</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="at">n_max =</span> <span class="cn">Inf</span>,               <span class="co"># Only read X lines</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="at">skip_empty_rows =</span> <span class="cn">TRUE</span>,    <span class="co"># Should empty lines be ignored?</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="at">comment =</span> <span class="st">""</span>,              <span class="co"># Should comment lines be ignored?</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="at">name_repair =</span> <span class="st">"unique"</span>,    <span class="co"># How should "broken" column names be fixed</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="what-does-readr-produce-the-tibble" class="level3" data-number="7.3.3">
<h3 data-number="7.3.3" class="anchored" data-anchor-id="what-does-readr-produce-the-tibble"><span class="header-section-number">7.3.3</span> What does <code>readr</code> produce? The <code>tibble</code></h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>sample_table_path <span class="ot">&lt;-</span> <span class="st">"/vol/volume/r-tidyverse/ancientmetagenome-hostassociated_samples.tsv"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>sample_table_url <span class="ot">&lt;-</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="st">"https://raw.githubusercontent.com/SPAAM-community/AncientMetagenomeDir/b187df6ebd23dfeb42935fd5020cb615ead3f164/</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="st">ancientmetagenome-hostassociated/samples/ancientmetagenome-hostassociated_samples.tsv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(sample_table_url)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The <code>tibble</code> is a “data frame”, a tabular data structure with rows and columns</li>
<li>Unlike a simple array, each column can have another data type</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(samples, <span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="how-to-look-at-a-tibble" class="level3" data-number="7.3.4">
<h3 data-number="7.3.4" class="anchored" data-anchor-id="how-to-look-at-a-tibble"><span class="header-section-number">7.3.4</span> How to look at a <code>tibble</code></h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>samples          <span class="co"># Typing the name of an object will print it to the console</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(samples)     <span class="co"># A structural overview of an object</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(samples) <span class="co"># A human-readable summary of an object</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(samples)    <span class="co"># RStudio's interactive data browser</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p>R provides a very flexible indexing operation for <code>data.frame</code>s and <code>tibble</code>s</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>samples[<span class="dv">1</span>,<span class="dv">1</span>]                         <span class="co"># Access the first row and column</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>samples[<span class="dv">1</span>,]                          <span class="co"># Access the first row</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>samples[,<span class="dv">1</span>]                          <span class="co"># Access the first column</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>samples[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>),<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>)]           <span class="co"># Access values from rows and columns</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>samples[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)]                    <span class="co"># Remove the first two columns</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>samples[,<span class="fu">c</span>(<span class="st">"site_name"</span>, <span class="st">"material"</span>)] <span class="co"># Columns can be selected by name</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>tibble</code>s are mutable data structures, so their content can be overwritten</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>samples[<span class="dv">1</span>,<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">"Cheesecake2015"</span>     <span class="co"># replace the first value in the first column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
</section>
<section id="plotting-data-in-tibbles" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="plotting-data-in-tibbles"><span class="header-section-number">7.4</span> Plotting data in <code>tibble</code>s</h2>
<section id="ggplot2-and-the-grammar-of-graphics" class="level3" data-number="7.4.1">
<h3 data-number="7.4.1" class="anchored" data-anchor-id="ggplot2-and-the-grammar-of-graphics"><span class="header-section-number">7.4.1</span> <code>ggplot2</code> and the “grammar of graphics”</h3>
<ul>
<li><p><code>ggplot2</code> offers an unusual, but powerful and logical interface</p></li>
<li><p>The following example describes a stacked bar chart</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co"># Loading a library to use its functions without ::</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(          <span class="co"># Every plot starts with a call to the ggplot() function</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="at">data =</span> samples <span class="co"># This function can also take the input tibble</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span>              <span class="co"># The plot consists of functions linked with +</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_bar</span>(        <span class="co"># "geoms" define the plot layers we want to draw</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="at">mapping =</span> <span class="fu">aes</span>( <span class="co"># The aes() function maps variables to visual properties</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> publication_year, <span class="co"># publication_year -&gt; x-axis</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">fill =</span> community_type <span class="co"># community_type -&gt; fill color</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>geom_*</code>: data + geometry (bars) + statistical transformation (sum)</p></li>
</ul>
</section>
<section id="ggplot2-and-the-grammar-of-graphics-1" class="level3" data-number="7.4.2">
<h3 data-number="7.4.2" class="anchored" data-anchor-id="ggplot2-and-the-grammar-of-graphics-1"><span class="header-section-number">7.4.2</span> <code>ggplot2</code> and the “grammar of graphics”</h3>
<ul>
<li><p>This is the plot described above: number of samples per community type through time</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> publication_year, <span class="at">fill =</span> community_type))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="ggplot2-features-many-geoms" class="level3" data-number="7.4.3">
<h3 data-number="7.4.3" class="anchored" data-anchor-id="ggplot2-features-many-geoms"><span class="header-section-number">7.4.3</span> <code>ggplot2</code> features many geoms</h3>
<p><img src="assets/images/chapters/r-tidyverse/geom_cheatsheet.png" class="img-fluid" alt="figures/geom_cheatsheet.png">(){width=55%}</p>
<ul>
<li>RStudio shares helpful cheatsheets for the tidyverse and beyond: <a href="https://www.rstudio.com/resources/cheatsheets">https://www.rstudio.com/resources/cheatsheets</a></li>
</ul>
</section>
<section id="scales-control-the-behaviour-of-visual-elements" class="level3" data-number="7.4.4">
<h3 data-number="7.4.4" class="anchored" data-anchor-id="scales-control-the-behaviour-of-visual-elements"><span class="header-section-number">7.4.4</span> <code>scale</code>s control the behaviour of visual elements</h3>
<ul>
<li><p>Another plot: Boxplots of sample age through time</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.factor</span>(publication_year), <span class="at">y =</span> sample_age))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>This is not well readable, because extreme outliers dictate the scale</p></li>
</ul>
</section>
<section id="scales-control-the-behaviour-of-visual-elements-1" class="level3" data-number="7.4.5">
<h3 data-number="7.4.5" class="anchored" data-anchor-id="scales-control-the-behaviour-of-visual-elements-1"><span class="header-section-number">7.4.5</span> <code>scale</code>s control the behaviour of visual elements</h3>
<ul>
<li><p>We can change the <strong>scale</strong> of different visual elements - e.g.&nbsp;the y-axis</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.factor</span>(publication_year), <span class="at">y =</span> sample_age)) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_y_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>The log-scale improves readability</p></li>
</ul>
</section>
<section id="scales-control-the-behaviour-of-visual-elements-2" class="level3" data-number="7.4.6">
<h3 data-number="7.4.6" class="anchored" data-anchor-id="scales-control-the-behaviour-of-visual-elements-2"><span class="header-section-number">7.4.6</span> <code>scale</code>s control the behaviour of visual elements</h3>
<ul>
<li><p>(Fill) color is a visual element of the plot and its scaling can be adjusted</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.factor</span>(publication_year), <span class="at">y =</span> sample_age,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">fill =</span> <span class="fu">as.factor</span>(publication_year))) <span class="sc">+</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="fu">scale_y_log10</span>() <span class="sc">+</span> <span class="fu">scale_fill_viridis_d</span>(<span class="at">option =</span> <span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="defining-plot-matrices-via-facets" class="level3" data-number="7.4.7">
<h3 data-number="7.4.7" class="anchored" data-anchor-id="defining-plot-matrices-via-facets"><span class="header-section-number">7.4.7</span> Defining plot matrices via <code>facet</code>s</h3>
<ul>
<li><p>Splitting up the plot by categories into <strong>facets</strong> is another way to visualize more variables at once</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_count</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.factor</span>(publication_year), <span class="at">y =</span> material)) <span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">facet_wrap</span>(<span class="sc">~</span>archive)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Unfortunately the x-axis became unreadable</p></li>
</ul>
</section>
<section id="setting-purely-aesthetic-settings-with-theme" class="level3" data-number="7.4.8">
<h3 data-number="7.4.8" class="anchored" data-anchor-id="setting-purely-aesthetic-settings-with-theme"><span class="header-section-number">7.4.8</span> Setting purely aesthetic settings with <code>theme</code></h3>
<ul>
<li><p>Aesthetic changes like this can be applied as part of the <code>theme</code></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">geom_count</span>(<span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">as.factor</span>(publication_year), <span class="at">y =</span> material)) <span class="sc">+</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">facet_wrap</span>(<span class="sc">~</span>archive) <span class="sc">+</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">vjust =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="exercise-1" class="level3" data-number="7.4.9">
<h3 data-number="7.4.9" class="anchored" data-anchor-id="exercise-1"><span class="header-section-number">7.4.9</span> Exercise 1</h3>
<ol type="1">
<li><p>Look at the <code>mtcars</code> dataset and read up on the meaning of its variables</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Visualize the relationship between <em>Gross horsepower</em> and <em>1/4 mile time</em></p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Integrate the <em>Number of cylinders</em> into your plot</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
<section id="possible-solutions-1" class="level3" data-number="7.4.10">
<h3 data-number="7.4.10" class="anchored" data-anchor-id="possible-solutions-1"><span class="header-section-number">7.4.10</span> Possible solutions 1</h3>
<ol type="1">
<li><p>Look at the <code>mtcars</code> dataset and read up on the meaning of its variables</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>?mtcars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Visualize the relationship between <em>Gross horsepower</em> and <em>1/4 mile time</em></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mtcars) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> hp, <span class="at">y =</span> qsec))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Integrate the <em>Number of cylinders</em> into your plot</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(mtcars) <span class="sc">+</span> <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> hp, <span class="at">y =</span> qsec, <span class="at">color =</span> <span class="fu">as.factor</span>(cyl)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
</section>
<section id="conditional-queries-on-tibbles" class="level2" data-number="7.5">
<h2 data-number="7.5" class="anchored" data-anchor-id="conditional-queries-on-tibbles"><span class="header-section-number">7.5</span> Conditional queries on tibbles</h2>
<section id="selecting-columns-and-filtering-rows-with-select-and-filter" class="level3" data-number="7.5.1">
<h3 data-number="7.5.1" class="anchored" data-anchor-id="selecting-columns-and-filtering-rows-with-select-and-filter"><span class="header-section-number">7.5.1</span> Selecting columns and filtering rows with <code>select</code> and <code>filter</code></h3>
<ul>
<li><p>The <code>dplyr</code> package includes powerful functions to subset data in tibbles based on conditions</p></li>
<li><p><code>dplyr::select</code> allows to select columns</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(samples, project_name, sample_age)   <span class="co"># reduce to two columns</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(samples, <span class="sc">-</span>project_name, <span class="sc">-</span>sample_age) <span class="co"># remove two columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>dplyr::filter</code> allows for conditional filtering of rows</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(samples, publication_year <span class="sc">==</span> <span class="dv">2014</span>)  <span class="co"># samples published in 2014</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(samples, publication_year <span class="sc">==</span> <span class="dv">2014</span> <span class="sc">|</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>                    publication_year <span class="sc">==</span> <span class="dv">2018</span>)  <span class="co"># samples from 2015 OR 2018</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(samples, publication_year <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">2014</span>, <span class="dv">2018</span>)) <span class="co"># match operator: %in%</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(samples, sample_host <span class="sc">==</span> <span class="st">"Homo sapiens"</span> <span class="sc">&amp;</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                    community_type <span class="sc">==</span> <span class="st">"oral"</span>)  <span class="co"># oral samples from modern humans</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="chaining-functions-together-with-the-pipe" class="level3" data-number="7.5.2">
<h3 data-number="7.5.2" class="anchored" data-anchor-id="chaining-functions-together-with-the-pipe"><span class="header-section-number">7.5.2</span> Chaining functions together with the pipe <code>%&gt;%</code></h3>
<ul>
<li><p>The pipe <code>%&gt;%</code> in the <code>magrittr</code> package is a clever infix operator to chain data and operations</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(publication_year <span class="sc">==</span> <span class="dv">2014</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>It forwards the LHS as the first argument of the function appearing on the RHS</p></li>
<li><p>That allows for sequences of functions (“tidyverse style”)</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">select</span>(sample_host, community_type) <span class="sc">%&gt;%</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(sample_host <span class="sc">==</span> <span class="st">"Homo sapiens"</span> <span class="sc">&amp;</span> community_type <span class="sc">==</span> <span class="st">"oral"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>() <span class="co"># count the rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>magrittr</code> also offers some more operators, among which the extraction <code>%$%</code> is particularly useful</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">filter</span>(material <span class="sc">==</span> <span class="st">"tooth"</span>) <span class="sc">%$%</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>sample_age <span class="sc">%&gt;%</span> <span class="co"># extract the sample_age column as a vector</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>()          <span class="co"># get the maximum of said vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="summary-statistics-in-base-r" class="level3" data-number="7.5.3">
<h3 data-number="7.5.3" class="anchored" data-anchor-id="summary-statistics-in-base-r"><span class="header-section-number">7.5.3</span> Summary statistics in <code>base</code> R</h3>
<ul>
<li><p>Summarising and counting data is indispensable and R offers all operations you would expect in its <code>base</code> package</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(samples)              <span class="co"># number of rows in a tibble</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(samples<span class="sc">$</span>site_name)  <span class="co"># length/size of a vector</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(samples<span class="sc">$</span>material)   <span class="co"># unique elements of a vector</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(samples<span class="sc">$</span>sample_age)    <span class="co"># minimum</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(samples<span class="sc">$</span>sample_age)    <span class="co"># maximum</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(samples<span class="sc">$</span>sample_age)   <span class="co"># mean</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(samples<span class="sc">$</span>sample_age) <span class="co"># median</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="fu">var</span>(samples<span class="sc">$</span>sample_age)    <span class="co"># variance</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(samples<span class="sc">$</span>sample_age)     <span class="co"># standard deviation</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="fu">quantile</span>(samples<span class="sc">$</span>sample_age, <span class="at">probs =</span> <span class="fl">0.75</span>) <span class="co"># sample quantiles for the given probs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>many of these functions can ignore missing values with an option <code>na.rm = TRUE</code></p></li>
</ul>
</section>
<section id="group-wise-summaries-with-group_by-and-summarise" class="level3" data-number="7.5.4">
<h3 data-number="7.5.4" class="anchored" data-anchor-id="group-wise-summaries-with-group_by-and-summarise"><span class="header-section-number">7.5.4</span> Group-wise summaries with <code>group_by</code> and <code>summarise</code></h3>
<ul>
<li><p>These summary statistics are particular useful when applied to conditional subsets of a dataset</p></li>
<li><p><code>dplyr</code> allows such summary operations with a combination of <code>group_by</code> and <code>summarise</code></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">group_by</span>(material) <span class="sc">%&gt;%</span>  <span class="co"># group the tibble by the material column</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">min_age =</span> <span class="fu">min</span>(sample_age),   <span class="co"># a new column: min age for each group</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">median_age =</span> <span class="fu">median</span>(sample_age), <span class="co"># a new column: median age for each group</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">max_age =</span> <span class="fu">max</span>(sample_age)    <span class="co"># a new column: max age for each group</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>grouping can be applied across multiple columns</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">group_by</span>(material, sample_host) <span class="sc">%&gt;%</span> <span class="co"># group by material and host</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">summarise</span>(</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">n =</span> dplyr<span class="sc">::</span><span class="fu">n</span>(),  <span class="co"># a new column: number of samples for each group</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">.groups =</span> <span class="st">"drop"</span> <span class="co"># drop the grouping after this summary operation</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="sorting-and-slicing-tibbles-with-arrange-and-slice" class="level3" data-number="7.5.5">
<h3 data-number="7.5.5" class="anchored" data-anchor-id="sorting-and-slicing-tibbles-with-arrange-and-slice"><span class="header-section-number">7.5.5</span> Sorting and slicing tibbles with <code>arrange</code> and <code>slice</code></h3>
<ul>
<li><p><code>dplyr</code> allows to <code>arrange</code> tibbles by one or multiple columns</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(publication_year)        <span class="co"># sort by publication year</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(publication_year,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                        sample_age)              <span class="co"># ... and sample age</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(sample_age)) <span class="co"># sort descending on sample age</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Sorting also works within groups and can be paired with <code>slice</code> to extract extreme values per group</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">group_by</span>(publication_year) <span class="sc">%&gt;%</span>       <span class="co"># group by publication year</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">arrange</span>(dplyr<span class="sc">::</span><span class="fu">desc</span>(sample_age)) <span class="sc">%&gt;%</span> <span class="co"># sort by age within (!) groups</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">slice_head</span>(<span class="at">n =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span>                <span class="co"># keep the first two samples per group</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">ungroup</span>()                            <span class="co"># remove the still lingering grouping</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Slicing is also the relevant operation to take random samples from the observations in a tibble</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="at">n =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="exercise-2" class="level3" data-number="7.5.6">
<h3 data-number="7.5.6" class="anchored" data-anchor-id="exercise-2"><span class="header-section-number">7.5.6</span> Exercise 2</h3>
<ol type="1">
<li><p>Determine the number of cars with four <em>forward gears</em> (<code>gear</code>) in the <code>mtcars</code> dataset</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Determine the mean <em>1/4 mile time</em> (<code>qsec</code>) per <em>Number of cylinders</em> (<code>cyl</code>) group</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Identify the least efficient cars for both <em>transmission types</em> (<code>am</code>)</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
<section id="possible-solutions-2" class="level3" data-number="7.5.7">
<h3 data-number="7.5.7" class="anchored" data-anchor-id="possible-solutions-2"><span class="header-section-number">7.5.7</span> Possible solutions 2</h3>
<ol type="1">
<li><p>Determine the number of cars with four <em>forward gears</em> (<code>gear</code>) in the <code>mtcars</code> dataset</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(gear <span class="sc">==</span> <span class="dv">4</span>) <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Determine the mean <em>1/4 mile time</em> (<code>qsec</code>) per <em>Number of cylinders</em> (<code>cyl</code>) group</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">group_by</span>(cyl) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">qsec_mean =</span> <span class="fu">mean</span>(qsec))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Identify the least efficient cars for both <em>transmission types</em> (<code>am</code>)</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co">#mtcars3 &lt;- tibble::rownames_to_column(mtcars, var = "car") %&gt;% tibble::as_tibble()</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">group_by</span>(am) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">arrange</span>(mpg) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">slice_head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
</section>
<section id="transforming-and-manipulating-tibbles" class="level2" data-number="7.6">
<h2 data-number="7.6" class="anchored" data-anchor-id="transforming-and-manipulating-tibbles"><span class="header-section-number">7.6</span> Transforming and manipulating tibbles</h2>
<section id="renaming-and-reordering-columns-and-values-with-rename-relocate-and-recode" class="level3" data-number="7.6.1">
<h3 data-number="7.6.1" class="anchored" data-anchor-id="renaming-and-reordering-columns-and-values-with-rename-relocate-and-recode"><span class="header-section-number">7.6.1</span> Renaming and reordering columns and values with <code>rename</code>, <code>relocate</code> and <code>recode</code></h3>
<ul>
<li><p>Columns in tibbles can be renamed with <code>dplyr::rename</code> and reordered with <code>dplyr::relocate</code></p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">country =</span> geo_loc_name) <span class="co"># rename a column</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">relocate</span>(site_name, <span class="at">.before =</span> project_name) <span class="co"># reorder columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Values in columns can also be changed with <code>dplyr::recode</code></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>samples<span class="sc">$</span>sample_host <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">recode</span>(<span class="st">`</span><span class="at">Homo sapiens</span><span class="st">`</span> <span class="ot">=</span> <span class="st">"modern human"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>R supports explicitly ordinal data with <code>factor</code>s, which can be reordered as well</p></li>
<li><p><code>factor</code>s can be handeld more easily with the <code>forcats</code> package</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(samples) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> community_type)) <span class="co"># bars are alphabetically ordered</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>sa2 <span class="ot">&lt;-</span> samples</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>sa2<span class="sc">$</span>cto <span class="ot">&lt;-</span> forcats<span class="sc">::</span><span class="fu">fct_reorder</span>(sa2<span class="sc">$</span>community_type, sa2<span class="sc">$</span>community_type, length)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># fct_reorder: reorder the input factor by a summary statistic on an other vector</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sa2) <span class="sc">+</span> <span class="fu">geom_bar</span>(<span class="fu">aes</span>(<span class="at">x =</span> community_type)) <span class="co"># bars are ordered by size</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="adding-columns-to-tibbles-with-mutate-and-transmute" class="level3" data-number="7.6.2">
<h3 data-number="7.6.2" class="anchored" data-anchor-id="adding-columns-to-tibbles-with-mutate-and-transmute"><span class="header-section-number">7.6.2</span> Adding columns to tibbles with <code>mutate</code> and <code>transmute</code></h3>
<ul>
<li><p>A common application of data manipulation is adding derived columns. <code>dplyr</code> offers that with <code>mutate</code></p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">mutate</span>(                                               <span class="co"># add a column that</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">archive_summary =</span> <span class="fu">paste0</span>(archive, <span class="st">": "</span>, archive_accession) <span class="co"># combines two other</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%$%</span> archive_summary                                        <span class="co"># columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>dplyr::transmute</code> removes all columns but the newly created ones</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">transmute</span>(</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">sample_name =</span> <span class="fu">tolower</span>(sample_name), <span class="co"># overwrite this columns</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    publication_doi                     <span class="co"># select this column</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>tibble::add_column</code> behaves as <code>dplyr::mutate</code>, but gives more control over column position</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">add_column</span>(., <span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(.), <span class="at">.before =</span> <span class="st">"project_name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="conditional-operations-with-ifelse-and-case_when" class="level3" data-number="7.6.3">
<h3 data-number="7.6.3" class="anchored" data-anchor-id="conditional-operations-with-ifelse-and-case_when"><span class="header-section-number">7.6.3</span> Conditional operations with <code>ifelse</code> and <code>case_when</code></h3>
<ul>
<li><p><code>ifelse</code> allows to implement conditional <code>mutate</code> operations, that consider information from other columns, but that gets cumbersome easily</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">hemi =</span> <span class="fu">ifelse</span>(latitude <span class="sc">&gt;=</span> <span class="dv">0</span>, <span class="st">"North"</span>, <span class="st">"South"</span>)) <span class="sc">%$%</span> hemi</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="at">hemi =</span> <span class="fu">ifelse</span>(<span class="fu">is.na</span>(latitude), <span class="st">"unknown"</span>, <span class="fu">ifelse</span>(latitude <span class="sc">&gt;=</span> <span class="dv">0</span>, <span class="st">"North"</span>, <span class="st">"South"</span>))</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%$%</span> hemi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p><code>dplyr::case_when</code> is a much more readable solution for this application</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="at">hemi =</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    latitude <span class="sc">&gt;=</span> <span class="dv">0</span> <span class="sc">~</span> <span class="st">"North"</span>,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>    latitude <span class="sc">&lt;</span> <span class="dv">0</span>  <span class="sc">~</span> <span class="st">"South"</span>,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span>          <span class="sc">~</span> <span class="st">"unknown"</span> <span class="co"># TRUE catches all remaining cases</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%$%</span> hemi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
</section>
<section id="long-and-wide-data-formats" class="level3" data-number="7.6.4">
<h3 data-number="7.6.4" class="anchored" data-anchor-id="long-and-wide-data-formats"><span class="header-section-number">7.6.4</span> Long and wide data formats</h3>
<ul>
<li>For different applications or to simplify certain analysis or plotting operations data often has to be transformed from a <strong>wide</strong> to a <strong>long</strong> format or vice versa</li>
</ul>
<p><img src="assets/images/chapters/r-tidyverse/pivot_longer_wider.png" class="img-fluid"></p>
<ul>
<li>A table in <strong>wide</strong> format has N key columns and N value columns</li>
<li>A table in <strong>long</strong> format has N key columns, one descriptor column and one value column</li>
</ul>
</section>
<section id="a-wide-dataset" class="level3" data-number="7.6.5">
<h3 data-number="7.6.5" class="anchored" data-anchor-id="a-wide-dataset"><span class="header-section-number">7.6.5</span> A wide dataset</h3>
<div class="sourceCode" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>carsales <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tribble</span>(</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="sc">~</span>brand, <span class="sc">~</span><span class="st">`</span><span class="at">2014</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2015</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2016</span><span class="st">`</span>, <span class="sc">~</span><span class="st">`</span><span class="at">2017</span><span class="st">`</span>,</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"BMW"</span>,  <span class="dv">20</span>,      <span class="dv">25</span>,      <span class="dv">30</span>,      <span class="dv">45</span>,</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"VW"</span>,   <span class="dv">67</span>,      <span class="dv">40</span>,     <span class="dv">120</span>,      <span class="dv">55</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a>carsales</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Wide format becomes a problem, when the columns are semantically identical. This dataset is in wide format and we can not easily plot it</li>
<li>We generally prefer data in long format, although it is more verbose with more duplication. “Long” format data is more “tidy”</li>
</ul>
</section>
<section id="making-a-wide-dataset-long-with-pivot_longer" class="level3" data-number="7.6.6">
<h3 data-number="7.6.6" class="anchored" data-anchor-id="making-a-wide-dataset-long-with-pivot_longer"><span class="header-section-number">7.6.6</span> Making a wide dataset long with <code>pivot_longer</code></h3>
<div class="sourceCode" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>carsales_long <span class="ot">&lt;-</span> carsales <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> tidyselect<span class="sc">::</span><span class="fu">num_range</span>(<span class="st">""</span>, <span class="at">range =</span> <span class="dv">2014</span><span class="sc">:</span><span class="dv">2017</span>), <span class="co"># set of columns to transform</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_to =</span> <span class="st">"year"</span>,            <span class="co"># the name of the descriptor column we want</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_transform =</span> as.integer, <span class="co"># a transformation function to apply to the names</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_to =</span> <span class="st">"sales"</span>           <span class="co"># the name of the value column we want</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>carsales_long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="making-a-long-dataset-wide-with-pivot_wider" class="level3" data-number="7.6.7">
<h3 data-number="7.6.7" class="anchored" data-anchor-id="making-a-long-dataset-wide-with-pivot_wider"><span class="header-section-number">7.6.7</span> Making a long dataset wide with <code>pivot_wider</code></h3>
<div class="sourceCode" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>carsales_wide <span class="ot">&lt;-</span> carsales_long <span class="sc">%&gt;%</span> tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">id_cols =</span> <span class="st">"brand"</span>,  <span class="co"># the set of id columns that should not be changed</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">names_from =</span> year,  <span class="co"># the descriptor column with the names of the new columns</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">values_from =</span> sales <span class="co"># the value column from which the values should be extracted</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>carsales_wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Applications of wide datasets are adjacency matrices to represent graphs, covariance matrices or other pairwise statistics</li>
<li>When data gets big, then wide formats can be significantly more efficient (e.g.&nbsp;for spatial data)</li>
</ul>
</section>
<section id="exercise-3" class="level3" data-number="7.6.8">
<h3 data-number="7.6.8" class="anchored" data-anchor-id="exercise-3"><span class="header-section-number">7.6.8</span> Exercise 3</h3>
<ol type="1">
<li><p>Move the column <code>gear</code> to the first position of the mtcars dataset</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Make a new dataset <code>mtcars2</code> with the column <code>mpg</code> and an additional column <code>am_v</code>, which encodes the <em>transmission type</em> (<code>am</code>) as either <code>"manual"</code> or <code>"automatic"</code></p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Count the number of cars per <em>transmission type</em> (<code>am_v</code>) and <em>number of gears</em> (<code>gear</code>). Then transform the result to a wide format, with one column per <em>transmission type</em>.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
<section id="possible-solutions-3" class="level3" data-number="7.6.9">
<h3 data-number="7.6.9" class="anchored" data-anchor-id="possible-solutions-3"><span class="header-section-number">7.6.9</span> Possible solutions 3</h3>
<ol type="1">
<li><p>Move the column <code>gear</code> to the first position of the mtcars dataset</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>mtcars <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">relocate</span>(gear, <span class="at">.before =</span> mpg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Make a new dataset <code>mtcars2</code> with the column <code>gear</code> and an additional column <code>am_v</code>, which encodes the <em>transmission type</em> (<code>am</code>) as either <code>"manual"</code> or <code>"automatic"</code></p>
<div class="sourceCode" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="ot">&lt;-</span> mtcars <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>gear, <span class="at">am_v =</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(am <span class="sc">==</span> <span class="dv">0</span> <span class="sc">~</span> <span class="st">"automatic"</span>, am <span class="sc">==</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"manual"</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Count the number of cars in <code>mtcars2</code> per <em>transmission type</em> (<code>am_v</code>) and <em>number of gears</em> (<code>gear</code>). Then transform the result to a wide format, with one column per <em>transmission type</em>.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>mtcars2 <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">group_by</span>(am_v, gear) <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">tally</span>() <span class="sc">%&gt;%</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> am_v, <span class="at">values_from =</span> n)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
</section>
<section id="combining-tibbles-with-join-operations" class="level2" data-number="7.7">
<h2 data-number="7.7" class="anchored" data-anchor-id="combining-tibbles-with-join-operations"><span class="header-section-number">7.7</span> Combining tibbles with join operations</h2>
<section id="types-of-joins" class="level3" data-number="7.7.1">
<h3 data-number="7.7.1" class="anchored" data-anchor-id="types-of-joins"><span class="header-section-number">7.7.1</span> Types of joins</h3>
<p>Joins combine two datasets x and y based on key columns</p>
<ul>
<li>Mutating joins add columns from one dataset to the other
<ul>
<li>Left join: Take observations from x and add fitting information from y</li>
<li>Right join: Take observations from y and add fitting information from x</li>
<li>Inner join: Join the overlapping observations from x and y</li>
<li>Full join: Join all observations from x and y, even if information is missing</li>
</ul></li>
<li>Filtering joins remove observations from x based on their presence in y
<ul>
<li>Semi join: Keep every observation in x that is in y</li>
<li>Anti join: Keep every observation in x that is not in y</li>
</ul></li>
</ul>
</section>
<section id="a-second-dataset" class="level3" data-number="7.7.2">
<h3 data-number="7.7.2" class="anchored" data-anchor-id="a-second-dataset"><span class="header-section-number">7.7.2</span> A second dataset</h3>
<div class="sourceCode" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>library_table_path <span class="ot">&lt;-</span> <span class="st">"/vol/volume/r-tidyverse/ancientmetagenome-hostassociated_libraries.tsv"</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>library_table_url <span class="ot">&lt;-</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="st">"https://raw.githubusercontent.com/SPAAM-community/AncientMetagenomeDir/b187df6ebd23dfeb42935fd5020cb615ead3f164/</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a><span class="st">ancientmetagenome-hostassociated/libraries/ancientmetagenome-hostassociated_libraries.tsv"</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>libraries <span class="ot">&lt;-</span> readr<span class="sc">::</span><span class="fu">read_tsv</span>(library_table_url)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(libraries, <span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="meaningful-subsets" class="level3" data-number="7.7.3">
<h3 data-number="7.7.3" class="anchored" data-anchor-id="meaningful-subsets"><span class="header-section-number">7.7.3</span> Meaningful subsets</h3>
<div class="sourceCode" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>samsub <span class="ot">&lt;-</span> samples <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(project_name, sample_name, sample_age)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>libsub <span class="ot">&lt;-</span> libraries <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(project_name, sample_name, library_name, read_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(samsub, <span class="at">n =</span> <span class="dv">3</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(libsub, <span class="at">n =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="left-join" class="level3" data-number="7.7.4">
<h3 data-number="7.7.4" class="anchored" data-anchor-id="left-join"><span class="header-section-number">7.7.4</span> Left join</h3>
<p>Take observations from x and add fitting information from y</p>
<p><img src="assets/images/chapters/r-tidyverse/left_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>left <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>) <span class="co"># the key columns by which to join</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(left, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Left joins are the most common join operation: Add information from another dataset</li>
</ul>
</section>
<section id="right-join" class="level3" data-number="7.7.5">
<h3 data-number="7.7.5" class="anchored" data-anchor-id="right-join"><span class="header-section-number">7.7.5</span> Right join</h3>
<p>Take observations from y and add fitting information from x</p>
<p><img src="assets/images/chapters/r-tidyverse/right_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>right <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">right_join</span>(</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(right, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Right joins are almost identical to left joins – only x and y have reversed roles</li>
</ul>
</section>
<section id="inner-join" class="level3" data-number="7.7.6">
<h3 data-number="7.7.6" class="anchored" data-anchor-id="inner-join"><span class="header-section-number">7.7.6</span> Inner join</h3>
<p>Join the overlapping observations from x and y</p>
<p><img src="assets/images/chapters/r-tidyverse/inner_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>inner <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">inner_join</span>(</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>)</span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(inner, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Inner joins are a fast and easy way to check, to which degree two dataset overlap</li>
</ul>
</section>
<section id="full-join" class="level3" data-number="7.7.7">
<h3 data-number="7.7.7" class="anchored" data-anchor-id="full-join"><span class="header-section-number">7.7.7</span> Full join</h3>
<p>Join all observations from x and y, even if information is missing</p>
<p><img src="assets/images/chapters/r-tidyverse/full_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>full <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">full_join</span>(</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(full, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Full joins allow to preserve every bit of information</li>
</ul>
</section>
<section id="semi-join" class="level3" data-number="7.7.8">
<h3 data-number="7.7.8" class="anchored" data-anchor-id="semi-join"><span class="header-section-number">7.7.8</span> Semi join</h3>
<p>Keep every observation in x that is in y</p>
<p><img src="assets/images/chapters/r-tidyverse/semi_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>semi <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">semi_join</span>(</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>)</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(semi, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Semi joins are underused operations to filter datasets</li>
</ul>
</section>
<section id="anti-join" class="level3" data-number="7.7.9">
<h3 data-number="7.7.9" class="anchored" data-anchor-id="anti-join"><span class="header-section-number">7.7.9</span> Anti join</h3>
<p>Keep every observation in x that is not in y</p>
<p><img src="assets/images/chapters/r-tidyverse/anti_join.png" class="img-fluid"></p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>anti <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">anti_join</span>(</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> samsub,                           <span class="co"># 1060 observations</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> libsub,                           <span class="co"># 1657 observations</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"project_name"</span>, <span class="st">"sample_name"</span>)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(anti, <span class="at">n =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>Anti joins allow to quickly specify incomplete datasets and missing information</li>
</ul>
</section>
<section id="exercise-4" class="level3" data-number="7.7.10">
<h3 data-number="7.7.10" class="anchored" data-anchor-id="exercise-4"><span class="header-section-number">7.7.10</span> Exercise 4</h3>
<p>Consider the following additional dataset:</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>gear_opinions <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">gear =</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">5</span>), <span class="at">opinion =</span> <span class="fu">c</span>(<span class="st">"boring"</span>, <span class="st">"wow"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ol type="1">
<li><p>Add my opinions about gears to the <code>mtcars</code> dataset</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Remove all cars from the dataset for which I don’t have an opinion</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>
</section>
<section id="possible-solutions-4" class="level3" data-number="7.7.11">
<h3 data-number="7.7.11" class="anchored" data-anchor-id="possible-solutions-4"><span class="header-section-number">7.7.11</span> Possible Solutions 4</h3>
<ol type="1">
<li><p>Add my opinions about gears to the <code>mtcars</code> dataset</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">left_join</span>(mtcars, gear_opinions, <span class="at">by =</span> <span class="st">"gear"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Remove all cars from the dataset for which I don’t have an opinion</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">anti_join</span>(mtcars, gear_opinions, <span class="at">by =</span> <span class="st">"gear"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ol>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./bare-bones-bash.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./python-pandas.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">© 2023 SPAAM Community &amp; Authors with ❤️. Available under <a href="http://creativecommons.org/licenses/by-sa/4.0/">CC-BY 4.0</a>. Source material <a href="https://github.com/SPAAM-community/intro-to-ancient-metagenomics-book">here</a>.</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>