<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Maxime Borry">

<title>Introduction to Ancient Metagenomics - 10&nbsp; Taxonomic Profiling, OTU Tables and Visualisation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./functional-profiling.html" rel="next">
<link href="./section-ancient-metagenomics.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">


<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section-ancient-metagenomics.html">Ancient Metagenomics</a></li><li class="breadcrumb-item"><a href="./taxonomic-profiling.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Introduction to Ancient Metagenomics</a> 
        <div class="sidebar-tools-main tools-wide">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Introduction-to-Ancient-Metagenomics.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./authors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Authors</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./acknowledgements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Acknowledgements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./before-you-start.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Before you Start</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Theory</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ngs-sequencing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction to NGS Sequencing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-ancient-dna.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction to Ancient DNA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Introduction to Metagenomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-microbial-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Introduction to Microbial Genomics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-evolutionary-biology.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Evolutionary Biology</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-useful-skills.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful Skills</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bare-bones-bash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-r.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Introduction to R and the Tidyverse</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-python.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Introduction to Python and Pandas</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./introduction-to-git.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Introduction to Git(Hub)</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./taxonomic-profiling.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional-profiling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./denovo-assembly.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Introduction to <em>de novo</em> Genome Assembly</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-genomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./genome-mapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Genome Mapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./phylogenomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Introduction to Phylogenomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./section-ancient-metagenomic-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Ancient Metagenomic Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./accessing-ancientmetagenomic-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Introduction to AncientMetagenomeDir</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ancient-metagenomic-pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Ancient Metagenomic Pipelines</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./appendices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendicies</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Tools</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#lecture" id="toc-lecture" class="nav-link active" data-scroll-target="#lecture"><span class="header-section-number">10.1</span> Lecture</a></li>
  <li><a href="#download-and-subsample" id="toc-download-and-subsample" class="nav-link" data-scroll-target="#download-and-subsample"><span class="header-section-number">10.2</span> Download and Subsample</a>
  <ul class="collapse">
  <li><a href="#subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial" id="toc-subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial" class="nav-link" data-scroll-target="#subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial"><span class="header-section-number">10.2.1</span> Subsampling the sequencing files to make the analysis quicker for this tutorial</a></li>
  </ul></li>
  <li><a href="#hands-on-introduction-to-ancient-microbiome-analysis" id="toc-hands-on-introduction-to-ancient-microbiome-analysis" class="nav-link" data-scroll-target="#hands-on-introduction-to-ancient-microbiome-analysis"><span class="header-section-number">10.3</span> Hands on introduction to ancient microbiome analysis</a>
  <ul class="collapse">
  <li><a href="#quick-intro-to-jupyter-notebooks" id="toc-quick-intro-to-jupyter-notebooks" class="nav-link" data-scroll-target="#quick-intro-to-jupyter-notebooks"><span class="header-section-number">10.3.1</span> 0. Quick intro to Jupyter Notebooks</a></li>
  <li><a href="#data-pre-processing" id="toc-data-pre-processing" class="nav-link" data-scroll-target="#data-pre-processing"><span class="header-section-number">10.3.2</span> 1. Data pre-processing</a></li>
  <li><a href="#adapter-sequence-trimming-and-low-quality-bases-trimming" id="toc-adapter-sequence-trimming-and-low-quality-bases-trimming" class="nav-link" data-scroll-target="#adapter-sequence-trimming-and-low-quality-bases-trimming"><span class="header-section-number">10.3.3</span> 2. Adapter sequence trimming and low-quality bases trimming</a></li>
  <li><a href="#visualizing-the-taxonomic-profile" id="toc-visualizing-the-taxonomic-profile" class="nav-link" data-scroll-target="#visualizing-the-taxonomic-profile"><span class="header-section-number">10.3.4</span> 4. Visualizing the taxonomic profile</a></li>
  <li><a href="#getting-modern-reference-data" id="toc-getting-modern-reference-data" class="nav-link" data-scroll-target="#getting-modern-reference-data"><span class="header-section-number">10.3.5</span> 5. Getting modern reference data</a></li>
  <li><a href="#bringing-together-ancient-and-modern-data" id="toc-bringing-together-ancient-and-modern-data" class="nav-link" data-scroll-target="#bringing-together-ancient-and-modern-data"><span class="header-section-number">10.3.6</span> 6. Bringing together ancient and modern data</a></li>
  <li><a href="#comparing-ancient-and-modern-samples" id="toc-comparing-ancient-and-modern-samples" class="nav-link" data-scroll-target="#comparing-ancient-and-modern-samples"><span class="header-section-number">10.3.7</span> 7. Comparing ancient and modern samples</a></li>
  <li><a href="#additional-steps" id="toc-additional-steps" class="nav-link" data-scroll-target="#additional-steps"><span class="header-section-number">10.3.8</span> 8. Additional steps</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Taxonomic Profiling, OTU Tables and Visualisation</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Maxime Borry </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>For this chapter’s exercises, if not already performed, you will need to create the <a href="./before-you-start.html#creating-a-conda-environment">conda environment</a> from the <code>yml</code> file in the following <a href="https://doi.org/10.5281/zenodo.6983165">archive</a>, and activate the environment:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate r-python</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<section id="lecture" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="lecture"><span class="header-section-number">10.1</span> Lecture</h2>
<iframe src="https://docs.google.com/viewer?url=https://raw.githubusercontent.com/SPAAM-community/wss-summer-school/main/docs/assets/slides/2022/3c-intro-to-taxprofiling/SPAAM%20Summer%20School%202022%20-%203C%20-%20Intro%20to%20microbial%20ecology%20for%20ancient%20DNA.pdf?embedded=true" width="100%" height="400px">
</iframe>
<p>PDF version of these slides can be downloaded from <a href="https://github.com/SPAAM-community/wss-summer-school/raw/main/docs/assets/slides/2022/3c-intro-to-taxprofiling/SPAAM%20Summer%20School%202022%20-%203C%20-%20Intro%20to%20microbial%20ecology%20for%20ancient%20DNA.pdf">here</a>.</p>
<p>This session is run using a Jupyter notebook. This can be found <a href="https://github.com/maxibor/microbiome_tutorial">here</a>. However, it will already be installed on compute nodes during the summer school.</p>
<iframe width="100%" height="400px" src="https://www.youtube.com/embed/zp3fmPAZnBo" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="">
</iframe>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>We highly recommend viewing this walkthrough via the Jupyter notebook above! The output of commands on the website for this walkthrough are displayed in their own code blocks - be wary of what you copy-paste!</p>
</div>
</div>
</section>
<section id="download-and-subsample" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="download-and-subsample"><span class="header-section-number">10.2</span> Download and Subsample</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> glob</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For this tutorial, we will be using the ERR5766177 library from the sample 2612 published by <a href="https://doi.org/10.1016/j.cub.2021.09.031">Maixner et al.&nbsp;2021</a></p>
<p><img src="assets/images/chapters/taxonomic-profiling/1-s2.0-S0960982221012719-fx1.jpg" class="img-fluid"></p>
<p><img src="assets/images/chapters/taxonomic-profiling/1-s2.0-S0960982221012719-gr1.jpg" class="img-fluid">`</p>
<section id="subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial" class="level3" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="subsampling-the-sequencing-files-to-make-the-analysis-quicker-for-this-tutorial"><span class="header-section-number">10.2.1</span> Subsampling the sequencing files to make the analysis quicker for this tutorial</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> subsample(filename, outdir, depth<span class="op">=</span><span class="dv">1000000</span>):</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    basename <span class="op">=</span> Path(filename).stem</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    cmd <span class="op">=</span> <span class="ss">f"seqtk sample -s42 </span><span class="sc">{</span>filename<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>depth<span class="sc">}</span><span class="ss"> &gt; </span><span class="sc">{</span>outdir<span class="sc">}</span><span class="ss">/</span><span class="sc">{</span>basename<span class="sc">}</span><span class="ss">_subsample_</span><span class="sc">{</span>depth<span class="sc">}</span><span class="ss">.fastq"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(cmd)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    subprocess.check_output(cmd, shell<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> f <span class="kw">in</span> glob.glob(<span class="st">"../data/raw/*"</span>):</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    outdir <span class="op">=</span> <span class="st">"../data/subsampled"</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    subsample(f, outdir)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>seqtk sample -s42 ../data/raw/ERR5766177_PE.mapped.hostremoved.fwd.fq.gz 1000000 &gt;
../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq
seqtk sample -s42 ../data/raw/ERR5766177_PE.mapped.hostremoved.rev.fq.gz 1000000 &gt;
../data/subsampled/ERR5766177_PE.mapped.hostremoved.rev.fq_subsample_1000000.fastq</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> gzip <span class="op">-</span>f ..<span class="op">/</span>data<span class="op">/</span>subsampled<span class="op">/*</span>.fastq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="hands-on-introduction-to-ancient-microbiome-analysis" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="hands-on-introduction-to-ancient-microbiome-analysis"><span class="header-section-number">10.3</span> Hands on introduction to ancient microbiome analysis</h2>
<p>Author: Maxime Borry<br>
Date: 12/08/2021</p>
<p>In this tutorial, we’re going to go through the steps necessary to:</p>
<ul>
<li>generate a taxonomic profile table with <a href="https://github.com/biobakery/MetaPhlAn">metaphlan v3.0</a></li>
<li>have a look at metaphlan results with <a href="https://github.com/fbreitwieser/pavian">Pavian</a> and generate a <a href="https://github.com/marbl/Krona/wiki">Krona plot</a></li>
<li>Compare the diversity of our samples vs the diversity of modern human gut samples</li>
<li>Compare the composition of our samples vs modern gut samples, and see where they fit in a lower dimensional space</li>
</ul>
<section id="quick-intro-to-jupyter-notebooks" class="level3" data-number="10.3.1">
<h3 data-number="10.3.1" class="anchored" data-anchor-id="quick-intro-to-jupyter-notebooks"><span class="header-section-number">10.3.1</span> 0. Quick intro to Jupyter Notebooks</h3>
<p>This a markdown cell</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"This is a python cell"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>This is a python cell</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> echo <span class="st">"This is how to run a single line bash command"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>This is how to run a single line bash command</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">%%bash</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"This how to run a multi"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="bu">echo</span> <span class="st">"lines bash command"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>This how to run a multi
lines bash command</code></pre>
</section>
<section id="data-pre-processing" class="level3" data-number="10.3.2">
<h3 data-number="10.3.2" class="anchored" data-anchor-id="data-pre-processing"><span class="header-section-number">10.3.2</span> 1. Data pre-processing</h3>
<p>Before starting to analyze our data, we will need to pre-process them to remove reads mapping to the host genome, here, <em>Homo sapiens</em></p>
<p>To do so, I’ve used <a href="https://github.com/nf-core/eager">nf-core/eager</a></p>
<p>I’ve already pre-processed the data, and the resulting cleaned files are available in the <a href="data/eager_cleaned">data/eager_cleaned</a>, but the basic eager command to do so is:</p>
<pre><code>nextflow run nf-core/eager -profile &lt;docker/singularity/podman/conda/institute&gt; --input '*_R{1,2}.fastq.gz' \
--fasta 'human_genome.fasta' --hostremoval_input_fastq</code></pre>
</section>
<section id="adapter-sequence-trimming-and-low-quality-bases-trimming" class="level3" data-number="10.3.3">
<h3 data-number="10.3.3" class="anchored" data-anchor-id="adapter-sequence-trimming-and-low-quality-bases-trimming"><span class="header-section-number">10.3.3</span> 2. Adapter sequence trimming and low-quality bases trimming</h3>
<p>Sequencing adapters are small DNA sequences adding prior to DNA sequencing to allow the DNA fragments to attach to the sequencing flow cells. Because these adapters could interfere with downtstream analyses, we need to remove them before proceeding any further. Furthermore, because the quality of the sequencing is not always optimal, we need to remove bases of lower sequencing quality to might lead to spurious results in downstream analyses.</p>
<p>To perform both of these tasks, we’ll use the program <a href="https://github.com/OpenGene/fastp">fastp</a>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> fastp <span class="op">-</span>h</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>option needs value: --html
usage: fastp [options] ...
options:
  -i, --in1                            read1 input file name (string [=])
  -o, --out1                           read1 output file name (string [=])
  -I, --in2                            read2 input file name (string [=])
  -O, --out2                           read2 output file name (string [=])
      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1.
                                       Default is to discard it. (string [=])
      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2.
                                       If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be
                                       written to this same file. (string [=])
      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched
                                       base. (string [=])
      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])
  -m, --merge                          for paired-end input, merge each pair of reads into a single read if they are
                                       overlapped. The merged reads will be written
                                       to the file given by --merged_out, the unmerged reads will be written to the
                                       files specified by --out1 and --out2. The merging mode is disabled by default.
      --merged_out                     in the merging mode, specify the file name to store merged output, or specify
                                       --stdout to stream the merged output (string [=])
      --include_unmerged               in the merging mode, write the unmerged or unpaired reads to the file specified
                                       by --merge. Disabled by default.
  -6, --phred64                        indicate the input is using phred64 scoring (it'll be converted to phred33,
                                       so the output will still be phred33)
  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])
      --stdin                          input from STDIN. If the STDIN is interleaved paired-end FASTQ, please also add --interleaved_in.
      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved
                                       FASTQ output for paired-end output. Disabled by default.
      --interleaved_in                 indicate that &lt;in1&gt; is an interleaved FASTQ which contains both read1 and read2.
                                       Disabled by default.
      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])
      --dont_overwrite                 don't overwrite existing files. Overwritting is allowed by default.
      --fix_mgi_id                     the MGI FASTQ ID format is not compatible with many BAM operation tools, enable this option to fix it.
  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).
  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled
  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected.
                                       For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])
      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped.
                                       If not specified, it will be the same as &lt;adapter_sequence&gt; (string [=auto])
      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])
      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this
                                    option to enable it for PE data.
  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])
  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])
  -b, --max_len1                       if read1 is longer than max_len1, then trim read1 at its tail to make it as
                                       long as max_len1. Default 0 means no limitation (int [=0])
  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])
  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])
  -B, --max_len2                       if read2 is longer than max_len2, then trim read2 at its tail to make it as long as max_len2.
                                       Default 0 means no limitation. If it's not specified, it will follow read1's settings (int [=0])
  -D, --dedup                          enable deduplication to drop the duplicated reads/pairs
      --dup_calc_accuracy              accuracy level to calculate duplication (1~6), higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G).
                                       Default 1 for no-dedup mode, and 3 for dedup mode. (int [=0])
      --dont_eval_duplication          don't evaluate duplication rate to save time and use less memory.
  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data
      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])
  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data
  -x, --trim_poly_x                    enable polyX trimming in 3' ends.
      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])
  -5, --cut_front                      move a sliding window from front (5') to tail, drop the bases in the window if
                                       its mean quality &lt; threshold, stop otherwise.
  -3, --cut_tail                       move a sliding window from tail (3') to front, drop the bases in the window if
                                       its mean quality &lt; threshold, stop otherwise.
  -r, --cut_right                      move a sliding window from front to tail, if meet one window with mean quality
                                       &lt; threshold, drop the bases in the window and the right part, and then stop.
  -W, --cut_window_size                the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])
  -M, --cut_mean_quality               the mean quality requirement option shared by cut_front, cut_tail or cut_sliding.
                                       Range: 1~36 default: 20 (Q20) (int [=20])
      --cut_front_window_size          the window size option of cut_front, default to cut_window_size if not specified (int [=4])
      --cut_front_mean_quality         the mean quality requirement option for cut_front, default to cut_mean_quality if not specified (int [=20])
      --cut_tail_window_size           the window size option of cut_tail, default to cut_window_size if not specified (int [=4])
      --cut_tail_mean_quality          the mean quality requirement option for cut_tail, default to cut_mean_quality if not specified (int [=20])
      --cut_right_window_size          the window size option of cut_right, default to cut_window_size if not specified (int [=4])
      --cut_right_mean_quality         the mean quality requirement option for cut_right, default to cut_mean_quality if not specified (int [=20])
  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled
  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality &gt;=Q15 is qualified. (int [=15])
  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])
  -n, --n_base_limit                   if one read's number of N base is &gt;n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])
  -e, --average_qual                   if one read's average quality score &lt;avg_qual, then this read/pair is discarded.
                                       Default 0 means no requirement (int [=0])
  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled
  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])
      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])
  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base
                                       that is different from its next base (base[i] != base[i+1]).
  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])
      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])
      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])
      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])
  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled
      --overlap_len_require            the minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge,
                                       adapter trimming and correction. 30 by default. (int [=30])
      --overlap_diff_limit             the maximum number of mismatched bases to detect overlapped region of PE reads.
                                       This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default. (int [=5])
      --overlap_diff_percent_limit     the maximum percentage of mismatched bases to detect overlapped region of PE reads.
                                       This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%. (int [=20])
  -U, --umi                            enable unique molecular identifier (UMI) preprocessing
      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])
      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])
      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e.
                                       prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])
      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])
  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.
  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation
                                       analysis (1~10000), smaller is slower, default is 20. (int [=20])
  -j, --json                           the json format report file name (string [=fastp.json])
  -h, --html                           the html format report file name (string [=fastp.html])
  -R, --report_title                   should be quoted with ' or ", default is "fastp report" (string [=fastp report])
  -w, --thread                         worker thread number, default is 3 (int [=3])
  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix
                                       will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])
  -S, --split_by_lines                 split output by limiting lines of each file with this option(&gt;=1000), a sequential number prefix will be
                                       added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])
  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as
                                       0001.xxx, 0 to disable padding (int [=4])
      --cut_by_quality5                DEPRECATED, use --cut_front instead.
      --cut_by_quality3                DEPRECATED, use --cut_tail instead.
      --cut_by_quality_aggressive      DEPRECATED, use --cut_right instead.
      --discard_unmerged               DEPRECATED, no effect now, see the introduction for merging.
  -?, --help                           print this message</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">%%bash</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="ex">fastp</span> <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--in1</span> ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz <span class="dt">\</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--in2</span> ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz <span class="dt">\</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--merge</span> <span class="dt">\</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--merged_out</span> ../results/fastp/ERR5766177.merged.fastq.gz <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--include_unmerged</span> <span class="dt">\</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">--dedup</span> <span class="dt">\</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">--json</span> ../results/fastp/ERR5766177.fastp.json <span class="dt">\</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">--html</span> ../results/fastp/ERR5766177.fastp.html <span class="dt">\</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Read1 before filtering:
total reads: 1000000
total bases: 101000000
Q20 bases: 99440729(98.4562%)
Q30 bases: 94683150(93.7457%)

Read2 before filtering:
total reads: 1000000
total bases: 101000000
Q20 bases: 99440729(98.4562%)
Q30 bases: 94683150(93.7457%)

Merged and filtered:
total reads: 1994070
total bases: 201397311
Q20 bases: 198330392(98.4772%)
Q30 bases: 188843169(93.7665%)

Filtering result:
reads passed filter: 1999252
reads failed due to low quality: 728
reads failed due to too many N: 20
reads failed due to too short: 0
reads with adapter trimmed: 282
bases trimmed due to adapters: 18654
reads corrected by overlap analysis: 0
bases corrected by overlap analysis: 0

Duplication rate: 0.2479%

Insert size peak (evaluated by paired-end reads): 31

Read pairs merged: 228
% of original read pairs: 0.0228%
% in reads after filtering: 0.0114339%


JSON report: ../results/fastp/ERR5766177.fastp.json
HTML report: ../results/fastp/ERR5766177.fastp.html

fastp --in1 ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz \
--in2 ../data/subsampled/ERR5766177_PE.mapped.hostremoved.fwd.fq_subsample_1000000.fastq.gz --merge \
--merged_out ../results/fastp/ERR5766177.merged.fastq.gz --include_unmerged --dedup \
--json ../results/fastp/ERR5766177.fastp.json --html ../results/fastp/ERR5766177.fastp.html
fastp v0.23.2, time used: 11 seconds</code></pre>
<section id="taxonomic-profiling-with-metaphlan" class="level5" data-number="10.3.3.0.1">
<h5 data-number="10.3.3.0.1" class="anchored" data-anchor-id="taxonomic-profiling-with-metaphlan"><span class="header-section-number">10.3.3.0.1</span> 3. Taxonomic profiling with Metaphlan</h5>
<div class="sourceCode" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> metaphlan  <span class="op">--</span><span class="bu">help</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>usage: metaphlan --input_type {fastq,fasta,bowtie2out,sam} [--force]
                 [--bowtie2db METAPHLAN_BOWTIE2_DB] [-x INDEX]
                 [--bt2_ps BowTie2 presets] [--bowtie2_exe BOWTIE2_EXE]
                 [--bowtie2_build BOWTIE2_BUILD] [--bowtie2out FILE_NAME]
                 [--min_mapq_val MIN_MAPQ_VAL] [--no_map] [--tmp_dir]
                 [--tax_lev TAXONOMIC_LEVEL] [--min_cu_len]
                 [--min_alignment_len] [--add_viruses] [--ignore_eukaryotes]
                 [--ignore_bacteria] [--ignore_archaea] [--stat_q]
                 [--perc_nonzero] [--ignore_markers IGNORE_MARKERS]
                 [--avoid_disqm] [--stat] [-t ANALYSIS TYPE]
                 [--nreads NUMBER_OF_READS] [--pres_th PRESENCE_THRESHOLD]
                 [--clade] [--min_ab] [-o output file] [--sample_id_key name]
                 [--use_group_representative] [--sample_id value]
                 [-s sam_output_file] [--legacy-output] [--CAMI_format_output]
                 [--unknown_estimation] [--biom biom_output] [--mdelim mdelim]
                 [--nproc N] [--install] [--force_download]
                 [--read_min_len READ_MIN_LEN] [-v] [-h]
                 [INPUT_FILE] [OUTPUT_FILE]

DESCRIPTION
 MetaPhlAn version 3.1.0 (25 Jul 2022):
 METAgenomic PHyLogenetic ANalysis for metagenomic taxonomic profiling.

AUTHORS: Francesco Beghini (francesco.beghini@unitn.it),Nicola Segata (nicola.segata@unitn.it), Duy Tin Truong,
Francesco Asnicar (f.asnicar@unitn.it), Aitor Blanco Miguez (aitor.blancomiguez@unitn.it)

COMMON COMMANDS

 We assume here that MetaPhlAn is installed using the several options available (pip, conda, PyPi)
 Also BowTie2 should be in the system path with execution and read permissions, and Perl should be installed)

========== MetaPhlAn clade-abundance estimation =================

The basic usage of MetaPhlAn consists in the identification of the clades (from phyla to species )
present in the metagenome obtained from a microbiome sample and their
relative abundance. This correspond to the default analysis type (-t rel_ab).

*  Profiling a metagenome from raw reads:
$ metaphlan metagenome.fastq --input_type fastq -o profiled_metagenome.txt

*  You can take advantage of multiple CPUs and save the intermediate BowTie2 output for re-running
   MetaPhlAn extremely quickly:
$ metaphlan metagenome.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq -o profiled_metagenome.txt

*  If you already mapped your metagenome against the marker DB (using a previous MetaPhlAn run), you
   can obtain the results in few seconds by using the previously saved --bowtie2out file and
   specifying the input (--input_type bowtie2out):
$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out -o profiled_metagenome.txt

*  bowtie2out files generated with MetaPhlAn versions below 3 are not compatibile.
   Starting from MetaPhlAn 3.0, the BowTie2 ouput now includes the size of the profiled metagenome and the average read length.
   If you want to re-run MetaPhlAn using these file you should provide the metagenome size via --nreads:
$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out --nreads 520000 -o profiled_metagenome.txt

*  You can also provide an externally BowTie2-mapped SAM if you specify this format with
   --input_type. Two steps: first apply BowTie2 and then feed MetaPhlAn with the obtained sam:
$ bowtie2 --sam-no-hd --sam-no-sq --no-unal --very-sensitive -S metagenome.sam -x \
  ${mpa_dir}/metaphlan_databases/mpa_v30_CHOCOPhlAn_201901 -U metagenome.fastq
$ metaphlan metagenome.sam --input_type sam -o profiled_metagenome.txt

*  We can also natively handle paired-end metagenomes, and, more generally, metagenomes stored in
  multiple files (but you need to specify the --bowtie2out parameter):
$ metaphlan metagenome_1.fastq,metagenome_2.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq

-------------------------------------------------------------------


========== Marker level analysis ============================

MetaPhlAn introduces the capability of characterizing organisms at the strain level using non
aggregated marker information. Such capability comes with several slightly different flavours and
are a way to perform strain tracking and comparison across multiple samples.
Usually, MetaPhlAn is first ran with the default -t to profile the species present in
the community, and then a strain-level profiling can be performed to zoom-in into specific species
of interest. This operation can be performed quickly as it exploits the --bowtie2out intermediate
file saved during the execution of the default analysis type.

*  The following command will output the abundance of each marker with a RPK (reads per kilo-base)
   higher 0.0. (we are assuming that metagenome_outfmt.bz2 has been generated before as
   shown above).
$ metaphlan -t marker_ab_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt
   The obtained RPK can be optionally normalized by the total number of reads in the metagenome
   to guarantee fair comparisons of abundances across samples. The number of reads in the metagenome
   needs to be passed with the '--nreads' argument

*  The list of markers present in the sample can be obtained with '-t marker_pres_table'
$ metaphlan -t marker_pres_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt
   The --pres_th argument (default 1.0) set the minimum RPK value to consider a marker present

*  The list '-t clade_profiles' analysis type reports the same information of '-t marker_ab_table'
   but the markers are reported on a clade-by-clade basis.
$ metaphlan -t clade_profiles metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt

*  Finally, to obtain all markers present for a specific clade and all its subclades, the
   '-t clade_specific_strain_tracker' should be used. For example, the following command
   is reporting the presence/absence of the markers for the B. fragilis species and its strains
   the optional argument --min_ab specifies the minimum clade abundance for reporting the markers

$ metaphlan -t clade_specific_strain_tracker --clade s__Bacteroides_fragilis metagenome_outfmt.bz2 --input_typ
  bowtie2out -o marker_abundance_table.txt

-------------------------------------------------------------------

positional arguments:
  INPUT_FILE            the input file can be:
                        * a fastq file containing metagenomic reads
                        OR
                        * a BowTie2 produced SAM file.
                        OR
                        * an intermediary mapping file of the metagenome generated by a previous MetaPhlAn run
                        If the input file is missing, the script assumes that the input is provided using the standard
                        input, or named pipes.
                        IMPORTANT: the type of input needs to be specified with --input_type
  OUTPUT_FILE           the tab-separated output file of the predicted taxon relative abundances
                        [stdout if not present]

Required arguments:
  --input_type {fastq,fasta,bowtie2out,sam}
                        set whether the input is the FASTA file of metagenomic reads or
                        the SAM file of the mapping of the reads against the MetaPhlAn db.

Mapping arguments:
  --force               Force profiling of the input file by removing the bowtie2out file
  --bowtie2db METAPHLAN_BOWTIE2_DB
                        Folder containing the MetaPhlAn database. You can specify the location by exporting the
                        DEFAULT_DB_FOLDER variable in the shell.
                        [default /Users/maxime/mambaforge/envs/summer_school_microbiome/lib/python3.9/site-packages/metaphlan/metaphlan_databases]
  -x INDEX, --index INDEX
                        Specify the id of the database version to use. If "latest", MetaPhlAn will get the latest version.
                        If an index name is provided, MetaPhlAn will try to use it, if available, and skip the online check.
                        If the database files are not found on the local MetaPhlAn installation they
                        will be automatically downloaded [default latest]
  --bt2_ps BowTie2 presets
                        Presets options for BowTie2 (applied only when a FASTA file is provided)
                        The choices enabled in MetaPhlAn are:
                         * sensitive
                         * very-sensitive
                         * sensitive-local
                         * very-sensitive-local
                        [default very-sensitive]
  --bowtie2_exe BOWTIE2_EXE
                        Full path and name of the BowTie2 executable. This option allowsMetaPhlAn to reach the
                        executable even when it is not in the system PATH or the system PATH is unreachable
  --bowtie2_build BOWTIE2_BUILD
                        Full path to the bowtie2-build command to use, deafult assumes that 'bowtie2-build is present in the system path
  --bowtie2out FILE_NAME
                        The file for saving the output of BowTie2
  --min_mapq_val MIN_MAPQ_VAL
                        Minimum mapping quality value (MAPQ) [default 5]
  --no_map              Avoid storing the --bowtie2out map file
  --tmp_dir             The folder used to store temporary files [default is the OS dependent tmp dir]

Post-mapping arguments:
  --tax_lev TAXONOMIC_LEVEL
                        The taxonomic level for the relative abundance output:
                        'a' : all taxonomic levels
                        'k' : kingdoms
                        'p' : phyla only
                        'c' : classes only
                        'o' : orders only
                        'f' : families only
                        'g' : genera only
                        's' : species only
                        [default 'a']
  --min_cu_len          minimum total nucleotide length for the markers in a clade for
                        estimating the abundance without considering sub-clade abundances
                        [default 2000]
  --min_alignment_len   The sam records for aligned reads with the longest subalignment
                        length smaller than this threshold will be discarded.
                        [default None]
  --add_viruses         Allow the profiling of viral organisms
  --ignore_eukaryotes   Do not profile eukaryotic organisms
  --ignore_bacteria     Do not profile bacterial organisms
  --ignore_archaea      Do not profile archeal organisms
  --stat_q              Quantile value for the robust average
                        [default 0.2]
  --perc_nonzero        Percentage of markers with a non zero relative abundance for misidentify a species
                        [default 0.33]
  --ignore_markers IGNORE_MARKERS
                        File containing a list of markers to ignore.
  --avoid_disqm         Deactivate the procedure of disambiguating the quasi-markers based on the
                        marker abundance pattern found in the sample. It is generally recommended
                        to keep the disambiguation procedure in order to minimize false positives
  --stat                Statistical approach for converting marker abundances into clade abundances
                        'avg_g'  : clade global (i.e. normalizing all markers together) average
                        'avg_l'  : average of length-normalized marker counts
                        'tavg_g' : truncated clade global average at --stat_q quantile
                        'tavg_l' : truncated average of length-normalized marker counts (at --stat_q)
                        'wavg_g' : winsorized clade global average (at --stat_q)
                        'wavg_l' : winsorized average of length-normalized marker counts (at --stat_q)
                        'med'    : median of length-normalized marker counts
                        [default tavg_g]

Additional analysis types and arguments:
  -t ANALYSIS TYPE      Type of analysis to perform:
                         * rel_ab: profiling a metagenomes in terms of relative abundances
                         * rel_ab_w_read_stats: profiling a metagenomes in terms of relative abundances and estimate
                                                the number of reads coming from each clade.
                         * reads_map: mapping from reads to clades (only reads hitting a marker)
                         * clade_profiles: normalized marker counts for clades with at least a non-null marker
                         * marker_ab_table: normalized marker counts (only when &gt; 0.0 and normalized by metagenome size if --nreads is specified)
                         * marker_counts: non-normalized marker counts [use with extreme caution]
                         * marker_pres_table: list of markers present in the sample (threshold at 1.0 if not differently specified with --pres_th
                         * clade_specific_strain_tracker: list of markers present for a specific clade, specified with --clade, and all its subclades
                        [default 'rel_ab']
  --nreads NUMBER_OF_READS
                        The total number of reads in the original metagenome. It is used only when
                        -t marker_table is specified for normalizing the length-normalized counts
                        with the metagenome size as well. No normalization applied if --nreads is not
                        specified
  --pres_th PRESENCE_THRESHOLD
                        Threshold for calling a marker present by the -t marker_pres_table option
  --clade               The clade for clade_specific_strain_tracker analysis
  --min_ab              The minimum percentage abundance for the clade in the clade_specific_strain_tracker analysis

Output arguments:
  -o output file, --output_file output file
                        The output file (if not specified as positional argument)
  --sample_id_key name  Specify the sample ID key for this analysis. Defaults to 'SampleID'.
  --use_group_representative
                        Use a species as representative for species groups.
  --sample_id value     Specify the sample ID for this analysis. Defaults to 'Metaphlan_Analysis'.
  -s sam_output_file, --samout sam_output_file
                        The sam output file
  --legacy-output       Old MetaPhlAn2 two columns output
  --CAMI_format_output  Report the profiling using the CAMI output format
  --unknown_estimation  Scale relative abundances to the number of reads mapping to known clades in order to estimate unknowness
  --biom biom_output, --biom_output_file biom_output
                        If requesting biom file output: The name of the output file in biom format
  --mdelim mdelim, --metadata_delimiter_char mdelim
                        Delimiter for bug metadata: - defaults to pipe. e.g. the pipe in k__Bacteria|p__Proteobacteria

Other arguments:
  --nproc N             The number of CPUs to use for parallelizing the mapping [default 4]
  --install             Only checks if the MetaPhlAn DB is installed and installs it if not. All other parameters are ignored.
  --force_download      Force the re-download of the latest MetaPhlAn database.
  --read_min_len READ_MIN_LEN
                        Specify the minimum length of the reads to be considered when parsing the input file with
                        'read_fastx.py' script, default value is 70
  -v, --version         Prints the current MetaPhlAn version and exit
  -h, --help            show this help message and exit</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">metaphlan</span> ../results/fastp/ERR5766177.merged.fastq.gz  <span class="dt">\</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">--input_type</span> fastq <span class="dt">\</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">--bowtie2out</span> ../results/metaphlan/ERR5766177.bt2.out  <span class="dt">\</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">--nproc</span> 4 <span class="dt">\</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">&gt;</span> ../results/metaphlan/ERR5766177.metaphlan_profile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The main results files that we’re interested in is located at <a href="../results/metaphlan/ERR5766177.metaphlan_profile.txt">../results/metaphlan/ERR5766177.metaphlan_profile.txt</a></p>
<p>It’s a tab separated file, with taxons in rows, with their relative abundance in the sample</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> head ..<span class="op">/</span>results<span class="op">/</span>metaphlan<span class="op">/</span>ERR5766177.metaphlan_profile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>#mpa_v30_CHOCOPhlAn_201901
#/home/maxime_borry/.conda/envs/maxime/envs/summer_school_microbiome/bin/metaphlan ../results/fastp/ERR5766177.merged.fastq.gz \
--input_type fastq --bowtie2out ../results/metaphlan/ERR5766177.bt2.out --nproc 8
#SampleID   Metaphlan_Analysis
#clade_name NCBI_tax_id relative_abundance  additional_species
k__Bacteria 2   82.23198
k__Archaea  2157    17.76802
k__Bacteria|p__Firmicutes   2|1239  33.47957
k__Bacteria|p__Bacteroidetes    2|976   28.4209
k__Bacteria|p__Actinobacteria   2|201174    20.33151
k__Archaea|p__Euryarchaeota 2157|28890  17.76802</code></pre>
</section>
</section>
<section id="visualizing-the-taxonomic-profile" class="level3" data-number="10.3.4">
<h3 data-number="10.3.4" class="anchored" data-anchor-id="visualizing-the-taxonomic-profile"><span class="header-section-number">10.3.4</span> 4. Visualizing the taxonomic profile</h3>
<section id="visualizing-metaphlan-taxonomic-profile-with-pavian" class="level4" data-number="10.3.4.1">
<h4 data-number="10.3.4.1" class="anchored" data-anchor-id="visualizing-metaphlan-taxonomic-profile-with-pavian"><span class="header-section-number">10.3.4.1</span> 4.1 Visualizing metaphlan taxonomic profile with Pavian</h4>
<p><a href="https://github.com/fbreitwieser/pavian">Pavian</a> is an interactive app to explore results of different taxonomic classifiers</p>
<p>There are different ways to run it:</p>
<ul>
<li>If you have <a href="https://www.docker.com/">docker</a> installed (and are somehow familiar with it)</li>
</ul>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull <span class="st">'florianbw/pavian'</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-p</span> 5000:80 florianbw/pavian</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then open your browser and visit <a href="localhost:5000">localhost:5000</a></p>
<ul>
<li>If you are familiar with <a href="https://www.r-project.org/">R</a></li>
</ul>
<div class="sourceCode" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(remotes)) { <span class="fu">install.packages</span>(<span class="st">"remotes"</span>) }</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"fbreitwieser/pavian"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>pavian<span class="sc">::</span><span class="fu">runApp</span>(<span class="at">port=</span><span class="dv">5000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then open your browser and visit <a href="localhost:5000">localhost:5000</a></p>
<ul>
<li>Otherwise, just visit <a href="https://fbreitwieser.shinyapps.io/pavian/.">fbreitwieser.shinyapps.io/pavian</a>.</li>
</ul>
</section>
<section id="visualizing-metaphlan-taxonomic-profile-with-krona" class="level4" data-number="10.3.4.2">
<h4 data-number="10.3.4.2" class="anchored" data-anchor-id="visualizing-metaphlan-taxonomic-profile-with-krona"><span class="header-section-number">10.3.4.2</span> 4.2 Visualizing metaphlan taxonomic profile with <a href="https://github.com/marbl/Krona/wiki">Krona</a></h4>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">%%bash</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="ex">python</span> ../scripts/metaphlan2krona.py <span class="at">-p</span> ../results/metaphlan/ERR5766177.metaphlan_profile.txt <span class="at">-k</span> ../results/krona/ERR5766177_krona.out</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="ex">ktImportText</span> <span class="at">-o</span> ../results/krona/ERR5766177_krona.html ../results/krona/ERR5766177_krona.out</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Writing ../results/krona/ERR5766177_krona.html...</code></pre>
</section>
</section>
<section id="getting-modern-reference-data" class="level3" data-number="10.3.5">
<h3 data-number="10.3.5" class="anchored" data-anchor-id="getting-modern-reference-data"><span class="header-section-number">10.3.5</span> 5. Getting modern reference data</h3>
<p>In order to compare our sample with modern reference samples, I used the curatedMetagenomicsData package, which provides both curated metadata, and pre-computed metaphlan taxonomic profiles for published modern human samples. The full R code to get these data is available in <a href="./curatedMetagenomics/get_sources.Rmd">curatedMetagenomics/get_sources.Rmd</a></p>
<p>I pre-selected 200 gut microbiome samples from non-westernized (100) and westernized (100) from healthy, non-antibiotic users donors.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedMetagenomicData)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>sampleMetadata <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(body_site<span class="sc">==</span><span class="st">'stool'</span> <span class="sc">&amp;</span> antibiotics_current_use  <span class="sc">==</span> <span class="st">'no'</span> <span class="sc">&amp;</span> disease <span class="sc">==</span> <span class="st">'healthy'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(non_westernized) <span class="sc">%&gt;%</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">100</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="ot">-&gt;</span> selected_samples</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>selected_samples <span class="sc">%&gt;%</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">returnSamples</span>(<span class="st">"relative_abundance"</span>) <span class="ot">-&gt;</span> rel_ab</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>data_ranks <span class="ot">=</span> <span class="fu">splitByRanks</span>(rel_ab)</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="fu">names</span>(data_ranks)){</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(r)</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>  assay_rank <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">assay</span>(data_ranks[[r]]))</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"../../data/curated_metagenomics/modern_sources_"</span>,<span class="fu">tolower</span>(r),<span class="st">".csv"</span>))</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">write.csv</span>(assay_rank, <span class="fu">paste0</span>(<span class="st">"../../data/curated_metagenomics/modern_sources_"</span>,<span class="fu">tolower</span>(r),<span class="st">".csv"</span>))</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>The resulting metaphlan taxonomic profiles (split by taxonomic ranks) are available at</li>
<li><a href="../data/curated_metagenomics">../data/curated_metagenomics</a></li>
<li>The associated metadata is available at</li>
<li><a href="../data/metadata/curated_metagenomics_modern_sources.csv">../data/metadata/curated_metagenomics_modern_sources.csv</a></li>
</ul>
<p><img src="assets/images/chapters/taxonomic-profiling/800px-Taxonomic_Rank_Graph.svg.png" class="img-fluid"></p>
</section>
<section id="bringing-together-ancient-and-modern-data" class="level3" data-number="10.3.6">
<h3 data-number="10.3.6" class="anchored" data-anchor-id="bringing-together-ancient-and-modern-data"><span class="header-section-number">10.3.6</span> 6. Bringing together ancient and modern data</h3>
<p>This is the moment where we will the <a href="https://pandas.pydata.org">Pandas</a> <a href="https://www.python.org/">Python</a> library to perform some data manipulation.<br>
We will also use the <a href="https://github.com/apcamargo/taxopy">Taxopy</a> library to work to taxonomic informations.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> pip install taxopy</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>Requirement already satisfied: taxopy in /Users/maxime/mambaforge/envs/summer_school_microbiome/lib/python3.9/site-packages (0.10.0)</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> taxopy</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pickle</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gzip</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> gzip.<span class="bu">open</span>(<span class="st">"../data/taxopy/taxdb.p.gz"</span>, <span class="st">'rb'</span>) <span class="im">as</span> tdb:</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    taxo_db <span class="op">=</span> pickle.load(tdb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span> head ..<span class="op">/</span>results<span class="op">/</span>metaphlan<span class="op">/</span>ERR5766177.metaphlan_profile.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>#mpa_v30_CHOCOPhlAn_201901
#/home/maxime_borry/.conda/envs/maxime/envs/summer_school_microbiome/bin/metaphlan ../results/fastp/ERR5766177.merged.fastq.gz
--input_type fastq --bowtie2out ../results/metaphlan/ERR5766177.bt2.out --nproc 8
#SampleID   Metaphlan_Analysis
#clade_name NCBI_tax_id relative_abundance  additional_species
k__Bacteria 2   82.23198
k__Archaea  2157    17.76802
k__Bacteria|p__Firmicutes   2|1239  33.47957
k__Bacteria|p__Bacteroidetes    2|976   28.4209
k__Bacteria|p__Actinobacteria   2|201174    20.33151
k__Archaea|p__Euryarchaeota 2157|28890  17.76802</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> pd.read_csv(<span class="st">"../results/metaphlan/ERR5766177.metaphlan_profile.txt"</span>,</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>                            comment<span class="op">=</span><span class="st">"#"</span>,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                            delimiter<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>                            names<span class="op">=</span>[<span class="st">'clade_name'</span>,<span class="st">'NCBI_tax_id'</span>,<span class="st">'relative_abundance'</span>,<span class="st">'additional_species'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
clade_name
</th>
<th>
NCBI_tax_id
</th>
<th>
relative_abundance
</th>
<th>
additional_species
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
k__Bacteria
</td>
<td>
2
</td>
<td>
82.23198
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
1
</th>
<td>
k__Archaea
</td>
<td>
2157
</td>
<td>
17.76802
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
2
</th>
<td>
k__Bacteria|p__Firmicutes
</td>
<td>
2|1239
</td>
<td>
33.47957
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
3
</th>
<td>
k__Bacteria|p__Bacteroidetes
</td>
<td>
2|976
</td>
<td>
28.42090
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
4
</th>
<td>
k__Bacteria|p__Actinobacteria
</td>
<td>
2|201174
</td>
<td>
20.33151
</td>
<td>
NaN
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>ancient_data.sample(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
clade_name
</th>
<th>
NCBI_tax_id
</th>
<th>
relative_abundance
</th>
<th>
additional_species
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
1
</th>
<td>
k__Archaea
</td>
<td>
2157
</td>
<td>
17.76802
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
46
</th>
<td>
k__Bacteria|p__Bacteroidetes|c_<em>Bacteroidia|o</em>…
</td>
<td>
2|976|200643|171549|171552|838|165179
</td>
<td>
25.75544
</td>
<td>
k__Bacteria|p__Bacteroidetes|c_<em>Bacteroidia|o</em>…
</td>
</tr>
<tr>
<th>
55
</th>
<td>
k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…
</td>
<td>
2|1239|186801|186802|186803|189330|88431
</td>
<td>
0.91178
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
18
</th>
<td>
k__Archaea|p__Euryarchaeota|c_<em>Halobacteria|o</em>…
</td>
<td>
2157|28890|183963|2235
</td>
<td>
0.71177
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
36
</th>
<td>
k__Bacteria|p__Actinobacteria|c__Actinobacteri…
</td>
<td>
2|201174|1760|85004|31953|1678
</td>
<td>
9.39377
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
65
</th>
<td>
k__Bacteria|p__Actinobacteria|c__Actinobacteri…
</td>
<td>
2|201174|1760|85004|31953|1678|216816
</td>
<td>
0.05447
</td>
<td>
k__Bacteria|p__Actinobacteria|c__Actinobacteri…
</td>
</tr>
<tr>
<th>
37
</th>
<td>
k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…
</td>
<td>
2|1239|186801|186802|186803|
</td>
<td>
2.16125
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
38
</th>
<td>
k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…
</td>
<td>
2|1239|186801|186802|541000|216851
</td>
<td>
1.24537
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
26
</th>
<td>
k__Bacteria|p__Actinobacteria|c__Actinobacteri…
</td>
<td>
2|201174|1760|85004|31953
</td>
<td>
9.39377
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
48
</th>
<td>
k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…
</td>
<td>
2|1239|186801|186802|541000|1263|40518
</td>
<td>
14.96816
</td>
<td>
k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo…
</td>
</tr>
</tbody>

</table>
</div>
<p>Because for this analysis, we’re only going to look at the relative abundance, we’ll only this column, an the <a href="https://www.ncbi.nlm.nih.gov/taxonomy">TAXID</a> information</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> (</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    ancient_data</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'NCBI_tax_id'</span>: <span class="st">'TAXID'</span>})</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>    .drop([<span class="st">'clade_name'</span>,<span class="st">'additional_species'</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Always investigate your data at first !</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>ancient_data.relative_abundance.<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>700.00007</code></pre>
<p><strong>Pause and think</strong>: A relative abundance of 700%, really ?</p>
<p>Let’s proceed further and try to understand what’s happening.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
TAXID
</th>
<th>
relative_abundance
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
2
</td>
<td>
82.23198
</td>
</tr>
<tr>
<th>
1
</th>
<td>
2157
</td>
<td>
17.76802
</td>
</tr>
<tr>
<th>
2
</th>
<td>
2|1239
</td>
<td>
33.47957
</td>
</tr>
<tr>
<th>
3
</th>
<td>
2|976
</td>
<td>
28.42090
</td>
</tr>
<tr>
<th>
4
</th>
<td>
2|201174
</td>
<td>
20.33151
</td>
</tr>
</tbody>

</table>
</div>
<p>To make sense of the TAXID, we will use taxopy to get all the taxonomic related informations such as:</p>
<ul>
<li>name of the taxon</li>
<li>rank of the taxon</li>
<li>lineage of the taxon</li>
</ul>
<div class="sourceCode" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">#### This function is here to help us get the taxon information</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co">#### from the metaphlan taxonomic ID lineage, of the following form</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co">#### 2|976|200643|171549|171552|838|165179</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> to_taxopy(taxid_entry, taxo_db):</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Returns a taxopy taxon object</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co">        taxid_entry(str): metaphlan TAXID taxonomic lineage</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co">        taxo_db(taxopy database)</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co">        (bool): Returns a taxopy taxon object</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>    taxid <span class="op">=</span> taxid_entry.split(<span class="st">"|"</span>)[<span class="op">-</span><span class="dv">1</span>] <span class="co"># get the last element</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">try</span>:</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">len</span>(taxid) <span class="op">&gt;</span> <span class="dv">0</span>:</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> taxopy.Taxon(<span class="bu">int</span>(taxid), taxo_db) <span class="co"># if it's not empty, get the taxon corresponding to the taxid</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> taxopy.Taxon(<span class="dv">12908</span>, taxo_db) <span class="co"># otherwise, return the taxon associated with unclassified sequences</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">except</span> taxopy.exceptions.TaxidError <span class="im">as</span> e:</span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> taxopy.Taxon(<span class="dv">12908</span>, taxo_db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>ancient_data[<span class="st">'taxopy'</span>] <span class="op">=</span> ancient_data[<span class="st">'TAXID'</span>].<span class="bu">apply</span>(to_taxopy, taxo_db<span class="op">=</span>taxo_db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>ancient_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
TAXID
</th>
<th>
relative_abundance
</th>
<th>
taxopy
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
2
</td>
<td>
82.23198
</td>
<td>
s__Bacteria
</td>
</tr>
<tr>
<th>
1
</th>
<td>
2157
</td>
<td>
17.76802
</td>
<td>
s__Archaea
</td>
</tr>
<tr>
<th>
2
</th>
<td>
2|1239
</td>
<td>
33.47957
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicutes
</td>
</tr>
<tr>
<th>
3
</th>
<td>
2|976
</td>
<td>
28.42090
</td>
<td>
s__Bacteria;c__FCB group;p__Bacteroidetes
</td>
</tr>
<tr>
<th>
4
</th>
<td>
2|201174
</td>
<td>
20.33151
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Actinoba…
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>ancient_data <span class="op">=</span> ancient_data.assign(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    rank <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.rank),</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    name <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.name),</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>    lineage <span class="op">=</span> ancient_data.taxopy.<span class="bu">apply</span>(<span class="kw">lambda</span> x: x.name_lineage),</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ancient_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
TAXID
</th>
<th>
relative_abundance
</th>
<th>
taxopy
</th>
<th>
rank
</th>
<th>
name
</th>
<th>
lineage
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
2
</td>
<td>
82.23198
</td>
<td>
s__Bacteria
</td>
<td>
superkingdom
</td>
<td>
Bacteria
</td>
<td>
[Bacteria, cellular organisms, root]
</td>
</tr>
<tr>
<th>
1
</th>
<td>
2157
</td>
<td>
17.76802
</td>
<td>
s__Archaea
</td>
<td>
superkingdom
</td>
<td>
Archaea
</td>
<td>
[Archaea, cellular organisms, root]
</td>
</tr>
<tr>
<th>
2
</th>
<td>
2|1239
</td>
<td>
33.47957
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicutes
</td>
<td>
phylum
</td>
<td>
Firmicutes
</td>
<td>
[Firmicutes, Terrabacteria group, Bacteria, ce…
</td>
</tr>
<tr>
<th>
3
</th>
<td>
2|976
</td>
<td>
28.42090
</td>
<td>
s__Bacteria;c__FCB group;p__Bacteroidetes
</td>
<td>
phylum
</td>
<td>
Bacteroidetes
</td>
<td>
[Bacteroidetes, Bacteroidetes/Chlorobi group, …
</td>
</tr>
<tr>
<th>
4
</th>
<td>
2|201174
</td>
<td>
20.33151
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Actinoba…
</td>
<td>
phylum
</td>
<td>
Actinobacteria
</td>
<td>
[Actinobacteria, Terrabacteria group, Bacteria…
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
62
</th>
<td>
2|1239|186801|186802|186803|572511|33039
</td>
<td>
0.24910
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicut…
</td>
<td>
species
</td>
<td>
[Ruminococcus] torques
</td>
<td>
[[Ruminococcus] torques, Mediterraneibacter, L…
</td>
</tr>
<tr>
<th>
63
</th>
<td>
2|201174|84998|84999|84107|1472762|1232426
</td>
<td>
0.17084
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Actinoba…
</td>
<td>
species
</td>
<td>
[Collinsella] massiliensis
</td>
<td>
[[Collinsella] massiliensis, Enorma, Coriobact…
</td>
</tr>
<tr>
<th>
64
</th>
<td>
2|1239|186801|186802|186803|189330|39486
</td>
<td>
0.07690
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicut…
</td>
<td>
species
</td>
<td>
Dorea formicigenerans
</td>
<td>
[Dorea formicigenerans, Dorea, Lachnospiraceae…
</td>
</tr>
<tr>
<th>
65
</th>
<td>
2|201174|1760|85004|31953|1678|216816
</td>
<td>
0.05447
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Actinoba…
</td>
<td>
species
</td>
<td>
Bifidobacterium longum
</td>
<td>
[Bifidobacterium longum, Bifidobacterium, Bifi…
</td>
</tr>
<tr>
<th>
66
</th>
<td>
2|1239|186801|186802|541000|1263|1262959
</td>
<td>
0.01440
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicut…
</td>
<td>
species
</td>
<td>
Ruminococcus sp. CAG:488
</td>
<td>
[Ruminococcus sp. CAG:488, environmental sampl…
</td>
</tr>
</tbody>

</table>
<p>
67 rows × 6 columns
</p>
</div>
<p>Because our modern data are split by ranks, we’ll first split our ancient sample by rank</p>
<p>Which of the entries are at the <code>species rank</code> level ?</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>ancient_species <span class="op">=</span> ancient_data.query(<span class="st">"rank == 'species'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>ancient_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
TAXID
</th>
<th>
relative_abundance
</th>
<th>
taxopy
</th>
<th>
rank
</th>
<th>
name
</th>
<th>
lineage
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
46
</th>
<td>
2|976|200643|171549|171552|838|165179
</td>
<td>
25.75544
</td>
<td>
s__Bacteria;c__FCB group;p__Bacteroidetes;c__B…
</td>
<td>
species
</td>
<td>
Prevotella copri
</td>
<td>
[Prevotella copri, Prevotella, Prevotellaceae,…
</td>
</tr>
<tr>
<th>
47
</th>
<td>
2157|28890|183925|2158|2159|2172|2173
</td>
<td>
17.05626
</td>
<td>
s__Archaea;p__Euryarchaeota;c__Methanomada gro…
</td>
<td>
species
</td>
<td>
Methanobrevibacter smithii
</td>
<td>
[Methanobrevibacter smithii, Methanobrevibacte…
</td>
</tr>
<tr>
<th>
48
</th>
<td>
2|1239|186801|186802|541000|1263|40518
</td>
<td>
14.96816
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicut…
</td>
<td>
species
</td>
<td>
Ruminococcus bromii
</td>
<td>
[Ruminococcus bromii, Ruminococcus, Oscillospi…
</td>
</tr>
<tr>
<th>
49
</th>
<td>
2|1239|186801|186802|186803|841|301302
</td>
<td>
13.57908
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Firmicut…
</td>
<td>
species
</td>
<td>
Roseburia faecis
</td>
<td>
[Roseburia faecis, Roseburia, Lachnospiraceae,…
</td>
</tr>
<tr>
<th>
50
</th>
<td>
2|201174|84998|84999|84107|102106|74426
</td>
<td>
9.49165
</td>
<td>
s__Bacteria;c__Terrabacteria group;p__Actinoba…
</td>
<td>
species
</td>
<td>
Collinsella aerofaciens
</td>
<td>
[Collinsella aerofaciens, Collinsella, Corioba…
</td>
</tr>
</tbody>

</table>
</div>
<p>Let’s do a bit of renaming to prepare for what’s coming next</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>ancient_species <span class="op">=</span> ancient_species[[<span class="st">'relative_abundance'</span>,<span class="st">'name'</span>]].set_index(<span class="st">'name'</span>).rename(columns<span class="op">=</span>{<span class="st">'relative_abundance'</span>:<span class="st">'ERR5766177'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>ancient_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
ERR5766177
</th>
</tr>
<tr>
<th>
name
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
Prevotella copri
</th>
<td>
25.75544
</td>
</tr>
<tr>
<th>
Methanobrevibacter smithii
</th>
<td>
17.05626
</td>
</tr>
<tr>
<th>
Ruminococcus bromii
</th>
<td>
14.96816
</td>
</tr>
<tr>
<th>
Roseburia faecis
</th>
<td>
13.57908
</td>
</tr>
<tr>
<th>
Collinsella aerofaciens
</th>
<td>
9.49165
</td>
</tr>
</tbody>

</table>
</div>
<div class="sourceCode" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>ancient_phylums <span class="op">=</span> ancient_data.query(<span class="st">"rank == 'phylum'"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>ancient_phylums <span class="op">=</span> ancient_phylums[[<span class="st">'relative_abundance'</span>,<span class="st">'name'</span>]].set_index(<span class="st">'name'</span>).rename(columns<span class="op">=</span>{<span class="st">'relative_abundance'</span>:<span class="st">'ERR5766177'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>ancient_phylums</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
ERR5766177
</th>
</tr>
<tr>
<th>
name
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
Firmicutes
</th>
<td>
33.47957
</td>
</tr>
<tr>
<th>
Bacteroidetes
</th>
<td>
28.42090
</td>
</tr>
<tr>
<th>
Actinobacteria
</th>
<td>
20.33151
</td>
</tr>
<tr>
<th>
Euryarchaeota
</th>
<td>
17.76802
</td>
</tr>
</tbody>

</table>
</div>
<p>Now, let’s go back to the 700% relative abundance issue…</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>ancient_data.groupby(<span class="st">'rank'</span>)[<span class="st">'relative_abundance'</span>].<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>rank
class            99.72648
family           83.49854
genus            97.56524
no rank          19.48331
order            99.72648
phylum          100.00000
species         100.00002
superkingdom    100.00000
Name: relative_abundance, dtype: float64</code></pre>
<p>Seems better, right ?</p>
<p><strong>Pause and think: why don’t we get exactly 100%</strong> ?</p>
<p>Now let’s load our modern reference samples</p>
<div class="sourceCode" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>modern_phylums <span class="op">=</span> pd.read_csv(<span class="st">"../data/curated_metagenomics/modern_sources_phylum.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>modern_phylums.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<th>
PNP_Main_283
</th>
<th>
PNP_Validation_55
</th>
<th>
G80275
</th>
<th>
PNP_Main_363
</th>
<th>
SAMEA7045572
</th>
<th>
SAMEA7045355
</th>
<th>
HD-13
</th>
<th>
EGAR00001420773_9002000001423910
</th>
<th>
SID5428-4
</th>
<th>
…
</th>
<th>
A46_02_1FE
</th>
<th>
TZ_87532
</th>
<th>
A94_01_1FE
</th>
<th>
KHG_7
</th>
<th>
LDK_4
</th>
<th>
KHG_9
</th>
<th>
A48_01_1FE
</th>
<th>
KHG_1
</th>
<th>
TZ_81781
</th>
<th>
A09_01_1FE
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
Bacteroidetes
</th>
<td>
0.00000
</td>
<td>
17.44332
</td>
<td>
82.86400
</td>
<td>
69.99087
</td>
<td>
31.93081
</td>
<td>
51.76204
</td>
<td>
53.32801
</td>
<td>
74.59667
</td>
<td>
8.81074
</td>
<td>
26.39694
</td>
<td>
…
</td>
<td>
1.97760
</td>
<td>
1.49601
</td>
<td>
67.21410
</td>
<td>
4.29848
</td>
<td>
68.16890
</td>
<td>
38.59709
</td>
<td>
14.81828
</td>
<td>
10.13908
</td>
<td>
57.14031
</td>
<td>
11.61544
</td>
</tr>
<tr>
<th>
Firmicutes
</th>
<td>
95.24231
</td>
<td>
60.47031
</td>
<td>
16.53946
</td>
<td>
22.81977
</td>
<td>
65.23075
</td>
<td>
41.96928
</td>
<td>
45.77661
</td>
<td>
23.51065
</td>
<td>
54.35341
</td>
<td>
62.23094
</td>
<td>
…
</td>
<td>
76.68499
</td>
<td>
78.13269
</td>
<td>
29.72394
</td>
<td>
33.51772
</td>
<td>
19.11149
</td>
<td>
46.87139
</td>
<td>
72.68136
</td>
<td>
35.43789
</td>
<td>
40.57101
</td>
<td>
24.72113
</td>
</tr>
<tr>
<th>
Proteobacteria
</th>
<td>
4.49959
</td>
<td>
0.77098
</td>
<td>
0.05697
</td>
<td>
4.07757
</td>
<td>
0.27316
</td>
<td>
3.33972
</td>
<td>
0.02001
</td>
<td>
1.72865
</td>
<td>
0.00000
</td>
<td>
1.81016
</td>
<td>
…
</td>
<td>
16.57250
</td>
<td>
0.76159
</td>
<td>
2.35058
</td>
<td>
9.83772
</td>
<td>
5.32392
</td>
<td>
0.19699
</td>
<td>
3.64655
</td>
<td>
17.64151
</td>
<td>
0.30580
</td>
<td>
56.20177
</td>
</tr>
<tr>
<th>
Actinobacteria
</th>
<td>
0.25809
</td>
<td>
10.27631
</td>
<td>
0.45187
</td>
<td>
1.11902
</td>
<td>
2.31075
</td>
<td>
2.92715
</td>
<td>
0.77667
</td>
<td>
0.16403
</td>
<td>
36.55138
</td>
<td>
1.19951
</td>
<td>
…
</td>
<td>
3.01814
</td>
<td>
19.20468
</td>
<td>
0.69913
</td>
<td>
46.99479
</td>
<td>
7.39093
</td>
<td>
14.26365
</td>
<td>
5.47750
</td>
<td>
36.77145
</td>
<td>
1.16426
</td>
<td>
7.40894
</td>
</tr>
<tr>
<th>
Verrucomicrobia
</th>
<td>
0.00000
</td>
<td>
0.00784
</td>
<td>
0.00000
</td>
<td>
1.99276
</td>
<td>
0.25451
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.09940
</td>
<td>
3.29795
</td>
<td>
…
</td>
<td>
0.05011
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
</tr>
</tbody>

</table>
<p>
5 rows × 200 columns
</p>
</div>
<div class="sourceCode" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>modern_species <span class="op">=</span> pd.read_csv(<span class="st">"../data/curated_metagenomics/modern_sources_species.csv"</span>, index_col<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>modern_species.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<th>
PNP_Main_283
</th>
<th>
PNP_Validation_55
</th>
<th>
G80275
</th>
<th>
PNP_Main_363
</th>
<th>
SAMEA7045572
</th>
<th>
SAMEA7045355
</th>
<th>
HD-13
</th>
<th>
EGAR00001420773_9002000001423910
</th>
<th>
SID5428-4
</th>
<th>
…
</th>
<th>
A46_02_1FE
</th>
<th>
TZ_87532
</th>
<th>
A94_01_1FE
</th>
<th>
KHG_7
</th>
<th>
LDK_4
</th>
<th>
KHG_9
</th>
<th>
A48_01_1FE
</th>
<th>
KHG_1
</th>
<th>
TZ_81781
</th>
<th>
A09_01_1FE
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
Bacteroides vulgatus
</th>
<td>
0.0
</td>
<td>
0.60446
</td>
<td>
1.59911
</td>
<td>
4.39085
</td>
<td>
0.04494
</td>
<td>
4.66505
</td>
<td>
2.99431
</td>
<td>
29.30325
</td>
<td>
1.48560
</td>
<td>
0.98818
</td>
<td>
…
</td>
<td>
0.20717
</td>
<td>
0.0
</td>
<td>
0.00309
</td>
<td>
0.48891
</td>
<td>
0.00000
</td>
<td>
0.02230
</td>
<td>
0.00000
</td>
<td>
0.15112
</td>
<td>
0.0
</td>
<td>
0.00836
</td>
</tr>
<tr>
<th>
Bacteroides stercoris
</th>
<td>
0.0
</td>
<td>
0.00546
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
2.50789
</td>
<td>
0.00000
</td>
<td>
20.57498
</td>
<td>
8.28443
</td>
<td>
1.23261
</td>
<td>
0.00000
</td>
<td>
…
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.00693
</td>
<td>
0.00000
</td>
<td>
0.02603
</td>
<td>
0.00000
</td>
<td>
0.19318
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
</tr>
<tr>
<th>
Acidaminococcus intestini
</th>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.32822
</td>
<td>
0.00000
</td>
<td>
…
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
</tr>
<tr>
<th>
Eubacterium sp CAG 38
</th>
<td>
0.0
</td>
<td>
0.06712
</td>
<td>
0.81149
</td>
<td>
0.05247
</td>
<td>
0.26027
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
2.62415
</td>
<td>
0.46585
</td>
<td>
0.23372
</td>
<td>
…
</td>
<td>
0.78140
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.00499
</td>
<td>
0.00000
</td>
<td>
0.02446
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
</tr>
<tr>
<th>
Parabacteroides distasonis
</th>
<td>
0.0
</td>
<td>
1.34931
</td>
<td>
2.00672
</td>
<td>
5.85067
</td>
<td>
0.59019
</td>
<td>
7.00027
</td>
<td>
1.28075
</td>
<td>
0.61758
</td>
<td>
0.07383
</td>
<td>
2.80355
</td>
<td>
…
</td>
<td>
0.11423
</td>
<td>
0.0
</td>
<td>
0.01181
</td>
<td>
0.01386
</td>
<td>
0.03111
</td>
<td>
0.07463
</td>
<td>
0.15597
</td>
<td>
0.07541
</td>
<td>
0.0
</td>
<td>
0.01932
</td>
</tr>
</tbody>

</table>
<p>
5 rows × 200 columns
</p>
</div>
<p>Now, let’s merge our ancient sample with the modern data in one single table</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>all_species <span class="op">=</span> ancient_species.merge(modern_species, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>, how<span class="op">=</span><span class="st">'outer'</span>).fillna(<span class="dv">0</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>all_phylums <span class="op">=</span> ancient_phylums.merge(modern_phylums, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>, how<span class="op">=</span><span class="st">'outer'</span>).fillna(<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, let’s load the metadata</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> pd.read_csv(<span class="st">"../data/metadata/curated_metagenomics_modern_sources.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>metadata.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
study_name
</th>
<th>
sample_id
</th>
<th>
subject_id
</th>
<th>
body_site
</th>
<th>
antibiotics_current_use
</th>
<th>
study_condition
</th>
<th>
disease
</th>
<th>
age
</th>
<th>
infant_age
</th>
<th>
age_category
</th>
<th>
…
</th>
<th>
hla_drb11
</th>
<th>
birth_order
</th>
<th>
age_twins_started_to_live_apart
</th>
<th>
zigosity
</th>
<th>
brinkman_index
</th>
<th>
alcohol_numeric
</th>
<th>
breastfeeding_duration
</th>
<th>
formula_first_day
</th>
<th>
ALT
</th>
<th>
eGFR
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
ShaoY_2019
</td>
<td>
de028ad4-7ae6-11e9-a106-68b59976a384
</td>
<td>
C01528_ba
</td>
<td>
stool
</td>
<td>
no
</td>
<td>
control
</td>
<td>
healthy
</td>
<td>
0.0
</td>
<td>
4.0
</td>
<td>
newborn
</td>
<td>
…
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
1
</th>
<td>
ZeeviD_2015
</td>
<td>
PNP_Main_283
</td>
<td>
PNP_Main_283
</td>
<td>
stool
</td>
<td>
no
</td>
<td>
control
</td>
<td>
healthy
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
adult
</td>
<td>
…
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
2
</th>
<td>
ZeeviD_2015
</td>
<td>
PNP_Validation_55
</td>
<td>
PNP_Validation_55
</td>
<td>
stool
</td>
<td>
no
</td>
<td>
control
</td>
<td>
healthy
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
adult
</td>
<td>
…
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
3
</th>
<td>
VatanenT_2016
</td>
<td>
G80275
</td>
<td>
T014806
</td>
<td>
stool
</td>
<td>
no
</td>
<td>
control
</td>
<td>
healthy
</td>
<td>
1.0
</td>
<td>
NaN
</td>
<td>
child
</td>
<td>
…
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
</tr>
<tr>
<th>
4
</th>
<td>
ZeeviD_2015
</td>
<td>
PNP_Main_363
</td>
<td>
PNP_Main_363
</td>
<td>
stool
</td>
<td>
no
</td>
<td>
control
</td>
<td>
healthy
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
adult
</td>
<td>
…
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
<td>
NaN
</td>
</tr>
</tbody>

</table>
<p>
5 rows × 130 columns
</p>
</div>
</section>
<section id="comparing-ancient-and-modern-samples" class="level3" data-number="10.3.7">
<h3 data-number="10.3.7" class="anchored" data-anchor-id="comparing-ancient-and-modern-samples"><span class="header-section-number">10.3.7</span> 7. Comparing ancient and modern samples</h3>
<section id="taxonomic-composition" class="level4" data-number="10.3.7.1">
<h4 data-number="10.3.7.1" class="anchored" data-anchor-id="taxonomic-composition"><span class="header-section-number">10.3.7.1</span> 7.1 Taxonomic composition</h4>
<p>One common plot in microbiome papers in a stacked barplot, often at the phylum or family level.</p>
<p>First, we’ll do some renaming, to make the value of the metadata variables a bit easier to understand</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>group_info <span class="op">=</span> (</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    metadata[<span class="st">'non_westernized'</span>]</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    .<span class="bu">map</span>({<span class="st">'no'</span>:<span class="st">'westernized'</span>,<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>}) <span class="co"># for the non_westernized in the modern sample metadata, rename the value levels</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>    .to_frame(name<span class="op">=</span><span class="st">'group'</span>).set_index(metadata[<span class="st">'sample_id'</span>]) <span class="co"># rename the column to group</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>    .reset_index()</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>    .append({<span class="st">'sample_id'</span>:<span class="st">'ERR5766177'</span>, <span class="st">'group'</span>:<span class="st">'ancient'</span>}, ignore_index<span class="op">=</span><span class="va">True</span>) <span class="co"># add the ancient sample</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>group_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/var/folders/1c/l1qb09f15jddsh65f6xv1n_r0000gp/T/ipykernel_40830/27419655.py:2:
FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version.
Use pandas.concat instead.
  metadata['non_westernized']</code></pre>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
sample_id
</th>
<th>
group
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
0
</th>
<td>
de028ad4-7ae6-11e9-a106-68b59976a384
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
1
</th>
<td>
PNP_Main_283
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
2
</th>
<td>
PNP_Validation_55
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
3
</th>
<td>
G80275
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
4
</th>
<td>
PNP_Main_363
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
196
</th>
<td>
A48_01_1FE
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
197
</th>
<td>
KHG_1
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
198
</th>
<td>
TZ_81781
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
199
</th>
<td>
A09_01_1FE
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
200
</th>
<td>
ERR5766177
</td>
<td>
ancient
</td>
</tr>
</tbody>

</table>
<p>
201 rows × 2 columns
</p>
</div>
<p>We need transform our data in <a href="https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html">tidy</a> format to plot with <a href="https://plotnine.readthedocs.io/en/stable/">plotnine</a>, a python clone of <a href="https://ggplot2.tidyverse.org/index.html">ggplot</a>.<br>
We then add the group information (Westernized, non westernized, or ancient sample), and compute the mean abundance for each phylum.</p>
<p>First we transpose the dataframe to have the samples as index, and the phylums as columns</p>
<p>We then add the metadata information</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>    all_phylums</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>    .transpose()</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>    .merge(group_info, left_index<span class="op">=</span><span class="va">True</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>)</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
Actinobacteria
</th>
<th>
Apicomplexa
</th>
<th>
Ascomycota
</th>
<th>
Bacteroidetes
</th>
<th>
Basidiomycota
</th>
<th>
Candidatus Melainabacteria
</th>
<th>
Chlamydiae
</th>
<th>
Chloroflexi
</th>
<th>
Cyanobacteria
</th>
<th>
Deferribacteres
</th>
<th>
…
</th>
<th>
Fusobacteria
</th>
<th>
Lentisphaerae
</th>
<th>
Planctomycetes
</th>
<th>
Proteobacteria
</th>
<th>
Spirochaetes
</th>
<th>
Synergistetes
</th>
<th>
Tenericutes
</th>
<th>
Verrucomicrobia
</th>
<th>
sample_id
</th>
<th>
group
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
200
</th>
<td>
20.33151
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
28.42090
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
ERR5766177
</td>
<td>
ancient
</td>
</tr>
<tr>
<th>
0
</th>
<td>
0.25809
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
4.49959
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
de028ad4-7ae6-11e9-a106-68b59976a384
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
1
</th>
<td>
10.27631
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
17.44332
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.01486
</td>
<td>
0.0
</td>
<td>
0.77098
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00784
</td>
<td>
PNP_Main_283
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
2
</th>
<td>
0.45187
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
82.86400
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.05697
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
PNP_Validation_55
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
3
</th>
<td>
1.11902
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
69.99087
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
4.07757
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
1.99276
</td>
<td>
G80275
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
195
</th>
<td>
14.26365
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
38.59709
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.19699
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
KHG_9
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
196
</th>
<td>
5.47750
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
14.81828
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
3.64655
</td>
<td>
0.09964
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
A48_01_1FE
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
197
</th>
<td>
36.77145
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
10.13908
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
17.64151
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
KHG_1
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
198
</th>
<td>
1.16426
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
57.14031
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.30580
</td>
<td>
0.70467
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
TZ_81781
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
199
</th>
<td>
7.40894
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
11.61544
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
56.20177
</td>
<td>
0.00000
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.00000
</td>
<td>
A09_01_1FE
</td>
<td>
non_westernized
</td>
</tr>
</tbody>

</table>
<p>
201 rows × 24 columns
</p>
</div>
<p>Now, we need it in the tidy format</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums <span class="op">=</span> (</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>    all_phylums</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>    .transpose()</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    .merge(group_info, left_index<span class="op">=</span><span class="va">True</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>)</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>    .melt(id_vars<span class="op">=</span>[<span class="st">'sample_id'</span>, <span class="st">'group'</span>], value_name<span class="op">=</span><span class="st">'relative_abundance'</span>, var_name<span class="op">=</span><span class="st">'Phylum'</span>, ignore_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Finally, we only want to keep the mean relative abundance for each phylum</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums <span class="op">=</span> tidy_phylums.groupby([<span class="st">'group'</span>, <span class="st">'Phylum'</span>]).mean().reset_index()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>tidy_phylums.groupby(<span class="st">'group'</span>)[<span class="st">'relative_abundance'</span>].<span class="bu">sum</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>group
ancient            100.000000
non_westernized     99.710255
westernized         99.905089
Name: relative_abundance, dtype: float64</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> plotnine <span class="im">import</span> <span class="op">*</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>ggplot(tidy_phylums, aes(x<span class="op">=</span><span class="st">'group'</span>, y<span class="op">=</span><span class="st">'relative_abundance'</span>, fill<span class="op">=</span><span class="st">'Phylum'</span>)) <span class="op">\</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_bar(position<span class="op">=</span><span class="st">'stack'</span>, stat<span class="op">=</span><span class="st">'identity'</span>) <span class="op">\</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> ylab(<span class="st">'mean abundance'</span>) <span class="op">\</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> xlab(<span class="st">""</span>) <span class="op">\</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_80_0.png" class="img-fluid"></p>
<pre><code>&lt;ggplot: (406187548)&gt;</code></pre>
</section>
<section id="ecological-diversity" class="level4" data-number="10.3.7.2">
<h4 data-number="10.3.7.2" class="anchored" data-anchor-id="ecological-diversity"><span class="header-section-number">10.3.7.2</span> 7.2 Ecological diversity</h4>
<section id="alpha-diversity" class="level5" data-number="10.3.7.2.1">
<h5 data-number="10.3.7.2.1" class="anchored" data-anchor-id="alpha-diversity"><span class="header-section-number">10.3.7.2.1</span> 7.2.1 Alpha diversity</h5>
<p>Alpha diversity is the measure of diversity withing each sample. It is used to estimate how many species are present in a sample, and how diverse they are.<br>
We’ll use the python library <a href="http://scikit-bio.org/">scikit-bio</a> to compute it, and the <a href="https://plotnine.readthedocs.io/">plotnine</a> library (a python port of <a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot2</a> to visualize the results).</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> skbio</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s compute the <a href="https://en.wikipedia.org/wiki/Species_richness">species richness</a>, the <a href="https://www.biologydiscussion.com/biodiversity/types/2-types-of-diversity-indices-of-biodiversity/8388">Shannon, and Simpson index of diversity</a> index</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>shannon <span class="op">=</span> skbio.diversity.alpha_diversity(metric<span class="op">=</span><span class="st">'shannon'</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>simpson <span class="op">=</span> skbio.diversity.alpha_diversity(metric<span class="op">=</span><span class="st">'simpson'</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>richness <span class="op">=</span> (all_species <span class="op">!=</span> <span class="dv">0</span>).astype(<span class="bu">int</span>).<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> (shannon.to_frame(name<span class="op">=</span><span class="st">'shannon'</span>)</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>                   .merge(simpson.to_frame(name<span class="op">=</span><span class="st">'simpson'</span>), left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>                   .merge(richness.to_frame(name<span class="op">=</span><span class="st">'richness'</span>), left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
shannon
</th>
<th>
simpson
</th>
<th>
richness
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
ERR5766177
</th>
<td>
3.032945
</td>
<td>
0.844769
</td>
<td>
21
</td>
</tr>
<tr>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<td>
0.798112
</td>
<td>
0.251280
</td>
<td>
11
</td>
</tr>
<tr>
<th>
PNP_Main_283
</th>
<td>
5.092878
</td>
<td>
0.954159
</td>
<td>
118
</td>
</tr>
<tr>
<th>
PNP_Validation_55
</th>
<td>
3.670162
</td>
<td>
0.812438
</td>
<td>
72
</td>
</tr>
<tr>
<th>
G80275
</th>
<td>
3.831358
</td>
<td>
0.876712
</td>
<td>
66
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
KHG_9
</th>
<td>
3.884285
</td>
<td>
0.861683
</td>
<td>
87
</td>
</tr>
<tr>
<th>
A48_01_1FE
</th>
<td>
4.377755
</td>
<td>
0.930024
</td>
<td>
53
</td>
</tr>
<tr>
<th>
KHG_1
</th>
<td>
3.733834
</td>
<td>
0.875335
</td>
<td>
108
</td>
</tr>
<tr>
<th>
TZ_81781
</th>
<td>
2.881856
</td>
<td>
0.719491
</td>
<td>
44
</td>
</tr>
<tr>
<th>
A09_01_1FE
</th>
<td>
2.982322
</td>
<td>
0.719962
</td>
<td>
75
</td>
</tr>
</tbody>

</table>
<p>
201 rows × 3 columns
</p>
</div>
<p>Let’s load the group information from the metadata</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> (</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>    alpha_diversity</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>    .merge(metadata[[<span class="st">'sample_id'</span>, <span class="st">'non_westernized'</span>]], left_index<span class="op">=</span><span class="va">True</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>, how<span class="op">=</span><span class="st">'outer'</span>)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>    .set_index(<span class="st">'sample_id'</span>)</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'non_westernized'</span>:<span class="st">'group'</span>})</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>alpha_diversity[<span class="st">'group'</span>] <span class="op">=</span> alpha_diversity[<span class="st">'group'</span>].replace({<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>,<span class="st">'no'</span>:<span class="st">'westernized'</span>, pd.NA:<span class="st">'ERR5766177'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
shannon
</th>
<th>
simpson
</th>
<th>
richness
</th>
<th>
group
</th>
</tr>
<tr>
<th>
sample_id
</th>
<th>
</th>
<th>
</th>
<th>
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
ERR5766177
</th>
<td>
3.032945
</td>
<td>
0.844769
</td>
<td>
21
</td>
<td>
ERR5766177
</td>
</tr>
<tr>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<td>
0.798112
</td>
<td>
0.251280
</td>
<td>
11
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
PNP_Main_283
</th>
<td>
5.092878
</td>
<td>
0.954159
</td>
<td>
118
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
PNP_Validation_55
</th>
<td>
3.670162
</td>
<td>
0.812438
</td>
<td>
72
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
G80275
</th>
<td>
3.831358
</td>
<td>
0.876712
</td>
<td>
66
</td>
<td>
westernized
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
KHG_9
</th>
<td>
3.884285
</td>
<td>
0.861683
</td>
<td>
87
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
A48_01_1FE
</th>
<td>
4.377755
</td>
<td>
0.930024
</td>
<td>
53
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
KHG_1
</th>
<td>
3.733834
</td>
<td>
0.875335
</td>
<td>
108
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
TZ_81781
</th>
<td>
2.881856
</td>
<td>
0.719491
</td>
<td>
44
</td>
<td>
non_westernized
</td>
</tr>
<tr>
<th>
A09_01_1FE
</th>
<td>
2.982322
</td>
<td>
0.719962
</td>
<td>
75
</td>
<td>
non_westernized
</td>
</tr>
</tbody>

</table>
<p>
201 rows × 4 columns
</p>
</div>
<div class="sourceCode" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity <span class="op">=</span> alpha_diversity.melt(id_vars<span class="op">=</span><span class="st">'group'</span>, value_name<span class="op">=</span><span class="st">'alpha diversity'</span>, var_name<span class="op">=</span><span class="st">'diversity_index'</span>, ignore_index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>alpha_diversity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
group
</th>
<th>
diversity_index
</th>
<th>
alpha diversity
</th>
</tr>
<tr>
<th>
sample_id
</th>
<th>
</th>
<th>
</th>
<th>
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
ERR5766177
</th>
<td>
ERR5766177
</td>
<td>
shannon
</td>
<td>
3.032945
</td>
</tr>
<tr>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<td>
westernized
</td>
<td>
shannon
</td>
<td>
0.798112
</td>
</tr>
<tr>
<th>
PNP_Main_283
</th>
<td>
westernized
</td>
<td>
shannon
</td>
<td>
5.092878
</td>
</tr>
<tr>
<th>
PNP_Validation_55
</th>
<td>
westernized
</td>
<td>
shannon
</td>
<td>
3.670162
</td>
</tr>
<tr>
<th>
G80275
</th>
<td>
westernized
</td>
<td>
shannon
</td>
<td>
3.831358
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
KHG_9
</th>
<td>
non_westernized
</td>
<td>
richness
</td>
<td>
87.000000
</td>
</tr>
<tr>
<th>
A48_01_1FE
</th>
<td>
non_westernized
</td>
<td>
richness
</td>
<td>
53.000000
</td>
</tr>
<tr>
<th>
KHG_1
</th>
<td>
non_westernized
</td>
<td>
richness
</td>
<td>
108.000000
</td>
</tr>
<tr>
<th>
TZ_81781
</th>
<td>
non_westernized
</td>
<td>
richness
</td>
<td>
44.000000
</td>
</tr>
<tr>
<th>
A09_01_1FE
</th>
<td>
non_westernized
</td>
<td>
richness
</td>
<td>
75.000000
</td>
</tr>
</tbody>

</table>
<p>
603 rows × 3 columns
</p>
</div>
<div class="sourceCode" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> ggplot(alpha_diversity, aes(x<span class="op">=</span><span class="st">'group'</span>, y<span class="op">=</span><span class="st">'alpha diversity'</span>, color<span class="op">=</span><span class="st">'group'</span>))</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> geom_violin()</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> geom_jitter()</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme_classic()</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> facet_wrap(<span class="st">'~diversity_index'</span>, scales <span class="op">=</span> <span class="st">'free'</span>)</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme(axis_text_x<span class="op">=</span>element_text(rotation<span class="op">=</span><span class="dv">45</span>, hjust<span class="op">=</span><span class="dv">1</span>))</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>g <span class="op">+=</span> theme(subplots_adjust<span class="op">=</span>{<span class="st">'wspace'</span>: <span class="fl">0.15</span>})</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_91_1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (407407577)&gt;</code></pre>
<p><strong>Pause and think: Why do we observe a smaller species richness and diversity in our sample ?</strong></p>
</section>
</section>
<section id="beta-diversity" class="level4" data-number="10.3.7.3">
<h4 data-number="10.3.7.3" class="anchored" data-anchor-id="beta-diversity"><span class="header-section-number">10.3.7.3</span> 7.2.2 Beta diversity</h4>
<p>The Beta diversity is the measure of diversity between a pair of samples. It is used to compare the diversity between samples and see how they relate.</p>
<p>We will compute the beta diversity using the <a href="https://en.wikipedia.org/wiki/Bray%E2%80%93Curtis_dissimilarity">bray-curtis</a> dissimilarity</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>beta_diversity <span class="op">=</span> skbio.diversity.beta_diversity(metric<span class="op">=</span><span class="st">'braycurtis'</span>, counts<span class="op">=</span>all_species.transpose(), ids<span class="op">=</span>all_species.columns, validate<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We get a distance matrix</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(beta_diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>201x201 distance matrix
IDs:
'ERR5766177', 'de028ad4-7ae6-11e9-a106-68b59976a384', 'PNP_Main_283', ...
Data:
[[0.         1.         0.81508134 ... 0.85716612 0.69790092 0.8303726 ]
 [1.         0.         0.99988327 ... 0.99853413 0.994116   0.99877258]
 [0.81508134 0.99988327 0.         ... 0.82311942 0.87202543 0.91363156]
 ...
 [0.85716612 0.99853413 0.82311942 ... 0.         0.84253376 0.76616679]
 [0.69790092 0.994116   0.87202543 ... 0.84253376 0.         0.82409272]
 [0.8303726  0.99877258 0.91363156 ... 0.76616679 0.82409272 0.        ]]</code></pre>
<p>To visualize this distance matrix in a lower dimensional space, we’ll use a <a href="https://en.wikipedia.org/wiki/Multidimensional_scaling#Types">PCoA</a>, which is is a method very similar to a PCA, but taking a distance matrix as input.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>pcoa <span class="op">=</span> skbio.stats.ordination.pcoa(beta_diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/Users/maxime/mambaforge/envs/summer_school_microbiome/lib/python3.9/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:143: RuntimeWarning:
The result contains negative eigenvalues. Please compare their magnitude with the magnitude of some of the largest positive eigenvalues.
If the negative ones are smaller, it's probably safe to ignore them, but if they are large in magnitude, the results won't be useful.
See the Notes section for more details. The smallest eigenvalue is -0.25334842745723996 and the largest is 10.204440747987945.</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>pcoa.samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div>


<table class="dataframe table table-sm table-striped small">
<thead>
<tr>
<th>
</th>
<th>
PC1
</th>
<th>
PC2
</th>
<th>
PC3
</th>
<th>
PC4
</th>
<th>
PC5
</th>
<th>
PC6
</th>
<th>
PC7
</th>
<th>
PC8
</th>
<th>
PC9
</th>
<th>
PC10
</th>
<th>
…
</th>
<th>
PC192
</th>
<th>
PC193
</th>
<th>
PC194
</th>
<th>
PC195
</th>
<th>
PC196
</th>
<th>
PC197
</th>
<th>
PC198
</th>
<th>
PC199
</th>
<th>
PC200
</th>
<th>
PC201
</th>
</tr>
</thead>
<tbody>
<tr>
<th>
ERR5766177
</th>
<td>
0.216901
</td>
<td>
-0.039778
</td>
<td>
0.107412
</td>
<td>
0.273272
</td>
<td>
0.020540
</td>
<td>
0.114876
</td>
<td>
-0.256332
</td>
<td>
-0.151069
</td>
<td>
0.097451
</td>
<td>
0.060211
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
de028ad4-7ae6-11e9-a106-68b59976a384
</th>
<td>
-0.099355
</td>
<td>
0.145224
</td>
<td>
-0.191676
</td>
<td>
0.127626
</td>
<td>
0.119754
</td>
<td>
-0.132209
</td>
<td>
-0.097382
</td>
<td>
0.036728
</td>
<td>
0.081294
</td>
<td>
-0.056686
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
PNP_Main_283
</th>
<td>
-0.214108
</td>
<td>
-0.147466
</td>
<td>
0.116027
</td>
<td>
0.090059
</td>
<td>
0.076644
</td>
<td>
0.111536
</td>
<td>
0.092115
</td>
<td>
0.026477
</td>
<td>
-0.006460
</td>
<td>
-0.018592
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
PNP_Validation_55
</th>
<td>
0.244827
</td>
<td>
-0.173996
</td>
<td>
-0.311197
</td>
<td>
-0.012836
</td>
<td>
0.031759
</td>
<td>
0.117548
</td>
<td>
0.148715
</td>
<td>
-0.135641
</td>
<td>
0.034730
</td>
<td>
-0.009395
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
G80275
</th>
<td>
-0.261358
</td>
<td>
-0.077147
</td>
<td>
-0.254374
</td>
<td>
-0.065932
</td>
<td>
0.088538
</td>
<td>
0.165970
</td>
<td>
-0.005260
</td>
<td>
-0.028739
</td>
<td>
-0.002016
</td>
<td>
0.015719
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
…
</th>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
<td>
…
</td>
</tr>
<tr>
<th>
KHG_9
</th>
<td>
0.296057
</td>
<td>
-0.150300
</td>
<td>
0.013941
</td>
<td>
0.032649
</td>
<td>
-0.147692
</td>
<td>
0.019663
</td>
<td>
-0.063120
</td>
<td>
-0.034453
</td>
<td>
-0.073514
</td>
<td>
0.070085
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
A48_01_1FE
</th>
<td>
0.110621
</td>
<td>
0.030971
</td>
<td>
0.154231
</td>
<td>
-0.185961
</td>
<td>
-0.008512
</td>
<td>
-0.103420
</td>
<td>
0.028169
</td>
<td>
-0.044530
</td>
<td>
0.041902
</td>
<td>
0.068597
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
KHG_1
</th>
<td>
-0.100009
</td>
<td>
0.167885
</td>
<td>
0.009915
</td>
<td>
0.076842
</td>
<td>
-0.405582
</td>
<td>
-0.039111
</td>
<td>
-0.006421
</td>
<td>
-0.009774
</td>
<td>
-0.072252
</td>
<td>
0.150000
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
TZ_81781
</th>
<td>
0.405716
</td>
<td>
-0.139297
</td>
<td>
-0.075026
</td>
<td>
-0.079716
</td>
<td>
-0.053264
</td>
<td>
-0.119271
</td>
<td>
0.068261
</td>
<td>
-0.018821
</td>
<td>
0.198152
</td>
<td>
-0.012792
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
<tr>
<th>
A09_01_1FE
</th>
<td>
0.089101
</td>
<td>
0.471135
</td>
<td>
0.069629
</td>
<td>
-0.125644
</td>
<td>
-0.036793
</td>
<td>
0.115151
</td>
<td>
0.060507
</td>
<td>
-0.000912
</td>
<td>
-0.027239
</td>
<td>
-0.138436
</td>
<td>
…
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
<td>
0.0
</td>
</tr>
</tbody>

</table>
<p>
201 rows × 201 columns
</p>
</div>
<p>Let’s look at the variance explained by the first axes by using a scree plot</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>var_explained <span class="op">=</span> pcoa.proportion_explained[:<span class="dv">9</span>].to_frame(name<span class="op">=</span><span class="st">'variance explained'</span>).reset_index().rename(columns<span class="op">=</span>{<span class="st">'index'</span>:<span class="st">'PC'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>ggplot(var_explained, aes(x<span class="op">=</span><span class="st">'PC'</span>, y<span class="op">=</span><span class="st">'variance explained'</span>, group<span class="op">=</span><span class="dv">1</span>)) <span class="op">\</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_point() <span class="op">\</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_line() <span class="op">\</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_102_0.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (407531271)&gt;</code></pre>
<p>In this scree plot, we’re looking for the “elbow”, where there is a drop in the slope. Here, it seems that most of the variance is captures by the 3 first principal components</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>pcoa_embed <span class="op">=</span> pcoa.samples[[<span class="st">'PC1'</span>,<span class="st">'PC2'</span>,<span class="st">'PC3'</span>]].rename_axis(<span class="st">'sample'</span>).reset_index()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>pcoa_embed <span class="op">=</span> (</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>    pcoa_embed</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>    .merge(metadata[[<span class="st">'sample_id'</span>, <span class="st">'non_westernized'</span>]], left_on<span class="op">=</span><span class="st">'sample'</span>, right_on<span class="op">=</span><span class="st">'sample_id'</span>, how<span class="op">=</span><span class="st">'outer'</span>)</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>    .drop(<span class="st">'sample_id'</span>, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>    .rename(columns<span class="op">=</span>{<span class="st">'non_westernized'</span>:<span class="st">'group'</span>})</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>pcoa_embed[<span class="st">'group'</span>] <span class="op">=</span> pcoa_embed[<span class="st">'group'</span>].replace({<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>,<span class="st">'no'</span>:<span class="st">'westernized'</span>, pd.NA:<span class="st">'ERR5766177'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Let’s first look at these components with 2D plots</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>ggplot(pcoa_embed, aes(x<span class="op">=</span><span class="st">'PC1'</span>, y<span class="op">=</span><span class="st">'PC2'</span>, color<span class="op">=</span><span class="st">'group'</span>)) <span class="op">\</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> geom_point() <span class="op">\</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> theme_classic() <span class="op">\</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a><span class="op">+</span> scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_107_0.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (407572134)&gt;</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>ggplot(pcoa_embed, aes(x<span class="op">=</span><span class="st">'PC1'</span>, y<span class="op">=</span><span class="st">'PC3'</span>, color<span class="op">=</span><span class="st">'group'</span>)) <span class="op">+</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>geom_point() <span class="op">+</span></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>theme_classic() <span class="op">+</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>scale_color_manual({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_108_0.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
<pre><code>&lt;ggplot: (407612651)&gt;</code></pre>
<p>Then with a 3d plot</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotly.express <span class="im">as</span> px</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> px.scatter_3d(pcoa_embed, x<span class="op">=</span><span class="st">"PC1"</span>, y<span class="op">=</span><span class="st">"PC2"</span>, z<span class="op">=</span><span class="st">"PC3"</span>,</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>                  color <span class="op">=</span> <span class="st">"group"</span>,</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>                  color_discrete_map<span class="op">=</span>{<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>},</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>                  hover_name<span class="op">=</span><span class="st">"sample"</span>)</span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>fig.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>3D PLOT HERE NOT DISPLAYED DUE TO RENDERING LIMITATIONS - PLEASE SEE JUPYTER NOTEBOOK</p>
</div>
</div>
<p><strong>Pause and think: How do you think this embedding represents how our sample relates to modern reference samples ?</strong></p>
<p>We can also visualize this distance matrix using a clustered heatmap, where pairs of sample with a small beta diversity are clustered together</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.spatial <span class="im">as</span> sp, scipy.cluster.hierarchy <span class="im">as</span> hc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>pcoa_embed[<span class="st">'colour'</span>] <span class="op">=</span> pcoa_embed[<span class="st">'group'</span>].<span class="bu">map</span>({<span class="st">'ERR5766177'</span>:<span class="st">'#DB5F57'</span>,<span class="st">'westernized'</span>:<span class="st">'#5F57DB'</span>,<span class="st">'non_westernized'</span>:<span class="st">'#57DB5E'</span>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>linkage <span class="op">=</span> hc.linkage(sp.distance.squareform(beta_diversity.to_data_frame()), method<span class="op">=</span><span class="st">'average'</span>)</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>sns.clustermap(</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>    beta_diversity.to_data_frame(),</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>    row_linkage<span class="op">=</span>linkage,</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>    col_linkage<span class="op">=</span>linkage,</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>    row_colors <span class="op">=</span> pcoa_embed[<span class="st">'colour'</span>].to_list()</span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>&lt;seaborn.matrix.ClusterGrid at 0x185b56100&gt;</code></pre>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="assets/images/chapters/taxonomic-profiling/analysis_115_1.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">png</figcaption>
</figure>
</div>
</section>
</section>
<section id="additional-steps" class="level3" data-number="10.3.8">
<h3 data-number="10.3.8" class="anchored" data-anchor-id="additional-steps"><span class="header-section-number">10.3.8</span> 8. Additional steps</h3>
<section id="source-tracking" class="level4" data-number="10.3.8.1">
<h4 data-number="10.3.8.1" class="anchored" data-anchor-id="source-tracking"><span class="header-section-number">10.3.8.1</span> 8.1 Source tracking</h4>
<p>Sourcetracker is a program that can estimate the proportion of different sources (reference biomes) contained in a sample (a sink). However, because of the statistical framework that it uses (MCMC with <a href="https://en.wikipedia.org/wiki/Gibbs_sampling#:~:text=In%20statistics%2C%20Gibbs%20sampling%20or,when%20direct%20sampling%20is%20difficult.">Gibbs sampling</a>), we recommend to limit the number of source samples to greatly reduce runtime</p>
<p>First, you will need to transform our relative abundance table to counts for SourceTracker</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>all_species_counts <span class="op">=</span> all_species.multiply(<span class="dv">1000000</span>).astype(<span class="bu">int</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>min_count <span class="op">=</span> all_species_counts.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">0</span>).<span class="bu">min</span>()</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>min_count</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>95327810</code></pre>
<p>Exporting the count table to <code>tsv</code></p>
<div class="sourceCode" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>all_species_counts.to_csv(<span class="st">"../results/sourcetracker2/all_species_counts.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index_label<span class="op">=</span><span class="st">'Taxon'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Converting to <a href="https://biom-format.org/"><code>biom</code> format</a></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>biom convert <span class="op">-</span>i ..<span class="op">/</span>results<span class="op">/</span>sourcetracker2<span class="op">/</span>all_species_counts.tsv <span class="op">\</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="op">-</span>o ..<span class="op">/</span>results<span class="op">/</span>sourcetracker2<span class="op">/</span>all_species_counts.biom <span class="op">\</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a><span class="op">--</span>table<span class="op">-</span><span class="bu">type</span><span class="op">=</span><span class="st">"Taxon table"</span> <span class="op">--</span>to<span class="op">-</span>hdf5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Converting the metadata to Sourtracker format</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>st2_metadata <span class="op">=</span> metadata[[<span class="st">'sample_id'</span>, <span class="st">'non_westernized'</span>]].rename(columns<span class="op">=</span>{<span class="st">'non_westernized'</span>:<span class="st">'Env'</span>, <span class="st">'sample_id'</span>:<span class="st">'#SampleID'</span>})</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>st2_metadata[<span class="st">'Env'</span>] <span class="op">=</span> st2_metadata[<span class="st">'Env'</span>].replace({<span class="st">'yes'</span>:<span class="st">'non_westernized'</span>,<span class="st">'no'</span>:<span class="st">'westernized'</span>})</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>st2_metadata[<span class="st">'SourceSink'</span>] <span class="op">=</span> [<span class="st">'source'</span>] <span class="op">*</span> st2_metadata.shape[<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>We subset it to select only 10 samples from each source</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>st2_metadata <span class="op">=</span> st2_metadata.groupby(<span class="st">'Env'</span>).sample(<span class="dv">10</span>).reset_index()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>st2_metadata <span class="op">=</span> st2_metadata.append({<span class="st">'#SampleID'</span>:<span class="st">'ERR5766177'</span>, <span class="st">'Env'</span>:<span class="st">'-'</span>,<span class="st">'SourceSink'</span>:<span class="st">'sink'</span>},</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>                                   ignore_index<span class="op">=</span><span class="va">True</span>)[[<span class="st">'#SampleID'</span>,<span class="st">'SourceSink'</span>,<span class="st">'Env'</span>]].set_index(<span class="st">'#SampleID'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<pre><code>/var/folders/1c/l1qb09f15jddsh65f6xv1n_r0000gp/T/ipykernel_40830/2882312005.py:1: FutureWarning:

The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>st2_metadata.to_csv(<span class="st">"../results/sourcetracker2/labels_st2.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, index_label<span class="op">=</span><span class="st">'#SampleID'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode" id="cb109"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="ex">sourcetracker2</span> gibbs <span class="dt">\</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">-i</span> ../results/sourcetracker2/all_species_counts.biom <span class="dt">\</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">-m</span> ../results/sourcetracker2/labels_st2.tsv <span class="dt">\</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">-o</span> ../results/sourcetracker2/st2 <span class="dt">\</span></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">--source_rarefaction_depth</span> 95327810 <span class="dt">\</span></span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">--sink_rarefaction_depth</span> 95327810 <span class="dt">\</span></span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">--jobs</span> 10</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Because SourceTracker is relying on MCMC sampling, it can very slow to run (which is why we won’t run it here)</strong></p>
<p>Among alternative faster solutions for source tracking are (among others):</p>
<ul>
<li>FEAST (<a href="https://doi.org/10.1038/s41592-019-0431-x">article</a>, <a href="https://github.com/cozygene/FEAST">code</a>),</li>
<li>Sourcepredict (<a href="https://doi.org/10.21105/joss.01540">article</a>, <a href="https://github.com/maxibor/sourcepredict">code</a>)</li>
</ul>
</section>
<section id="the-next-steps" class="level4" data-number="10.3.8.2">
<h4 data-number="10.3.8.2" class="anchored" data-anchor-id="the-next-steps"><span class="header-section-number">10.3.8.2</span> 8.2 The next steps:</h4>
<ul>
<li>Damage Analysis (<a href="https://ginolhac.github.io/mapDamage/">mapDamage</a>, <a href="https://github.com/Integrative-Transcriptomics/DamageProfiler">DamageProfiler</a>, <a href="https://github.com/maxibor/pydamage">PyDamage</a>)</li>
<li>Assembly (<a href="https://github.com/voutcn/megahit">megahit</a>, <a href="https://github.com/ablab/spades">metaSPAdes</a>), binning (<a href="https://bitbucket.org/berkeleylab/metabat/src/master/">metabat2</a>, <a href="https://sourceforge.net/projects/maxbin2/">maxbin2</a>, <a href="https://github.com/cmks/DAS_Tool">dastool</a>), and bin validation (<a href="https://github.com/Ecogenomics/CheckM">checkm</a>, <a href="https://github.com/grp-bork/gunc">gunc</a>)</li>
<li>Functional analysis (<a href="https://github.com/tseemann/prokka">Prokka</a>, <a href="https://github.com/biobakery/humann">Humann</a>)</li>
<li>Differential abundance (<a href="https://github.com/biobakery/Maaslin2">Maaslin2</a>, <a href="https://github.com/biobakery/lefse">Lefse</a>, <a href="https://github.com/biocore/songbird">Songbird</a>, GLM, Mixed effect models). Nice review by <a href="https://doi.org/10.1186/s12859-021-04193-6">Wallen 2021</a></li>
<li>genotyping</li>
<li>Phylogenies</li>
<li>…</li>
</ul>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./section-ancient-metagenomics.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Ancient Metagenomics</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./functional-profiling.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Functional Profiling</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">© 2022 SPAAM Community &amp; Authors. Available under <a href="http://creativecommons.org/licenses/by-sa/4.0/">CC-BY 4.0</a>.</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>